
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>GenBank parsing: both scripts - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/2008/03/14/genbank-parsing-both-scripts">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">GenBank parsing: both scripts</h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest to extract the DNA.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#!</span>
</span><span class='line'><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">env</span> <span class="n">python</span> <span class="s">&#39;&#39;&#39; input is a GenBank file. The script searches for</span>
</span><span class='line'><span class="s">gene annotations, extract all lines from the file and then parses these</span>
</span><span class='line'><span class="s">lines in order to extract protein sequences Ribosomal genes and other</span>
</span><span class='line'><span class="s">non-coding genes are not extracted - plan to do it later output is a</span>
</span><span class='line'><span class="s">fasta formatted file &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">Protein</span><span class="p">:</span> <span class="s">&#39;&#39;&#39;</span>
</span><span class='line'><span class="s">class that stores protein information &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span>
</span><span class='line'><span class="n">sequence</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; parses the entry received from the main</span>
</span><span class='line'><span class="s">function in order to extract information as protein id gi, etc &#39;&#39;&#39;</span>
</span><span class='line'><span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gene_data</span> <span class="o">=</span>
</span><span class='line'><span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span>
</span><span class='line'><span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span> <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span>
</span><span class='line'><span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span> <span class="k">if</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;sig_peptide&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">return</span>
</span><span class='line'><span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> \<span class="c">#only input is a genbank file</span>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">CDS</span> <span class="ow">and</span> <span class="n">appends</span> <span class="n">to</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span>
</span><span class='line'><span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found the</span>
</span><span class='line'><span class="n">DNA</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">we</span> <span class="n">can</span> <span class="n">stop</span> <span class="n">now</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">last</span> <span class="n">entry</span> <span class="n">after</span> <span class="n">leaving</span> <span class="n">the</span> <span class="n">loop</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>\<span class="c">#output for i in proteins: if len(i.gi) \&gt; 2: print i.gi, i.id output =</span>
</span><span class='line'><span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\\</span><span class="s">&quot;&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="mi">80</span><span class="p">))</span> <span class="k">print</span>
</span><span class='line'><span class="n">i</span><span class="o">.</span><span class="n">id</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="sd">&#39;&#39;&#39; script that extracts sequences from a GenBank file. Script reads the</span>
</span><span class='line'><span class="sd">gene CDS from the file and builds a list of start and end positions, and</span>
</span><span class='line'><span class="sd">if gene is complement outputs the 5&#39;3&#39; sequence and its reverse</span>
</span><span class='line'><span class="sd">complement only input is a GenBank file outputs a fasta file with the GI</span>
</span><span class='line'><span class="sd">ID as its name. &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">CDSinfo</span><span class="p">:</span> <span class="s">&#39;&#39;&#39; CDSinf</span>
</span><span class='line'><span class="s">class to store all the information from CDS &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span class='line'><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">=</span> <span class="n">gi_id</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'><span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span> <span class="bp">self</span><span class="o">.</span><span class="n">complement</span> <span class="o">=</span> <span class="n">complement</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; each CDS entry is obtained in the main</span>
</span><span class='line'><span class="s">function and a string of lines with information is passed to</span>
</span><span class='line'><span class="s">parse_entry to be parsed and have information extracted &#39;&#39;&#39;</span> <span class="n">gene_data</span>
</span><span class='line'><span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> \<span class="c">#changes a string to list, splitting at line</span>
</span><span class='line'><span class="n">ends</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">complement</span> <span class="o">=</span> <span class="bp">False</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> \<span class="c">#searches for regions annotated as CDS if line.find(&#39;</span>
</span><span class='line'><span class="n">CDS</span> <span class="s">&#39;) \&gt;=0: temp = line.split() \#checks for complement sequence, if</span>
</span><span class='line'><span class="n">true</span> <span class="n">remove</span> <span class="n">extra</span> <span class="n">characters</span> <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span>
</span><span class='line'><span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span> <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#checks for GI IDs elif line.find(&#39;GI:&#39;) \&gt;= 0: gi_id =</span>
</span><span class='line'><span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#get the gene name/function elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#and adds</span>
</span><span class='line'><span class="n">the</span> <span class="n">protein</span> <span class="nb">id</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">return</span> <span class="n">CDSinfo</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
</span><span class='line'><span class="n">complement</span><span class="p">)</span> \<span class="c">#only input is the genbank file with annotation and</span>
</span><span class='line'><span class="n">sequence</span> <span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">is_seq</span> <span class="o">=</span> <span class="bp">False</span> <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> \<span class="c">#reads the genbank file</span>
</span><span class='line'><span class="ow">and</span> <span class="n">whenever</span> <span class="n">finds</span> <span class="n">a</span> <span class="n">gene</span> <span class="n">annotation</span> \<span class="c">#concatenate the lines up to the</span>
</span><span class='line'><span class="nb">next</span> <span class="n">gene</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#if an entry is complete, send</span>
</span><span class='line'><span class="n">it</span> <span class="n">to</span> <span class="n">parse</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#appends to a list of CDSinfo objects</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found sequence start, set the flag on</span>
</span><span class='line'><span class="ow">and</span> <span class="n">parses</span> <span class="n">the</span> <span class="n">last</span> <span class="n">entry</span> <span class="n">is_seq</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="k">elif</span> <span class="n">is_seq</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> \<span class="c">#if flag is</span>
</span><span class='line'><span class="n">true</span> <span class="n">keep</span> <span class="n">going</span><span class="p">,</span> <span class="n">usually</span> <span class="n">sequences</span> <span class="n">are</span> <span class="n">store</span> <span class="n">at</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">the</span> <span class="nb">file</span> <span class="n">line</span>
</span><span class='line'><span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> \<span class="c">#this is an entry so append</span>
</span><span class='line'><span class="n">line</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="n">str_seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> \<span class="c">#make the sequence a string for i in</span>
</span><span class='line'><span class="n">sequence</span><span class="p">:</span> <span class="n">str_seq</span> <span class="o">+=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span> <span class="k">if</span>
</span><span class='line'><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span><span class="p">)</span> \<span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;.DNA.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="c"># if this is a complement, print both 5&#39;-3&#39; and reverse complement</span>
</span><span class='line'><span class="n">sequences</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">complement</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)]),</span>
</span><span class='line'><span class="mi">80</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;join&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)],</span>
</span><span class='line'><span class="mi">80</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>Checking the DNA extract script carefully, we notice
that we have an extra function in the <code>fasta</code> module, <code>invert</code>.
<strong>Update</strong>: A small change to the script is that <code>split('\n')</code> can be
replaced by <code>splitlines()</code> which is a built-in method for strings. This
hasn&#8217;t been added yet to the repository version of the file and an
explanation on its use will be added shortly.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">Paulo Nuin</span></span>

      








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
      


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.genedrift.org/blog/2008/03/14/genbank-parsing-both-scripts/" data-via="" data-counturl="http://zientzilaria.genedrift.org/blog/2008/03/14/genbank-parsing-both-scripts/" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
        <a class="basic-alignment left" href="/blog/2008/03/14/we-salute-you-2/" title="Previous Post: We salute you ...">&laquo; We salute you ...</a>
      
      
        <a class="basic-alignment right" href="/blog/2008/03/14/python-sets-and-intersections/" title="next Post: Python sets and intersections">Python sets and intersections &raquo;</a>
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere ...</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There's dumb, and there's ...</a>
      </li>
    
  </ul>
</section>




  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
