
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="We already have a function that reads the enzymes from a dataset in a
flat file (with one change: return) 1
2
3
4
5
6
7
8
9
10
11
12
def read_enzymes &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/34">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/restrinction-enzymes-second-take/">Restriction enzymes: second take</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T16:48:50-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We already have a function that reads the enzymes from a dataset in a
flat file (with one change: return)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">read_enzymes</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>  <span class="n">resenz</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>  <span class="n">start</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;Rich Roberts&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">start</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>          <span class="n">line</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</span><span class='line'>          <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
</span><span class='line'>              <span class="nb">buffer</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>              <span class="n">resenz</span><span class="p">[</span><span class="nb">buffer</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">buffer</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;\^&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>              
</span><span class='line'>  <span class="k">return</span> <span class="n">resenz</span>
</span></code></pre></td></tr></table></div></figure>


<p>We now need a function to write a function
that searches for the sites and a main function that accepts the
parameters, coordinate the search and return the results. Looks like we
are more than halfway there. Parameters input was shown before, starting
on section 3. We import the <code>sys</code> module and use the array inside
<code>sys.argv</code> to send the parameters to the script. A basic skeleton of our
main function would look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="c">#reading the ezyme dataset in one line and storing enzyme information in enzymeset </span>
</span><span class='line'><span class="n">enzymeset</span> <span class="o">=</span> <span class="n">read_enzymes</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;bionet.709&#39;</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">))</span>
</span><span class='line'><span class="c">#storing enzyme name on a string</span>
</span><span class='line'><span class="n">enzyme</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="c">#reading a FASTA file and storing the sequences</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">get</span>\<span class="n">_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s a start. Now we have to write a
function that will check for the enzyme name entered by the user in
order to check for the existence of such enzyme. Something like this
would work</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">check_enzyme</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="nb">set</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="nb">set</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="nb">input</span><span class="p">):</span>
</span><span class='line'>      <span class="k">return</span> <span class="bp">True</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="k">return</span> <span class="bp">False</span>
</span></code></pre></td></tr></table></div></figure>


<p>This
basically tests of the dictionary contains the name entered. If yes then
we return True, otherwise False is returned. This changes our main
script body</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="c">#reading the ezyme dataset in one line and storing enzyme information in enzymeset </span>
</span><span class='line'><span class="n">enzymeset</span> <span class="o">=</span> <span class="n">read</span>\<span class="n">_enzymes</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;bionet.709&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">))</span>
</span><span class='line'><span class="c">#storing enzyme name on a string </span>
</span><span class='line'><span class="n">enzyme</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="c">#check if the name entered is valid </span>
</span><span class='line'><span class="n">isname</span> <span class="o">=</span> <span class="n">check_enzyme</span><span class="p">(</span><span class="n">enzyme</span><span class="p">,</span> <span class="n">enzymeset</span><span class="p">)</span>
</span><span class='line'><span class="c">#if it is valid, continue, otherwise abort </span>
</span><span class='line'><span class="k">if</span> <span class="n">isname</span><span class="p">:</span>
</span><span class='line'>  <span class="c">#reading a FASTA file and storing the sequences </span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">get_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="s">&#39;Name invalid. Please try again.&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, we have a good idea on what to do now. We just need a function that creates a
regular expression and searches it on the sequence. Next time &#8230;</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/planning/">Planning</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T15:51:43-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Another aspect covered in the book that we haven&#8217;t seen yet is how to
plan, design out script or software. Usual ways to design a program
include writing use cases and drawing
<a href="http://en.wikipedia.org/wiki/Unified_Modeling_Language">UML</a> diagrams
(stands for Unified Modelling Language). Here we will scratch the
surface of <a href="http://en.wikipedia.org/wiki/Use_case">use cases</a>, where we
will try to determine how the program will interact with the user. How
to write use cases or design UML is not a subject taught in many biology
courses, so we won;t see much of the theory here. Consider this a more
informal way of planning a small script or application.</p>

<p>First thing
would be to set a goal:</p>

<p><em>What is the main objective of our script?</em></p>

<p>Create a simple restriction enzyme map of certain sequences Next, <em>what
do we need to make the program work?</em> We need restriction enzyme
information, such as names and sites and a sequence. That leads us <em>How
do I store information?</em> Restriction enzyme data (last entry) obtained
from a file can be stored in a dictionary, with the enzyme name as key
and the site as value.</p>

<p>Sequences are stored using our fasta class. This
will brings us to one important issue <em>How to interact with the user?</em>
The ideal way would be to present a list of enzymes for the user to
select, but we do not have a graphical interface to organize it nicely.
So, we will ask the user to enter the name of enzyme and a name of a
file to read the sequence where the enzyme site should be found. We can
do that interactively or by passing parameters to the script. We will do
it by parameters here, but the interactive way can be tested as a
homework for those following the blog.</p>

<p>And finally <em>What about an
output?</em> We will do the same way as the book: a list of positions,
indicating the restriction sites. I welcome other options in the
comments. It looks we have a plan. Let&#8217;s gather what we already have,
and our previous knowledge and meet on the other side of this post.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/creationism-and-third-world/">Creationism and third world</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T12:04:19-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I know third world countries are currently known as under-development, a more politically correct way to say the same thing. And I also know that news coverage about under development countries, at least in North American media, is very limited and brief. Usually you find reports in the business section or when some natural or man made catastrophe happens. So it is no surprise when you don&#8217;t find anything about creationism teaching and under-development countries (at least on the outlets I am used to check).</p>

<p>This week, the most important Brazilian magazine, Veja (like it or not) has a review on the educational quality of Adventist-supported schools. The title of the report is clear: <a href="http://veja.abril.com.br/120907/p_116.shtml">Thank God - and not Darwin</a> (in a free translation; warning: resource available to subscribers and in Portuguese). After introductory paragraphs putting in historical context church-supported schools in Brazil and the good level of formal education of students coming out these schools, the article gets to the point of evolution/creationism teaching that is common in such schools.</p>

<p>We learn that religion classes are a daily event, and in such classes religious hymns are chanted with fervour. And this starts in kindergarten, along with children&#8217;s versions of the Old Testament. And the most important, creationism is taught side-by-side with evolution.</p>

<blockquote><p>We are clear about our point of view, creationist, but we allow the students to know both sides. (&#8230;) We respect all customers. Furthermore, they need to know Darwin to be accepted in a university&#8221;</p></blockquote>

<p>Ivany Queiroga da Silva, principal of an Adventist school</p>

<p>In Brazil there is no current unified score to access university. Usually a test hosted by the university is applied once or twice a year with general questions on every subject. That&#8217;s why evolution has not been completely banned from their curriculum, but the article also points that such schools are the ones with the largest increase in number of students and profit. Also the Adventist church already has six universities and publish their own school books.</p>

<p>To me, it seems a pretty gloomy future. There is no obstacle for the universities controlled by the adventists to ban evolution from their entry test and from the regular curriculum. Even though they are clearly motivated by profit, it is not impossible that they ban evolution from their schools. When these children, that are taught from the early age with religion concepts, become adults and parents they might see Darwin&#8217;s Theory as a wrong and sinful concept. Then the circle is complete, and evolution will be a no-no in these schools and universities.</p>

<p>In a country where more than half of the population is functional illiterate and the science education is precarious, to say the least, these news are very sad. Instead of stepping into the future, it seems that things are slowly drifting back to the past.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/06/scaring/">Scaring </a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-06T13:11:56-04:00" pubdate data-updated="true">Sep 6<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><blockquote><p>&#8220;They teach evolution in the Ontario curriculum, but they also could teach the fact to the children that there are other theories that people have out there that are part of some Christian beliefs&#8221;</p></blockquote>

<p>(full report <a href="http://www.thestar.com/OntarioElection/article/253725">here</a>)</p>

<p>John Tory, Conservative Party leader trying to become Ontario&#8217;s Prime Minister. Yep, one day or another some Canadian politician would start the debate, for good or for worse. The Conservative Party is missing a good opportunity to become more modern and embrace the truly Canadian spirit, not supporting the non-sense of creationism teaching. Even on faith-based schools only.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/06/not-a-very-good-start/">Not a very good start &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-06T09:38:25-04:00" pubdate data-updated="true">Sep 6<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I subscribe to the job announcements list of Bioinformatics.org (as a post-doc you are never safe) and decided to check <a href="http://bioinformatics.org/forums/forum.php?forum_id=5692">this ad</a> which was sent yesterday. Basically is a Bioinformatician position at  Nestlé in Switzerland and look promising due to the size and history of the company. First thing you do is to check how to submit your information, and there is a <a href="http://www.careers.nestle.com">link</a> in the add. Great!</p>

<p>I open the page and check how to send an application. Ok, decided to search for <a href="https://nestle.recruitmax.com//main/careerportal/default.cfm?szUniqueCareerPortalID=9eaa1130-14c5-49ef-9ad1-4e40341f9454&amp;szIsJobBoard=0">all jobs</a>. As any Firefox user what I see is this</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ff2.jpg"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ff2.jpg" alt="nestle_ff2.jpg" /></a></p>

<p>basically a picture at the top and a blank page. Scrolling down</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ff1.jpg"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ff1.jpg" alt="nestle_ff1.jpg" /></a></p>

<p>I see the options, but where is the search button. This happened on Linux and Windows. Maybe the site is broken or maybe they forgot to make it compatible with Firefox. I fire up IE and</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ie.jpg"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/09/nestle_ie.jpg" alt="nestle_ie.jpg" /></a></p>

<p>a perfect site comes up. Checking the code a you see that JavaScript is there, but no reference to a cross-browser function.</p>

<p>I really wish to search for other positions and even apply to the advertised position from my favourite browser. I know my main page is not IE friendly, but I am just a person, not a corporation. And my site is not broken on IE 7, and if you are still using IE 6, good luck.</p>

<p>It wasn&#8217;t a good start for Nestlé, at least for me.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/31/restriction-enzymes-first-take/">Restriction enzymes: first take</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-31T16:08:12-04:00" pubdate data-updated="true">Aug 31<span>st</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We now jump with both feet on the main topic of the book&#8217;s chapter,
which is generating restriction maps of DNA sequences. First step is to
obtain restriction enzyme information, read it and format in a way that
our main script will understand. We will use the same dataset as the
book, the <a href="http://rebase.neb.com/rebase/rebase.html">Rebase</a> database at
New England Biolabs. In the book, Tisdall suggests the download of a
<em>bionet</em> format, which can be downloaded
<a href="ftp://ftp.neb.com/pub/rebase/">here</a> (scroll down to bionet.709).</p>

<p>This file looks like this:</p>

<pre><code>AaaI (XmaIII)                     C^GGCCG  
AacI (BamHI)                      GGATCC  
AaeI (BamHI)                      GGATCC  
AagI (ClaI)                       AT^CGAT  
AanI (PsiI)                       TTA^TAA  
AaqI (ApaLI)                      GTGCAC
</code></pre>

<p>where the first column contains the enzyme names and the second column
has the actual cleavage sites.</p>

<p>It won&#8217;t be difficult to parse this file
and create a dictionary with the names and sites, but it would be easier
if the file the data was tab-separated, but it is nothing we will have
problem dealing with. The file also has some header lines which we will
have to avoid. Next entry will deal with user cases and some other
aspects of coding, as we head to our most complex script until now. Noe
we will only create a function that reads the file and generates the
dictionary, what sounds very simple. There are two ways (maybe even
more) to discard the header lines: programatically or actively deleting
such lines. We won&#8217;t delete the lines this time, so we have to follow
the other path.</p>

<p>The easiest way to eliminate header lines from our
parsing function would be to count the lines and start the parsing
procedure after that certain number but that would mean a high
confidence of the format being constant in every release. So, first we
get rid of the header lines</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">read_enzymes</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>  <span class="n">resenz</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>  <span class="n">start</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;Rich Roberts&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">start</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>          <span class="n">line</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
</span></code></pre></td></tr></table></div></figure>


<p> where we already
declared the dictionary that will receive the name and sites. We have a
flag boolean that tells the script where the actual enzyme list starts
(<code>start</code>), declared as false and then modified to true when the line
containing <code>Rich Roberts</code> is found. The line <code>line = file.next()</code> tells
the script that the current line is equal to the next line of the file.
We do this to avoid starting the parsing of the file at the line we
found <code>Rich Roberts</code>. the <code>if</code> statement checks for the line size in
order to split and parse only the actual lines and discard empty ones.
Now, in order to get the sites and enzyme names we will use split,
differently than we used before. This time we won&#8217;t pass any arguments
to split as a separator. This will trigger another type of splitting
procedure, where some characters will be stripped from the string
(spaces, tabs, newlines, returns, and form feeds) and the resulting words
will be then separated by arbitrary length strings of white spaces.
Basically, this split will return a list with the words in that particular
line. Inspecting the file carefully, we notice that some enzyme names
have an extra id between parentheses. We will not consider this extra
id. After splitting we will get the first and last elements of the list
and put them in our dictionary.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">buffer</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'><span class="n">resenz</span><span class="p">[</span><span class="nb">buffer</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">buffer</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;\^&#39;</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Done. We have the dictionary ready. And using Python&#8217;s
included batteries we use the last line in the function to remove the
circumflex characters. Putting everything together we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">read_enzymes</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>  <span class="n">resenz</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>  <span class="n">start</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;Rich Roberts&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">start</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>      <span class="n">line</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
</span><span class='line'>          <span class="nb">buffer</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>          <span class="n">resenz</span><span class="p">[</span><span class="nb">buffer</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">buffer</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;\^&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/28/finding-motifs-iupac-and-regex-an-approach/">Finding motifs: IUPAC and RegEx, an approach</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-28T17:21:33-04:00" pubdate data-updated="true">Aug 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>After a long delay, we are back. Before entering in the next topic,
Restriction Enzymes, let&#8217;s take a look how to create a regex pattern
from user input and the dictionary of IUPAC code for nucleotides. We
will use the same dictionary from the previous entry</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">iupacdict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;M&#39;</span><span class="p">:</span><span class="s">&#39;[AC]&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">:</span><span class="s">&#39;[AG]&#39;</span><span class="p">,</span> <span class="s">&#39;W&#39;</span><span class="p">:</span><span class="s">&#39;[AT]&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;S&#39;</span><span class="p">:</span><span class="s">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">:</span><span class="s">&#39;[CT]&#39;</span><span class="p">,</span> <span class="s">&#39;K&#39;</span><span class="p">:</span><span class="s">&#39;[GT]&#39;</span><span class="p">,</span> <span class="s">&#39;V&#39;</span><span class="p">:</span><span class="s">&#39;[ACG]&#39;</span><span class="p">,</span> <span class="s">&#39;H&#39;</span><span class="p">:</span><span class="s">&#39;[ACT]&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;D&#39;</span><span class="p">:</span><span class="s">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">:</span><span class="s">&#39;[CGT]&#39;</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p>and the same consensus sequence of the GATA3 binding site NNGATARNG This
consensus sequence will be provided as a script parameter, along with
the filename</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="n">sequencefile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">motif</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>First line reads the sequence file from the
user input and second line stores the input motif in a string. Now we
have to get the motif string and check for each letter (IUPAC code) and
get the correspondent set of nucleotides. For this we will use a loop
and a the method <code>get</code> method of Python dictionaries. This method, as
its name implies, gets the <code>value of the key</code> in parentheses. Like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">motif</span><span class="p">:</span>
</span><span class='line'>  <span class="n">iupacdict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>If we combine both excerpts above and run
using as input the GATA3 model the result would like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="n">ACGT</span><span class="p">]</span> <span class="p">[</span><span class="n">ACGT</span><span class="p">]</span> <span class="p">[</span><span class="n">AG</span><span class="p">]</span> <span class="p">[</span><span class="n">ACGT</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>which is five nucleotides short of the motif length. How to correct it? Pretty simple
we just add to the dictionary the &#8220;regular&#8221; nucleotide codes</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">iupacdict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;M&#39;</span><span class="p">:</span><span class="s">&#39;[AC]&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">:</span><span class="s">&#39;[AG]&#39;</span><span class="p">,</span> <span class="s">&#39;W&#39;</span><span class="p">:</span><span class="s">&#39;[AT]&#39;</span><span class="p">,</span> <span class="s">&#39;S&#39;</span><span class="p">:</span><span class="s">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">:</span><span class="s">&#39;[CT]&#39;</span><span class="p">,</span> <span class="s">&#39;K&#39;</span><span class="p">:</span><span class="s">&#39;[GT]&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;V&#39;</span><span class="p">:</span><span class="s">&#39;[ACG]&#39;</span><span class="p">,</span> <span class="s">&#39;H&#39;</span><span class="p">:</span><span class="s">&#39;[ACT]&#39;</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">:</span><span class="s">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">:</span><span class="s">&#39;[CGT]&#39;</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;N&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p>This would solve our &#8220;problem&#8221; without adding
a line of code to the final script. Fastforwarding to creating the regex</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">mregex</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">motif</span><span class="p">:</span>
</span><span class='line'>  <span class="n">mregex</span> <span class="o">+=</span> <span class="n">iupacdict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">mregex</span>
</span><span class='line'><span class="c">#just to check </span>
</span><span class='line'><span class="n">tosearch</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">mregex</span><span class="p">))</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tosearch</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">sequencefile</span><span class="p">):</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">i</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple and quick. The final output is not really elaborated, but we can improve it. Next time.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/16/sciview-in-pdf/">SciView in PDF</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-16T16:52:12-04:00" pubdate data-updated="true">Aug 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I am preparing a PDF file with all five interviews so far, in a stylish format.</p>

<p>It will be ready next week and widely available for download.</p>

<p><em>Clearly it has been LONG delayed a bit. I will try soon.</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/11/thanks-tony/">Thanks Tony!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-11T10:08:13-04:00" pubdate data-updated="true">Aug 11<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Without you some days wouldn&#8217;t be as nice, and some nights wouldn&#8217;t be as depressive.</p>

<p><a href="http://news.bbc.co.uk/2/hi/uk_news/england/manchester/6941716.stm">Cheers!</a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/11/sciview-part-5-interview-with-alexei-drummond/">SciView part 5: interview with Alexei Drummond</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-11T09:53:56-04:00" pubdate data-updated="true">Aug 11<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are back with more interviews. This time it is Dr Alexei Drummond&#8217;s take on Science and everything that surrounds it. Apart from his academic work, Alexei is Chief Scientist of <a href="http://www.biomatters.com">Biomatters</a> the developers of <a href="http://blindscientist.genedrift.org/2007/03/02/geneious-one-year-or-almost-later/">Geneious</a> (and <a href="http://blindscientist.genedrift.org/2007/02/26/hello-world/">here</a>). Enjoy.</p>

<hr />

<p><em>You are Biomatters&#8217; Chief-Scientist, one of the founders of the company and the mastermind behind Geneious. What was your motivation to start the company and develop the program? How to balance these activities with your regular job as a university professor?</em></p>

<p>I would say that my main motivations for wanting to build Geneious were frustration and hope. I was frustrated at the number of molecular biologists who didn&#8217;t seem to have the software tools they needed to do their job. I was hopeful I could provide the framework for those tools, if only I had a few great programmers.</p>

<p>Since I was one of the only people in the department who could program, I was often asked to write scripts, or teach people how to program enough so they could do it themselves. When biologists start asking about where they can learn to program a computer, just so they can do their job you know something is wrong! Most of the things they wanted to do were reasonably straightforward. It seemed to me that these kind of basic productivity problems had been solved in other workplaces like the office &#8211; so why had they not been solved in the research laboratory? I initially wanted to build the system as an academic research project &#8211; but then realized that would fail at a number of levels. Firstly, software development isn&#8217;t science &#8211; but science now crucially relies on software development and support. Secondly, most academic programmers are not interested in (or good at) designing user interfaces, and certainly developing software is not a scientific outcome that gets recognized like publishing papers does. Thirdly, academics are quite bad at supporting software and documenting it. So it seemed to me, that for a lot of reasons, a professional software company was the best avenue to realize a software system that would dramatically improve the productivity of molecular biologists by putting bioinformatics at their fingertips. Really I just wanted to make science go faster &#8211; and with less hassle.</p>

<p>_ Geneious is a Java application? Why Java? Any special feature in the language or previous experience? Is there anything that you wish Java was capable of so you can include in Geneious?_</p>

<p>Java is a general-purpose programming language &#8211; so you can do in Java pretty much anything you can do in software. The main reason for choosing Java is that it is very easy to write sophisticated user interfaces that run on Windows, Linux and Mac OS X. This is really important in the field of molecular biology where about 20% of the users have Mac OS X. Earlier versions of Java didn&#8217;t integrate with some aspects of different operating systems as well as they could, but with Java 5 most of this integration is quite good. Also, more and more academic developers are using Java (because it is easy, cross-platform and OO) so that also works in our favor. Every language has its pros and cons, but  for our purposes Java was one of the best choices.</p>

<p><em>You are the first scientist outside North America that is interviewed here. You have worked for some years in Oxford, UK. What are the main differences of working in New Zealand in comparison to Europe? How your return to New Zealand has influenced your decision to start the company?</em></p>

<p>New Zealand is a very small country with just 4 million people. But we are a proud nation. But we have a long history of world firsts, like being the first country to give woman the right to vote (1893). A New Zealander was the first to &#8220;split the atom&#8221; (Ernest Rutherford), and a New Zealander was the first to the top of Mt Everest (Sir Edmund Hillary). We are also one of the first countries to declare ourselves a Nuclear-Free Zone (1985). So I would say that New Zealand is a place the is not afraid to try to lead the way, despite our small size. Our Prime Minister, Helen Clark, recently declared our Government&#8217;s aspiration to make New Zealand a carbon-neutral country. It is these kinds of radical ideas that we strive to make reality. So it is quite natural to try to build a unique software package in New Zealand &#8211; although we are still far from where we are aiming to be.</p>

<p>_In your opinion what are the main differences of developing applications in academia and in the private sector?
_</p>

<p>Academics develop software for the love of it &#8211; and their first aim is to make the software work for themselves, and secondarily to work for their colleagues, as long as that doesn&#8217;t take too much effort. There are a few academics that try very hard to produce easy-to-use well-supported software, but they tend not to have the resources to do what the private sector can - for example PAUP<em> still doesn&#8217;t have a manual after many many years. The challenge for the private sector developing </em>scientific* software, is how to keep abreast of the science, which is changing so rapidly. At Biomatters we really try to have the idea that we are part of the scientific community. We have invited a number of scientist to work with us on various projects and we find that in this way we learn a lot more about what is needed, and how best we can fit into the academic ecology. Our goal is a happy marriage where academic programmers can get on with developing great new algorithms, and Geneious can provide the interoperability, the user interface and the support.</p>

<p><em>What would you teach to an &#8220;academic developer&#8221;? What developers in academia lack when compared to developers working in companies?</em></p>

<p>I don&#8217;t think it is a question of what academic developers lack in terms of skills &#8211; I think it is more a question of motivation. Aside from Geneious I produce academic software (like BEAST). What motivates me to answer 100 technical email queries every month about the BEAST software? My main passion is science, not supporting software. So by creating a successful software package, I have created a heavy burden of support. I would rather that somebody else supported all the simple questions about user interface, file conversion, installation, et cetera&#8230; So in some sense, what I am hoping to &#8220;teach&#8221; academic developers is that maybe there is another way to do things. It we had one framework that made interoperability and user interface more seamless and standard, each academic could focus on the science, the new algorithms and the part of their job that they find really exciting &#8211; the discovery.</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. Geneious has a very nice GUI, but in my opinion still lacks some simple features. Can you tell us how the program&#8217;s GUI was first designed and what kind of user input is considered in the development? Which is more important: &#8220;what the user want&#8221; or &#8220;what we want&#8221;?</em></p>

<p>Most users didn&#8217;t say they wanted the mouse, or the iPod before the visionary told them they wanted it. So I do think there is room for a lot of innovation in software development. But of course, there are definitely things that users know that they need right now. Balancing these two things is the real trick. Biomatters is very serious about listening to our users and we have an active forum and a small group of power users that have been quite helpful in the past. Of course there is always more to do. We also think that we have introduced a few nice new ideas in our software and there are more on the way. Specifically we are interested in developing the concept of &#8220;live&#8221; research publications, where arrival of relevant new data automatically updates the published research in that area. As you can imagine, this is a long term goal. But I think you need these visions to aim for, if you are going to make real progress.</p>

<p><em>For the young bioinformatician: should s/he focus on publications, skills, languages? Or a balance of all? What is your advice?</em></p>

<p>A young bioinformatician should focus on the science. What are the outstanding questions that your skills enable you to take a step towards answering? Once you focus on the science the publications, skills and languages will come.</p>

<p>I used to program ten different languages very well. But 95% of my programming these days is done in Java (because I am usually building on stuff I have done previously). PERL is still quite important in bioinformatics &#8211; but I think that is because there is still a lot of need to write one-off scripts. Will we get to a point where there is more sophistication? I think that bioinformatics has to become a field where people without programming skills can contribute substantially. I would argue that all of the programmers in bioinformatics should be working very hard to program themselves out of their jobs (and into more satisfying jobs). Lets build software systems that take this field to the next level. That is what we am trying to do (slowly but surely) with Geneious. We want to create a system that ultimately makes the programmers redundant &#8211; at least for the menial tasks &#8211; so that any molecular biologist can construct complicated queries and analyses in an intuitive graphical manner. Then the programmers can focus on the hard questions like new algorithms and methods of analysis.</p>

<p><em>What in your opinion is your most important software?</em></p>

<p>I don&#8217;t really compare my software projects, because they are so different. I guess this is for other people to decide. But judging from you line of questioning, Geneious is currently winning ;-)</p>

<hr />

<p>I would like to thank Alexei for answering my questions.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/35/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/33/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
