
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="In Beginning Perl for Bioinformatics the chapter that covers simulated
mutations on a DNA sequence is quite verbose and the code examples
employ some &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/11">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/04/a-script-to-simulate-dna-sequence-sets/">A script to simulate DNA sequence sets</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-04T12:17:22-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In Beginning Perl for Bioinformatics the chapter that covers simulated
mutations on a DNA sequence is quite verbose and the code examples
employ some subroutines to do what we have done on the last post.
Basically the code example that generates a random DNA sequence is the
last one on the chapter, but it was the first one we covered. Also this
code example has a twist that our code from the last post does not have,
which is it allows you to generate a set of sequences with different
length instead of one sequence with fixed length that our script does.
So, the script that generates a user defined simulated DNA sequence set
is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>    <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>    <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple and efficient. First we
define a function that generates a simulated DNA sequence from the four
nucleotides, again using the <code>random.choice</code>. The sequence length is
based on the parameter received by the function. The sequence length is
another random number defined in the main body of the script. This
random number is generated by <code>random.randint</code> with a range based on the
arguments given by the user when running the script. And finally, the
number of sequences to be simulated is define by the first parameter.
Seventeen lines. Next we will see how to draw some scientific
information about the sequences, such as sequence identity and
nucleotide frequency.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/02/random-sequences-in-python/">Random sequences in Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-02T12:57:24-04:00" pubdate data-updated="true">Apr 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>There is a reason to say that Python has batteries included. On the last
post we have seen a small example of randomization in Python, generating
random integer values for a (extremely) simple dice game. We move to
another example, still simple which will allow us to generate random DNA
sequences. The script is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">dnaseq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s">&#39;ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">result</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dnaseq</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">result</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not fancy at all, just plain simple (yet again). We
import a couple of modules, <code>sys</code> and <code>random</code>, and ask for the sequence
length as a script parameter. <code>dnaseq</code> is a list containing a tandem
repeat of ACGT, and from there we will extract our random nucleotides.
In fact <code>dnaseq</code> could have been &#8216;ACGT&#8217; only. The fact that we create a
string and convert it to a list, is just for convenience of writing
<code>'ACGT...'</code> easier than <code>['A', 'C' ...]</code>. We then declare an empty
string that will be used to store the random sequence. And inside the
loop, the command that does all the magic: <code>random.choice()</code>. This
command will return a random element from the list passed as subject.
Using it inside a loop we will get a random nucleotide on each iteration
and add it to our string. This is a very simple command, but at the same
time extremely powerful and easy to implement. A good exercise from this
would be to modify the <code>dnaseq</code> string and see if there is any change in
the final random sequence.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/30/ramdomization/">Ramdomization</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-30T15:19:30-04:00" pubdate data-updated="true">Mar 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We start section 4 with a very short introduction to randomization.
Chapter 7 of BPB discuss the use of randomization to obtain mutations in
DNA and protein sequences. The example given in the book is at the same
time simple and interesting, as it creates a paragraph from random
selections of nouns, adjectives, verbs and other grammar elements. Here
we are going to to create a very (stress on very) simple dice game,
where each time you run the script it will throw two dices for you and
two dices for the computer. It then prints the sum of the dices and
tells who won the match. Randomization is an important feature of
computer languages. There are different researchers involved in the
creation of the best approaches to generate random number in computers.
Random number are important in the simulation of different natural
processes, such as genetic mutation, gene drift, epidemiology, weather
forecast, etc. The better the generator, the better the simulation.
Python has a random module included, with a myriad of functions that can
perform different randomizations. In this post we will see the <code>integer</code>
randomization, and in later entries we will see some other powerful
functions. Our script is quite simple, and the only new aspect for us
here is the random module and the <code>randint</code> function.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'>
</span><span class='line'><span class="n">dice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">dice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">mine</span> <span class="o">=</span> <span class="n">dice1</span> <span class="o">+</span> <span class="n">dice2</span>
</span><span class='line'>
</span><span class='line'><span class="n">his</span> <span class="o">=</span> <span class="n">computerdice1</span> <span class="o">+</span> <span class="n">computerdice2</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&#39;mine = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mine</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; vs. computer = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">his</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">if</span> <span class="n">mine</span> <span class="o">&gt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;I won&quot;</span>
</span><span class='line'><span class="k">elif</span> <span class="n">mine</span> <span class="o">&lt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Computer won&quot;</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Tie. Try again&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>random.randint</code> is a function that generates an integer random number
between a range specified by the number between parentheses. In the
above case, we are using a dice of 6 values. So, for every call of
<code>random.randint(1,6)</code> a random number between 1 and 6 will be generated.
We store this number in a variable, sum the user&#8217;s dices and the
computer&#8217;s and check with a <code>if</code> clause to see the winner. A good
exercise would be to make this script interactive, allowing multiple
matches. Next we will see another random module function and then we
will generate mutations on DNA sequences.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/26/command-line-arguments-and-a-second-take-on-functions/">Command line arguments and a second take on functions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-26T13:18:48-04:00" pubdate data-updated="true">Mar 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen, briefly, how to define and use a function in Python. Now
we are going to jump forward a bit and create a new function and at the
same time take a look on command line parameters that can be passed to
the script. If you have used command line applications before, you might
have encountered programs that asks for a file name, a calculation
parameter, etc to be passed in the command line. Python scripts are no
different, they accept such parameters. For this we have the <code>sys</code>
module that has system specific parameters and functions. We have used
before the <code>sys.exit</code>, imported as an extra module function.</p>

<p>Every operating system (even Windows) has arguments in its command line, and
programming languages usually call such arguments <code>argv</code> (in the C/C++
you have argv in the parameters of the main function). Lists in Python
start at 0 (zero), and for the argument list the first item is the
script/program name. Basically if we have this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'><span class="nv">$&gt;</span> python myscript.py DNA.txt
</span></code></pre></td></tr></table></div></figure>


<p><code>myscript.py</code> is the argument 0 in the
list and <code>DNA.txt</code> is the argument 1. So whenever we create a script that
receives arguments in the command line, we have to start (in most cases,
be aware) from 1.</p>

<p>In Python using system arguments in the CLI will look
like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">filename</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">valueone</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>We will a variation of our
previous script that counts the bases, now with command line arguments
and a function (with no &#8220;error&#8221; checking at first)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python import sys </span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">count_nucleotide_types</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">totalA</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalC</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalG</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalT</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequencefile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequencefile</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">values</span> <span class="o">=</span> <span class="n">count_nucleotide_types</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;As&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Cs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Gs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Ts&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>Few new things here. We created a function <code>count_nucleotide_types</code> that should receive a string
containing the sequence. The &#8220;real&#8221; first line of the program flow is
the one that gets the name of the file from the command line argument,
open and read it. We then convert the list to a string, modify it a but
and throw it to the function. Get the result back, and done. With
functions we actually don&#8217;t save coding time/length (at least here), we
make out code more organized, easier to read and somewhat easier to
someone else read and understand it. It is not a good coding practice to
have long programs/scripts with no functions, no subdivision, no
structure. Functions are sometimes good program nuggets that can be
reused in the same application or even ported/copied to other
applications and reused indefinitely. Soon we will see a function and
class that reads a FASTA file in Python that can be used anywhere in any
program that needs such feature. Try the code and come back later for more.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/22/python-functions-simple-example/">Python functions: simple example</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-22T09:04:10-04:00" pubdate data-updated="true">Mar 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Python subroutines do not exist. <a href="http://www.diveintopython.org/getting_to_know_python/declaring_functions.html">Everything is a function, all
functions return a value (even if it&#8217;s None), and all functions start
with <code>def</code></a>.
This statement is from Dive into Python, a book on Python programming
available for free. As mentioned Python functions start with the word
<code>def</code>, which is followed by the function name that is followed by the
arguments the function receives in between parentheses. Something like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span></code></pre></td></tr></table></div></figure>


<p>Usually Python coders
(sometime called Pythonistas, among others), following the Python coding
style (that states: <em>Function names should be lowercase, with words
separated by underscores as necessary to improve readability</em>) name
their functions with words separated by underscores. And we are going to
use this style here, whenever a function becomes handy. The parameters
passed to the function (above <code>somevalue</code>) do not have a datatype,
Python should handle it whatever is being passed. It is also attribute
of your code to handle the parameter/value passed inside the function
and avoid errors. Functions also follow the same indentation of normal
programming and the line after the declaration should be indented with
four spaces</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span><span class='line'>    <span class="n">do_something</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, let&#8217;s
warmup with functions. The following script is just the start: it adds a
poly-T tail to a DNA sequence. We are going to use our old friend
AY162388.seq. I will be back after the script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="k">def</span> <span class="nf">add_tail</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="n">seq</span> <span class="o">+</span> <span class="s">&#39;TTTTTTTTTTTTTTTTTTTTT&#39;</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&#39;AY162388.seq&#39;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="n">add</span>\<span class="n">_tail</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not very useful, at first sight, but gives us an
impression of what a function looks like. Basically we define a function
<code>add_tail</code> that receives <code>seq</code> as a parameter. Don&#8217;t worry about
variable scope now, we will see it later. The rest of the script is just
like things we saw before, except for the line
<code>sequence = add_tail(sequence)</code>. Here we are saving memory (yep, not
that much and not even impressive) by assigning the return value of the
function to the same string where we have the sequence stored. Run the
scritp and get ready for the command line arguments.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/15/debugging-in-python/">Debugging in Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-15T09:43:10-04:00" pubdate data-updated="true">Mar 15<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Beginning the third section in our tutorial/guide, we are going to see
the chapter six of BPB. This chapter discusses the topics of creating
subroutines (in Python&#8217;s case functions) and debugging the code. We are
going to start by the end.</p>

<p>In Python, code debugging can be done as in
any other programming language: Perl has pdb, C/C++ has gdb, etc. Python
also has a pdb module that can be imported and run to check for errors
in your code. Using this command line:</p>

<p><code>python -m pdb myscript</code></p>

<p>will
start the debug module and this will run your script. If you are an
experienced programmer, who is just starting Python, pdb usage might
look simple and straightforward. On the other hand, if you don&#8217;t have a
lot of experience in programming I would suggest a different approach,
as you become more comfortable with the language. Python has a great
advantage over some other interpreted languages, allowing you to
interactively code using the interpreter. So if your code is not working
properly, maybe a wrong output or a value that is not being correctly
calculated you have the options of coding the part of your script that
is not working using the interpreter or use the first rule of debugging:
include <code>print</code> statements that output the value of variables/objects.
Another option is to use a Python code editor, what will also help you
with highlight your code. I have little experience with Python code
editors, as I normally code in Linux and use
<a href="http://kate-editor.org/">Kate</a>. Lately I have been trying <a href="http://www.activestate.com/products/komodo_edit/">Komodo edit</a> which is a
cross-platform freeware from Active State. It looks pretty good but I
never tried debugging my code with it. So, these are my advices if you
are just starting to program. Maybe because of the age of Beginning Perl
for Bioinformatics (published in 2001), Perl&#8217;s pdb was the only option
back then. Thanks to major advances on open-source and free software
there are many other options nowadays to debug your code.
<code>python -m pdb myscript</code></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/14/section-2-a-nice-summary-script/">Section 2: a nice summary script</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-14T15:32:24-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Closing section two, let&#8217;s use everything we saw before and write a nice
script that will read a sequence file (DNA) and report us of any
&#8220;errors&#8221; and the number of different nucleotides. It is very simple, but
a good exercise. Depending on your needs you can easily modify it to
check for other characteristics of sequences, even change it to read
amino acids sequences. We start with the code, comments coming after it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="n">fileentered</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'><span class="k">while</span> <span class="n">fileentered</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'>    <span class="n">filename</span> <span class="o">=</span> <span class="nb">raw_input</span><span class="p">(</span><span class="s">&#39;Please enter a file to check: &#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>        <span class="k">try</span><span class="p">:</span>
</span><span class='line'>            <span class="n">seqlist</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>            <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seqlist</span><span class="p">)</span>
</span><span class='line'>            <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">totalA</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">totalC</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">totalG</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">totalT</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">otherletter</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&#39;[BDEFHIJKLMNOPQRSUVXZ]&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">extra</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">otherletter</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="o">+</span><span class="s">&#39;.count&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Count report for file &#39;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;A = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;C = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;G = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;T = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">extra</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Also were found &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extra</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; errors</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">extra</span><span class="p">:</span>
</span><span class='line'>                    <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="s">&#39; &#39;</span><span class="p">)</span>
</span><span class='line'>                <span class="k">else</span><span class="p">:</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;No error found&#39;</span><span class="p">)</span>
</span><span class='line'>            <span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</span><span class='line'>            <span class="k">print</span> <span class="s">&#39;Result file saved on &#39;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s">&#39;.count&#39;</span>
</span><span class='line'>        <span class="k">except</span><span class="p">:</span>
</span><span class='line'>            <span class="k">print</span> <span class="s">&#39;File not found. Please try again.&#39;</span>
</span><span class='line'>    <span class="k">else</span><span class="p">:</span>
</span><span class='line'>        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>The main change here is that we use a <code>while</code>
loop to control de program flow. Notice that we import <code>sys</code> and <code>re</code>.
Basically we ask for an user input, the filename, and depending on the
input given we process the file or exit the program. The early exit is
done with the <code>sys.exit</code> method which is a shortcut to get out of the
script processing. Very handy. If the input is valid we <code>try</code> to open
it. If the operation is successful, great, we read the file, count the
nucleotides and use a quite scary regular expression to search all the
&#8220;errors&#8221; in our sequence. The regex is compiled with the pattern
<code>'[BDEFHIJKLMNOPQRSUVXZ]'</code> which means &#8220;match any character in this
range&#8221;. Take a closer look at the pattern and you will see that is every
letter in the alphabet, except for ACGT. And when searching for these
&#8220;errors&#8221; instead of using <code>re.search</code> we use <code>re.findall</code>, which
conveniently returns a list with all the substrings found. That&#8217;s even
more handy. On the final part of the script we take care of the output,
opening a file called <code>.count</code> where we print the counts and the errors,
if they actually exist. One thing I left from the previous post, is that
we need to close the file opened to write. In some cases if the file is
not properly closed, errors might occur. So always remember to close the
files used for writing/appending, using <code>.close()</code>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/13/writing-to-files/">Writing to files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-13T14:58:58-04:00" pubdate data-updated="true">Mar 13<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We already know how to read from files, now we are going to see how to
write to them. We will start with a simple example, writing some content
to a &#8220;fresh&#8221; file that does not exist in the system. Sometime ago, we
have seen the <code>print</code> statement in Python, that prints to the system
standard output (usually the screen). This output can be redirected
using <code>&gt;</code> to a stream/file. But if we are going to create really
professional applications (even to our own use), usually stream
redirection is not really the nicest approach. In some cases the best
alternative is to save a file. Also, some posts ago, we covered the
methodology to <code>open</code> a file. We are going to use here the same command,
<code>open</code> to (in our case) create the file. Remember that to read the file
we used</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>where the <code>'r'</code> is the mode we are using to open the
file. In that case we used the <em>read</em> mode, now we are going to use the
<em>write</em> mode. When we open a file with this command in <em>write</em> mode if
the file (with the name with pass) does not exist it is created. If it
does exist, all the contents of the current file will be erased, so be
careful when opening a file from your script. Check for the location,
file name, etc before opening the file. So, basically we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>and
the file is open and ready to receive data. Let&#8217;s cheat and use the
previous script that counts nucleotides and modify it to save a
<code>count.txt</code> file wit the results:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python</span>
</span><span class='line'><span class="c">#let&#39;s keep the file fixed for now</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="c">#opening the file, reading the sequence and storing in a list </span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="c">#initialize a string to receive the data </span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>    <span class="c">#notice the strip, to remove n </span>
</span><span class='line'>    <span class="c">#&quot;exploding&quot; the sequence in a list </span>
</span><span class='line'>    <span class="n">seqlist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'><span class="c">#initializing integers to store the counts </span>
</span><span class='line'><span class="n">totalA</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalC</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalG</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalT</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="c">#checking each item in the list and updating counts </span>
</span><span class='line'><span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seqlist</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;A&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalA</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;C&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalC</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;G&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalG</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;T&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalT</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'><span class="c">#printing results</span>
</span><span class='line'><span class="n">resultfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;counts.txt&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">resultfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; As found </span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">resultfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Cs found </span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">resultfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Gs found </span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">resultfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Ts found </span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>The only difference is at the end of the script. Here instead of <code>print</code> we use
<code>write</code>. Notice that <code>write</code> is a method of the opened file.(in our case
called <code>resultfile</code>. This method only accepts strings, so we need to
convert everything to string before writing in the file. Notice also
that we need to add a carriage return/newline at the end of the string
to be written. Differently of <code>print</code>, <code>write</code> does not automatically
puts a new line at the end of the output.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/13/manipulating-strings-again/">Manipulating strings: again</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-13T14:30:54-04:00" pubdate data-updated="true">Mar 13<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As I have not updated the website in a couple of weeks, I will try to
catch up with the book instead of revisiting (for the time being) our
past scripts. I am going to finish the book&#8217;s chapter 5 (our section 2)
in the next post, where I will give a small introduction on how to
output data to a file in Python. On this post we will check some of the
methods that can be used to manipulate strings. The book gives only a
couple of methods to be used in perl on string, but here I will show a
longer list of Python methods that can be used on its immutable strings.</p>

<p>A full list of the methods can be found
<a href="http://docs.python.org/lib/string-methods.html">here</a> and I will will
give brief explanations on the ones I think are key for bioinformatics.</p>

<p>Remember that string cannot be changed in Python, so we will always
going to use a buffer/temp variable to store our changed string when
needed. - <strong>count</strong> this method returns the number of times you see a
substring (a letter/number, a word, etc) in another string. You can also
specify a start and an end positions to look for. This is handy if you
are counting nucleotides/aaminoacids in a sequence. This method returns
an integer - <strong>endswith</strong> this method checks the end of your string for
a determined substring. Very handy of you need to check the tail end of
your sequence right away. - <strong>find</strong> returns the position of the
substring being searched, and -1 if it is not found. It is similar to
the re.search we used before. This is very useful if you are looking for
a determined motif/subsequence in a hurry. - <strong>islower</strong> returns a true
bool (true or false) if all characters in the string are lowercase.
False if at least one of the characters is uppercase. Very handy if you
need to convert lowercase to UPPERCASE files for input in some
application. Works in conjunction with the <strong>isupper</strong> which is
basically the opposite. - <strong>join</strong>. We have seen this before: it
concatenates strings using a determined separator. - <strong>lower</strong> and
<strong>upper</strong>, that as their name might indicate return the string converted
to lowercase/uppercase.This is also good for the conversion of sequence
format in input files.</p>

<p>I will stop here. There are many other methods
that can be used. Some of the above were already covered here and in the
next posts we will take a look at the other ones, creating an
application that actually performs some useful function.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/22/counting-nucleotides-part-ii/">Counting nucleotides: part II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-22T14:51:58-05:00" pubdate data-updated="true">Feb 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Let&#8217;s check the &#8220;short&#8221; way, that is basically a method that avoid the
&#8220;explosion&#8221; of the string. Instead of transforming the sequence from the
file from a string to a list, we go and use the string directly,
applying one of the methods available to manipulate strings. This time
we read the file at once and convert the list to a string using <code>join</code>.
To count we simply use the method <code>count</code> on our string. Pretty nice.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="c">#still keep the file fixed </span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="c">#opening the file, reading the sequence and storing in a list</span>
</span><span class='line'><span class="n">seqlist</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="c">#let&#39;s join the the lines in a temporary string </span>
</span><span class='line'><span class="n">temp</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seqlist</span><span class="p">)</span>
</span><span class='line'><span class="c">#counting </span>
</span><span class='line'><span class="n">totalA</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalC</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalG</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalT</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'><span class="c">#printing results </span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; As found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Cs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Gs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Ts found&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s the short way: using <code>count</code>. This is a method that when applied
on a string counts the number of times the substring appears in our
string. We can even set a start and ending point to count. Notice one
difference in this script to the previous examples: after we join the
items of the list into a string we do not remove the carriage returns.
Because in this case we don&#8217;t need it, as it will be an extra character
there that won&#8217;t make any harm.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/12/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/10/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2010/12/01/why-i-left-biostar-but-i-still-like-stackoverflow/">Why I left Biostar, but I still like Stackoverflow</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/20/preview-of-django-1-1-testing-and-debugging/">Preview of Django 1.1 Testing and Debugging</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/01/initial-impressions-about-bioinformatics-programming-using-python/">Initial impressions about Bioinformatics Programming using Python</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/19/python-for-bioinformatics-by-sebastian-bassi-a-review/">Python for Bioinformatics by Sebastian Bassi: a (short) review</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/17/biostar-bioinformatics-community/">Biostar: bioinformatics community</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
