
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Geneious: a reply from Biomatters - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="On the Geneious discussion forum, Tobias, has replied to my review(s) with very interesting points that I think need some further argumentation. I &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/2007/03/02/geneious-a-reply-from-biomatters">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">Geneious: a reply from Biomatters</h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:27:54-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>On the Geneious discussion forum, Tobias, has replied to my review(s) with very interesting points that I think need some further argumentation. I start dissecting his post:</p>

<blockquote><p>I remember your old post from a year ago and just read your new review, and I thought I&#8217;d address a few of the points you raised.</p></blockquote>

<p>Although, I &#8220;raised more points&#8221; on my second review, which is based on a more mature version of the software, I had posted some interesting points on my initial assessment which haven&#8217;t been touched by Tobias on his reply. Maybe because my first review was less critical, less &#8220;negative&#8221; than the second one. Regardless, I still think (and I will have this opinion until I see some software better than) that Geneious is a great piece of software and deserve a very high praise.</p>

<blockquote><p>Of course I am 100% biased because I work at Biomatters, but I would like to give my personal view on some points that you raised. You may of course take my affiliation bias into account when reading those</p></blockquote>

<p>Bias acknowledged.</p>

<blockquote><ol>
<li>I agree that the free version gives you &#8220;limited resources&#8221; as you point out. But please do also consider that software is always limited, and in fact Geneious today gives you a lot more than the free version which you have positively reviewed a year ago. So the question here is whether you should compare today&#8217;s free version with Geneious Pro, or with last year&#8217;s free version and the other free software that is available. I lean towards the latter, and I think Geneious 2.5.3 scores very well in this comparison.</li>
</ol>
</blockquote>

<p>My first &#8220;positive&#8221; review praised the first version of Geneious. My second &#8220;negative&#8221; one &#8220;slammed&#8221; it, or maybe just gave a little bit of balance to the initial review. I agree that the number of free features in Geneious is good and software is always limiting. But with no bias, compare the pro and free versions.</p>

<blockquote><ol>
<li>The point of Geneious having &#8220;no groundbreaking or innovative feature that wasn&#8217;t seen in other molecular biology/phylogenetics package&#8221; is probably more true than others you raise, however it is also true that Geneious does offer some unique and innovative functionality that I don&#8217;t know from other bioinformatics software:</li>
<li>collaboration (obviously there is a chicken and egg problem,
and it will only become really useful once many people start
using it). The new version of collaboration for Geneious 3.0 has
much improved collaboration features such as one on one chat
and will hopefully greatly accelerate this development.</li>
<li>agents
Also, Geneious Pro offers many features otherwise found only in much more expensive software packages.</li>
</ol>
</blockquote>

<p>As I said, I could not test collaboration. I tried testing it with Daniel Batten, but we never found a right time to meet online. Regarding the interface, you can find similar or even better features in <a href="http://www.mbio.ncsu.edu/BioEdit/bioedit.html">BioEdit</a>, which is also free. But it has major disadvantages, as being Windows-only and being focused only on sequence alignment. Other software that can be comparable to Geneious is <a href="http://www.aborygen.com/downloads/biOpen/">bioOpen</a>, but falls in the same category as BioEdit: Mac-only and not even free. I don&#8217;t know the price of bioOpen but its ordering form gives you the option to select which functionalities you want.</p>

<blockquote><ol>
<li>You ask what incentive someone has to write a plugin for Geneious. If I was a bioinformatics student today and wrote software e.g. for my MSc project, I don&#8217;t see what reasons there are <em>not</em> to write it as a plugin for Geneious.</li>
</ol>
</blockquote>

<p>I was never a bioinformatics student, maybe because during my time there were not many bioinformatics courses available, even more where I did my graduate studies. So I cannot give you a personal view as a &#8220;bioinformatics student&#8221;. But first, we have to separate bioinformatics students in two classes (too few maybe): the computer scientist that understands biology and the biologist that programs computers. You can even include here a third class which is the computer-savvy biologist: he/she cannot program but have a high knowledge of computers.</p>

<p>With our categories defined I ask you again: what are the benefits of writing a plug-in for Geneious for each one of these classes of users?</p>

<blockquote><p>This is much easier than writing the functionality from scratch, because I don&#8217;t have to worry about the graphical user interface - which is one of the biggest workloads in writing software.</p></blockquote>

<p>I agree that writing plugins using a <strong>proprietary</strong> API is much easier than doing for scratch. I also agree that the interface is one of the biggest workloads in writing software. Being a interface programmer like myself, I understand how underrated the interface development is. But how many bioinformatics applications have interfaces? How many are cross-platform? And if I only need some functionality that Geneious does not give me but it would take a couple of hours to write in Perl/Python/Ruby (<a href="http://www.biomatters.com/userforum/comments.php?DiscussionID=87">like this</a>)? If I am a bench biologist with zero knowledge of programming do I have to wait for someone write me a plug-in or wait for Biomatters to include in Geneious?</p>

<blockquote><p>Also, it makes my program a lot more usable for a lot more people. Peer respect and publicity are among the strongest incentives - when people write free software, as many other people as possible using it is their <em>main</em> goal.</p></blockquote>

<p>Of course it makes a program more usable for a lot more people. People that are using Geneious, right? How do you measure peer respect? By doing a wonderful program from scratch or by adding to a <strong>proprietary</strong> software a plug-in? Or, wouldn&#8217;t be better to create a open source project (or join one) that works in bioinformatics development?</p>

<blockquote><p>I spent a considerable amount of time during my studies writing software, and I wish Geneious had been around back then - it would have made my software much easier to write and use, and it would have become known to a lot more people&lt;</p></blockquote>

<p>I did my Masters writing software and my PhD working on phylogenetics. Would Geneious help me in the first case? No (maybe with ideas). Would it help me in the latter? Definitely, yes. Would I spend time during my PhD to write plug-ins to Geneious? No. See the difference?</p>

<blockquote><ol>
<li>You suggest that we concentrate on writing plugins and sell licenses for using those. Really this isn&#8217;t very far from what we are already doing - the main difference between Geneious and Geneious Pro is that Geneious Pro has more features, and most of these are provided plugins. So really there are already <em>lots</em>of plugins - it&#8217;s just that most of them already come shipped with Geneious, instead of you having to download them in a separate file from our website. In fact, if other people were to write free plugins which provide all of the Geneious Pro functionality for free, the incentive for Geneious Pro would basically disappear. This is the only concern I personally have about our Public API - and it&#8217;s about the opposite from yours.</li>
</ol>
</blockquote>

<p>This is not the first time I suggested this. I had discussions with Daniel Batten about this subject, and even point him some licensing possibilities, such as <a href="http://www.mathworks.com/">MatLab</a>. I saw the list of add-ons when customizing Geneious. Regarding your concern, I don&#8217;t think it will be fulfilled.</p>

<blockquote><ol>
<li>The import feature hasn&#8217;t always worked correctly for you. We would like to keep Geneious as bug free as possible, so if you could please email any files upon which Geneious fails to e.g. tobias@biomatters.com, then that would be much appreciated and we will fix these problems as soon as possible.</li>
</ol>
</blockquote>

<p>I have sent samples, even add to other topics guidance to some biological file formats. I personally don&#8217;t have any file that I am working at the moment, but as everyone that releases a new bioinformatics application, releases a new filetype at the same time, the amount of filters and errors will be huge.</p>

<blockquote><p>We are typically quite responsive, even though we live in a different time zone to most other people.</p></blockquote>

<p>When I say that is a disadvantage to be in New Zealand, is that sometimes users (especially the ones that pay) tend to be quite demanding. As I am not a Geneious user <em>per se</em>, this does not affect me.</p>

<blockquote><ol>
<li>There are now 4 plugins available for download, and a fifth will possibly follow soon. Only three of the currently available ones work for Geneious 2.5.x, but this will probably be fixed at some point. As our API stabilises and more plugins become available, we hope to introduce a more comfortable way of installing them and keeping them up to date.</li>
</ol>
</blockquote>

<p>There were only two at the time of my review. It is good that it is increasing, 100% in a couple of days. And going back to the &#8220;peer respect&#8221;, I downloaded <a href="http://mrbayes.csit.fsu.edu/">MrBayes</a> plug-in and I couldn&#8217;t find the names of the developers. Also could not find any reference to MrBayes authors, papers, etc.</p>

<blockquote><ol>
<li>Scripting language support is already being developed. Update: It probably won&#8217;t make it into Geneious 3.0, but nevertheless Geneious will have scripting support at some point.</li>
</ol>
</blockquote>

<p>Great. That was another thing that Daniel and I discussed last year. I think this is nice for people trying to create molecular biology, phylogenetics, etc courses for a large audience or for some university program. I couldn&#8217;t find the Educational plug-in in Geneious.</p>

<p>Still, in my opinion Geneious is a great software.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">Paulo Nuin</span></span>

      








  


<time datetime="2007-03-02T17:27:54-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
      

<span class="categories">
  
    <a class='category' href='/blog/categories/bioinformatics-opinion/'>Bioinformatics - opinion</a>
  
</span>


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.genedrift.org/blog/2007/03/02/geneious-a-reply-from-biomatters/" data-via="" data-counturl="http://zientzilaria.genedrift.org/blog/2007/03/02/geneious-a-reply-from-biomatters/" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
        <a class="basic-alignment left" href="/blog/2007/03/02/geneious-one-year-or-almost-later/" title="Previous Post: Geneious: one year (or almost) later">&laquo; Geneious: one year (or almost) later</a>
      
      
        <a class="basic-alignment right" href="/blog/2007/03/02/global-warming-here-to-stay/" title="next Post: Global warming: here to stay">Global warming: here to stay &raquo;</a>
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/09/06/goodbye-and-so-long/">Goodbye and so long</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/03/06/testing-feed/">Testing feed</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere ...</a>
      </li>
    
  </ul>
</section>




  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
