
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>“Cutting” chromosomes with Python: a fast approach, take two, including precision - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="On the last post we saw how to &#8220;cut&#8221;/extract a segment of a chromosome
quickly in Python. But our last code had no precision in the &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/temp/2007-11-26-e2809ccuttinge2809d-chromosomes-with-python-a-fast-approach-take-two-including-precision.html">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">“Cutting” chromosomes with Python: a fast approach, take two, including precision</h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-26T16:41:57-05:00" pubdate data-updated="true">Nov 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>On the last post we saw how to &#8220;cut&#8221;/extract a segment of a chromosome
quickly in Python. But our last code had no precision in the
cutting/extracting, because it didn&#8217;t take in account the line size so
it would extract more information than actually requested. We need to
address precision them. First take into account the size of each line,
the number of nucleotides on each line. Then use this this information
to remove unwanted bases. Most of the changes we need to make on our
script is outside the main loop that cuts/extracts information. We need
to modify the post-processing of the sequence we nust read from the
chromosome. Also, instead of storing the sequence in a <code>string</code> it is
easier to use a <code>list</code> (in fact it is just a matter of personal taste,
and maybe a little be of being Pythonic). In the end our script would
look like [sourcecode language=&#8217;python&#8217;]#! /usr/bin/env python import
sys file = sys.argv[1] start = int(sys.argv[2]) end = int(sys.argv[3])
size = 0 linesize = 0 segment = [] for line in open(file, &#8216;r&#8217;): if not
line.startswith(&#8216;>&#8217;): size += len(line) else: name = line if size >=
start and size &lt;= end+linesize: segment.append(line.strip()) linesize =
len(line.strip()) startline = (start/linesize) + 1 endline =
(end/linesize)+1 if not start % linesize == 0 and not end % linesize ==
0: segment[0] = segment[0][startline*linesize-start:] segment[-1] =
segment[-1][endline*linesize-end:] elif not start %linesize == 0:
segment[0] = segment[0][startline*linesize-start:] elif not end %
linesize == 0: segment[-1] = segment[-1][endline*linesize-end:] print
name, &#8216;\n&#8217;.join(segment)[/sourcecode] Notice that we added a new
variable <code>linesize</code>, to store the size of the sequence line. The loop is
almost identical to the one before. The only changes are the line that
checks for each line size (can be optimized somewhere else and removed)
and instead of a segment string, it is a list. The bulk of
changes/additions are after the loop. We first need to calculate the
initial and final lines of our cut/extracted segment. We just divide the
value of the start and end positions entered by the size of the line and
add 1 to them in order to get a correct line in case we the entered
values are not multiple of the line size. Dividing integers in Python
will return us an integer. After this checking we need to add some
processing conditionals to make sure we are returning the exact segment
requested. The <code>%</code> operator in Python return the remainder of a
division, even if the two values are integers. We need to check that the
entered starting and ending positions are or aren&#8217;t multiple of the line
size. If one or both are multiples we don&#8217;t need to process the segment.
On the other hand some processing is needed. Clealy we only need to cut
extra nucleotides from the first and last returned lines. That&#8217;s why the
three clauses: one for both values not being multiple of line size, one
for just the start and another for the end position. And to remove
unwanted nucleotides we basically multiply the number of the
startting/ending line by the size of each line and subtract the actual
start/end position. This should give us the number of unwanted
nucleotides in those two lines. And the fact that we stored the segment
as a <code>list</code> makes our life easier as we access position 0 and -1 in our
list. On the final output, we print the name and join the list elements
by adding a new line between them.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">admin</span></span>

      








  


<time datetime="2007-11-26T16:41:57-05:00" pubdate data-updated="true">Nov 26<span>th</span>, 2007</time>
      

<span class="categories">
  
    Liquid error: undefined method `sort!' for nil:NilClass
  
</span>


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.genedrift.org/temp/2007-11-26-e2809ccuttinge2809d-chromosomes-with-python-a-fast-approach-take-two-including-precision.html" data-via="" data-counturl="http://zientzilaria.genedrift.org/temp/2007-11-26-e2809ccuttinge2809d-chromosomes-with-python-a-fast-approach-take-two-including-precision.html" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2010/12/01/why-i-left-biostar-but-i-still-like-stackoverflow/">Why I left Biostar, but I still like Stackoverflow</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/20/preview-of-django-1-1-testing-and-debugging/">Preview of Django 1.1 Testing and Debugging</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/01/initial-impressions-about-bioinformatics-programming-using-python/">Initial impressions about Bioinformatics Programming using Python</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/19/python-for-bioinformatics-by-sebastian-bassi-a-review/">Python for Bioinformatics by Sebastian Bassi: a (short) review</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/17/biostar-bioinformatics-community/">Biostar: bioinformatics community</a>
      </li>
    
  </ul>
</section>




  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
