
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Genetic code: part I - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="The Python dictionary data-type is like hash in Perl. It is a very
similar structure, where each element in the variable is composed of two
values, &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/temp/2007-04-19-genetic-code-part-i.html">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">Genetic code: part I</h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-19T16:53:29-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>The Python dictionary data-type is like hash in Perl. It is a very
similar structure, where each element in the variable is composed of two
values, more specifically a key-value pair. This is the ideal data type
to store the genetic code. As you might know the genetic code governs
the translation of DNA into proteins, where each codon (3 bases or
nucleotides in the DNA sequence) correspond to an amino acid in the
protein sequence. So for every sequence of 3 nucleotides (key) will
represent an amino acid (value). Important things: dictionaries do not
accept duplicated key values, and every time a new value is assigned to
a key the old value is erased. To create a new dictionary use the curly
brackets [sourcecode language=&#8217;python&#8217;]first_dictionary =
{}[/sourcecode] inside the curly braces we first assign a key and
separated by a colon (:), while multiple pairs should be separated by
comma. Both key and value have to be between single or double quotes.
Let&#8217;s see how we will represent the genetic code in a Python dictionary,
assigning values to keys [sourcecode language=&#8217;python&#8217;]gencode = {
&#8216;ATA&#8217;:&#8217;I&#8217;, #Isoleucine &#8216;ATC&#8217;:&#8217;I&#8217;, #Isoleucine &#8216;ATT&#8217;:&#8217;I&#8217;, # Isoleucine
&#8216;ATG&#8217;:&#8217;M&#8217;, # Methionine &#8216;ACA&#8217;:&#8217;T&#8217;, # Threonine &#8216;ACC&#8217;:&#8217;T&#8217;, # Threonine
&#8216;ACG&#8217;:&#8217;T&#8217;, # Threonine &#8216;ACT&#8217;:&#8217;T&#8217;, # Threonine &#8216;AAC&#8217;:&#8217;N&#8217;, # Asparagine
&#8216;AAT&#8217;:&#8217;N&#8217;, # Asparagine &#8216;AAA&#8217;:&#8217;K&#8217;, # Lysine &#8216;AAG&#8217;:&#8217;K&#8217;, # Lysine
&#8216;AGC&#8217;:&#8217;S&#8217;, # Serine &#8216;AGT&#8217;:&#8217;S&#8217;, # Serine &#8216;AGA&#8217;:&#8217;R&#8217;, # Arginine
&#8216;AGG&#8217;:&#8217;R&#8217;, # Arginine &#8216;CTA&#8217;:&#8217;L&#8217;, # Leucine &#8216;CTC&#8217;:&#8217;L&#8217;, # Leucine
&#8216;CTG&#8217;:&#8217;L&#8217;, # Leucine &#8216;CTT&#8217;:&#8217;L&#8217;, # Leucine &#8216;CCA&#8217;:&#8217;P&#8217;, # Proline
&#8216;CCC&#8217;:&#8217;P&#8217;, # Proline &#8216;CCG&#8217;:&#8217;P&#8217;, # Proline &#8216;CCT&#8217;:&#8217;P&#8217;, # Proline
&#8216;CAC&#8217;:&#8217;H&#8217;, # Histidine &#8216;CAT&#8217;:&#8217;H&#8217;, # Histidine &#8216;CAA&#8217;:&#8217;Q&#8217;, # Glutamine
&#8216;CAG&#8217;:&#8217;Q&#8217;, # Glutamine &#8216;CGA&#8217;:&#8217;R&#8217;, # Arginine &#8216;CGC&#8217;:&#8217;R&#8217;, # Arginine
&#8216;CGG&#8217;:&#8217;R&#8217;, # Arginine &#8216;CGT&#8217;:&#8217;R&#8217;, # Arginine &#8216;GTA&#8217;:&#8217;V&#8217;, # Valine
&#8216;GTC&#8217;:&#8217;V&#8217;, # Valine &#8216;GTG&#8217;:&#8217;V&#8217;, # Valine &#8216;GTT&#8217;:&#8217;V&#8217;, # Valine
&#8216;GCA&#8217;:&#8217;A&#8217;, # Alanine &#8216;GCC&#8217;:&#8217;A&#8217;, # Alanine &#8216;GCG&#8217;:&#8217;A&#8217;, # Alanine
&#8216;GCT&#8217;:&#8217;A&#8217;, # Alanine &#8216;GAC&#8217;:&#8217;D&#8217;, # Aspartic Acid &#8216;GAT&#8217;:&#8217;D&#8217;, # Aspartic
Acid &#8216;GAA&#8217;:&#8217;E&#8217;, # Glutamic Acid &#8216;GAG&#8217;:&#8217;E&#8217;, # Glutamic Acid &#8216;GGA&#8217;:&#8217;G&#8217;,
# Glycine &#8216;GGC&#8217;:&#8217;G&#8217;, # Glycine &#8216;GGG&#8217;:&#8217;G&#8217;, # Glycine &#8216;GGT&#8217;:&#8217;G&#8217;, #
Glycine &#8216;TCA&#8217;:&#8217;S&#8217;, # Serine &#8216;TCC&#8217;:&#8217;S&#8217;, # Serine &#8216;TCG&#8217;:&#8217;S&#8217;, # Serine
&#8216;TCT&#8217;:&#8217;S&#8217;, # Serine &#8216;TTC&#8217;:&#8217;F&#8217;, # Phenylalanine &#8216;TTT&#8217;:&#8217;F&#8217;, #
Phenylalanine &#8216;TTA&#8217;:&#8217;L&#8217;, # Leucine &#8216;TTG&#8217;:&#8217;L&#8217;, # Leucine &#8216;TAC&#8217;:&#8217;Y&#8217;, #
Tyrosine &#8216;TAT&#8217;:&#8217;Y&#8217;, # Tyrosine &#8216;TAA&#8217;:&#8217;_&#8217;, # Stop &#8216;TAG&#8217;:&#8217;_&#8217;, # Stop
&#8216;TGC&#8217;:&#8217;C&#8217;, # Cysteine &#8216;TGT&#8217;:&#8217;C&#8217;, # Cysteine &#8216;TGA&#8217;:&#8217;_&#8217;, # Stop
&#8216;TGG&#8217;:&#8217;W&#8217;, # Tryptophan }[/sourcecode] Simple, yet efficient. But this
is the type of functionality that would be great to have at hand
everytime you write a script to translate DNA into proteins. And it is
not something that you would like to type (or even copy-and-paste) all
the time. On the next post we will create the translation script and
will also create our first Python module.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">admin</span></span>

      








  


<time datetime="2007-04-19T16:53:29-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2007</time>
      

<span class="categories">
  
    Liquid error: undefined method `sort!' for nil:NilClass
  
</span>


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.genedrift.org/temp/2007-04-19-genetic-code-part-i.html" data-via="" data-counturl="http://zientzilaria.genedrift.org/temp/2007-04-19-genetic-code-part-i.html" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/09/06/goodbye-and-so-long/">Goodbye and so long</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/03/06/testing-feed/">Testing feed</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere ...</a>
      </li>
    
  </ul>
</section>




  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
