
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>GenBank parsing: both scripts - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/temp/2008-03-14-genbank-parsing-both-scripts.html">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">GenBank parsing: both scripts</h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest to extract the DNA. [sourcecode language=&#8217;python&#8217;]#!
/usr/bin/env python &#8221;&#8217; input is a GenBank file. The script searches for
gene annotations, extract all lines from the file and then parses these
lines in order to extract protein sequences Ribosomal genes and other
non-coding genes are not extracted - plan to do it later output is a
fasta formatted file &#8221;&#8217; import sys import fasta class Protein: &#8221;&#8217;
class that stores protein information &#8221;&#8217; def __init__(self, gi, id,
sequence): self.gi = gi self.id = id self.sequence = sequence def
parse_entry(gene_data): &#8221;&#8217; parses the entry received from the main
function in order to extract information as protein id gi, etc &#8221;&#8217;
prot_id = &#8221; sequence = &#8221; gi_id = &#8221; gene_data =
gene_data.splitlines() for line in gene_data: if line.find(&#8216;/product&#8217;)
>=0: prot_id = line[line.find(&#8216;=&#8217;) + 2:-1] elif
line.find(&#8216;protein_id&#8217;) >= 0: prot_id += &#8216;\t&#8217; + line[line.find(&#8216;=&#8217;)
+ 2: -1] elif line.find(&#8216;GI:&#8217;) >= 0: gi_id = &#8216;gi&#8217; +
line[line.find(&#8216;GI:&#8217;)+3:-1] elif line.find(&#8216;/translation&#8217;) >= 0:
sequence = line[line.find(&#8216;=&#8217;) + 2:] temp = gene_data.index(line) for i
in range(temp+1, len(gene_data)): if gene_data[i].find(&#8216;sig_peptide&#8217;)
>= 0: break else: sequence += gene_data[i].strip() return
Protein(gi_id, prot_id, sequence) #only input is a genbank file
gbfile = open(sys.argv[1]) proteins = [] index = 0 entry = &#8221; for line
in gbfile: if line.find(&#8217; gene &#8216;) >= 0: if index >= 1: #parses the
CDS and appends to a list proteins.append(parse_entry(entry)) entry =
&#8221; index += 1 entry += line elif line.find(&#8216;ORIGIN&#8217;) >= 0: #found the
DNA sequence, we can stop now break else: entry += line #parses the
last entry after leaving the loop proteins.append(parse_entry(entry))
#output for i in proteins: if len(i.gi) > 2: print i.gi, i.id output =
open(i.gi + &#8216;.fasta&#8217;, &#8216;w&#8217;) output.write(&#8216;>&#8217; + i.gi + &#8216;\t&#8217; + i.id +
&#8216;\n&#8217;) i.sequence = i.sequence.replace(&#8216;&#34;&#8217;, &#8221;)
output.write(fasta.format_output(i.sequence, 80)) print
i.id[/sourcecode] [sourcecode language=&#8217;python&#8217;]#! /usr/bin/env python
&#8221;&#8217; script that extracts sequences from a GenBank file. Script reads the
gene CDS from the file and builds a list of start and end positions, and
if gene is complement outputs the 5&#8217;3&#8217; sequence and its reverse
complement only input is a GenBank file outputs a fasta file with the GI
ID as its name. &#8221;&#8217; import sys import fasta class CDSinfo: &#8221;&#8217; CDSinf
class to store all the information from CDS &#8221;&#8217; def __init__(self,
gi_id, id, start, end, complement): self.gi_id = gi_id self.id = id
self.start = start self.end = end self.complement = complement def
parse_entry(gene_data): &#8221;&#8217; each CDS entry is obtained in the main
function and a string of lines with information is passed to
parse_entry to be parsed and have information extracted &#8221;&#8217; gene_data
= gene_data.splitlines() #changes a string to list, splitting at line
ends start, end = 0, 0 gi_id = &#8221; id = &#8221; complement = False for line
in gene_data: #searches for regions annotated as CDS if line.find(&#8217;
CDS &#8216;) >=0: temp = line.split() #checks for complement sequence, if
true remove extra characters if temp[1].find(&#8216;complement&#8217;) >= 0:
complement = True temp[1] = temp[1].replace(&#8216;complement(&#8216;, &#8221;) temp[1] =
temp[1].replace(&#8216;)&#8217;, &#8221;) temp2 = temp[1].split(&#8216;..&#8217;) start = temp2[0]
end = temp2[1] #checks for GI IDs elif line.find(&#8216;GI:&#8217;) >= 0: gi_id =
&#8216;gi&#8217; + line[line.find(&#8216;GI:&#8217;)+3:-1] #get the gene name/function elif
line.find(&#8216;/product&#8217;) >=0: id = line[line.find(&#8216;=&#8217;) + 2:-1] #and adds
the protein id elif line.find(&#8216;protein_id&#8217;) >= 0: id += &#8216;\t&#8217; +
line[line.find(&#8216;=&#8217;) + 2: -1] return CDSinfo(gi_id, id, start, end,
complement) #only input is the genbank file with annotation and
sequence gbfile = open(sys.argv[1]) index = 0 entry = &#8221; sequence = []
is_seq = False genes = [] for line in gbfile: #reads the genbank file
and whenever finds a gene annotation #concatenate the lines up to the
next gene if line.find(&#8217; gene &#8216;) >= 0: #if an entry is complete, send
it to parse if index >= 1: #appends to a list of CDSinfo objects
genes.append(parse_entry(entry)) entry = &#8221; index += 1 entry += line
elif line.find(&#8216;ORIGIN&#8217;) >= 0: #found sequence start, set the flag on
and parses the last entry is_seq = True
genes.append(parse_entry(entry)) elif is_seq == True: #if flag is
true keep going, usually sequences are store at the end of the file line
= line.split() sequence.append(line) else: #this is an entry so append
line entry += line str_seq = &#8221; #make the sequence a string for i in
sequence: str_seq += &#8221;.join(i[1:]).upper() for i in genes: if
len(i.gi_id) > 2: print i.id, i.start, i.end output = open(i.gi_id +
&#8216;.DNA.fasta&#8217;, &#8216;w&#8217;) output.write(&#8216;>&#8217; + i.gi_id + &#8216;\t&#8217; + i.id + &#8216;\n&#8217;)
# if this is a complement, print both 5&#8217;-3&#8217; and reverse complement
sequences if i.complement == True:
output.write(fasta.format_output(fasta.invert(str_seq[int(i.start)-1:int(i.end)]),
80) + &#8216;\n&#8217;) else: if not i.start.find(&#8216;join&#8217;) >= 0:
output.write(fasta.format_output(str_seq[int(i.start)-1:int(i.end)],
80))[/sourcecode] Checking the DNA extract script carefully, we notice
that we have an extra function in the <code>fasta</code> module, <code>invert</code>.
<strong>Update</strong>: A small change to the script is that <code>split('\n')</code> can be
replaced by <code>splitlines()</code> which is a built-in method for strings. This
hasn&#8217;t been added yet to the repository version of the file and an
explanation on its use will be added shortly.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">admin</span></span>

      








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
      

<span class="categories">
  
    Liquid error: undefined method `sort!' for nil:NilClass
  
</span>


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.heroku.com/temp/2008-03-14-genbank-parsing-both-scripts.html" data-via="nuin" data-counturl="http://zientzilaria.heroku.com/temp/2008-03-14-genbank-parsing-both-scripts.html" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere ...</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating...</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating...</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
