
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Off topic - Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Mr Daniel Swan accused me of not publishing his
comment or reply to his
comments. I have never seen his reply, either in the comments or as an
email &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/2007/10/18/off-topic">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
    
      <h1 class="entry-title">Off topic</h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-18T08:10:59-04:00" pubdate data-updated="true">Oct 18<span>th</span>, 2007</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>Mr Daniel Swan accused me of not publishing his
<a href="http://eridanus.net/blog/?p=233#respond">comment</a> or reply to his
comments. I have never seen his reply, either in the comments or as an
email to me, so to fulfill his desires it is published here.</p>

<blockquote><p>I admit that some of the things you have addressed in the past, whilst
‘reinventing’ the wheel, offer important insights into the general
problems of programming in general - this particular example was not
one of them. From following the course which I assume was outlined by
‘Beginning Perl for Bioinformatics’ (which I do not think is a good
reference for beginning Perl at all) a user should already be able to
perform the following operations - open a file, iterate over and parse
the contents, open a file for writing, and apply some rules for
formatting the output. Thus you have already given them the tools to
complete the operation should they wish to do so in a Python context.
The problem is that you are trying to educate potentially early
career/transitional bioinformaticians here - and showing them how to
complete tasks in X lines of Python when there is a single command
line application to do the trick I think is disingenuous. We wouldn’t
dream, for instance, of educating our MRes students in this way. I
also would take issue that diversity is a key. You can be as diverse
as you like but no-one is seriously going to suggest that you write
your next script to parse something out of a GenBank file in
assembler. One of the hallmarks of any computer professional in any
field is the ability to use the right tool for the job. You should be
technology agnostic and be prepared to string together applications or
scripts from anywhere to reach your goal - take a look at something
like Taverna for instance. I notice in the other post that you mention
there is a high proportion of Windows users to this site and therefore
command line *nix comments are of no interest. Although I know a
number of bioinformaticians who use Windows (generally to integrate
with an institutions existing Microsoft heavy infrastructure) they all
have logins to dozens of *nix machines, run *nix in a VM or at the
very least have a Cygwin installation with which to work with. Basing
the availability of technology on the OS of your page imprints is not
necessarily indicative. I hope you don’t mind me addressing these
points here rather in my blog, as my blog is personal (hence the lack
of bioinformatics content on there). I have actually been following
these tutorials for a while (as someone who came into Bioinformatics
some time ago - my first forays in programming were in Perl) it is
nice for me to see tasks that I find intuitive in one language written
in another. Consequently the fact that I felt need to comment, merely
suggests that I felt strongly on this issue, rather than having any
negative feelings towards what you are trying to achieve.</p></blockquote>

<p>I am sorry for the usual readers of this tutorial, for posting his
comments as an article here. We will be back to our uneducated, simple
and basic &#8220;tutorial for bioinformatics in X lines for the noobie in
Windows&#8221; soon. Thanks Daniel.</p>
</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">admin</span></span>

      








  


<time datetime="2007-10-18T08:10:59-04:00" pubdate data-updated="true">Oct 18<span>th</span>, 2007</time>
      


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://zientzilaria.genedrift.org/blog/2007/10/18/off-topic/" data-via="" data-counturl="http://zientzilaria.genedrift.org/blog/2007/10/18/off-topic/" >Tweet</a>
  
  
  <div class="g-plusone" data-size="medium"></div>
  
  
</div>

    
    <p class="meta">
      
        <a class="basic-alignment left" href="/blog/2007/10/17/splitting-a-fasta-file/" title="Previous Post: Splitting a FASTA file">&laquo; Splitting a FASTA file</a>
      
      
        <a class="basic-alignment right" href="/blog/2007/10/18/watson-yet-again/" title="next Post: Watson, yet again">Watson, yet again &raquo;</a>
      
    </p>
  </footer>
</article>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/09/06/goodbye-and-so-long/">Goodbye and so long</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/03/06/testing-feed/">Testing feed</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere ...</a>
      </li>
    
  </ul>
</section>




  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
