
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Last post received a lot of comments, mainly because I did not mention other languages on it. The reason of the post was to show that Perl and Python &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/32">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/16/apart-from-python-and-perl-in-bioinformatics/">Apart from Python and Perl in Bioinformatics</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-16T14:16:50-04:00" pubdate data-updated="true">Sep 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last post received a lot of comments, mainly because I did not mention other languages on it. The reason of the post was to show that Perl and Python are equivalent, just a matter of taste and personal preferences, and maybe of how comfortable one is with a programming language.</p>

<p>Also a <em>mea culpa</em> of missing Perl&#8217;s interactive shell. That would give Perl a slight advantage over Python, but in the end it really doesn&#8217;t matter.</p>

<p><strong>The case for Java</strong></p>

<p>A lot of people mentioned Java as a must for Bioinformatics, and I agree with them. But again it wasn&#8217;t in the scope of what I intended to write. Even though I don&#8217;t like Java (personal taste) I don&#8217;t deny its importance in the field, its robustness, its easy cross-platform development and power. Several large scale Java bioinformatics applications are available, showing all its features and beauty. Two good examples are Geneious and CLC Workbench, that are the closest you can get to project and database management in molecular biology.</p>

<p>I see Java dominating its niche in Bioinformatics, of the large scale, cross-platform GUI application development. At the same time Python and Perl &#8220;compete&#8221; for their niche of the small to medium applications to quickly deal with some data (sometimes large amounts), most of the time also cross-platform and sometimes only used <em>in house</em>. Ruby might be soon joining both for an extra competition. Just to be clear, I am not saying that either Perl or Python (you might include Ruby) are not engineered to large scale application development, or is not possible to use Java as an agile script language.</p>

<p><strong>Other languages</strong></p>

<p>I consider myself a generalist. I have tried many different languages and do not shy away of trying new ones now and then. In science the larger the diversity, the better, especially in bioinformatics. I would state the obvious by writing that each language is tailored to do different things, and it is also relevant in bioinformatics.</p>

<p>But as in any software development environment, you have to attach yourself to some rules. If you are a tenured professor with your own research program you might be able to use a more obscure language and impose on your lab such language.</p>

<p>On the other hand if you are a student or post-doc in a lab, you may not have that much liberty to use your language of choice. Sometimes, you will be working on someone else&#8217;s code, maybe going to upgrade an already released application. And in some cases, your boss don&#8217;t know that particular language and have no interest in learning it and have no interest in reviewing a code in a language she or he doesn&#8217;t understand. It is an identical situation for a developer or consultant starting a gig in a softhouse. You can bring your knowledge, but maybe you <strong>must</strong> code in C# even though you might not like it.</p>

<p>It also depends on the final use of the program. If it is going to be used in house just to generate results for a publication, you might have more liberty on what your decisions. But if if the application will be widely released, via publication or announcement, you probably need to follow the same rules of development and deployment as any commercial package.</p>

<p>And what do I consider obscure in bioinformatics? Maybe F#, OCaml, Haskell just to limit to the ones mentioned in the comments. I have F# and Haskell installed, but never toyed with them to be able to give a verdict. Never tried OCaml, but heard good things about it. R is a great environment, even though you might not do actual code in it and just use the functions and libraries available. I am starting with Ruby and giving a shot at Rails and will see how it goes.</p>

<p>There is space for most languages in Bioinformatics. It depends on your personal taste, the scope of your application and the environment where you working in. Just like any software development project.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/14/simplicity-sometimes-works/">Simplicity sometimes works</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-14T11:04:57-04:00" pubdate data-updated="true">Sep 14<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In science, <strong>sometimes</strong> to publish an article it seems that a very complicated methodology or results that were obtained through a maze of methods maybe your only chance of success. And this does not mean your new findings will be cited or your method will be used. But <strong>sometimes</strong> simplicity might do the work.</p>

<p>I am referring to a very simple software that I have developed in my spare time, just to force myself to learn some aspects of <a href="http://www.wxwidgets.org">wxWidgets</a> and C++ and help people with a user-friendly interface for a command-line application. This is a very <a href="http://genedrift.org/mtgui.php">software</a> started a shell for David Posada&#8217;s famous ModelTest, and was improved to allow automatic Paup runs and MrModeltest tests. It was really fun to do it and I never thought it would be so well received and get citations in scientific journals. Of course it is listed in my meager CV, but it makes me really happy that some researchers are mentioning it. It had more than 2000 downloads (87 in August 2007 alone) in the three years it has been available and now Google Scholar is reporting 7 citations. Wow! Thanks a lot for everyone that used the program, helped me with bugs and inconsistencies and cited it in their publication.</p>

<p>Simplicity sometimes works.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/13/restriction-enzymes-the-grand-finale/">Restriction enzymes: the grand finale</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-13T16:17:28-04:00" pubdate data-updated="true">Sep 13<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We get to the last piece of the puzzle. On the last four posts we have
seen each part of the restriction enzyme site searcher script, and now
we put everything together. If someone is also following the book, will
see that the Python code we are producing here is slightly different,
but in essence achieves the same result.</p>

<p>Ok, we saw last time that the
function which searches for the enzyme sites returns a tuple with the
actual sites and positions. Why is it better to return sites and
positions? Because one enzyme can have multiple sites and you might want
to know where are they. Also, it helps us practice some Python skills.
We will start by the last piece, and part of it we already saw before.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="s">&#39;</span>
</span><span class='line'><span class="k">if</span> <span class="n">isname</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="s">&#39;Name found&#39;</span>
</span><span class='line'>  <span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
</span><span class='line'>      <span class="n">sites</span><span class="p">,</span> <span class="n">positions</span> <span class="o">=</span> <span class="n">find_sites</span><span class="p">(</span><span class="n">enzyme</span><span class="p">,</span> <span class="n">enzymeset</span><span class="p">,</span> <span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>      <span class="k">print</span> <span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span><span class="o">+</span><span class="s">&#39;...&#39;</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sites</span><span class="p">,</span> <span class="n">positions</span><span class="p">):</span>
</span><span class='line'>          <span class="k">print</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="s">&#39;Enzyme name not found, please try again&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>This is the <code>if</code> that checks to see if
the input enzyme name was found in the list. Clearly we added a couple
of things. Already covered here, if <code>isname</code> is true we go and read the
sequence file, which can contain a single or multiple sequences. We
start a loop and call the <code>find_sites</code> and expect the return on a tuple
<code>sites, positions</code>. Next line, we print the sequence name for the
output. And then &#8230; Yes, and then we have something new: <code>zip</code>. This is
a function that returns a list of tuples from each one of the arguments
passed to it. In our case we are passing two lists (sites and positions)
and we know that each site has one position, and we also know that
because of the way we build both lists the i-th element in sites will be
equivalent to the i-th element in positions. Using <code>zip</code> we create n
tuples where the i-th tuple is equivalent to the i-th element in sites
and the i-th element in positions.</p>

<p>Confused? Let&#8217;s see. Imagine that sites has is composed of</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;AAA&#39;</span><span class="p">,</span> <span class="s">&#39;AAC&#39;</span><span class="p">,</span> <span class="s">&#39;ACA&#39;</span><span class="p">,</span> <span class="s">&#39;TAA&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>and positions is composed of</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;300&#39;</span><span class="p">,</span> <span class="s">&#39;454&#39;</span><span class="p">,</span> <span class="s">&#39;23&#39;</span><span class="p">,</span> <span class="s">&#39;345&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>and we want
to ouput this <code>AAA -&gt; 300 AAC -&gt; 454 ACA -&gt; 23 TAA -&gt; 345</code> The first
idea that comes to mind is to do a loop, with a range as we already know
that both lists will have the same size. Something in the lines of</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)):</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">sites</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">positions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>and it will work fine and it
is not that long (codewise). <code>zip</code> might not be an advantage here, but
it will be somewhere else, for sure. We print the results and we are
done. The full code will be posted in the repository soon. Next we will
move to the book&#8217;s chapter 10 and we start a new section here, checking
for GenBank files.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/13/restriction-enzymes-third-take/">Restriction enzymes, third take</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-13T12:08:42-04:00" pubdate data-updated="true">Sep 13<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We come to the penultimate part of our restriction enzyme site finder.
Just a couple of pieces lacking in the puzzle and we are there. First,
the most important: the function that searches for the sites, using
regex patterns. We called it <code>find_sites</code></p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">find</span>\<span class="n">_sites</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="nb">set</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>  <span class="n">iupacdict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="s">&#39;[A]&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="s">&#39;[C]&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="s">&#39;[G]&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="s">&#39;[T]&#39;</span><span class="p">,</span> <span class="s">&#39;M&#39;</span><span class="p">:</span><span class="s">&#39;[AC]&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">:</span><span class="s">&#39;[AG]&#39;</span><span class="p">,</span>
</span><span class='line'>  <span class="s">&#39;W&#39;</span><span class="p">:</span><span class="s">&#39;[AT]&#39;</span><span class="p">,</span> <span class="s">&#39;S&#39;</span><span class="p">:</span><span class="s">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">:</span><span class="s">&#39;[CT]&#39;</span><span class="p">,</span> <span class="s">&#39;K&#39;</span><span class="p">:</span><span class="s">&#39;[GT]&#39;</span><span class="p">,</span> <span class="s">&#39;V&#39;</span><span class="p">:</span><span class="s">&#39;[ACG]&#39;</span><span class="p">,</span>
</span><span class='line'>  <span class="s">&#39;H&#39;</span><span class="p">:</span><span class="s">&#39;[ACT]&#39;</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">:</span><span class="s">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">:</span><span class="s">&#39;[CGT]&#39;</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">}</span>
</span><span class='line'>  <span class="n">site</span> <span class="o">=</span> <span class="nb">set</span><span class="p">[</span><span class="nb">input</span><span class="p">]</span>
</span><span class='line'>  <span class="n">pattern</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">site</span><span class="p">:</span>
</span><span class='line'>      <span class="n">pattern</span> <span class="o">+=</span>
</span><span class='line'>      <span class="n">iupacdict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span><span class='line'>  <span class="n">searchpattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">pattern</span><span class="p">)</span>
</span><span class='line'>  <span class="n">sites</span> <span class="o">=</span> <span class="n">searchpattern</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>  <span class="n">temppos</span> <span class="o">=</span> <span class="n">searchpattern</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">temppos</span><span class="p">:</span>
</span><span class='line'>      <span class="n">begin</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">span</span><span class="p">()</span>
</span><span class='line'>      <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">begin</span><span class="p">)</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">return</span> <span class="n">sites</span><span class="p">,</span> <span class="n">positions</span>
</span></code></pre></td></tr></table></div></figure>


<p>We use the
IUPAC dictionary created previously to translate the nucleotides entries
in the restriction enzyme file. The function also receives three values:
the input name of the selected enzyme, the dictionary with all the
enzymes and sites and the sequence where to search. We could easily
remove one of those, but let&#8217;s leave it there.</p>

<p>First we get the site
from the dictionary and initialize an empty string to receive the patter
and a empty list to receive the positions. We will see why we don&#8217;t need
an empty list to store the found sites. We then iterate over the site
and create a pattern using the values for each letter of the site
(dictionary key). Created the patter, we compile the regex and with
<code>findall</code> we find every entry of the site in the sequence. As we already
have seen, using the regex <code>findall</code> will generate a list with all the
entries for that particular regex in the string we are searching. This
is pretty handy because some enzymes have degenerated restriction sites.
That&#8217;s why we don&#8217;t need a pre-initialized empty list for the sites.
Then we use the <code>finditer</code> to find the the exact position of each one of
the sites. Each iterator is a tuple with a start and end positions. In
our case we only need the start position, so in a small loop we iterate
over the temporary variable and just append to the positions empty list
the start value. We have two integers, <code>begin</code> and <code>end</code> that receive
the values from <code>i.span</code>, but we only use <code>begin</code>. The function then
returns two lists as a tuple: one for the sites and one for the
positions. If our programming is correct, both lists should have the
same size and are ready to generate the output.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/12/bioinformatic-perl-or-python/">&#8220;bioinformatic perl or python&#8221;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-12T21:16:02-04:00" pubdate data-updated="true">Sep 12<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I decided to create a new series of posts (a category to be more precise) that deals with some searches that reach the blog. I use a pretty nice logging <a href="http://haveamint.com">application</a> so it is easy to randomly select a Google search term to discuss about it.</p>

<p>About an hour ago someone in the interwebs searched for &#8220;bioinformatic perl or python&#8221; and my other <a href="http://python.genedrift.org">blog</a> is the third returned by Google. How can answer this question without praising too much Python or not despising too much perl?</p>

<p>My language of choice was always C/C++ even though it can explode in your face if not handled with care. I also tried Java (did not like it), Visual Basic, Pascal, Delphi and of course Perl, before arriving in Python. The are many differences between Python and Perl, and there aren&#8217;t many differences between Python and Perl. Specially for bioinformatics development.</p>

<p>Keep the score and please remind this comparison is for a bioinformatician perspective, or even a curious biologist that wants to learn programming.</p>

<p>In Python, OO (object orientation) is natural, while in Perl is an added layer. Is this important for a bioinformatician? It is relative, depends how much you know programming and what you want to do. OO is really nice even for some simple things, and if it is natural or hacked on the language, does not matter.
Tie: Python 1, Perl  1</p>

<p>Python syntax is easy to understand, Perl&#8217;s not so much. Python is naturally indented, Perl no. Cleaner syntax, easier to learn, more difficult to make simple mistakes (not 100% true, but let&#8217;s assume so).
Advantage Python: Python 2, Perl 1</p>

<p>Perl has natural support for regular expressions, which are very handy in a lot of biological tasks. Python has the re module but the regex is not natural. And you bet you are going to use a lot of regex in bioinformatics.
Advantage Perl: Python 2, Perl 2</p>

<p>Speed: very similar. In most cases speed is not a problem in bioinformatics applications. Both are interpreted languages, not compiled to bytecode, so if you want speed go check C/C++.
Tie: Python 3, Perl 3</p>

<p>Python has an interpreter that can be used interactively. That&#8217;s very handy when you are not sure if some technique will work and want to test before wasting hours implementing and failing. If you are learning to program or at least learning Python, that&#8217;s very useful. Perl does not have it.
Advantage Python: Python 4, Perl 3</p>

<p>Support: Perl has CPAN, Python has Cheese Shop. But Perl has much more third-party modules than Python. On the other hand, Python community is growing. Basically anything you want to do in Perl, might have been already implemented, so there is no reinventing-the-wheel-ism.
Advantage Perl: Python 4, Perl 4</p>

<p>Functional programming: according to Wikipedia, FP is a programming paradigm that treats computation as the evaluation of mathematical functions. Like OO, is something that you want to address at a higher level of programming knowledge. Perl has it, but Python&#8217;s is better. period.
Advantage Python: Python 5, Perl 4</p>

<p>Portability: it is a tie. Both run everywhere and anywhere.
Tie: Python 6, Perl 5</p>

<p>GUIs: Python has more possibilities in the GUI department, while Perl is a little behind. If you are a beginner, you probably won&#8217;t be making any graphical interfaces, and for most of the stuff in Biology (if you are computer savvy) interfaces are not important. In the end, you type faster than you click.
Tie: Python 7, Perl 6</p>

<p>And finally, support of the biological community: Perl, hands down. BioPerl is stronger than BioPython, and more complete, so most of the stuff you want are already there. BioPython is growing (not really that much), but it will take some time to reach the BioPerl level. Perl will also let you talk with more people at conferences, so it is an ice breaker.
Advantage Perl: Python 7, Perl 7</p>

<p>Conclusion:</p>

<p>It is just a matter of taste. Both languages are well balanced and powerful. Depending on your needs one will be more suited than the other, but in the end both can give you the desired result. Python has a slight advantage of being easier to learn, with a more English-like syntax, while Perl has the slight advantage of a slightly bigger community. You will be happy with both, and in the end it would be fun to know both. There are many ways of doing things, either in Perl or Python.</p>

<p>And remember, this is a perspective of a biologist/bioinformatician.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/12/clc-free-workbench-a-new-short-review/">CLC free workbench, a new (short) review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-12T19:16:12-04:00" pubdate data-updated="true">Sep 12<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I decided to give CLC free workbench a new look, just after the last update. Apart from the change in size, at least on Windows, there is not a lot of changes on the interface, which still looks a little bit more clumsier than Geneious.</p>

<p>But there was some nice changes behind the scene. I have tried version 3 again just to make sure I wasn&#8217;t doing anything wrong when aligning two sequences, and I wasn&#8217;t. So I decided to test aligning the same sequences on the new version 4 and I was pleasantly surprised to see that the alignment procedure runs fast and smoothly.</p>

<p>I will probably post a larger review and maybe a comparison with Geneious in the near future. Stay tuned.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/12/ratings/">Ratings</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-12T11:48:50-04:00" pubdate data-updated="true">Sep 12<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A new feature has been add to the blog: ratings. As I don&#8217;t receive many comments on the posts but there is a steady stream of viewers of the website, I decided to add the possibility of rating a post. So, if you feel the post was useful for you, well or badly written, just click the rate you feel the post deserves. It is an Ajax-like feature so there is no reloads or any other intrusion on your normal browsing.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/restrinction-enzymes-second-take/">Restriction enzymes: second take</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T16:48:50-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We already have a function that reads the enzymes from a dataset in a
flat file (with one change: return)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">read_enzymes</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>    <span class="n">resenz</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>    <span class="n">start</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;Rich Roberts&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>            <span class="n">start</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>            <span class="n">line</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span><span class="o">&lt;/</span><span class="n">pre</span><span class="o">&gt;</span>
</span><span class='line'>            <span class="nb">buffer</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>            <span class="n">resenz</span><span class="p">[</span><span class="nb">buffer</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">buffer</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;^&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'>    <span class="k">return</span> <span class="n">resenz</span>
</span><span class='line'>  
</span></code></pre></td></tr></table></div></figure>


<p>We now need a function to write a function
that searches for the sites and a main function that accepts the
parameters, coordinate the search and return the results. Looks like we
are more than halfway there. Parameters input was shown before, starting
on section 3. We import the <code>sys</code> module and use the array inside
<code>sys.argv</code> to send the parameters to the script. A basic skeleton of our
main function would look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="c">#reading the ezyme dataset in one line and storing enzyme information in enzymeset </span>
</span><span class='line'><span class="n">enzymeset</span> <span class="o">=</span> <span class="n">read_enzymes</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;bionet.709&#39;</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">))</span>
</span><span class='line'><span class="c">#storing enzyme name on a string</span>
</span><span class='line'><span class="n">enzyme</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="c">#reading a FASTA file and storing the sequences</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">get</span>\<span class="n">_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s a start. Now we have to write a
function that will check for the enzyme name entered by the user in
order to check for the existence of such enzyme. Something like this
would work</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">check_enzyme</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="nb">set</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="nb">set</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="nb">input</span><span class="p">):</span>
</span><span class='line'>      <span class="k">return</span> <span class="bp">True</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="k">return</span> <span class="bp">False</span>
</span></code></pre></td></tr></table></div></figure>


<p>This
basically tests of the dictionary contains the name entered. If yes then
we return True, otherwise False is returned. This changes our main
script body</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="c">#reading the ezyme dataset in one line and storing enzyme information in enzymeset </span>
</span><span class='line'><span class="n">enzymeset</span> <span class="o">=</span> <span class="n">read</span>\<span class="n">_enzymes</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&#39;bionet.709&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">))</span>
</span><span class='line'><span class="c">#storing enzyme name on a string </span>
</span><span class='line'><span class="n">enzyme</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="c">#check if the name entered is valid </span>
</span><span class='line'><span class="n">isname</span> <span class="o">=</span> <span class="n">check_enzyme</span><span class="p">(</span><span class="n">enzyme</span><span class="p">,</span> <span class="n">enzymeset</span><span class="p">)</span>
</span><span class='line'><span class="c">#if it is valid, continue, otherwise abort </span>
</span><span class='line'><span class="k">if</span> <span class="n">isname</span><span class="p">:</span>
</span><span class='line'>  <span class="c">#reading a FASTA file and storing the sequences </span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">get_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="s">&#39;Name invalid. Please try again.&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, we have a good idea on what to do now. We just need a function that creates a
regular expression and searches it on the sequence. Next time &#8230;</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/planning/">Planning</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T15:51:43-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Another aspect covered in the book that we haven&#8217;t seen yet is how to
plan, design out script or software. Usual ways to design a program
include writing use cases and drawing
<a href="http://en.wikipedia.org/wiki/Unified_Modeling_Language">UML</a> diagrams
(stands for Unified Modelling Language). Here we will scratch the
surface of <a href="http://en.wikipedia.org/wiki/Use_case">use cases</a>, where we
will try to determine how the program will interact with the user. How
to write use cases or design UML is not a subject taught in many biology
courses, so we won;t see much of the theory here. Consider this a more
informal way of planning a small script or application.</p>

<p>First thing
would be to set a goal:</p>

<p><em>What is the main objective of our script?</em></p>

<p>Create a simple restriction enzyme map of certain sequences Next, <em>what
do we need to make the program work?</em> We need restriction enzyme
information, such as names and sites and a sequence. That leads us <em>How
do I store information?</em> Restriction enzyme data (last entry) obtained
from a file can be stored in a dictionary, with the enzyme name as key
and the site as value.</p>

<p>Sequences are stored using our fasta class. This
will brings us to one important issue <em>How to interact with the user?</em>
The ideal way would be to present a list of enzymes for the user to
select, but we do not have a graphical interface to organize it nicely.
So, we will ask the user to enter the name of enzyme and a name of a
file to read the sequence where the enzyme site should be found. We can
do that interactively or by passing parameters to the script. We will do
it by parameters here, but the interactive way can be tested as a
homework for those following the blog.</p>

<p>And finally <em>What about an
output?</em> We will do the same way as the book: a list of positions,
indicating the restriction sites. I welcome other options in the
comments. It looks we have a plan. Let&#8217;s gather what we already have,
and our previous knowledge and meet on the other side of this post.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/07/creationism-and-third-world/">Creationism and third world</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-07T12:04:19-04:00" pubdate data-updated="true">Sep 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I know third world countries are currently known as under-development, a more politically correct way to say the same thing. And I also know that news coverage about under development countries, at least in North American media, is very limited and brief. Usually you find reports in the business section or when some natural or man made catastrophe happens. So it is no surprise when you don&#8217;t find anything about creationism teaching and under-development countries (at least on the outlets I am used to check).</p>

<p>This week, the most important Brazilian magazine, Veja (like it or not) has a review on the educational quality of Adventist-supported schools. The title of the report is clear: <a href="http://veja.abril.com.br/120907/p_116.shtml">Thank God - and not Darwin</a> (in a free translation; warning: resource available to subscribers and in Portuguese). After introductory paragraphs putting in historical context church-supported schools in Brazil and the good level of formal education of students coming out these schools, the article gets to the point of evolution/creationism teaching that is common in such schools.</p>

<p>We learn that religion classes are a daily event, and in such classes religious hymns are chanted with fervour. And this starts in kindergarten, along with children&#8217;s versions of the Old Testament. And the most important, creationism is taught side-by-side with evolution.</p>

<blockquote><p>We are clear about our point of view, creationist, but we allow the students to know both sides. (&#8230;) We respect all customers. Furthermore, they need to know Darwin to be accepted in a university&#8221;</p></blockquote>

<p>Ivany Queiroga da Silva, principal of an Adventist school</p>

<p>In Brazil there is no current unified score to access university. Usually a test hosted by the university is applied once or twice a year with general questions on every subject. That&#8217;s why evolution has not been completely banned from their curriculum, but the article also points that such schools are the ones with the largest increase in number of students and profit. Also the Adventist church already has six universities and publish their own school books.</p>

<p>To me, it seems a pretty gloomy future. There is no obstacle for the universities controlled by the adventists to ban evolution from their entry test and from the regular curriculum. Even though they are clearly motivated by profit, it is not impossible that they ban evolution from their schools. When these children, that are taught from the early age with religion concepts, become adults and parents they might see Darwin&#8217;s Theory as a wrong and sinful concept. Then the circle is complete, and evolution will be a no-no in these schools and universities.</p>

<p>In a country where more than half of the population is functional illiterate and the science education is precarious, to say the least, these news are very sad. Instead of stepping into the future, it seems that things are slowly drifting back to the past.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/33/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/31/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
