
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="When someone &#8220;preaches&#8221; atheism, what is the difference of preaching some kind of religion?
">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/29">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/08/preaching/">Preaching</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-08T12:59:07-05:00" pubdate data-updated="true">Jan 8<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>When someone &#8220;preaches&#8221; atheism, what is the difference of preaching some kind of religion?</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/12/19/a-bit-of-tnt-in-your-morning-coffee/">A bit of TNT in your morning coffee</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-12-19T11:24:20-05:00" pubdate data-updated="true">Dec 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.zmuc.dk/public/phylogeny/tnt/">TNT</a> is free!!! Long live TNT.</p>

<p>TNT stands for Tree analysis using New Technology. First published in 1999 by Nixon and Goloboff on Cladistics (where else?) it gives you the possibility doing phylogenetic analysis using parsimony and &#8230; No, that&#8217;s it. Only parsimony. But there is a bright side: you can actually run it on Linux and Mac, albeit without a user interface. That&#8217;s a major step for a program that when first released had an icon depicting an apple with the forbidden sign around it.</p>

<p>I am planning to test it and post a review soon.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/12/05/spam/">Spam</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-12-05T21:24:51-05:00" pubdate data-updated="true">Dec 5<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Getting spam/phishing is annoying, but getting spam/phishing with typos or poor grammar is even worse. IF you are not competent to write correctly, you shouldn&#8217;t be trying to scam someone.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/12/05/if-you-live-in-a-cave/">If you live in a cave</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-12-05T14:47:11-05:00" pubdate data-updated="true">Dec 5<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>and haven&#8217;t seen this</p>

<p><a href="http://xkcd.com/353/">
http://xkcd.com/353/</a></p>

<p><a href="http://blindscientist.genedrift.org/2007/12/05/if-you-live-in-a-cave/pythonpng/"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/12/python.png" alt="python.png" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/12/01/bioblogs-17-the-one-santa-brough-us/">Bio::Blogs #17 - The one Santa brought us</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-12-01T22:24:48-05:00" pubdate data-updated="true">Dec 1<span>st</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/12/santa.png" alt="santa.png" /></p>

<p>Yep, Santa dropped by before the busy Holiday Season to help with the 17th edition of the Bio::Blogs digest. Welcome and enjoy!</p>

<p><strong>New blogs</strong></p>

<p>This last month we welcomed one interesting blog that will have a lot to add to the community. Alexei Drummond is showing us his bag of goods at the <a href="http://bioinf.cs.auckland.ac.nz/">Computational Biology and Evolution</a>, self described as &#8220;a heady mix of computational science, evolutionary biology and other things that matter&#8221;.</p>

<p>Also Pedro suggested a good look at <a href="http://wrightfisher.wordpress.com/">Thirst for Science</a>, a blog that is surely &#8220;drinking the koolaid&#8221; and has opinions on a myriad of scientific subjects, from RNAs to population genetics.</p>

<p><strong>Not camera shy</strong></p>

<p>Deepak, from bbgm is definitely not camera shy, and he pitches his opinions on <a href="http://mndoci.com/blog/2007/11/04/scientific-intelligence-is-business-intelligence/">business and scientific intelligence</a>.</p>

<blockquote><p>I believe there will be a time when life science companies will begin looking at biological, chemical, etc information as a core enterprise asset, rather than something for the research crowd</p></blockquote>

<p>Watch the videocast too.</p>

<p><strong>Just rambling</strong></p>

<p>Publicly, Pedro rambles this month about the possibilities of predicting <a href="http://pbeltrao.blogspot.com/2007/11/predicting-functional-association-using.html">functional association by mRNA localization</a></p>

<blockquote><p>As expected, more specific mRNA localization terms (localizations 30) are more informative for prediction of functional association since fewer terms are required to obtain the same or higher likelihood of interaction.</p></blockquote>

<p>And he also rambles about the hot hyped topic of <a href="http://pbeltrao.blogspot.com/2007/11/linking-out-personalized-medicine.html">personalized medicine</a>. And he is concerned</p>

<blockquote><p>that all the attention the genomics side of personalized medicine will distort the relative importance of nature versus nurture.</p></blockquote>

<p>And he is not <a href="http://thegenesherpa.blogspot.com/2007/11/cottage-industry-cottage-cheese-is-more.html">the</a> <a href="http://www.palidwor.com/blog/?p=63">only</a> <a href="http://scienceroll.com/2007/11/10/navigenics-23andme-and-helix-health-the-review/">one</a> <a href="http://scienceroll.com/2007/11/18/23andme-the-re-review/">posting</a> <a href="http://pimm.wordpress.com/2007/11/18/23andmes-global-similarity-map-as-mission-connecting-all-people-on-the-dna-level/">about</a> <a href="http://mndoci.com/blog/2007/11/19/your-personal-health-social-networking-x0-revisited-and-some-killer-video/">it</a>.</p>

<p><strong>Frilas</strong></p>

<p>Frilas is the way freelancers are known in Brazilian newspapers, and Pavel, from Freelancing Science, blogs about the possibility of using <a href="http://freesci.wordpress.com/2007/11/27/software-portability-and-virtual-appliances/">Virtual Machines to distribute software</a>.</p>

<blockquote><p>So creating a virtual appliance for, let’s say, BLAST, would mean installing a fresh copy of our favourite linux under VMWare Server with nothing more than necessary libraries, copy of BLAST executables and possibly a web interface. Voilla. Virtual appliance for BLAST, anybody?</p></blockquote>

<p><strong>Say wha?</strong></p>

<p><a href="http://www.cs.man.ac.uk/~hulld/">Duncan</a> is still worried about the <a href="http://www.nodalpoint.org/2007/11/30/burn_semantic_web_burn">Semantic Web</a>. Nuf said.</p>

<p><strong>Is there a new Omic?</strong></p>

<p>Keith Robinson, from Omics! Omics!, wants <a href="http://omicsomics.blogspot.com/2007/11/incredible-shrinking-human-genome.html">to shrink us all</a>. Maybe it would be good to meet the aunts during Xmas. HE wonders about who or how they are shrinking our (mine, yours, theirs) genomes.</p>

<blockquote><p>The genome is finished &#8211; but not so finished. Bits and pieces are still getting polished up, and while they are generally dull and monotonous a gene or two might still hide there</p></blockquote>

<p><strong>Already tomorrow</strong></p>

<p>Yep, it is already tomorrow there (no matter what day is here!) and after posting about his perl variables  and the election, Neil, who is always doing something rather desperate, tells us about interesting <a href="http://nsaunders.wordpress.com/2007/11/21/stem-cells-and-science-blogs/">blogs that cover some stem cell subjects</a>. Check it out.</p>

<p><strong>My sardine</strong></p>

<p>This month I posted the last SciView interview in 2007, you might want to take a look. Dr Roderic Page was interviewed and he just said</p>

<blockquote><p>My sense is that large projects don’t always scale well. You don’t get 10x more out of a project simply because it has 10x as much money. You get administrative layers, and a lot of politics. If the big project is really lots of little ones, then it may well succeed. Barcoding is really lots of little projects — everybody grabs their favourite beastie and sequences it.</p></blockquote>

<p>among other things. <a href="http://blindscientist.genedrift.org/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">Take a look</a>.</p>

<p><strong>Happy Holidays</strong></p>

<p>Bio::Blogs #17 is very short. Everybody is busy finishing their papers and the shopping list. It will be back next month, maybe here, maybe somewhere else if you want to host it. On the meantime, please to not give <a href="http://blindscientist.genedrift.org/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">this </a>to any of your good friends.</p>

<p>Happy Holidays!!!!</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/30/clc-products-who-uses-them/">CLC products: who uses them?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-30T13:41:10-05:00" pubdate data-updated="true">Nov 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have a Google alerts with bioinformatics as its key word and this morning I received an email that points to <a href="http://www.medicalnewstoday.com/articles/90228.php">this</a> article on Medical News today. I know it has a huge amount to PR in the article, but the first phrase is symbolic:</p>

<p><em>&#8220;CLC bio, the world&#8217;s leading bioinformatics solution provider, announced the release of CLC Bioinformatics Database.&#8221;</em></p>

<p>So, I need to ask: have you ever used a CLC product or know someone that uses it? Leave a comment if so, because I never met someone that uses their products, and they are still the &#8220;world&#8217;s leading bioinformatics solution provider&#8221;.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/28/wow/">Wow!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-28T19:03:12-05:00" pubdate data-updated="true">Nov 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have been using <a href="http://www.repeatmasker.org/">RepeatMasker</a> to mask common repeats of my sequences. I have just upgraded it to the version open-3.1.8 from open-3.1.6 and the speed gain was incredible. The same 10 Mb sequence that was taking more than a day to run in 3.1.6 took a couple of minutes on the newest. Nice!</p>

<p><em>I need to correct myself: I was using version open-3.1.6, not 7</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/28/geneious-and-other-software/">Geneious and &#8220;specialized&#8221; software, featuring Dendroscope</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-28T13:59:36-05:00" pubdate data-updated="true">Nov 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last week a new phylogenetic tree viewing program was released/published. Called <a href="http://www-ab.informatik.uni-tuebingen.de/software/dendroscope/welcome.html">Dendroscope</a> it is a very nice piece of software that has lots of useful features to view, edit and manipulate phylogenetic trees.</p>

<p>Its release/publication coincided  with a period in the lab when two of the graduate students were obtaining new results, especially in the form of phylogenetic trees. There was some debate whether TreeView should be used, but we couldn&#8217;t compile for all Linux boxes we have, due to differences in their distro version. So, at first we checked Geneious, which has been reviewed here, and it is a do-it-all phyloegenetic/molecular biology/name-it software. Geneious has some very nice features, but I cannot call its tree viewing option excellent. Let&#8217;s say good defines it more than enough.</p>

<p>One particular nagging thing included in Geneious&#8217; latest releases were some extra features to expand/collapse branches, which are circles drawn over the tree&#8217;s nodes. One cannot turn off these circle, they are always there and sometimes with different diameters. When I first saw them I though it was showing some kind of support for the node but I was wrong. After some fidgeting we discovered that circles can be turned off, but only when you are exporting/printing the tree.</p>

<p>Bringing back Dendroscope. Intuititve interface, it delivers what you want and more than you expect for a tree viewing program. After one hour of playing one of the graduate students was already a pro-user of the software, showing everyone what can be accomplished with Dendroscope. And yes, there are no deceiving circles on the nodes. As Geneious, it is a Java based application, so easy to port to different systems. But differently it does one thing and does it better than the &#8220;competition&#8221;. That makes me question myself: why should you leave your &#8220;specialized&#8221; applications in favour of Geneious when the end product is below your expectation?</p>

<p>Maybe because of the competition Geneious got away of its almost sure path to be the &#8220;holy grail&#8221; of phylogenetics/molecular biology software. Trying to embrace every aspect of research at once might take its toll on the simpler details that are still not fixed/changed in the program. When Geneious was released it had the right amount of features and, as any other software, a little bit buggy but it delivered what you wanted. The program grew considerably since its first inception and it can do mostly of the common tasks of a small-medium biology lab requires in a day. Still, why would you want to trade your reliable do-one-thing-do-it-better application in favour of a do-it-all-do-it-so-so program that might have tons of features you will never use? Yep, I find a lot of things in MS Word that I don&#8217;t need, that&#8217;s why I use e, TextMate, Notepad++. And for tree viewing I will dump my always-loved TreeView and adopt the new kid on the block, Dendroscope.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/27/bioblogs-17-call-for-submissions/">Bio::Blogs #17 - call for submissions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-27T13:37:22-05:00" pubdate data-updated="true">Nov 27<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The next Bio::Blogs, currently at number 17, will be proudly hosted here. There is plenty of time for submissions, which should be sent directly to me, at nuin at genedrift dot org, or to the usual Bio::Blogs email address, at bioblogs at gmail dot com. Deadline is December 1st.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/26/e2809ccuttinge2809d-chromosomes-with-python-a-fast-approach-take-two-including-precision/">“Cutting” chromosomes with Python: a fast approach, take two, including precision</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-26T16:41:57-05:00" pubdate data-updated="true">Nov 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the last post we saw how to &#8220;cut&#8221;/extract a segment of a chromosome
quickly in Python. But our last code had no precision in the
cutting/extracting, because it didn&#8217;t take in account the line size so
it would extract more information than actually requested. We need to
address precision them. First take into account the size of each line,
the number of nucleotides on each line. Then use this this information
to remove unwanted bases. Most of the changes we need to make on our
script is outside the main loop that cuts/extracts information. We need
to modify the post-processing of the sequence we nust read from the
chromosome. Also, instead of storing the sequence in a <code>string</code> it is
easier to use a <code>list</code> (in fact it is just a matter of personal taste,
and maybe a little be of being Pythonic). In the end our script would
look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">size</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">linesize</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">segment</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span><span class="p">):</span>
</span><span class='line'>      <span class="n">size</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">name</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">size</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">size</span> <span class="o">&lt;=</span> <span class="n">end</span> <span class="o">+</span> <span class="n">linesize</span><span class="p">:</span>
</span><span class='line'>          <span class="n">segment</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
</span><span class='line'>          <span class="n">linesize</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
</span><span class='line'>          <span class="n">startline</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span><span class="o">/</span><span class="n">linesize</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
</span><span class='line'>          <span class="n">endline</span> <span class="o">=</span> <span class="p">(</span><span class="n">end</span><span class="o">/</span><span class="n">linesize</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
</span><span class='line'>      <span class="k">if</span> <span class="ow">not</span> <span class="n">start</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">end</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">startline</span> <span class="o">*</span> <span class="n">linesize</span><span class="o">-</span><span class="n">start</span><span class="p">:]</span>
</span><span class='line'>          <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">endline</span> <span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">end</span><span class="p">:]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="ow">not</span> <span class="n">start</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">startline</span><span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">start</span><span class="p">:]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="ow">not</span> <span class="n">end</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">endline</span> <span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">end</span><span class="p">:]</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">name</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">segment</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that we added a new
variable <code>linesize</code>, to store the size of the sequence line. The loop is
almost identical to the one before. The only changes are the line that
checks for each line size (can be optimized somewhere else and removed)
and instead of a segment string, it is a list.</p>

<p>The bulk of
changes/additions are after the loop. We first need to calculate the
initial and final lines of our cut/extracted segment. We just divide the
value of the start and end positions entered by the size of the line and
add 1 to them in order to get a correct line in case we the entered
values are not multiple of the line size. Dividing integers in Python
will return us an integer. After this checking we need to add some
processing conditionals to make sure we are returning the exact segment
requested. The <code>%</code> operator in Python return the remainder of a
division, even if the two values are integers. We need to check that the
entered starting and ending positions are or aren&#8217;t multiple of the line
size. If one or both are multiples we don&#8217;t need to process the segment.
On the other hand some processing is needed. Clearly we only need to cut
extra nucleotides from the first and last returned lines. That&#8217;s why the
three clauses: one for both values not being multiple of line size, one
for just the start and another for the end position. And to remove
unwanted nucleotides we basically multiply the number of the
starting/ending line by the size of each line and subtract the actual
start/end position.</p>

<p>This should give us the number of unwanted
nucleotides in those two lines. And the fact that we stored the segment
as a <code>list</code> makes our life easier as we access position 0 and -1 in our
list. On the final output, we print the name and join the list elements
by adding a new line between them.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/30/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/28/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
