
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Sometime ago we saw how to use sets and uniquify lists. This time we will see another example of the use of sets. Note: in the
previous example we &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/7">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/14/python-sets-and-intersections/">Python sets and intersections</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T14:36:11-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Sometime ago we saw how to use sets and
 uniquify lists. This time we will see another example of the use of sets.</p>

<p><em>Note: in the
previous example we saw sets imported as an extra module. This has to be
done for Python versions 2.3 and under. There is a difference between
both: when sets are imported the object is noted <code>Set</code> with capital S
while on Python 2.4 and above, set objects are noted with a small s,
<code>set</code>. In order to make the post as global as possible for people that
does not have the latest version of Python I will use the 2.3 notation
and import</em></p>

<p>This time we will see a different example of set usage that
includes some methods available for this type of objects. The initial
problem was how to determine which genomes are represented in
protein/DNA clusters obtained with <a href="http://bioinformatics.ljcrf.edu/cd-hi/">CD-HIT</a>. Basically CD-HIT uses
a multiple fasta file to generate clusters of proteins/DNA using their
sequence identity. It clusters the sequences, keeping their fasta title
and assigning an ID for each cluster obtained. We won&#8217;t see how is
CD-HIT&#8217;s output, not how to parse it. For this example, let&#8217;s say three
genomes (A, B and C) were analysed in CD-HIT and we want to determine
which clusters contain sequences from the combinations AB, AC and CB (we
won&#8217;t touch unique items and clusters with sequences from all genomes
this time). After reading the clusters and sequence IDs in each one of
them we basically need to create unique lists of cluster IDs for each
one of the genomes. That&#8217;s where <code>sets</code> are used</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span> <span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span> <span class="o">=</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([])</span>
</span></code></pre></td></tr></table></div></figure>


<p>We declare three emtpy sets, that will
store cluster IDs for each one of the genomes. The important part here
is that the sequences&#8217; fasta titles should have a unique identifier for
each one of the genomes, just to make it easier to read each cluster
contents. In each of our genomes the sequences were tagged with their
letter in the first character</p>

<p><code>&gt;genA_sequence1 ACGT &gt;genB_sequence1 ACGTT &gt;genC_sequence1 ACGTT ...</code></p>

<p>We run CD-HIT and parse the results, maybe creating a class to store
information about each cluster and its sequences. Then we analyse this
list</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
</span><span class='line'>  <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;&gt;genA&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genA</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;genB&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genB</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;genC&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genC</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p> Remembering
that <code>sets</code> are unordered unique lists, we don&#8217;t expect to see repeated
cluster IDs in each of the three sets. Now to the fun part. Our first
thought to determine how many clusters contain sequences of A and B
would be to create two loops and iterate checking for equalities in
cluster IDs. But with sets, our task is easier. What we need is the
intersection from genA and genB, and that&#8217;s the method available to do
that.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s it. Three lines of
code, one method. The output will be the number of clusters that contain
sequences of A and B, A and C and B and C, respectively. <em>PS: If anyone
is interested in the CD-HIT output parsing class/function just let me
know.</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/14/genbank-parsing-both-scripts/">GenBank parsing: both scripts</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest to extract the DNA.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#!</span>
</span><span class='line'><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">env</span> <span class="n">python</span> <span class="s">&#39;&#39;&#39; input is a GenBank file. The script searches for</span>
</span><span class='line'><span class="s">gene annotations, extract all lines from the file and then parses these</span>
</span><span class='line'><span class="s">lines in order to extract protein sequences Ribosomal genes and other</span>
</span><span class='line'><span class="s">non-coding genes are not extracted - plan to do it later output is a</span>
</span><span class='line'><span class="s">fasta formatted file &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">Protein</span><span class="p">:</span> <span class="s">&#39;&#39;&#39;</span>
</span><span class='line'><span class="s">class that stores protein information &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span>
</span><span class='line'><span class="n">sequence</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; parses the entry received from the main</span>
</span><span class='line'><span class="s">function in order to extract information as protein id gi, etc &#39;&#39;&#39;</span>
</span><span class='line'><span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gene_data</span> <span class="o">=</span>
</span><span class='line'><span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span>
</span><span class='line'><span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span> <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span>
</span><span class='line'><span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span> <span class="k">if</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;sig_peptide&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">return</span>
</span><span class='line'><span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> \<span class="c">#only input is a genbank file</span>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">CDS</span> <span class="ow">and</span> <span class="n">appends</span> <span class="n">to</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span>
</span><span class='line'><span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found the</span>
</span><span class='line'><span class="n">DNA</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">we</span> <span class="n">can</span> <span class="n">stop</span> <span class="n">now</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">last</span> <span class="n">entry</span> <span class="n">after</span> <span class="n">leaving</span> <span class="n">the</span> <span class="n">loop</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>\<span class="c">#output for i in proteins: if len(i.gi) \&gt; 2: print i.gi, i.id output =</span>
</span><span class='line'><span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\\</span><span class="s">&quot;&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="mi">80</span><span class="p">))</span> <span class="k">print</span>
</span><span class='line'><span class="n">i</span><span class="o">.</span><span class="n">id</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="sd">&#39;&#39;&#39; script that extracts sequences from a GenBank file. Script reads the</span>
</span><span class='line'><span class="sd">gene CDS from the file and builds a list of start and end positions, and</span>
</span><span class='line'><span class="sd">if gene is complement outputs the 5&#39;3&#39; sequence and its reverse</span>
</span><span class='line'><span class="sd">complement only input is a GenBank file outputs a fasta file with the GI</span>
</span><span class='line'><span class="sd">ID as its name. &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">CDSinfo</span><span class="p">:</span> <span class="s">&#39;&#39;&#39; CDSinf</span>
</span><span class='line'><span class="s">class to store all the information from CDS &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span class='line'><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">=</span> <span class="n">gi_id</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'><span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span> <span class="bp">self</span><span class="o">.</span><span class="n">complement</span> <span class="o">=</span> <span class="n">complement</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; each CDS entry is obtained in the main</span>
</span><span class='line'><span class="s">function and a string of lines with information is passed to</span>
</span><span class='line'><span class="s">parse_entry to be parsed and have information extracted &#39;&#39;&#39;</span> <span class="n">gene_data</span>
</span><span class='line'><span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> \<span class="c">#changes a string to list, splitting at line</span>
</span><span class='line'><span class="n">ends</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">complement</span> <span class="o">=</span> <span class="bp">False</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> \<span class="c">#searches for regions annotated as CDS if line.find(&#39;</span>
</span><span class='line'><span class="n">CDS</span> <span class="s">&#39;) \&gt;=0: temp = line.split() \#checks for complement sequence, if</span>
</span><span class='line'><span class="n">true</span> <span class="n">remove</span> <span class="n">extra</span> <span class="n">characters</span> <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span>
</span><span class='line'><span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span> <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#checks for GI IDs elif line.find(&#39;GI:&#39;) \&gt;= 0: gi_id =</span>
</span><span class='line'><span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#get the gene name/function elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#and adds</span>
</span><span class='line'><span class="n">the</span> <span class="n">protein</span> <span class="nb">id</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">return</span> <span class="n">CDSinfo</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
</span><span class='line'><span class="n">complement</span><span class="p">)</span> \<span class="c">#only input is the genbank file with annotation and</span>
</span><span class='line'><span class="n">sequence</span> <span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">is_seq</span> <span class="o">=</span> <span class="bp">False</span> <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> \<span class="c">#reads the genbank file</span>
</span><span class='line'><span class="ow">and</span> <span class="n">whenever</span> <span class="n">finds</span> <span class="n">a</span> <span class="n">gene</span> <span class="n">annotation</span> \<span class="c">#concatenate the lines up to the</span>
</span><span class='line'><span class="nb">next</span> <span class="n">gene</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#if an entry is complete, send</span>
</span><span class='line'><span class="n">it</span> <span class="n">to</span> <span class="n">parse</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#appends to a list of CDSinfo objects</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found sequence start, set the flag on</span>
</span><span class='line'><span class="ow">and</span> <span class="n">parses</span> <span class="n">the</span> <span class="n">last</span> <span class="n">entry</span> <span class="n">is_seq</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="k">elif</span> <span class="n">is_seq</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> \<span class="c">#if flag is</span>
</span><span class='line'><span class="n">true</span> <span class="n">keep</span> <span class="n">going</span><span class="p">,</span> <span class="n">usually</span> <span class="n">sequences</span> <span class="n">are</span> <span class="n">store</span> <span class="n">at</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">the</span> <span class="nb">file</span> <span class="n">line</span>
</span><span class='line'><span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> \<span class="c">#this is an entry so append</span>
</span><span class='line'><span class="n">line</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="n">str_seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> \<span class="c">#make the sequence a string for i in</span>
</span><span class='line'><span class="n">sequence</span><span class="p">:</span> <span class="n">str_seq</span> <span class="o">+=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span> <span class="k">if</span>
</span><span class='line'><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span><span class="p">)</span> \<span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;.DNA.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="c"># if this is a complement, print both 5&#39;-3&#39; and reverse complement</span>
</span><span class='line'><span class="n">sequences</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">complement</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)]),</span>
</span><span class='line'><span class="mi">80</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;join&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)],</span>
</span><span class='line'><span class="mi">80</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>Checking the DNA extract script carefully, we notice
that we have an extra function in the <code>fasta</code> module, <code>invert</code>.
<strong>Update</strong>: A small change to the script is that <code>split('\n')</code> can be
replaced by <code>splitlines()</code> which is a built-in method for strings. This
hasn&#8217;t been added yet to the repository version of the file and an
explanation on its use will be added shortly.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/11/fasta-module-transcribing-dna/">Fasta module: transcribing DNA</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-11T16:23:47-04:00" pubdate data-updated="true">Mar 11<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A long time ago when the blog was still based on the Perl book we have
seen <a href="http://python.genedrift.org/2007/01/23/transcribing-the-other-way/">how to transcribe DNA to RNA</a>.
This entry serves only to remember the method and add a function to the
<code>fasta</code> module in the repository. It is really simple to transcribe in
Python, by employing the <code>replace</code> method on strings. Our function looks
like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">transcribe</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span> <span class="n">RNA</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;U&#39;</span><span class="p">)</span> <span class="k">return</span> <span class="n">RNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>Some other functions will
be added to the module in the next couple of weeks. After that we will
divert the focus on optimizing our GenBank parsing script.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/11/fasta-module-generating-reverse-complement-of-dna-sequences/">Fasta module: generating reverse complement of DNA sequences</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-11T12:17:21-04:00" pubdate data-updated="true">Mar 11<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As shown in the GenBank DNA parser script, it is really useful to have
the ability to get the reverse complement of some DNA sequences. The
reverse complement of a 5&#8217;-3&#8217; DNA sequence is on its complementary
strand. Using our <code>fasta</code> module it is easy to implement a function to
generate the antiparallel sequence Basically we can do this in two step,
one - obtaining the complement and two - reversing it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>  <span class="n">abuffer</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s">&#39;A&#39;</span><span class="p">:</span>
</span><span class='line'>          <span class="n">abuffer</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="s">&#39;C&#39;</span><span class="p">:</span>
</span><span class='line'>          <span class="n">abuffer</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="s">&#39;T&#39;</span><span class="p">:</span>
</span><span class='line'>          <span class="n">abuffer</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="s">&#39;G&#39;</span><span class="p">:</span>
</span><span class='line'>          <span class="n">abuffer</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">return</span> <span class="n">abuffer</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s not a Pythonic approach, but it works
reasonably well for short sequences. What would be a Pythonic approach
to it? Using dictionaries and list comprehension. From the Python online
documentation: <em>List comprehensions provide a concise way to create
lists without resorting to use of map(), filter() and/or lambda. The
resulting list definition tends often to be clearer than lists built
using those constructs. Each list comprehension consists of an
expression followed by a for clause, then zero or more for or if
clauses. The result will be a list resulting from evaluating the
expression in the context of the for and if clauses which follow it.</em></p>

<p>It
is basically a specific way to modify lists using a loop and if clauses
when needed. In a couple of lines we can do what we accomplished in 10
lines with the above excerpt. First we need a dictionary to set how the
nucleotides are related</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">complement</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span>
</span><span class='line'><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span> <span class="s">&#39;A&#39;</span><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p>and then a list
comprehension to modify each nucleotide</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">complseq</span> <span class="o">=</span> <span class="p">[</span><span class="n">complement</span><span class="p">[</span><span class="n">base</span><span class="p">]</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>The list comprehension means: for each base in the
sequence get the dictionary value for each key (a nucleotide in the
initial sequence). The complete function would be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span> <span class="n">complement</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span> <span class="s">&#39;G&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;G&#39;</span><span class="p">:</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span> <span class="s">&#39;A&#39;</span><span class="p">}</span> <span class="n">complseq</span> <span class="o">=</span> <span class="p">[</span><span class="n">complement</span><span class="p">[</span><span class="n">base</span><span class="p">]</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">]</span> <span class="k">return</span>
</span><span class='line'><span class="n">complseq</span>
</span></code></pre></td></tr></table></div></figure>


<p>One step solved. We are already able to get the
complement, now we need to reverse it. Simple. A Python list method
automatically reverts list elements. Basically what we need then is
below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span> <span class="n">seq</span> <span class="o">=</span>
</span><span class='line'><span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="n">seq</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span> <span class="k">return</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">complement</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>The
call from any script to this function would be simply be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">fasta</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="s">&#39;AACCGGTTTTGGCCAA&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Batteries included, indeed.</p>

<p><em>more can be found
<a href="http://www.onlamp.com/pub/a/python/2002/10/17/biopython.html?page=4">here</a>.
that served as an inspiration for the function</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/11/genbank-parsing-part-ii/">GenBank parsing: part II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-11T08:49:47-04:00" pubdate data-updated="true">Mar 11<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last thing we saw in the previous entry was the function to parse a
GenBank file in order to get the DNA sequences of an annotated gene.
This time we will dissect the function.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span>
</span><span class='line'>  <span class="c">#changes a string to list, splitting at line ends </span>
</span><span class='line'>  <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span>  <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
</span><span class='line'>  <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">complement</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; CDS &#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">temp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>          <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>              <span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>              <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'>              <span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="nb">id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">CDSinfo</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>First things first. As in the CDS/protein parsing example, the function
receives a string and it is easier for us to operate with lists, so we
split the string at the carriage returns. This time we also need to
check for the start and end positions of the gene in the long string
that is the DNA sequence of a GenBank file, so we just declare a couple
of integers <code>start</code> and <code>end</code>. We also have a couple of strings to
receive the gi ID and the gene product (<code>id</code>). And finally a boolean,
that will tell us if the sequence in question is on the DNA complement
strand, meaning it is a reverse complement of the DNA that is stored in
the file. So we start our loop checking for the elements of each line.
If we find a <code>'  CDS  '</code> we know that we can extract the start and end
positions of the gene</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; CDS&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">temp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>  <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'>  <span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>We use temporary list and split the line at
the spaces and then move to get the information from it. If the line is
a reverse complement we remove the word from the entry and the
parentheses. Whenever the gene annotation relates to the complement
strand the entry is as such <code>complement(10728..11330)</code>. We then use
another temp object and split the second element in the split line and
separate it at the double dots. From this latter split, we end up with
two elements, the first is the start position and the second the end
position of the gene. To extract the other information everything works
as the previous script, basically checking for the presence of certain
words in the each line. In the end we return an instance of the class
<code>CDSinfo</code> which will be used to extract the DNA sequence.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/06/parsing-genbank-getting-dna-from-genome-entries-part-i/">Parsing GenBank: getting DNA from genome entries, part I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-06T17:17:03-05:00" pubdate data-updated="true">Mar 6<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Before we see how to improve the speed in our Python code (would it be
possible?), let&#8217;s go back to our GenBank parsing algorithm. Last time we
saw how to extract the proteins from the CDS entries along the file.
This time we will see how to get the DNA from the file. Usually DNA
sequence is stored at the bottom of the GenBank file, in a long sequence
(if the file is a genome, for instance). Basically, what we need to do
is to get the start and end position of each gene (or any annotated
piece) and cut the long sequence at the bottom. This will require a
couple of changes in our previous script, basically to get the DNA and
to cut it. Previously, we had this to read the file</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>      \<span class="c">#parses the CDS and appends to a list </span>
</span><span class='line'>          <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>          <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>          <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>          <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="c">#found the DNA sequence, we can stop now </span>
</span><span class='line'>      <span class="k">break</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice
that we stop when we reach the bottom DNA sequence. We need to go
further, but there is another issue: DNA sequence stored in a GenBank
file are stored in blocks of nucleotides and the initial position in
each line is also there. So we need to join these blocks and remove the
number from the beginning of the line. In the end we would have this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">is_seq</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'><span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>          <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>          <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>          <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>          <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">is_seq</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>      <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">is_seq</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'>      <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>      <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>      <span class="n">str_seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>      
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
</span><span class='line'>  <span class="n">str_seq</span> <span class="o">+=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>We added
a couple of things. Mainly a flag boolean variable (<code>is_seq</code>) that tells
the program that we have reached &#8220;DNA level&#8221; and a routine to join the
blocks from the second one to the end. <code>sequence</code> is list at first,
appending blocks of actual DNA sequence and numbers obtained from the
<code>split</code> method. We use a &#8220;blank&#8221; <code>split</code>, what means it will break the
string in the spaces. Our <code>join</code> method above joining list elements from
the element 1 and forward. Element 0 in this case would be the numbers
indicating sequence position. The function to parse the entries and the
class to store the information also need to be changed. Previously, our
class was simpler, with less objects. This time we need to extract more
information.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">CDSinfo</span><span class="p">:</span>
</span><span class='line'>  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">):</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">=</span> <span class="n">gi_id</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">complement</span> <span class="o">=</span> <span class="n">complement</span>
</span></code></pre></td></tr></table></div></figure>


<p>We removed the sequence object, but added a
start, end and a complement, for the cases where the gene is on the
reverse complement. The function to extract the info looks like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span>
</span><span class='line'>  <span class="c">#changes a string to list, splitting at line ends </span>
</span><span class='line'>  <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
</span><span class='line'>  <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">complement</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; CDS &#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">temp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
</span><span class='line'>          <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>              <span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>              <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span>
</span><span class='line'>              <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'>              <span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="nb">id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>      
</span><span class='line'>  <span class="k">return</span> <span class="n">CDSinfo</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>We will discuss the parsing function in the next entry and wrap up both codes, and add them to the repository.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/20/a-comparison-between-python-and-c-parsing-a-genbank-file/">A comparison between Python and C(++): parsing a GenBank file</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-20T14:07:11-05:00" pubdate data-updated="true">Feb 20<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last time we saw how to extract all the protein sequences from a GenBank
genome file. A program is available in the
<a href="http://blast.wustl.edu/">WU-Blast</a> package that does exactly the same
thing, called gt2fasta. Let&#8217;s compare the performance of our Python
script and this compiled C++ (I believe it is C++, the source is not
available). I am using Python 2.5 on the same machine I am running the
already compiled gt2fasta, so don&#8217;t worry about specs. The input file is
the <em>E. coli</em> genome which was linked in the previous post, and I am
timing with the <code>time</code> in Linux. Let&#8217;s see how Python does (less time,
the better):</p>

<p>gt2fasta     0m0.123s<br/>
Python  0m5.466s</p>

<p>Yep, that&#8217;s 44 times
slower for Python. It is widely known that compiled languages are faster
than interpreted ones, but maybe 44 times it is an exaggeration, what
makes me think that there are ways to improve out Python script to make
it a little bit faster. After next post we will see if this is possible.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/15/parsing-a-genbank-file-another-approach/">Parsing a GenBank file, another approach</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-15T11:35:13-05:00" pubdate data-updated="true">Feb 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of entries we learnt a little bit about <code>sets</code> and that
newer Python versions make its use a lost faster than <code>dictionaries</code> to
uniquify <code>lists</code>. And soon we will see good examples of <code>sets</code> usage.
And now for some completely different &#8230; we go back to parsing GenBank
files. We saw sometime ago how to extract some simple information
contained in this type of files. I am working on a project that requires
a lot of work in GenBank files, and instead of learning how to use an
available parser I decided to write a simple one on my own. The
requirements were simple: extract all protein sequences from a whole
genome GenBank file and also the DNA from the sequence available in it
(this time we will see how to get the proteins and next time the
latter). We all know that GenBank files have an annotation section, with
IDs, references, organism and some other information at the top. The DNA
sequence at the bottom, after ORIGIN, and the protein sequences are in
annotated regions of CDSs. An example is below:</p>

<p><code>     gene            8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /db_xref="GeneID:855380"      CDS             8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /experiment="experimental evidence, no additional details                      recorded"                      /note="Protein of unknown function, member of the DUP380                      subfamily of conserved, often subtelomerically-encoded                      proteins"                      /codon_start=1                      /product="Cos1p"                      /protein_id="NP_014063.1"                      /db_xref="GI:6323993"                      /db_xref="SGD:S000005280"                      /db_xref="GeneID:855380"                      /translation="MKENELKNEKSVDVLSFKQLESQKIVLPQDLFRSSFTWFCYEIY                      KSLAFRIWMLLWLPLSVWWKLSNNCIYPLIVSLLVLFLGPIFVLVICGLSRKRSLSKQ                      LIQFCKEVTENTPSSDPHDWEVVAANLNSYLYENKAWNTKYFFFNAMVCQEAFRTTLL                      EPFSLKKDEAAKVKSFKDSVPYIEEALGVYFREVEKQWKLFNSEKSWSPVGLEDAKLP                      KEAYRFKLTWFLKRISNIFMLIPFLNFLCCIYVSRGMCLLLRTFYLGWILFMLVQGFQ                      NMRMIVLSVKMEHKMQFLSTIINEQESGANGWDEIAKKMNRYLFEKKVWKNEEFFFDG
IDCEWFFSHFFYRVLSAKKSMRALSLNVELWPYIKEAQLSCSEESLA"</code></p>

<p>In a GenBank file each coding gene corresponds to at least one protein.
The first line, gene, contains the nucleotide region, as the CDS line.
The other lines, starting with a slash provide information about the
gene and the last piece of information is the DNA translation with the
whole protein sequence. In my case, the first objective was to obtain
the protein sequence, get the protein ID and the gi ID from each entry
in a genome file. An initial assessment, tell us that creating a class
for each protein in the file is a good start.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Protein</span><span class="p">:</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>So we declare the class and have elements for the
IDs and the amino acid sequence. The next problem is how do get each
gene/CDS information, extract what we need and store in a list of class
instances? This is much like we do with our FASTA entries, reading line
by line and looking for FASTA sequence names, which start with <code>&gt;</code>. For
GenBank files we will look for lines containing the word gene. Of course
this would mean that every gene that has this word in its name will be a
problem, so we match against <code>'  gene  '</code> (the word gene flanked by two
spaces).</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>      <span class="c">#parses the CDS and appends to a list</span>
</span><span class='line'>      <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>      <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="c">#found the DNA sequence, we can stop now </span>
</span><span class='line'>      <span class="k">break</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p> In the above excerpt we read
the file line by line, having an index just to keep the code aware of
which entry we are at. Every time we find a line with <code>'  gene  '</code>, we
check for the index and if it is larger than one we parse the entry that
has been compiled already and start a new one. If we find &#8216;ORIGIN&#8217; we
bail out, breaking the loop, and in every other case we add a line to
our current entry. When this is sent to the parsing function we clear
the string and start compiling the information again. Using a string for
an entry might be a second choice, as lists are easier to manipulate
after (and eventually we will transform our string in a list by
splitting it). Notice that we have a call to a function <code>parse_entry</code>,
let&#8217;s see how this function is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>      <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>          <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">return</span> <span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p> As mentioned
above, we split the string in a list (we will check if there is any
improvement in speed if we use a list from the beginning) and then
analyse each one of the lines. In this case, we only need the protein,
the gene name and, evidently, the protein sequence. Dissecting the
excerpt above we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>This will get us the protein ID from
<code>/protein_id="NP_014063.1"</code> and also the gene name/info from
<code>/product="Cos1p"</code>, which them can be stored in the <code>prot_id</code> string
that will be passed to the class instance. The gi ID can be extracted in
similar fashion. In order to get the sequence we have to do a small
trick, that can be seen on the excerpt below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>  <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p> We find the beginning of the
translation information by checking the line that contains
<code>/translation</code> and we extract the initial part of the sequence my
finding the equal sign and getting everything after it. Then the trick.
We know that the end of the translation information will be the last
line of the whole gene/CDS entry. So we check for the index of the first
sequence line and store it in a temp variable. Using this variable as
the first value of a range we are able to get every other line of the
sequence. That simple. To do the output, we can just iterate the list of
class objects and print all the information in FASTA format or any other
that we wish. Put everything together and we have a simple but effective
GenBank parser. Please notice that this script wouldn&#8217;t work in every
possible scenario, but with minor tweaks it should be applicable in most
cases. Next time, we will see how we can get the DNA sequence in an
effective way. <em>An example of GenBank file can be found
<a href="http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?db=nuccore&amp;id=117937805">here</a></em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/30/benchmarking-python-fastest-way-to-generate-unique-lists/">Benchmarking Python: fastest way to generate unique lists</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-30T13:13:29-05:00" pubdate data-updated="true">Jan 30<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Just for fun, let&#8217;s see if there is any advantage (apart from generating
a smaller code) in using either of the approaches to create an unique
list. A list of 741 gene IDs and another one with 1322 (that contained
all the 741 IDs from the first) were used. Instead of hard coding the
lists in the script, normal I/O was used and the files were read
automatically. Using a for loop the scripts were run 10 times and the
final time averaged. Here are the results Using dictionaries 0.08540
Using sets 0.15890 Almost twice as fast for the dictionaries. Why? One
thing, and one thing only: Python version.</p>

<p>My system&#8217;s box default
Python is 2.3.4, which has one of the first implementations of <code>set</code>. I
have a &#8220;personal&#8221; Python version 2.5 so the test was redone with the
newer Python Using dictionaries 0.04520 Using sets 0.03770 Ok. That
shows a little bit of advantage for <code>sets</code> what is expected. But it
shows us that different versions of Python have a huge difference
between them. This is a rather crude and simple test, with a small list
of entries, but there is a huge gain from version 2.3 to version 2.5,
either in <code>dicts</code> or <code>sets</code>. For a more comprehensive test and more
functions with a similar objective check this
<a href="http://www.peterbe.com/plog/uniqifiers-benchmark">page</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/29/uniquifying-lists-with-sets-and-dictionaries/">Uniquifying lists with sets and dictionaries</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-29T22:34:21-05:00" pubdate data-updated="true">Jan 29<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are going to use our previous example to compare the use of <code>sets</code>
and <code>dictionaries</code> to create unique lists. We&#8217;ve already seen that when
<code>sets</code> are used it is very simple to transform a <code>list</code> with repeated
items in a unique <code>list</code>. The only hassle is to create the <code>set</code> and
then transform it back into a <code>list</code>. Like last time (with one small
addition)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span><span class='line'><span class="n">cluster1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">cluster2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">cluster1</span> <span class="o">+</span> <span class="n">cluster2</span>
</span><span class='line'><span class="n">uniqueset</span> <span class="o">=</span> <span class="n">Set</span><span class="p">(</span><span class="n">allgenes</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="n">finalist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">uniqueset</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>We can accomplish identical result by using
a <code>dictionary</code>. We create a small function to make our code clearer and
pass a list to it and we return the <code>dictionary</code> keys. Rmember that
Python <code>dictionaries</code> have values and keys and the latter cannot be
repeated, so it is basically a list of unique entries. Our function
would look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'> <span class="k">def</span> <span class="nf">make_unique_list</span><span class="p">(</span><span class="n">mylist</span><span class="p">):</span>
</span><span class='line'>  <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">word</span> <span class="ow">in</span> <span class="n">mylist</span><span class="p">:</span>
</span><span class='line'>      <span class="nb">dict</span><span class="p">[</span><span class="n">word</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'>
</span><span class='line'><span class="k">return</span> <span class="nb">dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>In this function we declare the object
and in the loop, iterating over every list&#8217;s item we assign a value
(arbitrary) to the dictionary key. As pointed above, no repeated keys
are allowed, so every time a already checked item is seen by the
assignment it is not included in the dictionary. Finally we return only
the dictionary keys which is our final unique list. Our small script
would be:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">cluster1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">cluster2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">cluster1</span> <span class="o">+</span> <span class="n">cluster2</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">make_unique_list</span><span class="p">(</span><span class="n">allgenes</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Both methods are very
effective and usually fast. I will post some comparisons and benchmarks,
just for fun. Nathan posted in the comments another approach using
dictionaries. It is below with syntax highlighting</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="n">mylist</span><span class="p">)</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Basically in one line you pass a list of elements to dictionary and return all the
keys that are in the dic. Very Pythonic.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/8/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/6/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2010/12/01/why-i-left-biostar-but-i-still-like-stackoverflow/">Why I left Biostar, but I still like Stackoverflow</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/20/preview-of-django-1-1-testing-and-debugging/">Preview of Django 1.1 Testing and Debugging</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/01/initial-impressions-about-bioinformatics-programming-using-python/">Initial impressions about Bioinformatics Programming using Python</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/19/python-for-bioinformatics-by-sebastian-bassi-a-review/">Python for Bioinformatics by Sebastian Bassi: a (short) review</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/17/biostar-bioinformatics-community/">Biostar: bioinformatics community</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
