
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="But, I guess, it can be easily refactored/enhanced/despised by the
audience that read or have access to this blog via Planet Python.
Anyway, for &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/7">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/08/the-sickest-python-code-ive-ever-created-1/">The &#8220;sickest&#8221; Python code I&#8217;ve ever created 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-08T20:59:54-04:00" pubdate data-updated="true">Apr 8<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>But, I guess, it can be easily refactored/enhanced/despised by the
audience that read or have access to this blog via Planet Python.
Anyway, for someone like me, whose main task now is not to generate tons
of code and lines, I think the code (or part of it) that I will present
below is quite good. Feel free to comment, criticize and say bad and
good things about it. We needed a script that would take files coming
out from protein search engines that would be able to compare the
peptides and protein sequences, their abundance and some other
characteristics. We had a combination of protein and peptide files, with
a list of proteins (one protein per line in a tab delimited file) that
was related to a list of peptides in another file (one peptide per line,
with multiple peptides/lines related to one protein in the original
list). Also each line in both files had more than 50 columns, and 8 or
16 of them were the values we wanted to extract. I say 8 or 16 of them
because we didn&#8217;t know how many will output each time, as it would
depend on the number of samples per run (4 to 8 samples) . So, we had a
couple of issues: we didn&#8217;t know how many proteins would be output
(actually found) in each file, we wouldn&#8217;t know how many peptides for
each protein would be found and we didn&#8217;t know before hand how many
samples would be run at once. One good thing is that the 8-16 columns of
values were fixed, always in the same position and with empty cells if
no value was registered there. And we had a fourth problem: usually the
samples attributions would be random, meaning a control could come in
the first value column or could come in the last. And a fifth as we
didn&#8217;t know before hand (the tech knew) how many treatments would be run
each time. A treatment could be a different experimental condition, a
sample grouping or some other extraneous factor. An extra issue is that
we would need to compare multiple files, and get protein and peptide
abundances in all of them at the same time and finally compare each
treatment. Basically, in order to create an universal script we needed
something flexible enough that whatever the experiments threw at use we
would be able to handle. First step we decided to use a YAML file that
could be filled by the experimental researchers with sample assignments,
treatments, etc. The YAML would look like this B0: - 114: A - 115: D -
116: B - 117: C B1: - 114: C - 115: A - 116: D - 117: B In this file B0
and B1 would be the result file names, 114 is the column/channel where
the sample was run and and A, B, C and D the treatment. With this set,
out objective was to get all proteins and their peptides for treatment A
in files B0 and B1, do some calculation and them compare to all proteins
and peptides from treatment B, C and D extracted also from files B0 and
B1. First step was to get the names of the treatments from the YAML file
[sourcecode language=&#8217;python&#8217;]]def get_treatments(mapping): treats =
set([]) for entry in mapping: [treats.add(list(t.values())[0]) for t in
mapping[entry]] return treats[/sourcecode] where [code]mapping[/code] is
the YAML file name. We used a set to store and sets have unique items,
and treatment names can vary from file to file. In the code above we
basically read the YAML and the value for each entry. We then needed a
class to store protein information, and there was the story got hairy.
With all my (lack of ) experience, I decided to use [code]exec[/code]
statements to fix all the uncertainty of the experimental data details.
I didn&#8217;t have the treatment names before hand (or in a fixed immutable
list), and didn&#8217;t have the columns (channels) that were being used at
the time and I have to correctly assign each protein abundance (area) to
its place. In the end our class look like this [sourcecode
language=&#8217;python&#8217;]class Protein(): &#8220;&#8221;&#8220;Class Protein, stores all the
information about channels and areas, name and accession&#8221;&#8220;&#8221; def
__init__(self, accession, name, treatments): self.accession =
accession self.name = name #ratio channels are called based on their
name for i in treatments: exec(&#8216;self.%s = []&#8217; %i) exec(&#8216;self.area%s =
[]&#8217; %i) def add_to_channel(self, channel, peptide):
exec(&#8216;self.%s.append(peptide)&#8217; % channel) def add_to_area(self,
channel, area): exec(&#8216;self.area%s.append(area)&#8217; % channel)[/sourcecode]
In order to be faithful to this blog&#8217;s name, I will explain how the code
above is supposed to work. First, [code]exec[/code] is a Python
statement that support dynamic execution of code. In our case above it
was used to name the objects, so we would be able to access them by name
in subsequent functions. Let&#8217;s take this for example [sourcecode
language=&#8217;python&#8217;]for i in treatments: exec(&#8216;self.%s = []&#8217; %i)
exec(&#8216;self.area%s = []&#8217; %i)[/sourcecode] In this snippet we were trying
to create lists called (for the YAML file above) A, B, C and D, and
another set of lists called areaA, areaB, areaC and areaD. Let&#8217;s say for
another experiment we would have treatments &#8220;Control&#8221;, &#8220;Low&#8221; and &#8220;High&#8221;
and so on. The next two functions use the same approach, with exec, this
time appending to the freshly created lists. This way it&#8217;s easy to
control what the user is throwing at us. I don&#8217;f know if this the best
approach possible, or if it is or not harmful. Maybe experts reading
this might have better ideas, and I appreciate them. We check the rest
of the script next time.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/08/bioinformatics-needs-investment-ii/">Bioinformatics needs investment II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-08T13:25:54-04:00" pubdate data-updated="true">Apr 8<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Expanding a bit from what was posted yesterday on the Bioinformatics investment topic, the main investment we need to make is in personnel, courses. Is not admissible that in 2010 a graduate student doesn&#8217;t know some simple (I mean the really, really, really simple) Excel function/macro functionality. It doesn&#8217;t matter if the student is in a bioinfo/computational biology stream or not (it would be worse, then) or any other stream, this is basic stuff that most of the advanced students should know.</p>

<p>We have to forget apologies, such as &#8220;I don&#8217;t like computers&#8221;, &#8220;I&#8217;m not good with machines&#8221;, etc. These are present and future scientists, the top of the top students, they should know basic computer usage and management. I don&#8217;t want them at graduate school to be able to modify a makefile and compile some obscure boost-related library on Solaris, just use a &#8220;=IF A1=0, nothing, yes&#8221; in Excel. If we don&#8217;t teach them in the future I will still receive FASTA sequences in Word docs, and coloured tables in Excel. These are the best mind out there, they should devote maybe 5% of their time for these things, it&#8217;s not just beneficial for me but for them and for science in general too.</p>

<p>I don&#8217;t know where are the mistakes, the failures, what we can do, but we need to do something. And to do something we need money, and not a lot of it.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/07/bioinformatics-needs-investment/">Bioinformatics needs investment</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-07T13:56:14-04:00" pubdate data-updated="true">Apr 7<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>And a lot of it, but not to buy larger machines, or to create big departments or institute. Not even research investment. Bioinformatics needs investment in people, in courses, in workshops, in good practices for software development, in a knowledge base. Although, some people said sometime ago that &#8220;Bioinformatics is (was) dead&#8221;, the future is only bioiformatical ®. When was the last time you saw a bank branch with no computers?</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/06/open-science-and-pregnancy/">Open Science and pregnancy</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-06T22:05:04-04:00" pubdate data-updated="true">Apr 6<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Open Science is like pregnancy: either you are or you are not. There&#8217;s no sort-of Open Science, as there&#8217;s no sort-of pregnant woman.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/04/paid-software-vs-donating/">Paid software vs. donating</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-04T08:14:02-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.crunchbase.com/company/tuneup"><img src="http://www.crunchbase.com/assets/images/resized/0001/8408/18408v1-max-250x250.png" alt="Image representing TuneUp as depicted in Crunc..." /></a></p>

<pre><code>Image via [CrunchBase](http://www.crunchbase.com)
</code></pre>

<p>No, I don&#8217;t pirate them either. But whenever I buy something I get highly disappointed. Recently I bought <a href="http://www.tuneupmedia.com/blog/">TuneUp</a>, which is a iTunes plugin for Macs. It seems to work great, most of the time at least, bu every time I add one song to the library, it tries to re-sync with it, and it usually takes hours and hours. Really frustrating.</p>

<p>That&#8217;s why is better to stick to free stuff and donate when you feel like doing it. Sometimes it&#8217;s better.</p>

<p><a href="http://reblog.zemanta.com/zemified/c5096696-19b5-422e-bca7-8f130c114399/"><img src="http://img.zemanta.com/reblog_e.png?x-id=c5096696-19b5-422e-bca7-8f130c114399" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/01/bpb-on-github/">BPB on Github</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-01T21:09:07-04:00" pubdate data-updated="true">Apr 1<span>st</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I added a pledge button on the Beginning Python for Bioinformatics <a href="http://github.com/nuin/beginning-python-for-bioinformatics">repository on Github</a>. If you think it&#8217;s worth 10c of your money, it helps keep this site alive. Thanks.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/01/draft-blog-posts/">Draft blog posts</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-01T20:38:28-04:00" pubdate data-updated="true">Apr 1<span>st</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Have so many draft posts, no will to write. Ask me anything, I might answer.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/03/30/i-thought-i-was-alone-but-then-i-found-another-light/">I thought I was alone, but then I found another light</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-03-30T08:39:22-04:00" pubdate data-updated="true">Mar 30<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>&#8220;Mendeley sucks&#8221; ® is a common search that points to this humble site. I thought I was alone and my feeling was unique, or at least I was the lone voice in the crowd &#8230; But, then, someone, somewhere, some time searched &#8220;Mendeley sucks&#8221; ® on Google and I clicked on it and I found <a href="http://scripts.mit.edu/~birge/blog/mendeley-how-not-to-run-a-beta-preview-program/">this</a>. This is a MIT website, so it must be good, right?</p>

<p>But remember, you heard here first. &#8220;Mendeley sucks&#8221; ®</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/03/26/earth-hour-2/">Earth hour</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-03-26T13:52:37-04:00" pubdate data-updated="true">Mar 26<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Energy_Star_logo.svg"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/7/73/Energy_Star_logo.svg/300px-Energy_Star_logo.svg.png" alt="The Energy Star logo is placed on energy-effic..." /></a></p>

<pre><code>Image via [Wikipedia](http://commons.wikipedia.org/wiki/Image:Energy_Star_logo.svg)
</code></pre>

<p>This year during the <a href="http://www.earthhour.org/home/">Earth Hour</a> (I don&#8217;t know when it is), I will leave all my <a href="http://www.google.com/hostednews/ap/article/ALeqM5gST3p_0UZPNsD28hz7d4N3aHULBgD9EMF1203">EnergySta</a>r ® appliances on, including my gas-powered alarm clock.</p>

<p><a href="http://reblog.zemanta.com/zemified/7705219f-00d8-4dda-8e7b-3cfb4dc4eccb/"><img src="http://img.zemanta.com/reblog_e.png?x-id=7705219f-00d8-4dda-8e7b-3cfb4dc4eccb" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/03/03/python-testing-beginnere28099s-guide-review/">Python Testing Beginner’s Guide, review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-03-03T22:46:35-05:00" pubdate data-updated="true">Mar 3<span>rd</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I posted about a week ago that Packt Publishing had invited me to review
<strong><a href="http://www.packtpub.com/python-testing-beginners-guide/book/mid/240210aaphjg?utm_source=python.genedrift.org&amp;utm_medium=bookrev&amp;utm_content=blog&amp;utm_campaign=mdb_002570">Python Testing Beginner&#8217;s
Guide</a></strong>
by Daniel Arbuckle. Having finished reading the book (I must admit that
I haven&#8217;t tried all the code in it), I can say that I have an excellent
initial impression of the book. PTBG is not a long book and the topic is
divided in 10 chapters and one appendix. One of the first things that I
liked about the book is that there&#8217;s no <em>introduction</em> (or something
similar) to Python. It just goes straight to the point assuming that you
have some good understanding of the language and everything that
surrounds it. In the past I was frustrated with some &#8220;Introduction to X
with Python&#8221; that wasted precious space talking over and over about a
topic, learning Python, better covered in many other books. PTBG does
not waste time and space introducing its main topic which is testing,
and in my opinion that&#8217;s the best approach, even though it might look a
little bit abrupt by some. The language and text in the book is clear
and very pleasant. PTBG is a very well written book and I really enjoyed
its style. The first chapters of the book cover Python testing using
doctests. For someone like me that didn&#8217;t write so many tests in the
normal software development workflow (I know I should write more tests),
this section seems like a really nice introduction to the topic, with
well thought real-life like examples and a good flow on the explanation
of the different features. One small complain that I have is that for a
beginner sometimes the code listed in the examples might seem a little
bit confusing, and maybe the addition of line numbers might have helped
a bit here. But at the same I understand that this is normal style of
some Packt books. After the doctests section, PTBG gets into more
advanced techniques, covering a little bit mock objects with
<a href="http://labix.org/mocker">Mocker</a>, then moving into unittest and
<a href="http://code.google.com/p/python-nose/">nose</a>. The latter is a Python
tool that allows for managing, running and automating tests. Also
covered is <a href="%20http://pypi.python.org/pypi/%20twill/">Twill</a>, another
third-party library that allows for testing of web applications. One
full chapter is devoted to test-driven development, with a complete
walkthrough of this approach. This gives a wrap-up of most of the
techniques and modules covered in the book, but there&#8217;s still space for
another chapter that shows how beautifully doctests, unittest and nose
can be fully integrated and help the development of applications using
the test-driven approach. Overall, I really enjoyed PTBG. As I
mentioned, test driven development was never a high priority in the
application I usually developed with Python. But certainly this book can
be a good starting point for some Python test beginners to incorporate
these techniques in their usual development workflow. Scientific
software is also a perfect niche for this type of approach and we should
do what is possible in order to avoid the <a href="http://boscoh.com/protein/a-sign-a-flipped-structure-and-a-scientific-flameout-of-epic-proportions">nightmares of the
past</a>.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/8/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/6/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
