
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="A colleague came with a &#8220;problem&#8221;: what would be the most efficient way
to merge Pfam alignments? He had FASTA
files containing sequences &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/25">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/28/merging-sequences-from-a-pfam-alignment-using-sets/">Merging sequences from a Pfam alignment: using sets</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-28T20:58:24-04:00" pubdate data-updated="true">Mar 28<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A colleague came with a &#8220;problem&#8221;: what would be the most efficient way
to merge <a href="http://pfam.sanger.ac.uk/">Pfam</a> alignments? He had FASTA
files containing sequences and he wanted to find identical IDs in two
files and merge the related sequences from different domains of the same
protein. His C++ approach was taking too long to run so I jumped in to
help with some Python tricks.</p>

<p>Fasta headers of Pfam alignments look like
this <code>&gt;P00526.2|SRC_RSVP/147-229</code> where the first section, before the
pipe, is the protein family, the section between the pipe and the slash
is the protein ID and the after the slash are the start and end
positions. Basically we want to match the protein ID, between | and /,
which is the only section that should not change from one alignment from
the other, if there are similar sequences in both files. His dataset
size was around 8000 FASTA files and he wanted to compare
all-versus-all.</p>

<p>The first idea that comes to mind is to read the two
files and basically run nested loops checking for matches between two
sequence names. Simple and easy. But it would be good to find something
else to do in the meantime, instead of checking for its progress.
Another idea, to simplify the process, is to read both files, extract
the protein IDs and check for the intersection of the two sets of IDs
and then just extract the sequences from a list of matches.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">merge_seqs</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>  <span class="n">myset1</span><span class="p">,</span> <span class="n">myset2</span> <span class="o">=</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([])</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data1</span><span class="p">:</span>
</span><span class='line'>      <span class="n">myset1</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)])</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data2</span><span class="p">:</span>
</span><span class='line'>      <span class="n">myset2</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)])</span>
</span><span class='line'>      
</span><span class='line'>  <span class="n">mylist</span> <span class="o">=</span> <span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">myset1</span><span class="p">,</span> <span class="n">myset2</span><span class="p">)</span>
</span><span class='line'>  <span class="n">flist</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mylist</span><span class="p">:</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">data1</span><span class="p">:</span>
</span><span class='line'>          <span class="k">if</span> <span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]</span> <span class="o">==</span> <span class="n">i</span><span class="p">:</span>
</span><span class='line'>          <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">data2</span><span class="p">:</span>
</span><span class='line'>              <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">k</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]</span> <span class="o">==</span> <span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">j</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]:</span>
</span><span class='line'>                  <span class="n">tempname</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-&#39;</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-\&gt;&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">j</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
</span><span class='line'>                  <span class="n">tempseq</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>                  <span class="n">flist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tempname</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span> <span class="o">+</span> <span class="n">tempseq</span><span class="p">)</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">flist</span>
</span></code></pre></td></tr></table></div></figure>


<p>This
is the function that matches the sequence IDs. As we are reading the
FASTA file with the functions/classes in the <code>fasta</code> module there is
no* direct way to transform the list of sequence names into a set, that
would make easier to extract the intersection between two sets of IDs
from distinct files. In the above function, <code>data1</code> and <code>data2</code> are
instances of the <code>fasta</code> class, and we read these lists and add the
<code>fasta.name</code> to a couple of sets and get the intersection of them. The
<code>intersection</code> method returns a list that we use in the nested loops to
find the matching sequences, as we need to merge them, we add both names
and residues. Simple and easy. All merged sequences, and their
respective merged names, are appended to a list that is returned at the
end of the function. *<em>if there is a more direct way, just let me know</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/28/awk-and-sed-sed-and-awk/">Awk and sed, Sed and awk</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-28T09:44:54-04:00" pubdate data-updated="true">Mar 28<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I would like to thank Lee E. McMahon, Alfred Aho, Peter Weinberger and Brian Kernighan. Thanks a lot.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/26/ride-to-conquer-cancer/">Ride to Conquer Cancer</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-26T18:53:00-04:00" pubdate data-updated="true">Mar 26<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>My supervisor, Elisabeth Tillier, is part of a team on the Ride to Conquer Cancer benefiting the Princess Margaret Hospita, Toronto, Canada. The ride is a beautiful event where people ride their bikes from Toronto to Niagara Falls, during two days in June.</p>

<p>If you want to donate there is an easy way by going to <a href="http://www.conquercancer.ca/site/TR/Events/Toronto2008?px=1548939&amp;pg=personal&amp;fr_id=1201&amp;et=E-PMVS5Y4td-3Bf4EqAgug..&amp;s_tafId=96521">Elisabeth&#8217;s page on the ride website</a>. She is riding for her sister in law Carine, which was diagnosed an year ago with breast cancer, just after giving birth to her second son. Every little counts.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/18/sad-sad-day/">Sad, sad day</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-18T17:48:21-04:00" pubdate data-updated="true">Mar 18<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The guy who took me to Jupiter <a href="http://www.msnbc.msn.com/id/23697230/">has died</a>. No more monoliths.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/17/python-sets-intersections-reduce-and-more/">Python sets, intersections, reduce and more</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-17T16:31:43-04:00" pubdate data-updated="true">Mar 17<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last time we saw how to use a set intersection to check for clusters of
DNA/protein sequences and their genome intersections. We going to use
the same example but this time we will see how we can change it and
create a function that would calculate an arbitrary number of
intersections at a time and also be able to check the intersection of
more than two sets. Our previous code to calculate intersections was</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span><span class='line'><span class="c">#for Python 2.3 and below </span>
</span><span class='line'><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span> <span class="o">=</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([])</span>
</span><span class='line'><span class="c">#populate the sets ... </span>
</span><span class='line'>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>and that&#8217;s does not gives us the most important
piece of information: the intersection of A, B and C. Let&#8217;s see how we
can do it. Python has a function that can help us this time: <code>reduce</code>.
This function allows to reduce a series of values to a single one, by
employing a function defined in its parameters. From the file tutorial</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">reduce</span><span class="p">(</span><span class="n">function</span><span class="p">,</span> <span class="n">iterable</span><span class="p">[,</span><span class="n">initializer</span><span class="p">])</span>
</span></code></pre></td></tr></table></div></figure>


<p>Basically <code>reduce</code> will make the function
modify the values of a iterable (i.e. list). In our case <code>reduce</code> helps
us calculate the intersection of many sets at once (our sets are <code>genA</code>,
<code>genB</code> and <code>genC</code>)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">reduce</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">,</span> <span class="p">[</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">])</span>
</span></code></pre></td></tr></table></div></figure>


<p> where <code>reduce</code> is applying the <code>Set</code> method
intersection to a list of sets passed to the function. Python&#8217;s <code>reduce</code>
might not be available in 3.0, but it is worth knowing. Now that we have
a Pythonic way of checking the intersections we need to put <code>reduce</code> in
a function, to make our code more elegant.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">get_intersection</span><span class="p">(</span><span class="n">my_set1</span><span class="p">,</span> <span class="n">my_set2</span><span class="p">,</span> <span class="n">my_set3</span><span class="p">):</span>
</span><span class='line'>  <span class="k">return</span> <span class="nb">reduce</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">,</span> <span class="p">[</span><span class="n">my_set1</span><span class="p">,</span> <span class="n">my_set2</span><span class="p">,</span> <span class="n">my_set3</span><span class="p">])</span>
</span></code></pre></td></tr></table></div></figure>


<p>Done. In this function <code>reduce</code> will operate by
first checking the intersection between <code>my_set1</code> and <code>my_set2</code> and
storing it internally, then checking for the intersection between the
previous result and <code>my_set3</code>. But wait &#8230; <code>get_intersection</code> receives
three arguments, so it will be able to return the intersection only when
three sets are sent, thus we would have to create another function that
can receive two items in order to calculate the intersection between two
items. And if we have more than three genomes to compare &#8230; There is a
<a href="http://farmdev.com/thoughts/24/what-does-the-def-star-variable-or-def-asterisk-parameter-syntax-do-in-python-/">trick</a>.
If we define our function receiving only one argument and put a
star/asterisk on it, this means that an arbitrary number of arguments
will be provided to this function in a tuple (and a tuple is iterable,
so it can be used in a reduce). Changing our function accordingly, we
will have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">get_intersection</span><span class="p">(</span><span class="o">*</span><span class="n">my_sets</span><span class="p">):</span>
</span><span class='line'>  <span class="k">return</span> <span class="nb">reduce</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">,</span> <span class="n">my_sets</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>so we can send 2, 3, 4, 5 or <em>n</em> number of sets
to this function and it will return the intersection among all of them.
Our simple script would be (including the above function)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span><span class='line'><span class="c">#for Python 2.3 and below </span>
</span><span class='line'><span class="k">def</span> <span class="nf">get_intersection</span><span class="p">(</span><span class="o">*</span><span class="n">my_sets</span><span class="p">):</span>
</span><span class='line'>  <span class="k">return</span> <span class="nb">reduce</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">,</span> <span class="n">my_sets</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span> <span class="o">=</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([])</span>
</span><span class='line'>
</span><span class='line'><span class="c">#populate the sets ...</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">get_intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">get_intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">get_intersection</span><span class="p">(</span><span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">get_intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/17/how-realistic-is-10000-bc/">How realistic is 10,000 BC?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-17T14:48:28-04:00" pubdate data-updated="true">Mar 17<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Slate explainer really <a href="http://www.slate.com/id/2186730/">misses the point on this one</a>. Of course <a href="http://www.imdb.com/title/tt0443649/">10,000 BC</a> is NOT realistic. Everybody knows the Earth was created around 4,000 BC. Everybody. Don&#8217;t you?</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/14/python-sets-and-intersections/">Python sets and intersections</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T14:36:11-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Sometime ago we saw how to use sets and
 uniquify lists. This time we will see another example of the use of sets.</p>

<p><em>Note: in the
previous example we saw sets imported as an extra module. This has to be
done for Python versions 2.3 and under. There is a difference between
both: when sets are imported the object is noted <code>Set</code> with capital S
while on Python 2.4 and above, set objects are noted with a small s,
<code>set</code>. In order to make the post as global as possible for people that
does not have the latest version of Python I will use the 2.3 notation
and import</em></p>

<p>This time we will see a different example of set usage that
includes some methods available for this type of objects. The initial
problem was how to determine which genomes are represented in
protein/DNA clusters obtained with <a href="http://bioinformatics.ljcrf.edu/cd-hi/">CD-HIT</a>. Basically CD-HIT uses
a multiple fasta file to generate clusters of proteins/DNA using their
sequence identity. It clusters the sequences, keeping their fasta title
and assigning an ID for each cluster obtained. We won&#8217;t see how is
CD-HIT&#8217;s output, not how to parse it. For this example, let&#8217;s say three
genomes (A, B and C) were analysed in CD-HIT and we want to determine
which clusters contain sequences from the combinations AB, AC and CB (we
won&#8217;t touch unique items and clusters with sequences from all genomes
this time). After reading the clusters and sequence IDs in each one of
them we basically need to create unique lists of cluster IDs for each
one of the genomes. That&#8217;s where <code>sets</code> are used</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span> <span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">,</span> <span class="n">genC</span> <span class="o">=</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([]),</span> <span class="n">Set</span><span class="p">([])</span>
</span></code></pre></td></tr></table></div></figure>


<p>We declare three emtpy sets, that will
store cluster IDs for each one of the genomes. The important part here
is that the sequences&#8217; fasta titles should have a unique identifier for
each one of the genomes, just to make it easier to read each cluster
contents. In each of our genomes the sequences were tagged with their
letter in the first character</p>

<p><code>&gt;genA_sequence1 ACGT &gt;genB_sequence1 ACGTT &gt;genC_sequence1 ACGTT ...</code></p>

<p>We run CD-HIT and parse the results, maybe creating a class to store
information about each cluster and its sequences. Then we analyse this
list</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
</span><span class='line'>  <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;&gt;genA&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genA</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;genB&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genB</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span><span class='line'>      <span class="k">elif</span> <span class="nb">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;genC&#39;</span><span class="p">):</span>
</span><span class='line'>          <span class="n">genC</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">cluster_number</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p> Remembering
that <code>sets</code> are unordered unique lists, we don&#8217;t expect to see repeated
cluster IDs in each of the three sets. Now to the fun part. Our first
thought to determine how many clusters contain sequences of A and B
would be to create two loops and iterate checking for equalities in
cluster IDs. But with sets, our task is easier. What we need is the
intersection from genA and genB, and that&#8217;s the method available to do
that.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genB</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genA</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span><span class='line'><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">Set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genB</span><span class="p">,</span> <span class="n">genC</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s it. Three lines of
code, one method. The output will be the number of clusters that contain
sequences of A and B, A and C and B and C, respectively. <em>PS: If anyone
is interested in the CD-HIT output parsing class/function just let me
know.</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/14/genbank-parsing-both-scripts/">GenBank parsing: both scripts</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T13:40:26-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Both scripts to parse GenBank scripts are below and included in the
repository. First the one that extracts amino acid sequences, and then
the latest to extract the DNA.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#!</span>
</span><span class='line'><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">env</span> <span class="n">python</span> <span class="s">&#39;&#39;&#39; input is a GenBank file. The script searches for</span>
</span><span class='line'><span class="s">gene annotations, extract all lines from the file and then parses these</span>
</span><span class='line'><span class="s">lines in order to extract protein sequences Ribosomal genes and other</span>
</span><span class='line'><span class="s">non-coding genes are not extracted - plan to do it later output is a</span>
</span><span class='line'><span class="s">fasta formatted file &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">Protein</span><span class="p">:</span> <span class="s">&#39;&#39;&#39;</span>
</span><span class='line'><span class="s">class that stores protein information &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span>
</span><span class='line'><span class="n">sequence</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; parses the entry received from the main</span>
</span><span class='line'><span class="s">function in order to extract information as protein id gi, etc &#39;&#39;&#39;</span>
</span><span class='line'><span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">gene_data</span> <span class="o">=</span>
</span><span class='line'><span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span>
</span><span class='line'><span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span> <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span>
</span><span class='line'><span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span> <span class="k">if</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;sig_peptide&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">return</span>
</span><span class='line'><span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> \<span class="c">#only input is a genbank file</span>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">CDS</span> <span class="ow">and</span> <span class="n">appends</span> <span class="n">to</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span>
</span><span class='line'><span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found the</span>
</span><span class='line'><span class="n">DNA</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">we</span> <span class="n">can</span> <span class="n">stop</span> <span class="n">now</span> <span class="k">break</span> <span class="k">else</span><span class="p">:</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> \<span class="c">#parses the</span>
</span><span class='line'><span class="n">last</span> <span class="n">entry</span> <span class="n">after</span> <span class="n">leaving</span> <span class="n">the</span> <span class="n">loop</span> <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>\<span class="c">#output for i in proteins: if len(i.gi) \&gt; 2: print i.gi, i.id output =</span>
</span><span class='line'><span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\\</span><span class="s">&quot;&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="mi">80</span><span class="p">))</span> <span class="k">print</span>
</span><span class='line'><span class="n">i</span><span class="o">.</span><span class="n">id</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="sd">&#39;&#39;&#39; script that extracts sequences from a GenBank file. Script reads the</span>
</span><span class='line'><span class="sd">gene CDS from the file and builds a list of start and end positions, and</span>
</span><span class='line'><span class="sd">if gene is complement outputs the 5&#39;3&#39; sequence and its reverse</span>
</span><span class='line'><span class="sd">complement only input is a GenBank file outputs a fasta file with the GI</span>
</span><span class='line'><span class="sd">ID as its name. &#39;&#39;&#39;</span> <span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">fasta</span> <span class="nn">class</span> <span class="nn">CDSinfo</span><span class="p">:</span> <span class="s">&#39;&#39;&#39; CDSinf</span>
</span><span class='line'><span class="s">class to store all the information from CDS &#39;&#39;&#39;</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span class='line'><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">complement</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">=</span> <span class="n">gi_id</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'><span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span> <span class="bp">self</span><span class="o">.</span><span class="n">complement</span> <span class="o">=</span> <span class="n">complement</span> <span class="k">def</span>
</span><span class='line'><span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span> <span class="s">&#39;&#39;&#39; each CDS entry is obtained in the main</span>
</span><span class='line'><span class="s">function and a string of lines with information is passed to</span>
</span><span class='line'><span class="s">parse_entry to be parsed and have information extracted &#39;&#39;&#39;</span> <span class="n">gene_data</span>
</span><span class='line'><span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span> \<span class="c">#changes a string to list, splitting at line</span>
</span><span class='line'><span class="n">ends</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span> <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">complement</span> <span class="o">=</span> <span class="bp">False</span> <span class="k">for</span> <span class="n">line</span>
</span><span class='line'><span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span> \<span class="c">#searches for regions annotated as CDS if line.find(&#39;</span>
</span><span class='line'><span class="n">CDS</span> <span class="s">&#39;) \&gt;=0: temp = line.split() \#checks for complement sequence, if</span>
</span><span class='line'><span class="n">true</span> <span class="n">remove</span> <span class="n">extra</span> <span class="n">characters</span> <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;complement&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">complement</span> <span class="o">=</span> <span class="bp">True</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;complement(&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span>
</span><span class='line'><span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;)&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span> <span class="n">temp2</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;..&#39;</span><span class="p">)</span> <span class="n">start</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="n">temp2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#checks for GI IDs elif line.find(&#39;GI:&#39;) \&gt;= 0: gi_id =</span>
</span><span class='line'><span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#get the gene name/function elif</span>
</span><span class='line'><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> \<span class="c">#and adds</span>
</span><span class='line'><span class="n">the</span> <span class="n">protein</span> <span class="nb">id</span> <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="nb">id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span>
</span><span class='line'><span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">return</span> <span class="n">CDSinfo</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
</span><span class='line'><span class="n">complement</span><span class="p">)</span> \<span class="c">#only input is the genbank file with annotation and</span>
</span><span class='line'><span class="n">sequence</span> <span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">is_seq</span> <span class="o">=</span> <span class="bp">False</span> <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span> \<span class="c">#reads the genbank file</span>
</span><span class='line'><span class="ow">and</span> <span class="n">whenever</span> <span class="n">finds</span> <span class="n">a</span> <span class="n">gene</span> <span class="n">annotation</span> \<span class="c">#concatenate the lines up to the</span>
</span><span class='line'><span class="nb">next</span> <span class="n">gene</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#if an entry is complete, send</span>
</span><span class='line'><span class="n">it</span> <span class="n">to</span> <span class="n">parse</span> <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span> \<span class="c">#appends to a list of CDSinfo objects</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span> \<span class="c">#found sequence start, set the flag on</span>
</span><span class='line'><span class="ow">and</span> <span class="n">parses</span> <span class="n">the</span> <span class="n">last</span> <span class="n">entry</span> <span class="n">is_seq</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span> <span class="k">elif</span> <span class="n">is_seq</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> \<span class="c">#if flag is</span>
</span><span class='line'><span class="n">true</span> <span class="n">keep</span> <span class="n">going</span><span class="p">,</span> <span class="n">usually</span> <span class="n">sequences</span> <span class="n">are</span> <span class="n">store</span> <span class="n">at</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">the</span> <span class="nb">file</span> <span class="n">line</span>
</span><span class='line'><span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> \<span class="c">#this is an entry so append</span>
</span><span class='line'><span class="n">line</span> <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span> <span class="n">str_seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> \<span class="c">#make the sequence a string for i in</span>
</span><span class='line'><span class="n">sequence</span><span class="p">:</span> <span class="n">str_seq</span> <span class="o">+=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span> <span class="k">if</span>
</span><span class='line'><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span><span class="p">)</span> \<span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span> <span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">end</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span>
</span><span class='line'><span class="s">&#39;.DNA.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">gi_id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">t&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span>
</span><span class='line'>\<span class="c"># if this is a complement, print both 5&#39;-3&#39; and reverse complement</span>
</span><span class='line'><span class="n">sequences</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">complement</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)]),</span>
</span><span class='line'><span class="mi">80</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span> <span class="k">else</span><span class="p">:</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;join&#39;</span><span class="p">)</span> \<span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">str_seq</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">end</span><span class="p">)],</span>
</span><span class='line'><span class="mi">80</span><span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>


<p>Checking the DNA extract script carefully, we notice
that we have an extra function in the <code>fasta</code> module, <code>invert</code>.
<strong>Update</strong>: A small change to the script is that <code>split('\n')</code> can be
replaced by <code>splitlines()</code> which is a built-in method for strings. This
hasn&#8217;t been added yet to the repository version of the file and an
explanation on its use will be added shortly.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/14/we-salute-you-2/">We salute you &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-14T12:09:40-04:00" pubdate data-updated="true">Mar 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>That&#8217;s for all the bioinformatics developers that make websites not compatible with Linux and Firefox (and Konqueror too). Thanks.</p>

<p><em>Adding insult to injury, there is also the use of the blink tag. Could you format the help with Comic Sans?</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/11/fasta-module-transcribing-dna/">Fasta module: transcribing DNA</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-11T16:23:47-04:00" pubdate data-updated="true">Mar 11<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A long time ago when the blog was still based on the Perl book we have
seen <a href="http://python.genedrift.org/2007/01/23/transcribing-the-other-way/">how to transcribe DNA to RNA</a>.
This entry serves only to remember the method and add a function to the
<code>fasta</code> module in the repository. It is really simple to transcribe in
Python, by employing the <code>replace</code> method on strings. Our function looks
like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">transcribe</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span> <span class="n">RNA</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;U&#39;</span><span class="p">)</span> <span class="k">return</span> <span class="n">RNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>Some other functions will
be added to the module in the next couple of weeks. After that we will
divert the focus on optimizing our GenBank parsing script.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/26/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/24/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/09/06/goodbye-and-so-long/">Goodbye and so long</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/03/06/testing-feed/">Testing feed</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
