
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="I&#8217;d pay nothing, would never pay a dime to get my genome sequenced. But, Dan Stoicescu is paying US$ 350,000 to get his. He is also retired at &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/25">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/04/how-much-would-you-pay/">How much would you pay? </a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-04T12:15:35-05:00" pubdate data-updated="true">Mar 4<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I&#8217;d pay nothing, would never pay a dime to get my genome sequenced. But, <a href="http://www.nytimes.com/2008/03/04/health/research/04geno.html?ex=1362286800&amp;en=1a8fb705f5970ec0&amp;ei=5090&amp;partner=rssuserland&amp;emc=rss&amp;pagewanted=all">Dan Stoicescu</a> is paying US$ 350,000 to get his. He is also retired at the age of 56, and apparently owned <a href="http://www.actavis.com.ro/ro/default.htm">Actavis</a>.</p>

<blockquote><p>“I’d rather spend my money on my genome than a Bentley or an airplane,”</p></blockquote>

<p>(&#8230;)</p>

<blockquote><p>Mr. Stoicescu said he worried about being seen as self-indulgent (though he donates much more each year to philanthropic causes), egotistical (for obvious reasons) or stupid (the cost of the technology, he knows, is dropping so fast that he would have certainly paid much less by waiting a few months).</p></blockquote>

<p>In my opinion he shouldn&#8217;t worry about being seen as a self-indulgent egotistical stupid person. He already is. Whoever drops the philanthropy excuse is a sure winner for the self-indulgent prize, anytime.</p>

<p>Maybe the airplane or the Bentley would be more fulfilling.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/03/the-sun-will-shine-on-python/">The Sun will shine on Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-03T15:21:39-05:00" pubdate data-updated="true">Mar 3<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Sun hired some good Python programmers/advocates/developers/thinkers, <a href="http://www.sauria.com/blog/2008/03/03/the-sun-is-going-to-shine-on-python/">Ted Leung</a> and <a href="http://fwierzbicki.blogspot.com/">Frank Wierzbicki</a>, of Jython fame. I hope it does no end as Solaris, the movie.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/03/03/best-and-worst/">Best and worst</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-03-03T10:27:28-05:00" pubdate data-updated="true">Mar 3<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>If I had a vote in these polls that decide for the best and (maybe) worst inventions I think these would be mine:</p>

<p><strong>Best</strong>: Wheel
<strong>Worst</strong>: Cellphone</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/21/burst-baby-burst/">Burst, baby burst</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-21T18:23:07-05:00" pubdate data-updated="true">Feb 21<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content">
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/20/a-comparison-between-python-and-c-parsing-a-genbank-file/">A comparison between Python and C(++): parsing a GenBank file</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-20T14:07:11-05:00" pubdate data-updated="true">Feb 20<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last time we saw how to extract all the protein sequences from a GenBank
genome file. A program is available in the
<a href="http://blast.wustl.edu/">WU-Blast</a> package that does exactly the same
thing, called gt2fasta. Let&#8217;s compare the performance of our Python
script and this compiled C++ (I believe it is C++, the source is not
available). I am using Python 2.5 on the same machine I am running the
already compiled gt2fasta, so don&#8217;t worry about specs. The input file is
the <em>E. coli</em> genome which was linked in the previous post, and I am
timing with the <code>time</code> in Linux. Let&#8217;s see how Python does (less time,
the better):</p>

<p>gt2fasta     0m0.123s<br/>
Python  0m5.466s</p>

<p>Yep, that&#8217;s 44 times
slower for Python. It is widely known that compiled languages are faster
than interpreted ones, but maybe 44 times it is an exaggeration, what
makes me think that there are ways to improve out Python script to make
it a little bit faster. After next post we will see if this is possible.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/19/empires/">Empires</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-19T16:03:01-05:00" pubdate data-updated="true">Feb 19<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Most of those that are fighting against an empire, are trying to create one of their own.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/18/bosc-2008/">BOSC 2008</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-18T13:33:47-05:00" pubdate data-updated="true">Feb 18<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have tried to attend BOSC 2005 in Detroit but run out funds at the time and could only be at ISMB. I missed 06 and 07 meetings, no papers, posters ready. But 2008 ISMB and <a href="http://news.open-bio.org/archives/2008_02.html#000093">BOSC</a> will be in my &#8220;hometown&#8221;, Toronto. If you are coming let me know we can definitely hang out.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/15/for-every-problem-there-is-a-solution-thankfully/">For every &#8220;problem&#8221; there is a &#8220;solution&#8221;. Thankfully.</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-15T14:07:48-05:00" pubdate data-updated="true">Feb 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>If you don&#8217;t like Snap preview tooltips (oh, they are <em>so useful</em>) you came to the right place. No, in fact it wasn&#8217;t me that invented it, the company that created Snap did it. You don&#8217;t need to thank me.</p>

<blockquote><p>How can I turn Snap Shots on or off?</p></blockquote>

<p>Click here (on their <a href="http://www.snap.com/snapshots_faq.php#2">website</a> FAQ) to disable Snap Shots. Please note you must reload the page or clear your cache for the change to take effect. Deactivation uses cookies, so if you delete your cookies, Snap Shots will be re-activated for all sites.</p>

<p>The JavaScript will save a cookie on your browser and you are free to surf without any further annoyance. Thankfully.</p>

<p><em>from <a href="http://www.ofzenandcomputing.com/zanswers/304">1</a> (Of Zen and Computing) and <a href="http://www.lifehacker.com/software/annoyances/disable-snap-previews-on-web-sites-232846.php">2</a> (Lifehacker)</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/15/parsing-a-genbank-file-another-approach/">Parsing a GenBank file, another approach</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-15T11:35:13-05:00" pubdate data-updated="true">Feb 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of entries we learnt a little bit about <code>sets</code> and that
newer Python versions make its use a lost faster than <code>dictionaries</code> to
uniquify <code>lists</code>. And soon we will see good examples of <code>sets</code> usage.
And now for some completely different &#8230; we go back to parsing GenBank
files. We saw sometime ago how to extract some simple information
contained in this type of files. I am working on a project that requires
a lot of work in GenBank files, and instead of learning how to use an
available parser I decided to write a simple one on my own. The
requirements were simple: extract all protein sequences from a whole
genome GenBank file and also the DNA from the sequence available in it
(this time we will see how to get the proteins and next time the
latter). We all know that GenBank files have an annotation section, with
IDs, references, organism and some other information at the top. The DNA
sequence at the bottom, after ORIGIN, and the protein sequences are in
annotated regions of CDSs. An example is below:</p>

<p><code>     gene            8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /db_xref="GeneID:855380"      CDS             8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /experiment="experimental evidence, no additional details                      recorded"                      /note="Protein of unknown function, member of the DUP380                      subfamily of conserved, often subtelomerically-encoded                      proteins"                      /codon_start=1                      /product="Cos1p"                      /protein_id="NP_014063.1"                      /db_xref="GI:6323993"                      /db_xref="SGD:S000005280"                      /db_xref="GeneID:855380"                      /translation="MKENELKNEKSVDVLSFKQLESQKIVLPQDLFRSSFTWFCYEIY                      KSLAFRIWMLLWLPLSVWWKLSNNCIYPLIVSLLVLFLGPIFVLVICGLSRKRSLSKQ                      LIQFCKEVTENTPSSDPHDWEVVAANLNSYLYENKAWNTKYFFFNAMVCQEAFRTTLL                      EPFSLKKDEAAKVKSFKDSVPYIEEALGVYFREVEKQWKLFNSEKSWSPVGLEDAKLP                      KEAYRFKLTWFLKRISNIFMLIPFLNFLCCIYVSRGMCLLLRTFYLGWILFMLVQGFQ                      NMRMIVLSVKMEHKMQFLSTIINEQESGANGWDEIAKKMNRYLFEKKVWKNEEFFFDG
IDCEWFFSHFFYRVLSAKKSMRALSLNVELWPYIKEAQLSCSEESLA"</code></p>

<p>In a GenBank file each coding gene corresponds to at least one protein.
The first line, gene, contains the nucleotide region, as the CDS line.
The other lines, starting with a slash provide information about the
gene and the last piece of information is the DNA translation with the
whole protein sequence. In my case, the first objective was to obtain
the protein sequence, get the protein ID and the gi ID from each entry
in a genome file. An initial assessment, tell us that creating a class
for each protein in the file is a good start.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Protein</span><span class="p">:</span>
</span><span class='line'>  <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'>      <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>So we declare the class and have elements for the
IDs and the amino acid sequence. The next problem is how do get each
gene/CDS information, extract what we need and store in a list of class
instances? This is much like we do with our FASTA entries, reading line
by line and looking for FASTA sequence names, which start with <code>&gt;</code>. For
GenBank files we will look for lines containing the word gene. Of course
this would mean that every gene that has this word in its name will be a
problem, so we match against <code>'  gene  '</code> (the word gene flanked by two
spaces).</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;  gene &#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>            <span class="c">#parses the CDS and appends to a list</span>
</span><span class='line'>            <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>            <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>        <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>        <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="c">#found the DNA sequence, we can stop now</span>
</span><span class='line'>        <span class="k">break</span>
</span><span class='line'>    <span class="k">else</span><span class="p">:</span>
</span><span class='line'>        <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p> In the above excerpt we read
the file line by line, having an index just to keep the code aware of
which entry we are at. Every time we find a line with <code>'  gene  '</code>, we
check for the index and if it is larger than one we parse the entry that
has been compiled already and start a new one. If we find &#8216;ORIGIN&#8217; we
bail out, breaking the loop, and in every other case we add a line to
our current entry. When this is sent to the parsing function we clear
the string and start compiling the information again. Using a string for
an entry might be a second choice, as lists are easier to manipulate
after (and eventually we will transform our string in a list by
splitting it). Notice that we have a call to a function <code>parse_entry</code>,
let&#8217;s see how this function is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span>
</span><span class='line'>    <span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>            <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>            <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>            <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>            <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>            <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>                    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'>    <span class="k">return</span> <span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p> As mentioned
above, we split the string in a list (we will check if there is any
improvement in speed if we use a list from the beginning) and then
analyse each one of the lines. In this case, we only need the protein,
the gene name and, evidently, the protein sequence. Dissecting the
excerpt above we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>This will get us the protein ID from
<code>/protein_id="NP_014063.1"</code> and also the gene name/info from
<code>/product="Cos1p"</code>, which them can be stored in the <code>prot_id</code> string
that will be passed to the class instance. The gi ID can be extracted in
similar fashion. In order to get the sequence we have to do a small
trick, that can be seen on the excerpt below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>  <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p> We find the beginning of the translation information by checking the line that contains
<code>/translation</code> and we extract the initial part of the sequence my
finding the equal sign and getting everything after it. Then the trick.
We know that the end of the translation information will be the last
line of the whole gene/CDS entry. So we check for the index of the first
sequence line and store it in a temp variable. Using this variable as
the first value of a range we are able to get every other line of the
sequence. That simple. To do the output, we can just iterate the list of
class objects and print all the information in FASTA format or any other
that we wish. Put everything together and we have a simple but effective
GenBank parser. Please notice that this script wouldn&#8217;t work in every
possible scenario, but with minor tweaks it should be applicable in most
cases. Next time, we will see how we can get the DNA sequence in an
effective way. <em>An example of GenBank file can be found
<a href="http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?db=nuccore&amp;id=117937805">here</a></em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/14/3-masterpieces/">3 masterpieces</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-14T15:04:10-05:00" pubdate data-updated="true">Feb 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Leaving the Science behind, and thanks to the &#8220;beauty&#8221; of open office spaces fro bioinformatics development, I am able to enjoy three great music masterpieces (AKA the records I would take to an isolated island)</p>

<p>1- <a href="http://www.amazon.com/Black-Love-Afghan-Whigs/dp/B000002HLG/ref=pd_bbs_1?ie=UTF8&amp;s=music&amp;qid=1203019110&amp;sr=8-1">Black Love - Afghan Whighs</a></p>

<p>2- <a href="http://www.amazon.com/Heaven-Up-Here-Echo-Bunnymen/dp/B0000E2Y8Z/ref=sr_1_2?ie=UTF8&amp;s=music&amp;qid=1203019254&amp;sr=1-2">Heaven up Here - Echo and the Bunnymen</a></p>

<p>3- <a href="http://www.amazon.com/Starfish-Church/dp/B0009NR7VI/ref=sr_1_1?ie=UTF8&amp;s=music&amp;qid=1203019320&amp;sr=1-1">Starfish - The Church</a></p>

<p><em>Disclaimer: none of the links above are connected to the Amazon Associates program. I searched Amazon for the title and copied the url that appeared on my address bar.</em></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/26/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/24/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
