
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="About eight months ago I started using Biostar as I saw it as a great
opportunity to exchange some ideas, concepts, tips in biology and &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/12/01/why-i-left-biostar-but-i-still-like-stackoverflow/">Why I left Biostar, but I still like Stackoverflow</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-12-01T10:44:08-05:00" pubdate data-updated="true">Dec 1<span>st</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>About eight months ago I started using Biostar as I saw it as a great
opportunity to exchange some ideas, concepts, tips in biology and
bioinformatics. I even mentioned the website in this space, trying to
bring more people to the mix; at the time the community wasn&#8217;t big
enough, and some days went by without any question being posted. But a
couple of months ago my interest started to go down the drain. I don&#8217;t
know if it was the constant next-generation sequencing barrage of
questions every day, if it was the infantile blog/twitter posts from
members competing for points or maybe the lack of votes for some answers
that I posted (that&#8217;s selfish on my part, I admit). But at some point it
seemed that the website turned into a competition of CVs or knowledge,
very different from what I could see in different Stackoverflow
spin-offs or in the main site. I guess the turning-point, or the moment
I realized that the scientific community (at least in bioinformatics and
related fields) will never the be the same as the programming and
statistical ones, was the time I gave an answer that had less votes that
the one saying &#8220;it&#8217;s not possible&#8221;. Maybe the problem is myself, I don&#8217;t
like cliques, don&#8217;t mind helping people for nothing, don&#8217;t care about
reputation. I didn&#8217;tt care about how many points I had, and used the
down-vote to actually vote down answers that I didn&#8217;t see as pertinent
(if you never used those sites, every down-vote removes one point from
your score). I still think that Biostar is a great idea, and I wish it
becomes a great resource for all the bio fields. Maybe if the community
gets big enough, maybe if don&#8217;t see the same group of people that you
see every where else it might become a better place to hang-out online.
But right now, I&#8217;m over it.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/05/20/preview-of-django-1-1-testing-and-debugging/">Preview of Django 1.1 Testing and Debugging</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-05-20T16:34:41-04:00" pubdate data-updated="true">May 20<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Packt Publishing invited me to review <a href="http://www.packtpub.com/django-1-1-testing-and-debugging/book/mid/280410xne4fe?utm_source=python.genedrift.org&amp;utm_medium=affiliate&amp;utm_content=blog&amp;utm_campaign=mdb_003140">Django 1.1 Testing and
Debugging</a>by
Karen M. Tracey. They also kindly provided a free chapter that you can
download from the link below. A full review will be posted as soon as I
finish the book. preview chapter - <a href="http://python.genedrift.org/7566_Django%201.1%20Testing%20and%20Debugging_SampleChapter_1.pdf">Chapter No.3 &#8220;Testing 1, 2, 3: Basic
Unit
Testing&#8221;</a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/05/01/initial-impressions-about-bioinformatics-programming-using-python/">Initial impressions about Bioinformatics Programming using Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-05-01T20:59:17-04:00" pubdate data-updated="true">May 1<span>st</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last week I made a 5 book order at Amazon and one of them was
<a href="http://www.amazon.com/gp/product/059615450X?ie=UTF8&amp;tag=genedrift-20&amp;linkCode=as2&amp;camp=1789&amp;creative=9325&amp;creativeASIN=059615450X">Bioinformatics Programming Using Python: Practical Programming for
Biological Data (Animal
Guide)</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=as2&amp;o=1&amp;a=059615450X" alt="image" />
by Mitchell L Model. I started reading the book late Friday night, and
I&#8217;m on the third chapter, where there&#8217;s an introduction to sequences. So
far, I found the book very confusing, especially as it claims to be a
book for people with no programming background. The examples are OK, but
there&#8217;s a very messy mixture of Python interpreter and standalone script
usage, as the author jumps back and forth them. Another thing is that
some examples are explained in detail including the line number, while
others you depend on the code&#8217;s docstring to understand it. So far, I&#8217;m
not impressed. The initial Python sequence example is a set and in this
chapter there already are some functional programming concepts, what can
be quite challenging to someone that has never programmed in their life.
And in the second chapter the reader sees a ternary operator. Another
criticism, is that in the preface the author suggests using Python 3,
instead of 2, what might add to the frustration of the beginner when a
module cannot be installed. I will continue reading it and post whenever
I have a more complete overview of the book.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/19/python-for-bioinformatics-by-sebastian-bassi-a-review/">Python for Bioinformatics by Sebastian Bassi: a (short) review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-19T09:01:08-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I promised some time ago to post a complete review of <a href="http://www.amazon.com/gp/product/1584889292?ie=UTF8&amp;tag=genedrift-20&amp;linkCode=as2&amp;camp=1789&amp;creative=9325&amp;creativeASIN=1584889292">Python for
Bioinformatics (Chapman &amp; Hall/CRC Mathematical &amp; Computational
Biology)</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=as2&amp;o=1&amp;a=1584889292" alt="image" />
by Sebastian Bassi. It&#8217;s long overdue, but the delay allowed me to get
more acquainted to the book and its contents. I can only say that I
highly recommend this book, especially for the biologist that is
beginning in bioinformatics or python (or both). I cannot compare it to
any other Python and Bioinformatics books (I&#8217;m planning to buy the
another one), but I can say that I could learn a thing or two from
Sebastian&#8217;s book. Evidently is not a perfect book, as some of the
explanations are a little bit rushed and might be difficult for a
beginner. At the same time this is a very carefully thought and planned
book and has more than enough for one to learn Python and apply the
language to solve biological problems. I really liked the BioPython
section, and this section made me use BioPython for the first time. Some
of BioPython&#8217;s examples in the book are light years ahead of the
examples in the tool&#8217;s website. Lastly, I would like to congratulate
Sebastian for his work and effort in putting together a nice tome for
Python and Bioinformatics. It&#8217;s a valuable resource for everyone in the
field and certainly will help spread Python in the community.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/17/biostar-bioinformatics-community/">Biostar: bioinformatics community</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-17T13:26:08-04:00" pubdate data-updated="true">Apr 17<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://biostar.stackexchange.com/">Biostar</a> is a bioinformatics
community on the StackExchange network. It&#8217;s still small and not a lot
of questions are asked and answered every day, so we need more people
participating. If you are new to bioinformatics, or are just curious
about the newest trends in the field, help us grow.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/10/the-real-value-of-blogging/">The real value of blogging</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-10T13:45:03-04:00" pubdate data-updated="true">Apr 10<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A couple of days ago I posted on here an entry called &#8217;<a href="http://python.genedrift.org/2010/04/08/the-sickest-python-code-ive-ever-created-1/">The &#8220;sickest&#8221;
Python code I&#8217;ve ever
created</a>&#8217;.
It&#8217;s a code that does some file management for proteomics data, with a
different set of inputs each time you run it. The &#8220;sickest&#8221; part of the
title is that it was a small challenge to me. I&#8217;ve been away of actual
hard-core coding for quite sometime, and you lose some of the gist of
the thing with time. Mostly, nowadays, I make simple scripts that don&#8217;t
require any kind of advanced skills (in any language) and I don&#8217;t worry
that much of releasing code or about ultra fast performance. I knew from
the time I posted that a lot of people would jump and help and teach me,
as I was aware it wasn&#8217;t the most elegant code out there, not even the
most Pythonic one too. What also helped me is that my
Python/Bioinformatics blog is indexed on Planet Python, and the audience
is far more hard-core Python that I ever dreamed of getting by myself
alone. But the real deal is that I believe it would be much more
difficult for me to get some positive feedback or even an answer if I
had posted bits of my code on a online forum or community or list. Every
time I used one of these methods, I either got no answer, or got
schooled for not posting in the right format or somebody replied that no
one knew how to do it. There&#8217;s the real deal of blogging, and the value
is even higher if your audience knows more than you do. I appreciate
every comment I got on that post and on others too, I learned things
that I wasn&#8217;t able to learn from computer books and online tutorials
(yes, I searched sometimes before reading the comments, and sometimes
after).</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/04/08/the-sickest-python-code-ive-ever-created-1/">The &#8220;sickest&#8221; Python code I&#8217;ve ever created 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-04-08T20:59:54-04:00" pubdate data-updated="true">Apr 8<span>th</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>But, I guess, it can be easily refactored/enhanced/despised by the
audience that read or have access to this blog via Planet Python.
Anyway, for someone like me, whose main task now is not to generate tons
of code and lines, I think the code (or part of it) that I will present
below is quite good. Feel free to comment, criticize and say bad and
good things about it. We needed a script that would take files coming
out from protein search engines that would be able to compare the
peptides and protein sequences, their abundance and some other
characteristics. We had a combination of protein and peptide files, with
a list of proteins (one protein per line in a tab delimited file) that
was related to a list of peptides in another file (one peptide per line,
with multiple peptides/lines related to one protein in the original
list). Also each line in both files had more than 50 columns, and 8 or
16 of them were the values we wanted to extract. I say 8 or 16 of them
because we didn&#8217;t know how many will output each time, as it would
depend on the number of samples per run (4 to 8 samples) . So, we had a
couple of issues: we didn&#8217;t know how many proteins would be output
(actually found) in each file, we wouldn&#8217;t know how many peptides for
each protein would be found and we didn&#8217;t know before hand how many
samples would be run at once. One good thing is that the 8-16 columns of
values were fixed, always in the same position and with empty cells if
no value was registered there. And we had a fourth problem: usually the
samples attributions would be random, meaning a control could come in
the first value column or could come in the last. And a fifth as we
didn&#8217;t know before hand (the tech knew) how many treatments would be run
each time. A treatment could be a different experimental condition, a
sample grouping or some other extraneous factor. An extra issue is that
we would need to compare multiple files, and get protein and peptide
abundances in all of them at the same time and finally compare each
treatment. Basically, in order to create an universal script we needed
something flexible enough that whatever the experiments threw at use we
would be able to handle. First step we decided to use a YAML file that
could be filled by the experimental researchers with sample assignments,
treatments, etc. The YAML would look like this B0: - 114: A - 115: D -
116: B - 117: C B1: - 114: C - 115: A - 116: D - 117: B In this file B0
and B1 would be the result file names, 114 is the column/channel where
the sample was run and and A, B, C and D the treatment. With this set,
out objective was to get all proteins and their peptides for treatment A
in files B0 and B1, do some calculation and them compare to all proteins
and peptides from treatment B, C and D extracted also from files B0 and
B1. First step was to get the names of the treatments from the YAML file
[sourcecode language=&#8217;python&#8217;]]def get_treatments(mapping): treats =
set([]) for entry in mapping: [treats.add(list(t.values())[0]) for t in
mapping[entry]] return treats[/sourcecode] where [code]mapping[/code] is
the YAML file name. We used a set to store and sets have unique items,
and treatment names can vary from file to file. In the code above we
basically read the YAML and the value for each entry. We then needed a
class to store protein information, and there was the story got hairy.
With all my (lack of ) experience, I decided to use [code]exec[/code]
statements to fix all the uncertainty of the experimental data details.
I didn&#8217;t have the treatment names before hand (or in a fixed immutable
list), and didn&#8217;t have the columns (channels) that were being used at
the time and I have to correctly assign each protein abundance (area) to
its place. In the end our class look like this [sourcecode
language=&#8217;python&#8217;]class Protein(): &#8220;&#8221;&#8220;Class Protein, stores all the
information about channels and areas, name and accession&#8221;&#8220;&#8221; def
__init__(self, accession, name, treatments): self.accession =
accession self.name = name #ratio channels are called based on their
name for i in treatments: exec(&#8216;self.%s = []&#8217; %i) exec(&#8216;self.area%s =
[]&#8217; %i) def add_to_channel(self, channel, peptide):
exec(&#8216;self.%s.append(peptide)&#8217; % channel) def add_to_area(self,
channel, area): exec(&#8216;self.area%s.append(area)&#8217; % channel)[/sourcecode]
In order to be faithful to this blog&#8217;s name, I will explain how the code
above is supposed to work. First, [code]exec[/code] is a Python
statement that support dynamic execution of code. In our case above it
was used to name the objects, so we would be able to access them by name
in subsequent functions. Let&#8217;s take this for example [sourcecode
language=&#8217;python&#8217;]for i in treatments: exec(&#8216;self.%s = []&#8217; %i)
exec(&#8216;self.area%s = []&#8217; %i)[/sourcecode] In this snippet we were trying
to create lists called (for the YAML file above) A, B, C and D, and
another set of lists called areaA, areaB, areaC and areaD. Let&#8217;s say for
another experiment we would have treatments &#8220;Control&#8221;, &#8220;Low&#8221; and &#8220;High&#8221;
and so on. The next two functions use the same approach, with exec, this
time appending to the freshly created lists. This way it&#8217;s easy to
control what the user is throwing at us. I don&#8217;f know if this the best
approach possible, or if it is or not harmful. Maybe experts reading
this might have better ideas, and I appreciate them. We check the rest
of the script next time.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/03/03/python-testing-beginnere28099s-guide-review/">Python Testing Beginner’s Guide, review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-03-03T22:46:35-05:00" pubdate data-updated="true">Mar 3<span>rd</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I posted about a week ago that Packt Publishing had invited me to review
<strong><a href="http://www.packtpub.com/python-testing-beginners-guide/book/mid/240210aaphjg?utm_source=python.genedrift.org&amp;utm_medium=bookrev&amp;utm_content=blog&amp;utm_campaign=mdb_002570">Python Testing Beginner&#8217;s
Guide</a></strong>
by Daniel Arbuckle. Having finished reading the book (I must admit that
I haven&#8217;t tried all the code in it), I can say that I have an excellent
initial impression of the book. PTBG is not a long book and the topic is
divided in 10 chapters and one appendix. One of the first things that I
liked about the book is that there&#8217;s no <em>introduction</em> (or something
similar) to Python. It just goes straight to the point assuming that you
have some good understanding of the language and everything that
surrounds it. In the past I was frustrated with some &#8220;Introduction to X
with Python&#8221; that wasted precious space talking over and over about a
topic, learning Python, better covered in many other books. PTBG does
not waste time and space introducing its main topic which is testing,
and in my opinion that&#8217;s the best approach, even though it might look a
little bit abrupt by some. The language and text in the book is clear
and very pleasant. PTBG is a very well written book and I really enjoyed
its style. The first chapters of the book cover Python testing using
doctests. For someone like me that didn&#8217;t write so many tests in the
normal software development workflow (I know I should write more tests),
this section seems like a really nice introduction to the topic, with
well thought real-life like examples and a good flow on the explanation
of the different features. One small complain that I have is that for a
beginner sometimes the code listed in the examples might seem a little
bit confusing, and maybe the addition of line numbers might have helped
a bit here. But at the same I understand that this is normal style of
some Packt books. After the doctests section, PTBG gets into more
advanced techniques, covering a little bit mock objects with
<a href="http://labix.org/mocker">Mocker</a>, then moving into unittest and
<a href="http://code.google.com/p/python-nose/">nose</a>. The latter is a Python
tool that allows for managing, running and automating tests. Also
covered is <a href="%20http://pypi.python.org/pypi/%20twill/">Twill</a>, another
third-party library that allows for testing of web applications. One
full chapter is devoted to test-driven development, with a complete
walkthrough of this approach. This gives a wrap-up of most of the
techniques and modules covered in the book, but there&#8217;s still space for
another chapter that shows how beautifully doctests, unittest and nose
can be fully integrated and help the development of applications using
the test-driven approach. Overall, I really enjoyed PTBG. As I
mentioned, test driven development was never a high priority in the
application I usually developed with Python. But certainly this book can
be a good starting point for some Python test beginners to incorporate
these techniques in their usual development workflow. Scientific
software is also a perfect niche for this type of approach and we should
do what is possible in order to avoid the <a href="http://boscoh.com/protein/a-sign-a-flipped-structure-and-a-scientific-flameout-of-epic-proportions">nightmares of the
past</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/02/22/preview-of-python-testing-beginners-guide/">Preview of Python Testing Beginner&#8217;s Guide</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-02-22T11:00:53-05:00" pubdate data-updated="true">Feb 22<span>nd</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I was invited by Packt Publishing to review <strong><a href="http://www.packtpub.com/python-testing-beginners-guide/book/mid/240210aaphjg">Python Testing Beginner&#8217;s
Guide</a></strong>
by Daniel Arbuckle. This is a book on one of the most important aspects
of scientific programming (even though the majority of scientific
software don&#8217;t have any testing routines): code testing, checking if
your code actually does what is intended to do. I can say I&#8217;m not really
an expert on testing so I guess I&#8217;m the right audience for it:</p>

<blockquote><p>You&#8217;ll learn about several of Python&#8217;s automated testing tools, and
you&#8217;ll learn about the philosophies and methodologies that they were
designed to support, like unit testing and test-driven development.
When you&#8217;re done, you&#8217;ll be able to produce thoroughly tested code
faster and more easily than ever before, and you&#8217;ll be able to do it
in a way that doesn&#8217;t distract you from your &#8220;real&#8221; programming.</p></blockquote>

<p>Packt also supplied a preview/sample chapter that you can download
<a href="http://python.genedrift.org/wordpress/wp-content/uploads/2010/02/8846-python-testing-beginners-guide-sample-chapter-5-when-doctest-isnt-enough-unittest-to-the-rescue.pdf">here</a>.
I hope to get a review ready by the end of the week. before the Ontario
Institute of Cancer Research retreat, otherwise I will try to post a
full review next week.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2010/01/03/preliminary-review-of-python-for-bioinformatics-by-sebastian-bassi/">Preliminary review of Python for Bioinformatics by Sebastian Bassi</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2010-01-03T14:25:00-05:00" pubdate data-updated="true">Jan 3<span>rd</span>, 2010</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Let me start by saying that <a href="http://www.amazon.com/gp/product/1584889292?ie=UTF8&amp;tag=genedrift-20&amp;linkCode=as2&amp;camp=1789&amp;creative=9325&amp;creativeASIN=1584889292">Python for Bioinformatics (Chapman &amp;
Hall/Crc Mathematical &amp; Computational
Biology)</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=as2&amp;o=1&amp;a=1584889292" alt="image" />
is a massive book, massive in a way that it contains a lot of material.
I still didn&#8217;t have enough time to check everything, but I&#8217;m well into
the first section of the book that gives an initial view of Python and
how to program it. The initial section of the book is well written (I&#8217;m
not going criticize the book in terms of good/poor English, as I&#8217;m not
well qualified to do that), and gives a clear perspective on how to
program Python for scientists, who are the main target demographic of
the book. Of course, it always help to have some basic knowledge of
command line shells, but the book also includes some explanations of
IDLE and other Python-capable IDEs. I cannot say that I read this
section with the enough care and attention, but what I can say is that
you won&#8217;t miss a beat with PfB, as it has more material than I expected.
I still have to start with the more advanced topics, like
<a href="http://biopython.org/" title="Biopython">BioPython</a> and so forth, what I plan
to do in the coming month, and as I don&#8217;t have a lot of experience with
BioPython, I&#8217;m looking forward to it. On the other hand I have a
small-ish complaint, that maybe is more about style than substance. I
don&#8217;t like the design of the book, the way the code interleaves with the
text and the way the code explanations are presented. Most of the code
blocks are followed by a careful explanation, but this explanation works
as a figure label for the code block. That is quite annoying because
there are too many stops in the text fluidity as one tends to lose
attention to it (my case, not exactly everyone&#8217;s). Another minor detail
is the use of &#8220;he&#8221; every time scientists are referred (one example is on
page 3 on the second phrase of the introduction). The (politically)
correct would be to use &#8220;he or she&#8221; or &#8220;she or he&#8221; (but that&#8217;s OK with
me). I will try to post more complete reviews of the sections that I
don&#8217;t master. I would also like to thank Sebastian for sending me a copy
of the book.
<a href="http://reblog.zemanta.com/zemified/99930d7a-5cef-4e74-b46c-71e031c4d627/" title="Reblog this post [with Zemanta]"><img src="http://img.zemanta.com/reblog_e.png?x-id=99930d7a-5cef-4e74-b46c-71e031c4d627" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/2/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2010/12/01/why-i-left-biostar-but-i-still-like-stackoverflow/">Why I left Biostar, but I still like Stackoverflow</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/20/preview-of-django-1-1-testing-and-debugging/">Preview of Django 1.1 Testing and Debugging</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/05/01/initial-impressions-about-bioinformatics-programming-using-python/">Initial impressions about Bioinformatics Programming using Python</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/19/python-for-bioinformatics-by-sebastian-bassi-a-review/">Python for Bioinformatics by Sebastian Bassi: a (short) review</a>
      </li>
    
      <li class="post">
        <a href="/blog/2010/04/17/biostar-bioinformatics-community/">Biostar: bioinformatics community</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
