
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="After a long hiatus we are (almost) back on track in order to get our
scripts to determine overrepresented motifs in DNA sequences. Last time
we &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/18">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/08/12/obtaining-overrepresented-motifs-in-dna-sequences-part-11/">Obtaining overrepresented motifs in DNA sequences, part 11</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-08-12T20:50:17-04:00" pubdate data-updated="true">Aug 12<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>After a long hiatus we are (almost) back on track in order to get our
scripts to determine overrepresented motifs in DNA sequences. Last time
we checked we defined the &#8220;best&#8221; <a href="http://en.wikipedia.org/wiki/Factorial" title="Factorial">factorial function</a> in Python</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">fac_01</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
</span><span class='line'>        <span class="n">result</span> <span class="o">*=</span> <span class="n">i</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span></code></pre></td></tr></table></div></figure>


<p> and Andrew Dalke
sent a couple of links pointing out to a binomial calculation function,
one of them is below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c"># This file contains the Python code from Program 14.10 of</span>
</span><span class='line'><span class="c"># &quot;Data Structures and Algorithms with Object-Oriented Design Patterns in Python&quot; </span>
</span><span class='line'><span class="c"># by Bruno R. Preiss. </span>
</span><span class='line'><span class="c">## Copyright (c) 2003 by Bruno R. Preiss, P.Eng. All rights reserved. </span>
</span><span class='line'><span class="c">## http://www.brpreiss.com/books/opus7/programs/pgm14_10.txt # </span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">binom</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">m</span><span class="p">):</span>
</span><span class='line'>  <span class="n">b</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
</span><span class='line'>  <span class="n">b</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
</span><span class='line'>      <span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'>      <span class="n">j</span> <span class="o">=</span> <span class="n">i</span> <span class="o">-</span> <span class="mi">1</span>
</span><span class='line'>      <span class="k">while</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>          <span class="n">b</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+=</span> <span class="n">b</span><span class="p">[</span><span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
</span><span class='line'>          <span class="n">j</span> <span class="o">-=</span> <span class="mi">1</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">return</span> <span class="n">b</span><span class="p">[</span><span class="n">m</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>There is a similar implementation in the Python Cookbook
online and it is clearly more convenient using this function than
actually coding to calculate an identical value using factorials. But
anyway, let&#8217;s see how a function that calculates one binomial
coefficient from the Hypergeometric Distribution (HD) by using the
factorial function. Later we benchmark both methods. Each <a href="http://en.wikipedia.org/wiki/Binomial_coefficient" title="Binomial coefficient">binomial coefficient</a>
can be expanded as
<img src="http://codecogs.izyba.com/png.latex?%5C200dpi%20%5Cbegin%7Bpmatrix%7Dn%5C%5Cr%5Cend%7Bpmatrix%7D%20=%20%5Cfrac%7Bn!%7D%7Br!(nr" alt="image" />!%7D)
and the HD has three of them. From the Wikipedia &#8220;In probability theory
and statistics, the <a href="http://en.wikipedia.org/wiki/Hypergeometric_distribution" title="Hypergeometric distribution">hypergeometric distribution</a>
is a discrete probability distribution that describes the number of
successes in a sequence of n draws from a finite population without
replacement.&#8221; In the next post we will define each term of the HD for
the motifs case. In each <a href="http://en.wikipedia.org/wiki/Binomial_theorem" title="Binomial theorem">binomial expansion</a>
we would have to calculate three factorial values, nine in total. With
the binomial function, only three values need to be calculated. So,
using the factorial function we would need to code something like this
in order to calculate one of the binomials</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>
</span><span class='line'><span class="c">#let&#39;s say the motif quorum in the foreground is 7</span>
</span><span class='line'><span class="c">#and the total number of sequences is 3000 </span>
</span><span class='line'><span class="c">#we won&#39;t touch the other required values </span>
</span><span class='line'><span class="n">fore</span> <span class="o">=</span> <span class="mi">7</span> <span class="n">total</span> <span class="o">=</span> <span class="mi">3000</span>
</span><span class='line'><span class="n">hd</span> <span class="o">=</span> <span class="n">fac_01</span><span class="p">(</span><span class="mi">3000</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">fac_01</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span> <span class="o">*</span> <span class="n">fac_01</span><span class="p">(</span><span class="mi">2993</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">hd</span>
</span></code></pre></td></tr></table></div></figure>


<p>Next, we will benchmark and see if there is an advantage to either method.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/31/these-are-really-my-sequences/">These are really my sequences!!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-31T16:45:00-04:00" pubdate data-updated="true">Jul 31<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Just an update to my previous post: I submitted around 40 sequences to NCBI, all of them from South American frogs and toads. You can find all of them <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=nucleotide&amp;cmd=&amp;term=nuin+[AUTH]&amp;go=Go">here</a>. If you ever use them, I&#8217;d be happy to know and we can even share a beer sometime.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/31/those-are-my-sequences/">Those are my sequences!!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-31T16:37:53-04:00" pubdate data-updated="true">Jul 31<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Yesterday I was lazily browsing the web, reading my RSS feeds when I found <a href="http://treethinkers.blogspot.com/2008/07/phylota-my-mind-is-blown.html">this</a>:</p>

<blockquote><p>As a side note, I talked to a colleague who got harassed at the Ichs and Herps meeting for&#8230; gasp&#8230; downloading sequences from GenBank and using them without asking the author&#8217;s permission! Good lord, what is the world coming to? I&#8217;m surprised to hear of such active resistance to public availability of information.</p></blockquote>

<p>Oh, boy, that brought me back memories of the time I spent at the Museu de Zoologia (MZUSP) doing my PhD. I think that people that make such kind of remarks are the same that say Clustal W masks the &#8220;parsimony&#8221; of the sequence as it generates a Neighbour Joining tree to guide the alignment. Strange times, why would you complain that someone used your sequences without asking permission if your sequences are stored in a public repository? If you don&#8217;t want anyone to use don&#8217;t submit to NCBI or keep renewing the deadline to make them public. It might give you time to actually learn them by heart &#8230;</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/29/shame/">Shame</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-29T17:33:41-04:00" pubdate data-updated="true">Jul 29<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content">
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/23/staring-at-the-clouds/">Staring at the clouds</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-23T21:54:00-04:00" pubdate data-updated="true">Jul 23<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Us-nasa-columbia.jpg"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/3/3d/Us-nasa-columbia.jpg/202px-Us-nasa-columbia.jpg" alt="An example of a computer cluster" /></a>Image via <a href="http://commons.wikipedia.org/wiki/Image:Us-nasa-columbia.jpg">Wikipedia</a>In my opinion the best aspect of the so-called Web 2.0 (a catchy name) is the interaction we get from tools such <a href="http://friendfeed.com/">FriendFeed</a> and <a href="http://www.twitter.com/">Twitter</a>. Accessing these sites, one can be overwhelmed by the amount of information, all kinds of information, ranging from the new photos of tropical fruits, to the end of science as we now it, being liveblogged from ISMB 2008.</p>

<p>In the past months I have been accessing these services/tools more and more, and in the past months by the power of the crowd (at the least the small scientific crowd) a <a href="http://friendfeed.com/rooms/the-life-scientists">small group of scientists</a> have been exchanging information, ideas, links, tools, comments, learning and teaching anything that comes to mind. There I learned more about the computer cloud (an example <a href="http://www.amazon.com/gp/browse.html?node=201590011">here</a>) and one idea leading to another I thought about the possibility of creating an startup that provides bioinformatics computing services on the cloud.</p>

<p>Say you are in a small lab with limited or no access to a large cluster and you have some really interesting data that can analyzed quickly by a cluster, but slowly by a desktop PC, and generate a good set of results for a nice publication. You can always throw the data on a qsub in a university cluster and wait to the big shots to calculate the next prime number before you start your analysis or you can have a personalized virtual image with only the software you need and use the cloud to generate your results and have them when you need. Of course there is a cost to this use, but not close to the cost of applying for a grant to get your cluster, dealing with the sys admin to install the applications you want, dealing with the bureaucracy red tape and finding the right location to put your server.</p>

<p>It has become a cliche to say that (bio)informatics is becoming prevalent in scientific life and research (true but a cliche nonetheless), and it is not difficult to find scientist that don&#8217;t have available the infrastructure of a <a href="http://en.wikipedia.org/wiki/High-performance_computing">HPC</a> node or a personal cluster. The same way people outsource their sequencing os mass-spec needs, why not outsource their (bio)informatics needs? Many labs have clusters or paid access to clusters even though their use is limited or inconstant. Why invest in a cluster or pay an annual fee while you are using a couple of CPU-hours a week?</p>

<p>When I threw the idea on Twitter and FriendFeed, I didn&#8217;t know what to expect. Surprisingly for me there was a good reception to it with some people jumping right into it. I got a nice email from <a href="http://freesci.wordpress.com/">Pawel</a> with some more comments that I think were right on the spot. I don&#8217;t see many hurdles to create a startup to provide these kind of services, especially because we live in a virtual world at the moment. There is no need for a physical office, computer infrastructure or employees. The founders, as long as you have a nice and diverse group of people with broad knowledge and interests, can provide all the service and grow together. I am not saying that it will be easy, but after learning a lot from the bright group of people that make our small online community, I think this is a feasible project that can be on the street in a short time.</p>

<p>I am going to prepare a small plan (some ideas on a virtual paper) and anyone interested would be able to join, contribute and start the &#8220;company&#8221;.</p>

<h6>Related articles by Zemanta</h6>

<ul>
<li><a href="http://yihongs-research.blogspot.com/2008/07/failure-and-success-of-web-startups.html">Failure and Success of Web Startups</a></li>
</ul>


<p><a href="http://reblog.zemanta.com/zemified/a6329016-177e-4379-a14d-591db006964d/"><img src="http://img.zemanta.com/reblog_e.png?x-id=a6329016-177e-4379-a14d-591db006964d" alt="Zemanta Pixie" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/08/plos-computational-biology-t-shirt/">PLoS Computational Biology t-shirt</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-08T16:19:45-04:00" pubdate data-updated="true">Jul 8<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.flickr.com/photos/13172875@N08/2650908386/"><img src="http://farm4.static.flickr.com/3024/2650908386_c180e7d745_m.jpg" alt="The Island" /></a>Image by <a href="http://www.flickr.com/photos/13172875@N08/2650908386/">Aaronyx</a> via FlickrI was at gym, proudly wearing my <a href="http://compbiol.plosjournals.org/">PLoS Computational Biology</a> t-shirt when a person tripped in one of the corrifors me and started shouting that I was &#8220;supporting a failed model of scientific publication&#8221; and that I should start &#8220;supporting the <a href="http://www.nature.com/npg">Nature Publishing Group</a> model of publication&#8221;. I didn&#8217;t know what to say. I just know that it was the last time I wore this t-shirt in public.</p>

<h6>Related articles by Zemanta</h6>

<ul>
<li><a href="http://scholarlykitchen.sspnet.org/2008/07/07/bulk-publishing-keeps-plos-afloat/">Bulk Publishing Keeps PLoS Afloat</a></li>
</ul>


<p><a href="http://reblog.zemanta.com/zemified/a02f51b3-aef1-4540-bbbb-8971861a2fa3/"><img src="http://img.zemanta.com/reblog_e.png?x-id=a02f51b3-aef1-4540-bbbb-8971861a2fa3" alt="Zemanta Pixie" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/07/01/bioinformatics-carrer-survey/">Bioinformatics career survey</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-07-01T15:44:23-04:00" pubdate data-updated="true">Jul 1<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Via <a href="http://www.bioinformaticszen.com/2008/07/creating-a-picture-of-different-careers-in-bioinformatics/">Bioinformatics Zen</a>:</p>

<p>Loading&#8230;</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/06/29/what-a-happy-day/">What a happy day!!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-06-29T16:14:56-04:00" pubdate data-updated="true">Jun 29<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Although I was born in Brazil, my heart (football and everything else) was always Spanish. I saw and suffer with losses to Yugoslavia, France, Italy, Belgium, Korea (what a robbery!!), etc, but today was a day of enjoyment and “good” suffering in the final minutes. Thanks to the Niño and to a wonderful 11 we leave the underachievers banner behind and hope to even better days ahead. See you in South Africa.</p>

<p>Aupa España!</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/06/18/wordle/">Wordle</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-06-18T12:36:04-04:00" pubdate data-updated="true">Jun 18<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://wordle.net/gallery/nuinondelicious"><img src="http://wordle.net/thumb/nuinondelicious" alt="" /></a></p>

<p>As <a href="http://nsaunders.wordpress.com/2008/06/19/well-everyone-else-was-doing-it/">Neil</a> said about his, it is difficult to guess that I am (was?) a biologist.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/06/14/the-new-cult-is-coming-to-life/">The &#8220;new&#8221; cult is coming to life</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-06-14T21:00:03-04:00" pubdate data-updated="true">Jun 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Repent sinners, beware of the <a href="http://scienceblogs.com/pharyngula/2008/06/not_even_tempted.php">knights of the cult</a>. The &#8220;new <a href="http://en.wikipedia.org/wiki/Scientist">scientists</a>&#8221;.</p>

<p>It&#8217;s getting really boring to read all the posts and the ever growing ignorance in them. I bet you will get much better science from the related articles in this post. Or not, who knows.Related articles</p>

<ul>
<li><a href="http://kotaku.com/391451/blow-135000-on-this-bogus-crystal-ps3">Blow $135,000 On This Bogus, Crystal PS3 [Playstation 3]</a> [via Zemanta]</li>
<li><a href="http://mndoci.com/blog/2008/05/29/programming-and-science-education/">Programming and science education</a> [via Zemanta]</li>
</ul>


<p><a href="http://reblog.zemanta.com/zemified/d515eb08-2be7-4f9e-8a59-c5bf010a9c71/"><img src="http://img.zemanta.com/reblog_a.png?x-id=d515eb08-2be7-4f9e-8a59-c5bf010a9c71" alt="Zemanta Pixie" /></a></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/19/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/17/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
