
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Last couple of posts we started with some functional programming aspects
of Python. I was away last week and couldn&#8217;t create anything related &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/30">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/20/cutting-chromosomes-with-python-a-fast-approach-take-one/">&#8220;Cutting&#8221; chromosomes with Python: a fast approach, take one</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-20T16:24:11-05:00" pubdate data-updated="true">Nov 20<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of posts we started with some functional programming aspects
of Python. I was away last week and couldn&#8217;t create anything related to
FP in the meantime, so I decided to post about a quick way to &#8220;cut&#8221;, or
extract segments, from chromosomes stored as FASTA files. This is a
subject that I have been investigating for some time, as I needed a tool
that would get some segment from a chromosome for further analysis. I
have a short C++ code that does the trick and is really fast, but I
wanted to find a Pythonic way of doing it, and quick. I have been trying
different methods but no one were fast. Chromosomes are usually large,
and stored in large files, so reading line by line and using <code>readlines</code>
were extremely slow.</p>

<p>Inspired by a
<a href="http://www.goldb.org/goldblog/2007/11/07/PythonProcessingLargeTextFilesOneLineAtATime.aspx">post</a>
from Corey Goldberg, I decide to test the method explained there, with
wonderful results. Basically what we need to do is to use an iterator
protocol on the file object</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;foo.txt&#39;</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span></code></pre></td></tr></table></div></figure>


<p>This processes the human chromosome 1
(the largest) in less than 2 seconds. Our first take will be how to
extract a segment, but not precisely (that we will see in a future
post). Basically we need a start and an end point and a FASTA file. All
three would be used as parameters and a loop should take care of the
rest. Our script should look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">size</span> <span class="o">=</span> <span class="mi">0</span> <span class="n">segment</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;\&gt;&#39;</span><span class="p">):</span>
</span><span class='line'>      <span class="n">size</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">name</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">size</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">size</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="p">:</span>
</span><span class='line'>          <span class="n">segment</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>          
</span><span class='line'><span class="k">print</span> <span class="n">name</span><span class="p">,</span> <span class="n">segment</span>
</span></code></pre></td></tr></table></div></figure>


<p>We start getting the parameters from
the command line, assigning a couple of variables and we start the loop.
We open the file and use the file object iterator to check each line. We
check each line for a FASTA sequence title. If found we store tht title
in the variable <code>name</code>, and if not we check the size of the line and
increment the variable that will guide us on the file position. We use
<code>size</code> as a line by line increment and if it is between <code>start</code> and
<code>end</code> we store the read line into the <code>segment</code> variable. That&#8217;s it. A
few drawbacks with this script: it will fail on FASTA files where the
sequence is stored in one line, there is no precision to get the actual
start and end points and it can only run on unique sequence files. On
the next post we will address some of these concerns.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">SciView, part 6: interview with Dr Roderic Page</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-19T13:29:47-05:00" pubdate data-updated="true">Nov 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><h2>This edition of SciView features Dr Roderic Page from the University of Glasgow. Among other things, Dr Page was the editor of the Current Protocols in Bioinformatics, is the Editor in Chief of Systematic Biology, is an avid <a href="http://taxonomy.zoology.gla.ac.uk/rod/rod.html">blogger</a> and is the developer of TreeView, a phylogenetic tree visualization software. I had the opportunity to ask him about his bioinformatics work and his views on science. I would like to thank Dr Page. </h2>

<p><em>In your blog (<a href="http://iphylo.blogspot.com/">iPhylo</a>) an usual topic is the management and flow of information in Biology/Phylogenetics. In your opinion what are the main  hurdles of this subject in science? With the ever increasing amount of  generated information, databases, scientific publications, do you think that we will reach a stage where it won&#8217;t be possible to stay afloat in a sea of data?</em></p>

<p>RP- Well I think we are well past the floating stage! What I think would be cool would be a service like Google News that automatically aggregates<br/>
stories (for &#8220;stories&#8221; read &#8220;papers and data)&#8221; and summarises the Zeitgeist. This wouldn&#8217;t be overly hard to do. <a href="http://www.ubio.org/index.php?pagename=ubioRSS">uBio</a> has a service that classifies RSS feeds from journals by taxon. Imagine taking that, together with newly added GenBank sequences, and clustering the feeds by various topics. Then display that using a treemap (analogous to  Marcos Weskamp&#8217;s <a href="http://marumushi.com/apps/newsmap/">newsmap</a>). You could get an &#8220;at a glance&#8221; view of what is going on.</p>

<p>This is a bit of a half-baked idea, but it relies on things that are becoming more and more available: stable identifiers, and services that<br/>
provide data (and/or links between data). I think these are the two key things that will help us manage the flood of data. Without identifiers we will struggle to realise when we are talking about the same thing (witness the frustration with <a href="http://www.nodalpoint.org/2006/12/15/buggotea_redundant_links_in_connotea">redundant links in Connotea</a>, even though much of Connotea&#8217;s core functionality relies on stable identifiers), and services give us means to extract and manipulate data. We are slowly moving away from relying sole on web links, or thinking that merely &#8220;putting something on the web&#8221; is good enough.</p>

<p>Lastly, I&#8217;ve often argued that in some fields, such as phylogenetics, the days of the journal are numbered. I think what we really need is a database that can generate &#8220;reports&#8221; that function as a journal article (i.e., the contributor gets a citable paper), but where the submission process is primarily one of submitting data and protocols. Many papers in phylogenetics are of the sort &#8220;I think group x is way kewl, I grabbed gene y, used program z, and here is my tree&#8221;. Why not submit these straight to a database, instead of what we have now, which is the original data (e.g., alignments) and trees are usually lost? Databases such as TreeBASE that largely collect data after publication are sparsely populated, hence much valuable data and results are locked up in PDFs on journal web sites. Every month a weighty copy of Molecular Phylogenetics and Evolution lands on my desk, and the wasted data and trees makes me want to cry.</p>

<p><em>This is an identical question I asked Joe Felsenstein (with a twist): Lately two big projects related to phylogenetics are getting some steam: the Tree of Life and the DNA barcoding. I am more aware of the latter, due to the fact that I attended a presentation about it and through a friend that woks on the project. I think that the goals of the DNA barcoding are great for the field, wet-lab and theoretical scientists. What is your opinion about this type of project that tries to address all subjects in a big umbrella? Wouldn&#8217;t the support of a larger number of small projects be better? How would this type of projects can contribute to the information management and flow?</em></p>

<p>RP - Barcoding is the simplest of the two &#8220;big projects&#8221; &#8211; it&#8217;s a simple idea (and one that has been employed by microbiologists for decades), applied to things that we can see with the naked eye. As with all of these things, there&#8217;s been much hype, controversy, and heated debate. My own sense is that much of this is dying down, and certainly most articles in journals like Systematic Biology are investigating the performance of barcoding methods, rather than engaging in philosophical or political debate. I think barcoding will generate a lot of data, much of which will be useful, probably in ways we don&#8217;t expect. We will have standardised genetic samples taken at a known place at a known time from organisms whose identity is either known or can be inferred. I&#8217;m struggling to see why this isn&#8217;t a good thing.</p>

<p>My sense is that large projects don&#8217;t always scale well. You don&#8217;t get 10x more out of a project simply because it has 10x as much money. You get administrative layers, and a lot of politics. If the big project is really lots of little ones, then it may well succeed. Barcoding is really lots of little projects &#8211; everybody grabs their favourite beastie and sequences it. There aren&#8217;t any great technological or intellectual challenges. I think the real problem is when the project has big goals, articulates a grand vision, then struggles to deliver. I suspect <a href="http://www.phylo.org">CIPRES</a> is an example of a large project that hasn&#8217;t delivered quite as much as it promised. My biggest fear is that the <a href="http://eol.org">Encyclopedia of Life</a> may end up in this category.</p>

<p><em>You have three different blogs, iPhylo, iSpecies and bioGUID, but none of them are, what I can call, Open Science. Would you consider endeavouring in this new trend and opening up your scientific activities?</em></p>

<p>RP - You missed one blog &#8211; <a href="http://semant.blogspot.com">http://semant.blogspot.com</a>. I guess it depends what you mean by &#8220;Open Science.&#8221; These blogs are pretty open about the success and failure of various projects, and I don&#8217;t shy away from thinking out aloud. In some cases suggestions I&#8217;ve made in the blogs have been taken up and explored further by others. The Pygmybrowse widget to display large trees in a small space is a recent example.</p>

<p>For some time I&#8217;ve done things like post final reports on research grants online, as well as manuscripts that didn&#8217;t make it into print (these are now available in Nature Precedings).</p>

<p>Not all the code I write is available. Some projects, like the <a href="http://taxonomicsearch.sourceforge.net/">Taxonomic Search Engine</a> and <a href="http://myphpbib.sourceforge.net/">MyPHPBib</a> are hosted on SourceForge, and <a href="http://darwin.zoology.gla.ac.uk/~rpage/treeviewx/">TreeView X</a> is also open source (code available from my site, and as part of the Debian repository). Other stuff is sometimes so experimental and fluid that it would be more trouble than it is worth to package it nicely. Some legacy code uses libraries that are not Open Source and hence can&#8217;t be distributed. That said, I&#8217;m looking at hosting current projects (such as <a href="http://bioguid.info">http://bioguid.info</a>) on Google Code. Since 2005 anything which I&#8217;ve published as first author has been Open Access, and any software or data described in those papers is Open Source.</p>

<p><em>Sometime ago, in the Nature Network&#8217;s Bioinformatics group someone was asking about good protocols in the field. Can you tell us about the Current Protocols in Bioinformatics, which is was edited by you? How content is added to it and how frequent are the updates?</em></p>

<p>RP - I&#8217;m no longer involved with <a href="http://www.currentprotocols.com/WileyCDA/CPTitle/isbn-0471250937.html">Current Protocols</a>. It was a great project to be involved with as it opened my eyes to the breadth of the field, and I got free trips to the US to meet with the other editors (I&#8217;m easily bought). Eventually I just had too many other commitments (it overlapped with being editor of <a href="http://systbiol.org">Systematic Biology</a>), so I left the board.</p>

<p>Content is commissioned by the editorial board, who each handle 2-3 topic areas. Updates come out every three months. If you have the print version (currently two massive ring binders), then you have the fun task of inserting new sections and removing old content.</p>

<p><em>What do you think is the importance of protocols in Bioinformatics? Isn&#8217;t bioinformatics a field more dynamic than cytogenetics and molecular biology for instance?</em></p>

<p>RP - This is an interesting point. I think there are some protocols which are  widely used and are reasonably stable over time (BLAST searching, for example), and others which either are short lived (the software keeps changing, or the question being addressed by the protocol looses its relevance). In some ways Current Protocols is an old fashioned way of providing information on how to do things. A wiki may be more appropriate, certainly it could be more dynamic, and ideally free. However, one could ask whether a wiki would have acquired the same degree of authority as the Current Protocols, which could assemble a board of people with broad expertise and experience, pay contributors for their work, provide copy editing and other editorial assistance, plus give a commitment to maintain and update the publication.</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. TreeView(X) is a very important application in Phylogenetics and it is basically interface only, as the main focus is to display phylogenetic trees. Usually, a good percentage of messages sent to Paup&#8217;s discussion list are somewhat related to TreeView. What in your opinion is the importance of a graphical interface for a biologist?</em></p>

<p>RP - Well, unless you only use Lynx as your web browser, and vi to edit text, then I think the importance of the graphical interface is pretty much self evident! I think graphical user interfaces are desirable in two sets of circumstances: (1) the software is infrequently used, or (2) the user is not a &#8220;power user&#8221;. When I was at Oxford a colleague told me that he liked graphical interfaces because, after weeks in the field, the last thing he needed was to have to learn (or relearn) some garbled syntax in order to analyse his hard-won data. I&#8217;ve also seen the difference a graphical interface can make when teaching students &#8211; you want to keep the mechanics of using the software out of the way and let them focus on what they are actually doing (and why they are doing it).</p>

<p>Of course, graphical interfaces can get in the way as soon as you want to do something not originally thought of by the person who wrote the program. This is one reason why David Swofford&#8217;s PAUP* is a model of flexible software. There is an elegant GUI (sadly fully available on Mac Classic only, although the Windows version has some of the essential features such as the spinning cursor and the PAUP Monaco font, both things I ported across a decade ago), and the command-line interface for batch processing.</p>

<p><em>Continuing on TreeView(X), your latest version is pure wxWidgets. Why did you choose this framework to develop TreeViewX? Why not QT or Java or another GUI framework? How TreeViewX is different to the initial version of TreeView?</em></p>

<p>The short answer is that I have a lot of code written in C++, and I haven&#8217;t the time nor inclination to learn Java. When Java come along in the early &#8216;90&#8217;s it&#8217;s big selling point was cross-platform portability. Well, it didn&#8217;t support the Mac terribly well, and I had my own C++ class libraries that worked natively on Windows and Macs, so there was no incentive to use it. Until fairly recently most Java applications I&#8217;ve seen are sluggish and/or poorly designed. When the time came to look at supporting Linux (around 2000), Qt wasn&#8217;t an option because it wasn&#8217;t free on Windows, and didn&#8217;t support the Mac. So, wxWidgets was the logical choice, although it took a while for the Mac version to be usable.</p>

<p>TreeView X differs from TreeView in being Open Source, and available for Windows, Mac OS X, Linux, and other forms of Unix. It lacks some of the nicer features of the original TreeView (which only runs on Windows and Macs), mainly because I haven&#8217;t had time to add them, and my own work has focussed more on web application development. Like many people who program, the software I write addresses problems I&#8217;m interested in at the time, so if my interests change, the software may languish.</p>

<p><em>In your opinion what is your greatest contribution to science. A software, a publication&#8230;</em></p>

<p>RP - I don&#8217;t think there&#8217;s any easy way to answer that. Let me respond by breaking it down into three categories. What contribution was the most fun, what was the most satisfying, and what do I think people will associate with me?</p>

<p>The most fun I&#8217;ve had is blogging, or to be more precise, trying to make connections between different areas such as bioinformatics, biodiversity informatics, systematics, digital libraries, and the Semantic Web. My hope is that readers of blogs like iPhylo will get something out of it, but ultimately I don&#8217;t particularly care because, hell, I&#8217;m having a ball.</p>

<p>In terms of satisfaction, I enjoyed working on mincut supertrees. This was a change of topic (aggregating small phylogenies into one large &#8220;supertree&#8221;), required programming algorithms (rather than, say, graphical interfaces), and was my first introduction to the world of publishing in the computer science literature (the joys of LaTeX, and the requirement of having your paper accepted for publication BEFORE you could give a talk). It was also one of the best refereeing experiences I&#8217;ve had, because one diligent referee read my source code and pointed out that what I said I was doing in the manuscript and what my code actually did were not quite the same thing (I guess you&#8217;d call this an example of &#8220;Open Science&#8221;).</p>

<p>But, if your question is what will people associate with me, then I think that is the work on &#8220;historical associations&#8221;, namely genes and species, hosts and parasites, and areas and phylogenies. Making the case that these are all equivalent questions, and could be approached using the same algorithm, is probably the thing I could point to as being the thing that I have have done that was &#8220;significant&#8221;.</p>

<hr />

<p>I would like to thank Dr Page for taking the time to answer my questions.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/06/goodbye-john/">Goodbye John</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-06T14:07:40-05:00" pubdate data-updated="true">Nov 6<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I discovered John Rebus in 2002, on a trip to Canada. It was almost winter and a cool place to hang with my wife was the Eaton Centre and the Chapters (or Indigo) inside it. There I discovered <a href="http://www.amazon.com/gp/redirect.html?ie=UTF8&amp;location=http%3A%2F%2Fwww.amazon.com%2FFalls-Ian-Rankin%2Fdp%2F0752877259%3Fie%3DUTF8%26s%3Dbooks%26qid%3D1194375773%26sr%3D8-3&amp;tag=genedrift-20&amp;linkCode=ur2&amp;camp=1789&amp;creative=9325">The Falls</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=ur2&amp;o=1" alt="" /> there, attracted by the stylish cover and for the good reviews on the back. I decided to give it a try and after that I was hooked. Read all the other books by <a href="http://www.amazon.com/gp/search?ie=UTF8&amp;keywords=ian%20rankin&amp;tag=genedrift-20&amp;index=blended&amp;linkCode=ur2&amp;camp=1789&amp;creative=9325">Ian Rankin</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=ur2&amp;o=1" alt="" />.</p>

<p>Sadly he decided to retire John Rebus in his latest book. I will miss Edinburgh (for a while maybe) and I will miss John and Siobham. Thanks for the ride, it was fun.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/02/can-you-point-your-country-on-a-map/">Can you point your country on a map?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-02T16:07:26-04:00" pubdate data-updated="true">Nov 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This week the largest Brazilian magazine (Newsweek-, Time-style) has a <a href="http://veja.abril.uol.com.br/071107/p_108.shtml">report</a> (closed source, in Portuguese) about the percentage of Brazilians that can find Brazil on a World Map. The report&#8217;s title can be freely translated to &#8220;And we mock the Americans &#8230;&#8221;, referencing the usual stories of the ignorance of the Americans and other nationalities on geographical knowledge. At least 1000 people were interviewed in different Brazilian state capitals and 50% were not able to point the country on a map and, not surprisingly, only 18% can find US.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/01/functional-programming-in-python-using-filter-take-one/">Functional programming in Python: using filter, take one</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-01T16:08:24-04:00" pubdate data-updated="true">Nov 1<span>st</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This time we check another functional programming function from Python:
<code>filter</code>. As the name implies, <code>filter</code> returns items from a sequence
(list, string, etc) that are true to a certain condition defined by the
function. The syntax is very similar to <code>map</code></p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">filter</span> <span class="p">(</span><span class="n">function</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>and as <code>map</code> it returns a list (except when the sequence is a string or tuple). In
the example here we will use <code>lambda</code> to define a one-line function.
Let&#8217;s say we want to quickly find the sequences that contain a motif,
sequences that are stored in a FASTA file (again, this is a very simple
example, just a primer). Of course we can use another
<a href="http://python.genedrift.org/2007/08/28/finding-motifs-iupac-and-regex-an-approach/">method</a>,
but this time we want to use a functional programming approach. <code>filter</code>
suits us best here. Again we will reuse the newly created function that
returns only the sequences from a FASTA file and we end up with a script
that looks like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="n">motif</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
</span><span class='line'><span class="k">print</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">sequences</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>We skip the part we already seen and check the last line. Basically each
item from the list (sequences) is a string and we are applying the
<code>find</code> method in order to find a motif on a position larger or equal to
0. Notice again, the syntax similar to <code>map</code>. Of course this is not very
useful as it is, so next time we will change this script to make it more
meaningful.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/31/functional-programming-in-python-using-map/">Functional programming in Python: using map</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-31T13:33:59-04:00" pubdate data-updated="true">Oct 31<span>st</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Functional programming in Python. First we need to define what is
functional programming. Quoting
<a href="http://en.wikipedia.org/wiki/Functional_programming">Wikipedia</a>: &#8220;is a
programming paradigm that treats computation as the evaluation of
mathematical functions and avoids state and mutable data. It emphasizes
the application of functions, in contrast with the imperative
programming style that emphasizes changes in state.&#8221; Rather complex, eh?</p>

<p>Another explanation can be found
<a href="http://www.cs.nott.ac.uk/~gmh/faq.html#general-topics">here</a>:
&#8220;Functional programming is a style of programming that emphasizes the
evaluation of expressions, rather than execution of commands. The
expressions in these language are formed by using functions to combine
basic values. A functional language is a language that supports and
encourages programming in a functional style.&#8221; Basically any functional
programming language tries to minimize the code, reducing it, avoiding
the use of variables and relying on the use of functions and
expressions. Python, clearly, is a not a functional programming language
<em>per se</em>, but it has some functions that allow a functional approach. We
will see a couple of them in this entry. If you are interested in
learning more about functional programming and Python, check Alan
Gauld&#8217;s
<a href="http://www.freenetpages.co.uk/hp/alan.gauld/tutfctnl.htm">tutorial</a>.</p>

<p>There are many ways to use functional programming in bioinformatics. We
will start with a very simple example, but first we need to modify the
FASTA module we created before. Currently, this module only has two
functions, one that reads sequences and their names and one that formats
their output. We will add a function that reads only the sequences and
returns a list containing them. This function is very similar to the
<code>read_fasta</code> function</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">read_seqs</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>  <span class="n">items</span> <span class="o">=</span> <span class="p">[]</span> <span class="n">seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;&gt;&quot;</span><span class="p">):</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>          <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</span><span class='line'>          <span class="n">seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>          <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span> <span class="k">else</span><span class="p">:</span>
</span><span class='line'>              <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>          <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</span><span class='line'>      <span class="k">return</span> <span class="n">items</span>
</span></code></pre></td></tr></table></div></figure>


<p>and it uses an identical approach. Instead of creating an instance of
<code>Fasta</code> class we only create a simple list, and ignore the sequence
title. This will also be useful for a future code reuse, as sometimes we
are only interested in the sequence itself and not the name. Might save
us some typing and decrease code size depending on the situation. In
this entry we will see one Python&#8217;s functional programming function:
<code>map</code>. But we will need to use another one too, which is called
<code>lambda</code>. <code>lambda</code> is the function that defines what is called an
anonymous function, or a nameless function that can fit in one line. We
have seen that functional programming is all about functions, so
<code>lambda</code> is the aid to implement the small and short functions we use in
this type of approach. One example of <code>lambda</code> use would be on the
calculation of number exponentials, which normally can be achieved like
this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">exp</span><span class="p">(</span><span class="n">value</span><span class="p">):</span> <span class="k">return</span>
</span><span class='line'><span class="n">value</span>\<span class="o">*</span>\<span class="o">*</span><span class="mi">2</span>
</span></code></pre></td></tr></table></div></figure>


<p>(exponentials in Python are defined by **)</p>

<p>Using <code>lambda</code> we can rewrite the above function like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">exp</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">value</span><span class="p">:</span><span class="n">value</span>\<span class="o">*</span>\<span class="o">*</span><span class="mi">2</span>
</span></code></pre></td></tr></table></div></figure>


<p>with the
following syntax:</p>

<p><code>lambda [parameters] : [expression to be used on the parameters]</code></p>

<p>In
both cases you can use the functions the same way, by calling <code>exp(10)</code>
for example. Ok, let&#8217;s move to <code>map</code> then. <code>map</code> applies a function,
usually a <code>lambda</code>, to every item in a list of items. This list can be
of any sequence type, but <code>map</code> always returns a list. Using the above
exponential example, let&#8217;s say we want to calculate the cube of a series
of values. Using <code>map</code> and <code>lambda</code> we would have something like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">value</span><span class="p">:</span><span class="n">value</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">34</span><span class="p">,</span><span class="mi">56</span><span class="p">])</span>
</span><span class='line'><span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">144</span><span class="p">,</span> <span class="mi">1156</span><span class="p">,</span> <span class="mi">3136</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>And that&#8217;s it.
Dissecting this line of code, we have a <code>lambda</code> function that due to
the use of the <code>map</code> function is applied on every item of the list that
follows. This list can be anything, previously set or not. So, applying
this to a bioinformatics example would be simple (this might not be the
most useful or real-world example ever, but it should give us a primer
to endeavour in more advanced stuff). Let&#8217;s say we want to check
sequence size for all sequences in a FASTA file. Emphasizing code reuse,
we created a new function that returns a list of the actual sequences
only. Without functional programming a loop to read the sequence lengths
would be very simple</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>With a functional programming approach,
we can do it in one line (not a large advantage here)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Yep, not
a lot of gain with this example, as mentioned. The full script would
look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">data</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="k">print</span> <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">seq</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>data</code> is our list of sequences, and inside the <code>map</code> function we have a
<code>lambda</code> that checks for the size of each sequence in that list. Quite
simple.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/30/new-sciview-interview-coming/">New SciView interview coming.</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-30T16:01:13-04:00" pubdate data-updated="true">Oct 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Stay tuned.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/30/python-on-osx-leopard/">Python on OSX Leopard</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-30T09:46:30-04:00" pubdate data-updated="true">Oct 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Rumors that Python wouldn&#8217;t be included in the new Mac OSX Leopard (10.5) were unfounded. Anyway, Apple decided not to include less important languages.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/29/merging-single-or-multiple-sequence-fasta-files/">Merging single (or multiple) sequence FASTA files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-29T15:17:02-04:00" pubdate data-updated="true">Oct 29<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of entries we saw how to split a multiple FASTA file. And
how can we achieve the opposite, merge single (or sometimes multiples)
sequence files in on larger multiple sequence file? In Python is very
simple, but we need to introduce a new module that would allow us to get
the desired result. In most scripts we have seen in previous posts, we
end up reading a single file from the input (in some cases two) and
that&#8217;s not enough in the case of merging multiple files.</p>

<p>Python has a
module called <code>glob</code> that is very powerful for reading contents of a
directory. We can use any type of pattern matching to find only the
files we want, wildcards allowed. Let&#8217;s test glob and see what we can
get. We start with a simple script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">glob</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&#39;\*.txt&#39;</span><span class="p">):</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">i</span>
</span></code></pre></td></tr></table></div></figure>


<p>If we run this
script in a directory (Windows or Linux or Mac) we should see a list of
the filenames that match the patter passed to glob.glob. In this case
every file with a txt extension. This should be enough for us, for now.
Basically we need to create a list of all FASTA files, that match a
certain name pattern, and then take care of the sequences using our
already known FASTA module. To make things more flexible, we should ask
in the command line for a pattern to match, and maybe a output filename.
Our script would look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">glob</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="n">filepattern</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="c">#initialize lists </span>
</span><span class='line'><span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="c">#glob.glob returns a list of files that match the pattern </span>
</span><span class='line'><span class="k">for</span> <span class="nb">file</span> <span class="ow">in</span>  <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">filepattern</span><span class="p">):</span>
</span><span class='line'>  <span class="c">#we read the contents and an instance of the class is returned </span>
</span><span class='line'>  <span class="n">contents</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'>  <span class="c">#a file can contain more than one sequence so we read them in a loop</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">contents</span><span class="p">:</span>
</span><span class='line'>      <span class="n">names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span class='line'>      <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="c">#we print the output </span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">names</span><span class="p">)):</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">seqs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Except for the
glob part, we&#8217;ve already seen the rest in previous entries. Here, the
input <code>filepattern</code> should be enclosed in single or double quotes in
order to be transmitted properly (for instance when you include
wildcards). In the first <code>for</code> loop we access the list of files created
by <code>glob.glob</code> and extract the sequence name(s) and sequence(s) from
each file and parse them to a list which will serve to print the final
output. An example command line would look like
<code>python merge.py '*.fas' merged.fa</code> Of course, there are several ways of
merging FASTA files, but not every time you have Linux&#8217;s <code>cat</code> available.
Next time we will (try) to have fun with Python functional programming.
Stay tuned.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/29/splitting-a-fasta-file-using-awk-no-sed-required-or-do-we-care-about-csplit/">Splitting a FASTA file using awk (no sed required), or do we care about csplit?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-29T11:37:07-04:00" pubdate data-updated="true">Oct 29<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We saw that &#8221;<a href="http://eridanus.net/">top notch bioinformaticians</a>&#8221; use
csplit to split FASTA files, so I decided to
<a href="http://python.genedrift.org/2007/10/10/alternative-methods-to-split-a-fasta-file/">post as many as possible alternatives</a>
to split these files. As csplit, awk is something found with more
frequency in Linux machines than Windows, but it can be installed on
Windows (even Vista) and it runs fine.</p>

<p>Awk is usually shown in parallel
with sed, as both utilities have the main objective of text processing.
I browsed the web searching for a sed/awk solution to split FASTA files
something that I thought could be done in one line. And I found
<a href="http://www.unix.com/shell-programming-scripting/14576-split-file-using-awk.html">it</a>.</p>

<p>With a little tweak I ended up with a one-liner using awk that split the
sequences in multiple files:</p>

<p><code>awk '/^&gt;/{f="fasta."++d} {print &gt; f}'</code></p>

<p>Awk breaks each line of its input into fields that can be checked,
printed or ignored. In the command line above. awk&#8217;s commands are in
between the single quotes. The first command is a regex that matches
initial <code>&gt;</code>. Inside the curly braces we assign a variable <code>f</code> to store
the output filename we want and also have an integer that is incremented
every time we pass through it. The second curly braces are for the
actual output, where we use awk&#8217;s <code>print</code> command and redirect output to
a file named after the variable we created in the first commands in
between curly braces. Finally, we assign an input file, outside the
region delimited by single quotes. And that&#8217;s it. As output we will have
a series of files called fasta each one containing one sequence.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/31/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/29/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
