---
date: '2007-03-02 17:27:24'
layout: post
slug: geneious-one-year-or-almost-later
status: publish
title: 'Geneious: one year (or almost) later'
wordpress_id: '3'
categories:
- Bioinformatics - opinion
---

I intend to cover bioinformatics things in this space. At least mainly. I will try to post here and there stuff about science in general, more specifically biology and genetics. On my first post I reused an old overview of Geneious, and now I will take a fresher look at the program.

Geneious, at the time of this post, is on its version 2.5.3. Since the first release the program has evolved a lot: new functions were added, a cleaner and leaner interface and major bug killing. My first review claimed that Geneious is the Holy Grail of bioinformatics application, and I still think it is. But ...

Well, all stories have a "but", and Geneious' could not be different. Indeed, its new interface is cleaner, buttons were redesigned, some menus received illustrative icons. But there is (or was) no groundbreaking  or innovative feature that wasn't seen in other molecular biology/phylogenetics packages. The strength is still having all features and functions in one cross-platform application. Also the power of having database-like capabilities is another plus: all your sequences, protein structures, references, etc is there in one place, easily accessible and at hand. Sometimes the search is a little bit slow, sometimes the import feature does not work correctly and even sometimes the program indexes entries in a strange fashion.

About new functions, one that calls attention is the collaboration possibilities. Using the free [Jabber](http://www.jabber.org) network it is quite easy to register a new user and connect in the wild. But the niceties stop there. In my lab I am the only using Geneious at the moment, so my options of collaboration are minimal. I have tried setting up a virtual meeting with the developers in order to try the feature, but the fact that Biomatters is on the other side of the world makes a little bit difficult to set up a decent time to connect. And this is another small disadvantage of Geneious: the soft-house is located in New Zealand so any hope of fast support for people in North America are at least thin.

The last releases were followed by an introduction of several small (but very useful) features. More types of files are now supported, small bells and whistles were also added. But (always the but) one of the most promising features seems to be underused. Last time I checked Geneious API does not have the same success as the main program. I have tried creating a plugin to share with its userbase, but not being a Java developer I found more obstacles than solutions and gave up on trying. Geneious website gives links to two (yes two) plugins, and this is the result of almost six months of the API in the wild (if you know any other Geneious plugin please let me know).

And here I have to draw a line. Geneious is "free", but this version gives you limited resources. A pro version will make you at least $166 poorer and will give you a lot of power. So my question is: why would you spend time and effort to learn the API and create a plugin to a mostly paid software? I understand people developing Firefox extensions, Linux open-source packages, Python libraries, etc (they are free as in freedom), but I can't see scientists creating add-ons to Geneious. I might be wrong and would like to be proven wrong, but at least until now I can say that the plug-in feature hasn't taken off.

Maybe the whole plug-in thing wasn't a good move for Geneious and Biomatters. Maybe they should invest more time making Geneious even better and dedicate its programming resources to create the plug-ins themselves and charge for them. Something that MatLab does very well, staying in the scientific software realm only. You might argue that MatLab is also a programming language, has a compiler and it light-years more powerful than Geneious. I agree. But MatLab tries to cover many different fields, while Geneious is "just" a molecular biology/phylogenetics package.  Most of bench biologists are not interested in understanding how computer programs work, they want a black-box-like application that uses an easy and user friendly input and spits out a pre-formatted result, simply and efficiently. Maybe a macro tool or similar script/automating programming in Geneious would be more than enough for most people.

Overall, I still think that Geneious is the Holy Grail of bioinformatics applications, as mentioned above. But like the quest for the Holy Grail, we still have to wait sometime to finally find it completely. Geneious is great and a little bit of fine tuning would certainly improve it. I think Biomatters should focus on some points:

- improving its coverage of biological formats, maybe even making the bold move of standardize the myriad of different file types out there
- improving a little bit more the interface of Geneious, adding some minor features like right-click in some areas
- add a simple and effective scripting/automating feature to the program, instead of relying on the API and the "heavy" Java to create plugins
- add some database-like search and access to the data stored, something like a grep for finding motifs on sequences, or even sequences itself when the data stored is massive

Although excellent, Geneious has room for improvement. We will keep searching for the Holy Grail.
