
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Yep, Santa dropped by before the busy Holiday Season to help with the 17th edition of the Bio::Blogs digest. Welcome and enjoy! New blogs This last &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/28">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/12/01/bioblogs-17-the-one-santa-brough-us/">Bio::Blogs #17 - The one Santa brought us</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-12-01T22:24:48-05:00" pubdate data-updated="true">Dec 1<span>st</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/12/santa.png" alt="santa.png" /></p>

<p>Yep, Santa dropped by before the busy Holiday Season to help with the 17th edition of the Bio::Blogs digest. Welcome and enjoy!</p>

<p><strong>New blogs</strong></p>

<p>This last month we welcomed one interesting blog that will have a lot to add to the community. Alexei Drummond is showing us his bag of goods at the <a href="http://bioinf.cs.auckland.ac.nz/">Computational Biology and Evolution</a>, self described as &#8220;a heady mix of computational science, evolutionary biology and other things that matter&#8221;.</p>

<p>Also Pedro suggested a good look at <a href="http://wrightfisher.wordpress.com/">Thirst for Science</a>, a blog that is surely &#8220;drinking the koolaid&#8221; and has opinions on a myriad of scientific subjects, from RNAs to population genetics.</p>

<p><strong>Not camera shy</strong></p>

<p>Deepak, from bbgm is definitely not camera shy, and he pitches his opinions on <a href="http://mndoci.com/blog/2007/11/04/scientific-intelligence-is-business-intelligence/">business and scientific intelligence</a>.</p>

<blockquote><p>I believe there will be a time when life science companies will begin looking at biological, chemical, etc information as a core enterprise asset, rather than something for the research crowd</p></blockquote>

<p>Watch the videocast too.</p>

<p><strong>Just rambling</strong></p>

<p>Publicly, Pedro rambles this month about the possibilities of predicting <a href="http://pbeltrao.blogspot.com/2007/11/predicting-functional-association-using.html">functional association by mRNA localization</a></p>

<blockquote><p>As expected, more specific mRNA localization terms (localizations 30) are more informative for prediction of functional association since fewer terms are required to obtain the same or higher likelihood of interaction.</p></blockquote>

<p>And he also rambles about the hot hyped topic of <a href="http://pbeltrao.blogspot.com/2007/11/linking-out-personalized-medicine.html">personalized medicine</a>. And he is concerned</p>

<blockquote><p>that all the attention the genomics side of personalized medicine will distort the relative importance of nature versus nurture.</p></blockquote>

<p>And he is not <a href="http://thegenesherpa.blogspot.com/2007/11/cottage-industry-cottage-cheese-is-more.html">the</a> <a href="http://www.palidwor.com/blog/?p=63">only</a> <a href="http://scienceroll.com/2007/11/10/navigenics-23andme-and-helix-health-the-review/">one</a> <a href="http://scienceroll.com/2007/11/18/23andme-the-re-review/">posting</a> <a href="http://pimm.wordpress.com/2007/11/18/23andmes-global-similarity-map-as-mission-connecting-all-people-on-the-dna-level/">about</a> <a href="http://mndoci.com/blog/2007/11/19/your-personal-health-social-networking-x0-revisited-and-some-killer-video/">it</a>.</p>

<p><strong>Frilas</strong></p>

<p>Frilas is the way freelancers are known in Brazilian newspapers, and Pavel, from Freelancing Science, blogs about the possibility of using <a href="http://freesci.wordpress.com/2007/11/27/software-portability-and-virtual-appliances/">Virtual Machines to distribute software</a>.</p>

<blockquote><p>So creating a virtual appliance for, let’s say, BLAST, would mean installing a fresh copy of our favourite linux under VMWare Server with nothing more than necessary libraries, copy of BLAST executables and possibly a web interface. Voilla. Virtual appliance for BLAST, anybody?</p></blockquote>

<p><strong>Say wha?</strong></p>

<p><a href="http://www.cs.man.ac.uk/~hulld/">Duncan</a> is still worried about the <a href="http://www.nodalpoint.org/2007/11/30/burn_semantic_web_burn">Semantic Web</a>. Nuf said.</p>

<p><strong>Is there a new Omic?</strong></p>

<p>Keith Robinson, from Omics! Omics!, wants <a href="http://omicsomics.blogspot.com/2007/11/incredible-shrinking-human-genome.html">to shrink us all</a>. Maybe it would be good to meet the aunts during Xmas. HE wonders about who or how they are shrinking our (mine, yours, theirs) genomes.</p>

<blockquote><p>The genome is finished &#8211; but not so finished. Bits and pieces are still getting polished up, and while they are generally dull and monotonous a gene or two might still hide there</p></blockquote>

<p><strong>Already tomorrow</strong></p>

<p>Yep, it is already tomorrow there (no matter what day is here!) and after posting about his perl variables  and the election, Neil, who is always doing something rather desperate, tells us about interesting <a href="http://nsaunders.wordpress.com/2007/11/21/stem-cells-and-science-blogs/">blogs that cover some stem cell subjects</a>. Check it out.</p>

<p><strong>My sardine</strong></p>

<p>This month I posted the last SciView interview in 2007, you might want to take a look. Dr Roderic Page was interviewed and he just said</p>

<blockquote><p>My sense is that large projects don’t always scale well. You don’t get 10x more out of a project simply because it has 10x as much money. You get administrative layers, and a lot of politics. If the big project is really lots of little ones, then it may well succeed. Barcoding is really lots of little projects — everybody grabs their favourite beastie and sequences it.</p></blockquote>

<p>among other things. <a href="http://blindscientist.genedrift.org/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">Take a look</a>.</p>

<p><strong>Happy Holidays</strong></p>

<p>Bio::Blogs #17 is very short. Everybody is busy finishing their papers and the shopping list. It will be back next month, maybe here, maybe somewhere else if you want to host it. On the meantime, please to not give <a href="http://blindscientist.genedrift.org/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">this </a>to any of your good friends.</p>

<p>Happy Holidays!!!!</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/30/clc-products-who-uses-them/">CLC products: who uses them?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-30T13:41:10-05:00" pubdate data-updated="true">Nov 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have a Google alerts with bioinformatics as its key word and this morning I received an email that points to <a href="http://www.medicalnewstoday.com/articles/90228.php">this</a> article on Medical News today. I know it has a huge amount to PR in the article, but the first phrase is symbolic:</p>

<p><em>&#8220;CLC bio, the world&#8217;s leading bioinformatics solution provider, announced the release of CLC Bioinformatics Database.&#8221;</em></p>

<p>So, I need to ask: have you ever used a CLC product or know someone that uses it? Leave a comment if so, because I never met someone that uses their products, and they are still the &#8220;world&#8217;s leading bioinformatics solution provider&#8221;.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/28/wow/">Wow!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-28T19:03:12-05:00" pubdate data-updated="true">Nov 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have been using <a href="http://www.repeatmasker.org/">RepeatMasker</a> to mask common repeats of my sequences. I have just upgraded it to the version open-3.1.8 from open-3.1.6 and the speed gain was incredible. The same 10 Mb sequence that was taking more than a day to run in 3.1.6 took a couple of minutes on the newest. Nice!</p>

<p><em>I need to correct myself: I was using version open-3.1.6, not 7</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/28/geneious-and-other-software/">Geneious and &#8220;specialized&#8221; software, featuring Dendroscope</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-28T13:59:36-05:00" pubdate data-updated="true">Nov 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last week a new phylogenetic tree viewing program was released/published. Called <a href="http://www-ab.informatik.uni-tuebingen.de/software/dendroscope/welcome.html">Dendroscope</a> it is a very nice piece of software that has lots of useful features to view, edit and manipulate phylogenetic trees.</p>

<p>Its release/publication coincided  with a period in the lab when two of the graduate students were obtaining new results, especially in the form of phylogenetic trees. There was some debate whether TreeView should be used, but we couldn&#8217;t compile for all Linux boxes we have, due to differences in their distro version. So, at first we checked Geneious, which has been reviewed here, and it is a do-it-all phyloegenetic/molecular biology/name-it software. Geneious has some very nice features, but I cannot call its tree viewing option excellent. Let&#8217;s say good defines it more than enough.</p>

<p>One particular nagging thing included in Geneious&#8217; latest releases were some extra features to expand/collapse branches, which are circles drawn over the tree&#8217;s nodes. One cannot turn off these circle, they are always there and sometimes with different diameters. When I first saw them I though it was showing some kind of support for the node but I was wrong. After some fidgeting we discovered that circles can be turned off, but only when you are exporting/printing the tree.</p>

<p>Bringing back Dendroscope. Intuititve interface, it delivers what you want and more than you expect for a tree viewing program. After one hour of playing one of the graduate students was already a pro-user of the software, showing everyone what can be accomplished with Dendroscope. And yes, there are no deceiving circles on the nodes. As Geneious, it is a Java based application, so easy to port to different systems. But differently it does one thing and does it better than the &#8220;competition&#8221;. That makes me question myself: why should you leave your &#8220;specialized&#8221; applications in favour of Geneious when the end product is below your expectation?</p>

<p>Maybe because of the competition Geneious got away of its almost sure path to be the &#8220;holy grail&#8221; of phylogenetics/molecular biology software. Trying to embrace every aspect of research at once might take its toll on the simpler details that are still not fixed/changed in the program. When Geneious was released it had the right amount of features and, as any other software, a little bit buggy but it delivered what you wanted. The program grew considerably since its first inception and it can do mostly of the common tasks of a small-medium biology lab requires in a day. Still, why would you want to trade your reliable do-one-thing-do-it-better application in favour of a do-it-all-do-it-so-so program that might have tons of features you will never use? Yep, I find a lot of things in MS Word that I don&#8217;t need, that&#8217;s why I use e, TextMate, Notepad++. And for tree viewing I will dump my always-loved TreeView and adopt the new kid on the block, Dendroscope.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/27/bioblogs-17-call-for-submissions/">Bio::Blogs #17 - call for submissions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-27T13:37:22-05:00" pubdate data-updated="true">Nov 27<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The next Bio::Blogs, currently at number 17, will be proudly hosted here. There is plenty of time for submissions, which should be sent directly to me, at nuin at genedrift dot org, or to the usual Bio::Blogs email address, at bioblogs at gmail dot com. Deadline is December 1st.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/26/e2809ccuttinge2809d-chromosomes-with-python-a-fast-approach-take-two-including-precision/">“Cutting” chromosomes with Python: a fast approach, take two, including precision</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-26T16:41:57-05:00" pubdate data-updated="true">Nov 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the last post we saw how to &#8220;cut&#8221;/extract a segment of a chromosome
quickly in Python. But our last code had no precision in the
cutting/extracting, because it didn&#8217;t take in account the line size so
it would extract more information than actually requested. We need to
address precision them. First take into account the size of each line,
the number of nucleotides on each line. Then use this this information
to remove unwanted bases. Most of the changes we need to make on our
script is outside the main loop that cuts/extracts information. We need
to modify the post-processing of the sequence we nust read from the
chromosome. Also, instead of storing the sequence in a <code>string</code> it is
easier to use a <code>list</code> (in fact it is just a matter of personal taste,
and maybe a little be of being Pythonic). In the end our script would
look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">size</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">linesize</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">segment</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span><span class='line'>    <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;&gt;&#39;</span><span class="p">):</span>
</span><span class='line'>        <span class="n">size</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>    <span class="k">else</span><span class="p">:</span>
</span><span class='line'>        <span class="n">name</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">size</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">size</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="o">+</span><span class="n">linesize</span><span class="p">:</span>
</span><span class='line'>            <span class="n">segment</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
</span><span class='line'>            <span class="n">linesize</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
</span><span class='line'>
</span><span class='line'><span class="n">startline</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span><span class="o">/</span><span class="n">linesize</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
</span><span class='line'><span class="n">endline</span> <span class="o">=</span> <span class="p">(</span><span class="n">end</span><span class="o">/</span><span class="n">linesize</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>
</span><span class='line'>
</span><span class='line'><span class="k">if</span> <span class="ow">not</span> <span class="n">start</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">end</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>    <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">startline</span><span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">start</span><span class="p">:]</span>
</span><span class='line'>    <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">endline</span><span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">end</span><span class="p">:]</span>
</span><span class='line'><span class="k">elif</span> <span class="ow">not</span> <span class="n">start</span> <span class="o">%</span><span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>    <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">startline</span><span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">start</span><span class="p">:]</span>
</span><span class='line'><span class="k">elif</span> <span class="ow">not</span> <span class="n">end</span> <span class="o">%</span> <span class="n">linesize</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>    <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">segment</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">endline</span><span class="o">*</span><span class="n">linesize</span><span class="o">-</span><span class="n">end</span><span class="p">:]</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">name</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">segment</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that we added a new
variable <code>linesize</code>, to store the size of the sequence line. The loop is
almost identical to the one before. The only changes are the line that
checks for each line size (can be optimized somewhere else and removed)
and instead of a segment string, it is a list.</p>

<p>The bulk of
changes/additions are after the loop. We first need to calculate the
initial and final lines of our cut/extracted segment. We just divide the
value of the start and end positions entered by the size of the line and
add 1 to them in order to get a correct line in case we the entered
values are not multiple of the line size. Dividing integers in Python
will return us an integer. After this checking we need to add some
processing conditionals to make sure we are returning the exact segment
requested. The <code>%</code> operator in Python return the remainder of a
division, even if the two values are integers. We need to check that the
entered starting and ending positions are or aren&#8217;t multiple of the line
size. If one or both are multiples we don&#8217;t need to process the segment.
On the other hand some processing is needed. Clearly we only need to cut
extra nucleotides from the first and last returned lines. That&#8217;s why the
three clauses: one for both values not being multiple of line size, one
for just the start and another for the end position. And to remove
unwanted nucleotides we basically multiply the number of the
starting/ending line by the size of each line and subtract the actual
start/end position.</p>

<p>This should give us the number of unwanted
nucleotides in those two lines. And the fact that we stored the segment
as a <code>list</code> makes our life easier as we access position 0 and -1 in our
list. On the final output, we print the name and join the list elements
by adding a new line between them.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/22/what-happened/">What happened?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-22T10:10:42-05:00" pubdate data-updated="true">Nov 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I thought that <a href="http://www.soccerlounge.tv/2007/05/epl_best_in_the_world.html">EPL</a> <a href="http://www.youtube.com/watch?v=BqWQSOee_LQ&amp;feature=related">was</a> <a href="http://www.tv.com/english-premier-league/show/70376/reviews.html?review_id=423803&amp;flag=">the</a> <a href="http://www.sportsfilter.com/comments.cfm/4413">best</a> <a href="http://goliath.ecnext.com/coms2/gi_0199-4986247/Americans-Vote-Barclays-English-Premier.html">league in the world</a>. Being the best, England should have the best players, thus the best national team. I was wrong. They will always have 2012 to try.</p>

<p>I think everything can be summarized with this article: &#8221; Americans Vote Barclays English Premier League Best Soccer League in the World&#8221;. Oh, irony.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/20/cutting-chromosomes-with-python-a-fast-approach-take-one/">&#8220;Cutting&#8221; chromosomes with Python: a fast approach, take one</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-20T16:24:11-05:00" pubdate data-updated="true">Nov 20<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of posts we started with some functional programming aspects
of Python. I was away last week and couldn&#8217;t create anything related to
FP in the meantime, so I decided to post about a quick way to &#8220;cut&#8221;, or
extract segments, from chromosomes stored as FASTA files. This is a
subject that I have been investigating for some time, as I needed a tool
that would get some segment from a chromosome for further analysis. I
have a short C++ code that does the trick and is really fast, but I
wanted to find a Pythonic way of doing it, and quick. I have been trying
different methods but no one were fast. Chromosomes are usually large,
and stored in large files, so reading line by line and using <code>readlines</code>
were extremely slow.</p>

<p>Inspired by a
<a href="http://www.goldb.org/goldblog/2007/11/07/PythonProcessingLargeTextFilesOneLineAtATime.aspx">post</a>
from Corey Goldberg, I decide to test the method explained there, with
wonderful results. Basically what we need to do is to use an iterator
protocol on the file object</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;foo.txt&#39;</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span></code></pre></td></tr></table></div></figure>


<p>This processes the human chromosome 1
(the largest) in less than 2 seconds. Our first take will be how to
extract a segment, but not precisely (that we will see in a future
post). Basically we need a start and an end point and a FASTA file. All
three would be used as parameters and a loop should take care of the
rest. Our script should look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">size</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">segment</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">):</span>
</span><span class='line'>    <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;&gt;&#39;</span><span class="p">):</span>
</span><span class='line'>       <span class="n">size</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>    <span class="k">else</span><span class="p">:</span>
</span><span class='line'>        <span class="n">name</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">size</span> <span class="o">&gt;=</span> <span class="n">start</span> <span class="ow">and</span> <span class="n">size</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="p">:</span>
</span><span class='line'>        <span class="n">segment</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">name</span><span class="p">,</span> <span class="n">segment</span>
</span></code></pre></td></tr></table></div></figure>


<p>We start getting the parameters from
the command line, assigning a couple of variables and we start the loop.
We open the file and use the file object iterator to check each line. We
check each line for a FASTA sequence title. If found we store tht title
in the variable <code>name</code>, and if not we check the size of the line and
increment the variable that will guide us on the file position. We use
<code>size</code> as a line by line increment and if it is between <code>start</code> and
<code>end</code> we store the read line into the <code>segment</code> variable. That&#8217;s it. A
few drawbacks with this script: it will fail on FASTA files where the
sequence is stored in one line, there is no precision to get the actual
start and end points and it can only run on unique sequence files. On
the next post we will address some of these concerns.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/19/sciview-part-6-interview-with-dr-roderic-page/">SciView, part 6: interview with Dr Roderic Page</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-19T13:29:47-05:00" pubdate data-updated="true">Nov 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><h2>This edition of SciView features Dr Roderic Page from the University of Glasgow. Among other things, Dr Page was the editor of the Current Protocols in Bioinformatics, is the Editor in Chief of Systematic Biology, is an avid <a href="http://taxonomy.zoology.gla.ac.uk/rod/rod.html">blogger</a> and is the developer of TreeView, a phylogenetic tree visualization software. I had the opportunity to ask him about his bioinformatics work and his views on science. I would like to thank Dr Page. </h2>

<p><em>In your blog (<a href="http://iphylo.blogspot.com/">iPhylo</a>) an usual topic is the management and flow of information in Biology/Phylogenetics. In your opinion what are the main  hurdles of this subject in science? With the ever increasing amount of  generated information, databases, scientific publications, do you think that we will reach a stage where it won&#8217;t be possible to stay afloat in a sea of data?</em></p>

<p>RP- Well I think we are well past the floating stage! What I think would be cool would be a service like Google News that automatically aggregates<br/>
stories (for &#8220;stories&#8221; read &#8220;papers and data)&#8221; and summarises the Zeitgeist. This wouldn&#8217;t be overly hard to do. <a href="http://www.ubio.org/index.php?pagename=ubioRSS">uBio</a> has a service that classifies RSS feeds from journals by taxon. Imagine taking that, together with newly added GenBank sequences, and clustering the feeds by various topics. Then display that using a treemap (analogous to  Marcos Weskamp&#8217;s <a href="http://marumushi.com/apps/newsmap/">newsmap</a>). You could get an &#8220;at a glance&#8221; view of what is going on.</p>

<p>This is a bit of a half-baked idea, but it relies on things that are becoming more and more available: stable identifiers, and services that<br/>
provide data (and/or links between data). I think these are the two key things that will help us manage the flood of data. Without identifiers we will struggle to realise when we are talking about the same thing (witness the frustration with <a href="http://www.nodalpoint.org/2006/12/15/buggotea_redundant_links_in_connotea">redundant links in Connotea</a>, even though much of Connotea&#8217;s core functionality relies on stable identifiers), and services give us means to extract and manipulate data. We are slowly moving away from relying sole on web links, or thinking that merely &#8220;putting something on the web&#8221; is good enough.</p>

<p>Lastly, I&#8217;ve often argued that in some fields, such as phylogenetics, the days of the journal are numbered. I think what we really need is a database that can generate &#8220;reports&#8221; that function as a journal article (i.e., the contributor gets a citable paper), but where the submission process is primarily one of submitting data and protocols. Many papers in phylogenetics are of the sort &#8220;I think group x is way kewl, I grabbed gene y, used program z, and here is my tree&#8221;. Why not submit these straight to a database, instead of what we have now, which is the original data (e.g., alignments) and trees are usually lost? Databases such as TreeBASE that largely collect data after publication are sparsely populated, hence much valuable data and results are locked up in PDFs on journal web sites. Every month a weighty copy of Molecular Phylogenetics and Evolution lands on my desk, and the wasted data and trees makes me want to cry.</p>

<p><em>This is an identical question I asked Joe Felsenstein (with a twist): Lately two big projects related to phylogenetics are getting some steam: the Tree of Life and the DNA barcoding. I am more aware of the latter, due to the fact that I attended a presentation about it and through a friend that woks on the project. I think that the goals of the DNA barcoding are great for the field, wet-lab and theoretical scientists. What is your opinion about this type of project that tries to address all subjects in a big umbrella? Wouldn&#8217;t the support of a larger number of small projects be better? How would this type of projects can contribute to the information management and flow?</em></p>

<p>RP - Barcoding is the simplest of the two &#8220;big projects&#8221; &#8211; it&#8217;s a simple idea (and one that has been employed by microbiologists for decades), applied to things that we can see with the naked eye. As with all of these things, there&#8217;s been much hype, controversy, and heated debate. My own sense is that much of this is dying down, and certainly most articles in journals like Systematic Biology are investigating the performance of barcoding methods, rather than engaging in philosophical or political debate. I think barcoding will generate a lot of data, much of which will be useful, probably in ways we don&#8217;t expect. We will have standardised genetic samples taken at a known place at a known time from organisms whose identity is either known or can be inferred. I&#8217;m struggling to see why this isn&#8217;t a good thing.</p>

<p>My sense is that large projects don&#8217;t always scale well. You don&#8217;t get 10x more out of a project simply because it has 10x as much money. You get administrative layers, and a lot of politics. If the big project is really lots of little ones, then it may well succeed. Barcoding is really lots of little projects &#8211; everybody grabs their favourite beastie and sequences it. There aren&#8217;t any great technological or intellectual challenges. I think the real problem is when the project has big goals, articulates a grand vision, then struggles to deliver. I suspect <a href="http://www.phylo.org">CIPRES</a> is an example of a large project that hasn&#8217;t delivered quite as much as it promised. My biggest fear is that the <a href="http://eol.org">Encyclopedia of Life</a> may end up in this category.</p>

<p><em>You have three different blogs, iPhylo, iSpecies and bioGUID, but none of them are, what I can call, Open Science. Would you consider endeavouring in this new trend and opening up your scientific activities?</em></p>

<p>RP - You missed one blog &#8211; <a href="http://semant.blogspot.com">http://semant.blogspot.com</a>. I guess it depends what you mean by &#8220;Open Science.&#8221; These blogs are pretty open about the success and failure of various projects, and I don&#8217;t shy away from thinking out aloud. In some cases suggestions I&#8217;ve made in the blogs have been taken up and explored further by others. The Pygmybrowse widget to display large trees in a small space is a recent example.</p>

<p>For some time I&#8217;ve done things like post final reports on research grants online, as well as manuscripts that didn&#8217;t make it into print (these are now available in Nature Precedings).</p>

<p>Not all the code I write is available. Some projects, like the <a href="http://taxonomicsearch.sourceforge.net/">Taxonomic Search Engine</a> and <a href="http://myphpbib.sourceforge.net/">MyPHPBib</a> are hosted on SourceForge, and <a href="http://darwin.zoology.gla.ac.uk/~rpage/treeviewx/">TreeView X</a> is also open source (code available from my site, and as part of the Debian repository). Other stuff is sometimes so experimental and fluid that it would be more trouble than it is worth to package it nicely. Some legacy code uses libraries that are not Open Source and hence can&#8217;t be distributed. That said, I&#8217;m looking at hosting current projects (such as <a href="http://bioguid.info">http://bioguid.info</a>) on Google Code. Since 2005 anything which I&#8217;ve published as first author has been Open Access, and any software or data described in those papers is Open Source.</p>

<p><em>Sometime ago, in the Nature Network&#8217;s Bioinformatics group someone was asking about good protocols in the field. Can you tell us about the Current Protocols in Bioinformatics, which is was edited by you? How content is added to it and how frequent are the updates?</em></p>

<p>RP - I&#8217;m no longer involved with <a href="http://www.currentprotocols.com/WileyCDA/CPTitle/isbn-0471250937.html">Current Protocols</a>. It was a great project to be involved with as it opened my eyes to the breadth of the field, and I got free trips to the US to meet with the other editors (I&#8217;m easily bought). Eventually I just had too many other commitments (it overlapped with being editor of <a href="http://systbiol.org">Systematic Biology</a>), so I left the board.</p>

<p>Content is commissioned by the editorial board, who each handle 2-3 topic areas. Updates come out every three months. If you have the print version (currently two massive ring binders), then you have the fun task of inserting new sections and removing old content.</p>

<p><em>What do you think is the importance of protocols in Bioinformatics? Isn&#8217;t bioinformatics a field more dynamic than cytogenetics and molecular biology for instance?</em></p>

<p>RP - This is an interesting point. I think there are some protocols which are  widely used and are reasonably stable over time (BLAST searching, for example), and others which either are short lived (the software keeps changing, or the question being addressed by the protocol looses its relevance). In some ways Current Protocols is an old fashioned way of providing information on how to do things. A wiki may be more appropriate, certainly it could be more dynamic, and ideally free. However, one could ask whether a wiki would have acquired the same degree of authority as the Current Protocols, which could assemble a board of people with broad expertise and experience, pay contributors for their work, provide copy editing and other editorial assistance, plus give a commitment to maintain and update the publication.</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. TreeView(X) is a very important application in Phylogenetics and it is basically interface only, as the main focus is to display phylogenetic trees. Usually, a good percentage of messages sent to Paup&#8217;s discussion list are somewhat related to TreeView. What in your opinion is the importance of a graphical interface for a biologist?</em></p>

<p>RP - Well, unless you only use Lynx as your web browser, and vi to edit text, then I think the importance of the graphical interface is pretty much self evident! I think graphical user interfaces are desirable in two sets of circumstances: (1) the software is infrequently used, or (2) the user is not a &#8220;power user&#8221;. When I was at Oxford a colleague told me that he liked graphical interfaces because, after weeks in the field, the last thing he needed was to have to learn (or relearn) some garbled syntax in order to analyse his hard-won data. I&#8217;ve also seen the difference a graphical interface can make when teaching students &#8211; you want to keep the mechanics of using the software out of the way and let them focus on what they are actually doing (and why they are doing it).</p>

<p>Of course, graphical interfaces can get in the way as soon as you want to do something not originally thought of by the person who wrote the program. This is one reason why David Swofford&#8217;s PAUP* is a model of flexible software. There is an elegant GUI (sadly fully available on Mac Classic only, although the Windows version has some of the essential features such as the spinning cursor and the PAUP Monaco font, both things I ported across a decade ago), and the command-line interface for batch processing.</p>

<p><em>Continuing on TreeView(X), your latest version is pure wxWidgets. Why did you choose this framework to develop TreeViewX? Why not QT or Java or another GUI framework? How TreeViewX is different to the initial version of TreeView?</em></p>

<p>The short answer is that I have a lot of code written in C++, and I haven&#8217;t the time nor inclination to learn Java. When Java come along in the early &#8216;90&#8217;s it&#8217;s big selling point was cross-platform portability. Well, it didn&#8217;t support the Mac terribly well, and I had my own C++ class libraries that worked natively on Windows and Macs, so there was no incentive to use it. Until fairly recently most Java applications I&#8217;ve seen are sluggish and/or poorly designed. When the time came to look at supporting Linux (around 2000), Qt wasn&#8217;t an option because it wasn&#8217;t free on Windows, and didn&#8217;t support the Mac. So, wxWidgets was the logical choice, although it took a while for the Mac version to be usable.</p>

<p>TreeView X differs from TreeView in being Open Source, and available for Windows, Mac OS X, Linux, and other forms of Unix. It lacks some of the nicer features of the original TreeView (which only runs on Windows and Macs), mainly because I haven&#8217;t had time to add them, and my own work has focussed more on web application development. Like many people who program, the software I write addresses problems I&#8217;m interested in at the time, so if my interests change, the software may languish.</p>

<p><em>In your opinion what is your greatest contribution to science. A software, a publication&#8230;</em></p>

<p>RP - I don&#8217;t think there&#8217;s any easy way to answer that. Let me respond by breaking it down into three categories. What contribution was the most fun, what was the most satisfying, and what do I think people will associate with me?</p>

<p>The most fun I&#8217;ve had is blogging, or to be more precise, trying to make connections between different areas such as bioinformatics, biodiversity informatics, systematics, digital libraries, and the Semantic Web. My hope is that readers of blogs like iPhylo will get something out of it, but ultimately I don&#8217;t particularly care because, hell, I&#8217;m having a ball.</p>

<p>In terms of satisfaction, I enjoyed working on mincut supertrees. This was a change of topic (aggregating small phylogenies into one large &#8220;supertree&#8221;), required programming algorithms (rather than, say, graphical interfaces), and was my first introduction to the world of publishing in the computer science literature (the joys of LaTeX, and the requirement of having your paper accepted for publication BEFORE you could give a talk). It was also one of the best refereeing experiences I&#8217;ve had, because one diligent referee read my source code and pointed out that what I said I was doing in the manuscript and what my code actually did were not quite the same thing (I guess you&#8217;d call this an example of &#8220;Open Science&#8221;).</p>

<p>But, if your question is what will people associate with me, then I think that is the work on &#8220;historical associations&#8221;, namely genes and species, hosts and parasites, and areas and phylogenies. Making the case that these are all equivalent questions, and could be approached using the same algorithm, is probably the thing I could point to as being the thing that I have have done that was &#8220;significant&#8221;.</p>

<hr />

<p>I would like to thank Dr Page for taking the time to answer my questions.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/11/06/goodbye-john/">Goodbye John</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-11-06T14:07:40-05:00" pubdate data-updated="true">Nov 6<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I discovered John Rebus in 2002, on a trip to Canada. It was almost winter and a cool place to hang with my wife was the Eaton Centre and the Chapters (or Indigo) inside it. There I discovered <a href="http://www.amazon.com/gp/redirect.html?ie=UTF8&amp;location=http%3A%2F%2Fwww.amazon.com%2FFalls-Ian-Rankin%2Fdp%2F0752877259%3Fie%3DUTF8%26s%3Dbooks%26qid%3D1194375773%26sr%3D8-3&amp;tag=genedrift-20&amp;linkCode=ur2&amp;camp=1789&amp;creative=9325">The Falls</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=ur2&amp;o=1" alt="" /> there, attracted by the stylish cover and for the good reviews on the back. I decided to give it a try and after that I was hooked. Read all the other books by <a href="http://www.amazon.com/gp/search?ie=UTF8&amp;keywords=ian%20rankin&amp;tag=genedrift-20&amp;index=blended&amp;linkCode=ur2&amp;camp=1789&amp;creative=9325">Ian Rankin</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=ur2&amp;o=1" alt="" />.</p>

<p>Sadly he decided to retire John Rebus in his latest book. I will miss Edinburgh (for a while maybe) and I will miss John and Siobham. Thanks for the ride, it was fun.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/29/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/27/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
