
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Last week I registered to use a large online database of biological data. So far, so good, I started checking out what I was able to do, how to get &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/28">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/31/internet-explorer-7-are-you-sure/">Internet Explorer 7? Are you sure?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-31T16:50:12-05:00" pubdate data-updated="true">Jan 31<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last week I registered to use a large online database of biological data. So far, so good, I started checking out what I was able to do, how to get data, the whole nine yards. Until I got to the main page, listing most of the database entries. I noticed that Firefox wasn&#8217;t showing the vertical  scrollbar, so, evidently, I wasn&#8217;t able to scroll down and up the page to see the data.</p>

<p>Problems like this happen when the scrollbar is attached to a <code>div</code> instead of being set for the whole page. As a diligent user I dug for the web admin email address and sent a short message with the possible error, including a couple of screenshots, one from &#8220;buggy&#8221; Firefox and another from Konqueror (which displayed the page with no &#8220;errors&#8221;). I also mentioned that Firefox under Linux and Windows had problems showing the page.</p>

<p>A couple of days later I receive a polite reply explaining that Firefox isn&#8217;t supported to display their pages. Sorry!? I cleared my eyes and then I found a link to the requirements page where I saw a second time that they suggest you to use Internet Explorer 6 or 7. Next thing I did was to check a GPS just to make sure I wasn&#8217;t in Punxsutawney.</p>

<p>My humble internet &#8220;portal&#8221; served by the good guys at <a href="http://geekisp.com">GeekISP</a> are accessed by the following browsers
Firefox     49%
Internet Explorer   32%
Safari  10%
Opera   3%</p>

<p>I heard a rumor that it is 2008.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/31/we-salute-you/">We salute you &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-31T10:29:09-05:00" pubdate data-updated="true">Jan 31<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>That&#8217;s for all the bioinformatics developers that distribute executables with spaces in their names. Thanks.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/30/things-you-can-only-learn-from-science-the-journal-not-science-itself/">Things you can only learn from Science (the journal, not science itself)</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-30T15:28:55-05:00" pubdate data-updated="true">Jan 30<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>From this <a href="http://wrightfisher.wordpress.com/2008/01/25/20/">post</a> I became aware of the paper &#8220;Alignment Uncertainty and Genomic Analysis&#8221; by Wong <em>et al.</em>.</p>

<p>We learn valuable lessons, some evident and some implied. I am covering only the evident here, the implied I leave for the imagination of who&#8217;s reading the paper. To the evident lessons (conclusions):</p>

<ul>
<li><p>Alignments are important for genomics studies</p></li>
<li><p>Some genes are more difficult to align than others</p></li>
<li><p>Some alignment programs align some genes better than others, while other alignment programs align other genes better than some</p></li>
<li><p>Most genes in an organism are under the same DNA model of substitution, hence they all have identical selective pressure, drift, migration, recombination and mutation rates</p></li>
</ul>


<p>There are things you can only learn from Science.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/30/tnt-a-comprehensive-review/">TNT: a(n) (almost) comprehensive review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-30T14:41:52-05:00" pubdate data-updated="true">Jan 30<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As promised I&#8217;m writing about the free version of TNT. So far I was able to test the Linux version which does not have a graphical interface. I am planning to download a Windows version and test it, but not until next week. So far here are my impressions.</p>

<p>When you start TNT you are greeted by its own prompt, in identical fashion to PAUP and MrBayes. As a regular user (or most of the regular users) I fired up TNT without reading the manual, so my first attempt was to write help at the prompt and the prompt answered by changing itself from &#8220;>tnt*&#8221; to &#8220;>help&#8221;. One point against TNT, as either PAUP of MrBayes at least display a list of possible commands. Feeling helpless I decided to access the manual.</p>

<p><strong>The Manual</strong></p>

<p>Oh, my eyes!!! Whoever chose the background colour for the html manual might have an agreement with some optometrist. I cannot read a paragraph without averting my eyes for a &#8220;refresh&#8221; (if you are interested search for bgcolor=aqua). Anyway, the introductory part of the manual tells you to do exactly what I did, by entering &#8220;help&#8221; at the prompt. Second thing it tells you is that if you don&#8217;t have a Windows version it is not possible to convert your files (i.e. fasta, phylip, nexus) to TNT&#8217;s own format. Usual, it follows the rule, <em>one application</em> - <em>one format</em>, they just forgot to include <em>one converter</em>. Just adding insult to injure they suggest that you use a text editor, anyone, and check their example files. Yeah, right. Basically TNT read Hennig86 matrices. Just get them from your 5 1/4&#8221; floppies.</p>

<blockquote><p>update: there is a way to display the commands at the prompt. &#8220;help&#8221; alone does not work, it has to be followed by a &#8220;;&#8221;, but the manual says &#8220;help &#8221;</p>

<p>update II: according to the manual TNT is able to read NEXUS files and can even include TNT commands in a block. Good.</p></blockquote>

<p><strong>Using TNT</strong></p>

<p>Apparently TNT has a good selection of commands, from what can be seen from its help. It covers most of the things you want to study by using parsimony. It is mentioned in the manual that TNT does not do maximum likelihood (don&#8217;t say!) or bayesian analysis (geez!), but they will incorporate some ML algorithms, even though rudimentary (I bet!).</p>

<p>So, from my shell I started TNT and as I would do in the PAUP CLI version just put the filename as a parameter</p>

<p>./tnt primate-mtDNA.nex (if you really noticed I was using one of PAUP&#8217;s example files)</p>

<p>No dice. So I dig deeper in the manual and find a section dedicated to us poor Linux and OS X users</p>

<blockquote><p>Linux and Mac OS X versions of TNT have a couple differences with the standard versions.  First, those versions can take a comma ( , ) instead of a semicolon ( ; ).  This is because the shells there use semicolons to separate commands, and giving arguments to TNT from the command line becomes difficult. Second difference is that those versions can run in the background.  If you want to start TNT in the background, use bground as first argument, then give the normal arguments you would give the program (do not activate reports and do not change the silent status, but you can set report+ to report every certain number of trees, replications, or seconds, so that you can monitor progress by inspecting the output file; make sure you do have output files, otherwise the program would run and exit without doing anything!), and end with the usual &#8216;&amp;&#8217;.  TNT will not write to the console, or expect console input (at least if no errors occur). It will write its results to a file and exit smoothly (exit code is 0; if the program runs out of commands while in the background it exits with code 1). Should some error occur during the search, the program writes a message to the ouptut file, and exits with code 255.  Try</p></blockquote>

<pre><code> ./tnt bground p zilla.tnt, echo=, log peep, rep+1, mu10=ho3, le, ne, quit, &amp;
</code></pre>

<p>and do &#8220;tail â€“f peep&#8221; to monitor search progress.</p>

<p>Very simple and straightforward. But at least I got a clue on how to open a file or use it as a parameter. So I repeat my first command line with the parameter p and then the filename</p>

<p>./tnt p primate-mtDNA.nex</p>

<p>Success &#8230; no not so fast. An error is reported:</p>

<p>PISH (Phylogenetic Inference SHell)</p>

<p>Reading from primate-mtDNA.nex
Reading file primate-mtDNA.nex as NEXUS</p>

<blockquote><p>Data from:</p>

<pre><code>    Hayasaka, K., T. Gojobori, and S. Horai. 1988. Molecular phylogeny
            and evolution of primate mitochondrial DNA. Mol. Biol. Evol.
            5:626-644.
</code></pre></blockquote>

<p>  &#8230; Skipping taxa block &#8230;
  &#8230; Skipping characters block &#8230;
Error reading NEXUS file: expected cost
Error reading primate-mtDNA.nex
g = coding;    usertype 2_1 = 4  [weights transversions 2 times transitions</p>

<p>OK, fine. But what&#8217;s PISH? No clue. I edited the file and removed the assumptions block, as TNT seems buggy to read square-bracket enclosed comments on PAUP blocks. I tried again and it worked but skipped both the characters and taxa blocks. Time to use one TNT&#8217;s example files. So I chose the more evident <em>example.tnt</em>. I repeated the above command now using the tnt file as input. Everything works fine, data is read, stored and ready to go. Basically you have a number of commands that you can use to obtain your trees.  I used only <strong>mult</strong> which does multiple random addition sequences plus TBR (RAS+TBR) to find the most parsimonious tree(s). The program is fast for a morphology matrix of 84 characters and 112 taxa. In less than a second the tree is calculated.</p>

<p>But there is a major problem visualizing the tree(s): the so called PISH (Phylogenetic Inference SHell) does not allow scroll. So after obtaining the tree and using the <strong>ne</strong> command to visualize it you are left with a &#8220;tail&#8221; of the tree only. After some fidgeting I found a way to export trees to NEXUS format and could check the trees obtained. By using the command <strong>export</strong> with a dash and filenames as parameters.</p>

<p>tnt*>export -</p>

<p><strong>Conclusions</strong></p>

<p>TNT seems to be a powerful software and should be able to please anyone intending to employ parsimony as their phylogenetic method of choice. It is fast and, according to the documentation, is loaded.</p>

<p>On the other hand it is still (and probably will be for some time) limited and the CLI version for Linux is not the poster version of a user friendly application. There are many drawbacks and it is very difficult to learn the program without a deep study of the manual. Also the documentation is more focused to the GUI Windows version and the commands that can be used in Linux/OS X are not clear at first. In my opinion the worst feature of PISH is the lack of scroll (if there is scroll I wasn&#8217;t able to turn it on) especially when the tree visualization is selected.</p>

<p>Overall is TNT is a fair application, geared to its already known audience. Even being free, I don&#8217;t see it occupying the place of PAUP, Phylip and MrBayes in the phyloege. Not without major changes.</p>

<p><strong>_Update: another review, of the GUI version is available <a href="http://blindscientist.genedrift.org/2008/02/03/tnt-review-of-the-graphical-interface/">here</a>.</strong>_</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/30/benchmarking-python-fastest-way-to-generate-unique-lists/">Benchmarking Python: fastest way to generate unique lists</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-30T13:13:29-05:00" pubdate data-updated="true">Jan 30<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Just for fun, let&#8217;s see if there is any advantage (apart from generating
a smaller code) in using either of the approaches to create an unique
list. A list of 741 gene IDs and another one with 1322 (that contained
all the 741 IDs from the first) were used. Instead of hard coding the
lists in the script, normal I/O was used and the files were read
automatically. Using a for loop the scripts were run 10 times and the
final time averaged. Here are the results Using dictionaries 0.08540
Using sets 0.15890 Almost twice as fast for the dictionaries. Why? One
thing, and one thing only: Python version.</p>

<p>My system&#8217;s box default
Python is 2.3.4, which has one of the first implementations of <code>set</code>. I
have a &#8220;personal&#8221; Python version 2.5 so the test was redone with the
newer Python Using dictionaries 0.04520 Using sets 0.03770 Ok. That
shows a little bit of advantage for <code>sets</code> what is expected. But it
shows us that different versions of Python have a huge difference
between them. This is a rather crude and simple test, with a small list
of entries, but there is a huge gain from version 2.3 to version 2.5,
either in <code>dicts</code> or <code>sets</code>. For a more comprehensive test and more
functions with a similar objective check this
<a href="http://www.peterbe.com/plog/uniqifiers-benchmark">page</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/29/uniquifying-lists-with-sets-and-dictionaries/">Uniquifying lists with sets and dictionaries</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-29T22:34:21-05:00" pubdate data-updated="true">Jan 29<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are going to use our previous example to compare the use of <code>sets</code>
and <code>dictionaries</code> to create unique lists. We&#8217;ve already seen that when
<code>sets</code> are used it is very simple to transform a <code>list</code> with repeated
items in a unique <code>list</code>. The only hassle is to create the <code>set</code> and
then transform it back into a <code>list</code>. Like last time (with one small
addition)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span><span class='line'><span class="n">cluster1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">cluster2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">cluster1</span> <span class="o">+</span> <span class="n">cluster2</span>
</span><span class='line'><span class="n">uniqueset</span> <span class="o">=</span> <span class="n">Set</span><span class="p">(</span><span class="n">allgenes</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="n">finalist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">uniqueset</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>We can accomplish identical result by using
a <code>dictionary</code>. We create a small function to make our code clearer and
pass a list to it and we return the <code>dictionary</code> keys. Rmember that
Python <code>dictionaries</code> have values and keys and the latter cannot be
repeated, so it is basically a list of unique entries. Our function
would look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'> <span class="k">def</span> <span class="nf">make_unique_list</span><span class="p">(</span><span class="n">mylist</span><span class="p">):</span>
</span><span class='line'>  <span class="nb">dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">word</span> <span class="ow">in</span> <span class="n">mylist</span><span class="p">:</span>
</span><span class='line'>      <span class="nb">dict</span><span class="p">[</span><span class="n">word</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'>
</span><span class='line'><span class="k">return</span> <span class="nb">dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>In this function we declare the object
and in the loop, iterating over every list&#8217;s item we assign a value
(arbitrary) to the dictionary key. As pointed above, no repeated keys
are allowed, so every time a already checked item is seen by the
assignment it is not included in the dictionary. Finally we return only
the dictionary keys which is our final unique list. Our small script
would be:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">cluster1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">cluster2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">cluster1</span> <span class="o">+</span> <span class="n">cluster2</span>
</span><span class='line'>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">make_unique_list</span><span class="p">(</span><span class="n">allgenes</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Both methods are very
effective and usually fast. I will post some comparisons and benchmarks,
just for fun. Nathan posted in the comments another approach using
dictionaries. It is below with syntax highlighting</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="n">mylist</span><span class="p">)</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Basically in one line you pass a list of elements to dictionary and return all the
keys that are in the dic. Very Pythonic.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/28/python-sets/">Python sets</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-28T16:55:35-05:00" pubdate data-updated="true">Jan 28<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>After a long hiatus we are back with more Python tips, tricks, codes and
snippets. This time we will check how <code>set</code> works in Python. <code>set</code> is
another object type available in Python (version 2.3 and up) that brings
a lot of features to the language. From the Python Library Reference: &#8220;A
set object is an unordered collection of immutable values. Common uses
include membership testing, removing duplicates from a sequence, and
computing mathematical operations such as intersection, union,
difference, and symmetric difference.&#8221;</p>

<p>Yep, all these are possible with
<code>set</code>, and let me stress that it is not duplicated. So, a <code>set</code> is
basically a collection of item, but its unordered and not indexed,
because it does not record element position or insertion order. Methods
available for a <code>set</code> include <code>union()</code>, <code>intersection()</code>,
<code>difference()</code> that we will check next time.</p>

<p>First let&#8217;s see some basic
set functionality. Differently from other commonly available Python
object types, we need to import a library in order to use <code>set</code></p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span></code></pre></td></tr></table></div></figure>


<p>should
work. A first use for <code>set</code> would be to uniquify a list. Let&#8217;s say that
you have the gene IDs of two different clusters and you want to merge
these lists and keep only the unique ones, eliminating possible
duplicates IDs. We could do that with a <code>dictionary</code> and a simple
function (we will also check this later on) but a <code>set</code> makes our life
easier.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">sets</span> <span class="kn">import</span> <span class="n">Set</span>
</span><span class='line'>
</span><span class='line'><span class="n">cluster1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">cluster2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">allgenes</span> <span class="o">=</span> <span class="n">cluster1</span> <span class="o">+</span> <span class="n">cluster2</span>
</span><span class='line'>
</span><span class='line'><span class="n">uniqueset</span> <span class="o">=</span> <span class="n">Set</span><span class="p">(</span><span class="n">allgenes</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>and that&#8217;s all. Of course we won&#8217;t have a flexibility of a <code>list</code>, but
we can easily convert the <code>set</code> to a list and manipulate as before. Next
time we will see what else we can do with <code>sets</code> and gene lists.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/19/a-leopard-almost-ate-my-imac/">A leopard almost ate my iMac</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-19T22:01:53-05:00" pubdate data-updated="true">Jan 19<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Painful installation process for the latest Apple OS. Multiple failures during install, DVD trapped inside the drive, old system wiped out without warning, computer unusable for hours and hours. Installing Leopard reminded me when I first installed Windows 95 on my 40 Gb disk. I had the original 95 that couldn&#8217;t handle disks larger than 4 Gb (I think)  so it was a painful process to install, install SP1, repartition the drive and then finish. Hours and hours, just like Leopard yesterday.</p>

<p>And it is 2008, not 1996. Worse than that only OS/2. Thanks Apple: you blasted the last atoms of apple-fanboy in me.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/15/impressive-but-how-do-you-call-it/">Impressive, but how do you call it?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-15T14:02:05-05:00" pubdate data-updated="true">Jan 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.theglobeandmail.com/servlet/story/RTGAM.20080115.wgtmacworld0115/BNStory/Technology/home?cid=al_gam_mostemail">Very impressive, thin, light, powerful.</a>. But do they need to called MacBook Air? What a lame-name.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/01/08/preaching/">Preaching</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-01-08T12:59:07-05:00" pubdate data-updated="true">Jan 8<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>When someone &#8220;preaches&#8221; atheism, what is the difference of preaching some kind of religion?</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/29/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/27/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
