
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="We are back with more interviews. This time it is Dr Alexei Drummond&#8217;s take on Science and everything that surrounds it. Apart from his &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/35">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/08/11/sciview-part-5-interview-with-alexei-drummond/">SciView part 5: interview with Alexei Drummond</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-08-11T09:53:56-04:00" pubdate data-updated="true">Aug 11<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are back with more interviews. This time it is Dr Alexei Drummond&#8217;s take on Science and everything that surrounds it. Apart from his academic work, Alexei is Chief Scientist of <a href="http://www.biomatters.com">Biomatters</a> the developers of <a href="http://blindscientist.genedrift.org/2007/03/02/geneious-one-year-or-almost-later/">Geneious</a> (and <a href="http://blindscientist.genedrift.org/2007/02/26/hello-world/">here</a>). Enjoy.</p>

<hr />

<p><em>You are Biomatters&#8217; Chief-Scientist, one of the founders of the company and the mastermind behind Geneious. What was your motivation to start the company and develop the program? How to balance these activities with your regular job as a university professor?</em></p>

<p>I would say that my main motivations for wanting to build Geneious were frustration and hope. I was frustrated at the number of molecular biologists who didn&#8217;t seem to have the software tools they needed to do their job. I was hopeful I could provide the framework for those tools, if only I had a few great programmers.</p>

<p>Since I was one of the only people in the department who could program, I was often asked to write scripts, or teach people how to program enough so they could do it themselves. When biologists start asking about where they can learn to program a computer, just so they can do their job you know something is wrong! Most of the things they wanted to do were reasonably straightforward. It seemed to me that these kind of basic productivity problems had been solved in other workplaces like the office &#8211; so why had they not been solved in the research laboratory? I initially wanted to build the system as an academic research project &#8211; but then realized that would fail at a number of levels. Firstly, software development isn&#8217;t science &#8211; but science now crucially relies on software development and support. Secondly, most academic programmers are not interested in (or good at) designing user interfaces, and certainly developing software is not a scientific outcome that gets recognized like publishing papers does. Thirdly, academics are quite bad at supporting software and documenting it. So it seemed to me, that for a lot of reasons, a professional software company was the best avenue to realize a software system that would dramatically improve the productivity of molecular biologists by putting bioinformatics at their fingertips. Really I just wanted to make science go faster &#8211; and with less hassle.</p>

<p>_ Geneious is a Java application? Why Java? Any special feature in the language or previous experience? Is there anything that you wish Java was capable of so you can include in Geneious?_</p>

<p>Java is a general-purpose programming language &#8211; so you can do in Java pretty much anything you can do in software. The main reason for choosing Java is that it is very easy to write sophisticated user interfaces that run on Windows, Linux and Mac OS X. This is really important in the field of molecular biology where about 20% of the users have Mac OS X. Earlier versions of Java didn&#8217;t integrate with some aspects of different operating systems as well as they could, but with Java 5 most of this integration is quite good. Also, more and more academic developers are using Java (because it is easy, cross-platform and OO) so that also works in our favor. Every language has its pros and cons, but  for our purposes Java was one of the best choices.</p>

<p><em>You are the first scientist outside North America that is interviewed here. You have worked for some years in Oxford, UK. What are the main differences of working in New Zealand in comparison to Europe? How your return to New Zealand has influenced your decision to start the company?</em></p>

<p>New Zealand is a very small country with just 4 million people. But we are a proud nation. But we have a long history of world firsts, like being the first country to give woman the right to vote (1893). A New Zealander was the first to &#8220;split the atom&#8221; (Ernest Rutherford), and a New Zealander was the first to the top of Mt Everest (Sir Edmund Hillary). We are also one of the first countries to declare ourselves a Nuclear-Free Zone (1985). So I would say that New Zealand is a place the is not afraid to try to lead the way, despite our small size. Our Prime Minister, Helen Clark, recently declared our Government&#8217;s aspiration to make New Zealand a carbon-neutral country. It is these kinds of radical ideas that we strive to make reality. So it is quite natural to try to build a unique software package in New Zealand &#8211; although we are still far from where we are aiming to be.</p>

<p>_In your opinion what are the main differences of developing applications in academia and in the private sector?
_</p>

<p>Academics develop software for the love of it &#8211; and their first aim is to make the software work for themselves, and secondarily to work for their colleagues, as long as that doesn&#8217;t take too much effort. There are a few academics that try very hard to produce easy-to-use well-supported software, but they tend not to have the resources to do what the private sector can - for example PAUP<em> still doesn&#8217;t have a manual after many many years. The challenge for the private sector developing </em>scientific* software, is how to keep abreast of the science, which is changing so rapidly. At Biomatters we really try to have the idea that we are part of the scientific community. We have invited a number of scientist to work with us on various projects and we find that in this way we learn a lot more about what is needed, and how best we can fit into the academic ecology. Our goal is a happy marriage where academic programmers can get on with developing great new algorithms, and Geneious can provide the interoperability, the user interface and the support.</p>

<p><em>What would you teach to an &#8220;academic developer&#8221;? What developers in academia lack when compared to developers working in companies?</em></p>

<p>I don&#8217;t think it is a question of what academic developers lack in terms of skills &#8211; I think it is more a question of motivation. Aside from Geneious I produce academic software (like BEAST). What motivates me to answer 100 technical email queries every month about the BEAST software? My main passion is science, not supporting software. So by creating a successful software package, I have created a heavy burden of support. I would rather that somebody else supported all the simple questions about user interface, file conversion, installation, et cetera&#8230; So in some sense, what I am hoping to &#8220;teach&#8221; academic developers is that maybe there is another way to do things. It we had one framework that made interoperability and user interface more seamless and standard, each academic could focus on the science, the new algorithms and the part of their job that they find really exciting &#8211; the discovery.</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. Geneious has a very nice GUI, but in my opinion still lacks some simple features. Can you tell us how the program&#8217;s GUI was first designed and what kind of user input is considered in the development? Which is more important: &#8220;what the user want&#8221; or &#8220;what we want&#8221;?</em></p>

<p>Most users didn&#8217;t say they wanted the mouse, or the iPod before the visionary told them they wanted it. So I do think there is room for a lot of innovation in software development. But of course, there are definitely things that users know that they need right now. Balancing these two things is the real trick. Biomatters is very serious about listening to our users and we have an active forum and a small group of power users that have been quite helpful in the past. Of course there is always more to do. We also think that we have introduced a few nice new ideas in our software and there are more on the way. Specifically we are interested in developing the concept of &#8220;live&#8221; research publications, where arrival of relevant new data automatically updates the published research in that area. As you can imagine, this is a long term goal. But I think you need these visions to aim for, if you are going to make real progress.</p>

<p><em>For the young bioinformatician: should s/he focus on publications, skills, languages? Or a balance of all? What is your advice?</em></p>

<p>A young bioinformatician should focus on the science. What are the outstanding questions that your skills enable you to take a step towards answering? Once you focus on the science the publications, skills and languages will come.</p>

<p>I used to program ten different languages very well. But 95% of my programming these days is done in Java (because I am usually building on stuff I have done previously). PERL is still quite important in bioinformatics &#8211; but I think that is because there is still a lot of need to write one-off scripts. Will we get to a point where there is more sophistication? I think that bioinformatics has to become a field where people without programming skills can contribute substantially. I would argue that all of the programmers in bioinformatics should be working very hard to program themselves out of their jobs (and into more satisfying jobs). Lets build software systems that take this field to the next level. That is what we am trying to do (slowly but surely) with Geneious. We want to create a system that ultimately makes the programmers redundant &#8211; at least for the menial tasks &#8211; so that any molecular biologist can construct complicated queries and analyses in an intuitive graphical manner. Then the programmers can focus on the hard questions like new algorithms and methods of analysis.</p>

<p><em>What in your opinion is your most important software?</em></p>

<p>I don&#8217;t really compare my software projects, because they are so different. I guess this is for other people to decide. But judging from you line of questioning, Geneious is currently winning ;-)</p>

<hr />

<p>I would like to thank Alexei for answering my questions.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/26/sciview-part-4-interview-with-rosie-redfield/">SciView: part 4, interview with Rosie Redfield</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-26T21:05:18-04:00" pubdate data-updated="true">Jul 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Dr Redfield, from the University of British Columbia is a microbiologist and an advocate of Open Science, a rather recent movement that is trying to open the doors (and computers) of every scientific lab. I took the opportunity to ask her about her views of Open Science and why to create an scientific diary online. I would like to thank Dr Redfield for the opportunity.</p>

<hr />

<p><em>My initial questions are about Open Science. What are the main benefits of opening your whole scientific life to the public? And the drawbacks of doing this? How is the best way to manage a lab under Open Science?</em></p>

<p>RR: There are practical consequences and philosophical consequences to doing open science the way we are.  So far the practical consequences have been good - I get helpful suggestions from people I&#8217;ve never met. And I can tell people who wonder what we&#8217;re doing to take a look at the stuff on our web pages.  I&#8217;d love to also be able to say that our open approach has attracted collaborators and new lab members, but that hasn&#8217;t happened yet.  Another big practical consequence is that writing blog posts about the research I&#8217;m working on helps me think clearly about it. So whenever I&#8217;m feeling muddled I post the problem on my blog.</p>

<p>Philosophically it just feels right. Science is fundamentally a collaborative endeavour; we all build on what others have done.  Just as the web gives us new and better ways to publish our findings in peer-reviewed journals, it lets us find out communicate about our work in progress.  Furthermore, the taxpayer pays for our work, and deserves to be able to see what&#8217;s going on.  So I like to think of random people stumbling onto my blog and thinking &#8220;Hmm, so that&#8217;s what a scientist really does.&#8221;.</p>

<p>Also, openness is much easier and more comfortable than secrecy. In my teaching (mostly first-year biology) I give only open-book exams, and it&#8217;s much easier for me than worrying about cheaters.</p>

<p><em>I can see on your webpage that you have grant applications, results and even future plans listed. Do you fear having your ideas and plans &#8220;stolen&#8221;? You might have some data and results that end up on the webpage/blog only after the publication or everything is uploaded as soon as it out of the oven?</em></p>

<p>RR: I blog about the day-to-day research we&#8217;re doing, and I post our grant proposals as soon as they&#8217;re submitted.  So everything is open from the start.</p>

<p>I don&#8217;t worry much about people stealing our ideas, but it&#8217;s not a big risk because nobody else is working on what we&#8217;re working on. And most of the people doing related work seem to think I&#8217;m more-or-less misguided in my interests and approaches. I try not to post exactly the same figures I&#8217;m putting into a manuscript (only because I fear the journal editors might get cranky), but otherwise my goal is to give anyone who&#8217;s interested an accurate picture of what we&#8217;re doing.</p>

<p><em>The visibility of scientific blogs has increased considerably in the past years. Yours is up for almost an year now (at least your archives go back to August 2006) and it seems that you support that your whole group of students and post-docs do the same. What have influenced you to start recording your daily routine on a website? Does it help the interaction in the lab since everyone has its own personal page?</em></p>

<p>RR: I started blogging about research in response to a post by Pedro Beltrao (http://pbeltrao.blogspot.com/2006/07/opening-up-scientific-process-during.html) about the desirability of opening up the scientific process (I found it through Tangled Bank #59). As soon as I read it I realized that I wanted to have a blog about my lab&#8217;s research.</p>

<p>After a couple of weeks I told everyone in my lab that I wanted them to have blogs too.  I love reading their blogs - they have such thoughtful things to say about their research.  They don&#8217;t always post as often as I&#8217;d like them to; they say it&#8217;s because they feel there&#8217;s a high standard that they can only sometimes reach. But I&#8217;d like them to at least scribble a few lines every day.</p>

<p><em>From your blog it is easy to see some frustration with bioinformatics. How important is bioinformatics in your research and what the major problems that you have with it?</em></p>

<p>RR: It&#8217;s a collection of important tools and resources, and I love what I can do. And I enjoy learning new things, like babytalk Perl and standalone BLAST. But Unix has always driven me crazy - I keep hoping to attract a Unix/Perl geek to the lab, but no luck so far.</p>

<p>My bioinformatics frustrations arise in two ways.  First, it&#8217;s only part of what we do, so I&#8217;m not doing it nearly enough to get really good at it.  If bioinformatics was the main focus of our work I would invest more of my brain in improving my skills.  But we&#8217;re primarily an experimental lab (we get our hands wet). And each time I go back to it,  NCBI or TIGR has changed their interfaces.</p>

<p>The other source of frustration is that we&#8217;re always trying to do new things, that nobody has ever done before,  Sometimes we can find existing tools that can do these things, but sometimes we have to write our own and we don&#8217;t really know what we&#8217;re doing.  For a while we were lucky enough to have Mario Pineda-Krch working with us, but he&#8217;s gone on to better things (see http://pineda-krch.com/).</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. From what I could see your lab generates most of its scripts in Perl, but some work is done with graphical interfaces. What do you and your group look for in a program and its graphic interface? What are the main drawbacks of the command line interface that could be easily mitigated by a GUI?</em></p>

<p>RR: By graphic interfaces do you mean things like the web interfaces for BLAST? Basically we&#8217;ll take whatever we can get. If the resource has a nice point-and-click interface, great. If we have to submit a string of instructions to get what we need, we can cope. If we&#8217;d need to write some software ourselves, we usually move on to a different problem unless we&#8217;re highly motivated and the problem looks simple. I hate the frustration of not getting line commands exactly right, so much prefer menu-driven interfaces.</p>

<p>The only good interface we&#8217;ve generated was done by Mario, working on the fabulous hotspot paradox.</p>

<p><em>How do you see the &#8220;publish or perish&#8221; in science, with the increasing number of retracted papers, fake results, rushed publications and sometimes publication of extremely similar results in different papers and journals? Being an advocate of Open Science, do you think this new movement can help alleviate some of the known problems of the &#8220;publish or perish&#8221; culture?</em></p>

<p>RR: There&#8217;s always been a lot of bad science mixed in with the good; I&#8217;m not convinced that the proportion has gotten higher as the total number  of papers has increased.</p>

<p><em>Do you have an advice for the young scientist? And, as you are the first woman interviewed here, any special advice for the girls that are considering a scientific career?</em></p>

<p>RR: Work by women is unconsciously but systematically undervalued by everyone (including the women who do it), so don&#8217;t go into science for the glory.  But if you love doing research it&#8217;s wonderful.  I&#8217;m still delighted that I can get paid to do this.</p>

<p><em>What in your opinion is your most important publication?</em></p>

<p>RR: That&#8217;s a hard one.  I don&#8217;t get much recognition for what I think is important, mostly because my conclusions don&#8217;t fit what others want to hear.</p>

<hr />

<p>I would like to thank again for Dr Redfield&#8217;s answers. I am planning some changes to SciView soon.</p>

<p>If you want to follow the RRResearch blog, here is the <a href="http://rrresearch.blogspot.com/">link</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/26/call-for-collaboration-an-open-science-first-for-me/">Call for collaboration: an Open Science (?) first for me</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-26T08:31:20-04:00" pubdate data-updated="true">Jul 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Following the new trend in science (i.e. <a href="http://nsaunders.wordpress.com/2007/07/13/open-science-discussion/">here</a> and <a href="http://wiki.nodalpoint.org/open_research_discussion">here</a>) I decided to make a scientific experiment.</p>

<p>For quite some time I have been writing my interface programs by myself - <a href="http://genedrift.org/mtgui.php">one</a> and <a href="http://genedrift.org/winpop.php">two</a> among others that have never seen the light of the day or left the sketch pad. I don&#8217;t know the penetration of this space, as most people come here for the interviews, but I am willing to give it a shot and ask if anyone is interested in developing interfaces for bioinformatics programs.</p>

<p>Something that you you are willing to learn how to use, something that you use in command line but rather see it with a nice GUI, something that you are developing and want to include an interface or even if you never programmed an interface and want to learn. I already have some ideas on what to work, and the ideal languages (at least for me) are the combos C++/<a href="http://www.wxwidgets.org">wxWidgets </a>and Python/<a href="http://www.wxpython.org/">wxPython</a>, but anything goes here.</p>

<p>My idea is to define a small project and to report the progress in an &#8220;Open Science&#8221; website/blog, showing the whole process with its frustrations and achievements along the way.</p>

<p>Just send me an email at nuin at genedrift dot org or leave a comment below.</p>

<p>_<strong>Addendum:</strong>I would be interested in other GUI frameworks too, such as QT, PyQt and Cocoa. _</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/16/finding-motifs-iupac-and-regex-an-overview/">Finding motifs: IUPAC and RegEx, an overview</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-16T11:59:56-04:00" pubdate data-updated="true">Jul 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>End of Section 5, moving to Section 6. For anyone also following the
book there will be a jump at the end of chapter 8, so we are not
touching the final script that deals with different reading frames here.
We are going straight, or almost, to another take of Regular
Expressions. We are going to check some aspects of restriction enzymes,
but first we are going to touch base with motif finding in DNA
sequences. We already saw how to use the <code>re</code> module and to do some
simple regular expression searches. Basically our motif search was very
simple, with run-time user input using</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">motif</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&#39;</span><span class="si">%s</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">inmotif</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>where <code>inmotif</code> was the short string sequence entered. Enhancing this a little
bit we will first create a more advanced regex search using mismatches
in the sequences and after we will see (as in the book) how to translate
IUPAC code to regex. The IUPAC table is</p>

<pre><code>A - Adenine 
C - Cytosine 
G - Guanine 
T - Thymine 
U - Uracil 
M - A or C 
R - A or G 
W - A or T 
S - C or G 
Y - C or T 
K - G or T 
V - A or C or G 
H - A or C or T 
D - A or G or T
B - C or G or T 
X - A or C or G or T 
N - A or C or G or T 
</code></pre>

<p>You are going
to use basic Python to do this. There are mode advanced ways to reach
the same goals, but as we haven&#8217;t seen a lot of dictionaries it would be
great to check them again. The first idea for the table above is to use
a dictionary for the codes that represent two or more nucleotides, as
dictionary keys would be ideal to include in regex elements. Our basic
dictionary would look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">iupacdict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;M&#39;</span><span class="p">:</span><span class="s">&#39;[AC]&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">:</span><span class="s">&#39;[AG]&#39;</span><span class="p">,</span> <span class="s">&#39;W&#39;</span><span class="p">:</span><span class="s">&#39;[AT]&#39;</span><span class="p">,</span> <span class="s">&#39;S&#39;</span><span class="p">:</span><span class="s">&#39;[CG]&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">:</span><span class="s">&#39;[CT]&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;K&#39;</span><span class="p">:</span><span class="s">&#39;[GT]&#39;</span><span class="p">,</span> <span class="s">&#39;V&#39;</span><span class="p">:</span><span class="s">&#39;[ACG]&#39;</span><span class="p">,</span> <span class="s">&#39;H&#39;</span><span class="p">:</span><span class="s">&#39;[ACT]&#39;</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">:</span><span class="s">&#39;[AGT]&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">:</span><span class="s">&#39;[CGT]&#39;</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="s">&#39;[ACGT]&#39;</span><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p>We include the square brackets
around each key because these brackets inside a regex indicate a match
in one of the characters. This dictionary would be really advantageous
for restriction enzymes too. Now in order to find a motif we need to
create a regular expression from the consensus sequence of the motif
model, that can be for instance a transcription factor binding site.
Let&#8217;s say we have a GATA3 model of binding site which has a model
NNGATARNG what would give us a regular expression</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">]</span><span class="n">motif</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&#39;[ACGT][ACGT]GATA[AG][ACGT]G&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>that can be extracted directly from the <code>iupacdict</code>. Next time, we will check how to
create these regex on the fly, from input and the dictionary.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/11/translating-dna-into-proteins-second-approach-now-using-fasta-files/">Translating DNA into proteins: second approach, now using FASTA files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-11T15:10:09-04:00" pubdate data-updated="true">Jul 11<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen before how to translate DNA sequences into amino acids
sequences. We have even created a module that contains the dictionary
for the genetic code. Now we are going to combine both (very simple)
modules we created in one nice script for day-to-day use. So, we have
the <a href="http://python.genedrift.org/code/dnatranslate.py">dnatranslate.py</a>
and the <a href="http://python.genedrift.org/code/fasta.py">fasta.py</a> that we
are going to import into our script. And that&#8217;s basically it: calling
function already created, stored in modules that can be reused anytime.
In the end our script that translates DNA sequences to proteins takes a
little bit more than a handful of lines.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">dnatranslate</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'>
</span><span class='line'><span class="n">dna</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">dna</span><span class="p">:</span>
</span><span class='line'>    <span class="n">protein</span> <span class="o">=</span> <span class="n">dnatranslate</span><span class="o">.</span><span class="n">translate_dna</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">item</span><span class="o">.</span><span class="n">name</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">protein</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s it. A good example of
reusable code, that once created fits everywhere and handles most type
of data. We read the FASTA file in the first line, and iterate over the
items created translating them as we go. As an extra exercise, we can
include the output formatting function. First we need to update the
<code>fasta.py</code> module (already on the repository) and slightly change the
formatting function, that ends up looking like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">format_output</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">temp</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">temp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="n">length</span><span class="p">])</span>
</span><span class='line'>    <span class="k">return</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>For this case the ideal formatting
function would go through the &#8220;longer&#8221; route mentioned before, because
the final printing should be done by the main script and not by the
imported module. This gives us more control on what we want to do with
the resulting string. The <code>format_output</code> function receives two
arguments: the first is the actual DNA/protein sequence to be formatted
and the length we want to output it. We had to remove the loop too, so
only one sequence can be sent to the function and, as pointed, the
function returns a string with the formatted sequence. In the end our
post&#8217;s initial sequence has one modification only</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">dnatranslate</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'>
</span><span class='line'><span class="n">dna</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">dna</span><span class="p">:</span>
</span><span class='line'>    <span class="n">protein</span> <span class="o">=</span> <span class="n">dnatranslate</span><span class="o">.</span><span class="n">translate_dna</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">item</span><span class="o">.</span><span class="n">name</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">fasta</span><span class="o">.</span><span class="n">format_output</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>the last line, that instead of printing directly the result of the
translation, sends the sequence to the formatting function before.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/10/formatting-output-of-fasta-files/">Formatting output of FASTA files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-10T11:30:15-04:00" pubdate data-updated="true">Jul 10<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The Beginning Perl for Bioinformatics book shows a script to print
formatted sequence data, specifying that no more than 80 characters
(either nucleotides or amino acids) should be printed across a page.
Here, we will see a similar script in Python and will include it in our
<code>fasta.py</code> module in order to use it a default output, taking advantage
of its usability. Basically, our FASTA reader strips all the carriage
returns/new lines (if existent) from the input file, so the sequence
itself is stored continuously in a string. Of course we could do it
differently and store the sequence as is in the file, but then they
might differ on the number of characters per line and other features.
Our goal then is to <em>break</em> the sequence in chunks, making it look like
a justified paragraph. We use the code below to obtain this, back after
it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="n">temp</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">name</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">),</span><span class="mi">80</span><span class="p">):</span>
</span><span class='line'>        <span class="n">temp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">80</span><span class="p">])</span>
</span><span class='line'>        <span class="k">print</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>
</span><span class='line'>        <span class="n">temp</span> <span class="o">=</span> <span class="p">[]</span>
</span></code></pre></td></tr></table></div></figure>


<p>The script, as always, is very
simple and does the job. A new feature of the range method is
introduced: the <em>step</em>. From previous posts we have seen that function
range generates a sequence of integers based either on one value</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</span><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>or two values</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">15</span><span class="p">)</span>
</span><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">14</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>The <em>step</em> parameter makes the generated to jump a
certain number of values.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</span><span class='line'><span class="o">&gt;&gt;&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">18</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>This is similar to the
<code>for</code> loop in C/C++ where the loop is define by three parameters on a
construction like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='c'><span class='line'><span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">;</span> <span class="n">i</span><span class="o">++</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>where the first parameter sets the initial value of
<code>i</code>, the second value sets the maximum value of <code>i</code> and the last one
sets the <em>step</em> that <code>i</code> will be incremented. In our small script above,
we read the sequence (this time putting everything in one line), get the
sequences and then iterate along the instances of our class, printing
first the name of the sequence. Then we iterate on the whole length of
each sequence, this time generating a range that starts on zero and goes
up to the final nucleotide/amino acid and has a <em>step</em> that defines the
number of characters we want to print in one line. To print the sequence
to the screen we had two options: directly print using</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">80</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>or use the
longer route shown in our script. Trying this script will output the
sequences from the original file with 80 characters along the line. Next
following the book we will write a script to translate DNA to protein
using our FASTA reading function and modify our FASTA module to include
the formatted output function.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/04/reading-fasta-files-2/">Reading FASTA files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-04T12:33:41-04:00" pubdate data-updated="true">Jul 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We now have an idea on how to create a class in Python. For our FASTA
reader we will use a very similar approach as the employee class created
before. Based on the simplicity of the FASTA format it is easy to see
what attributes we need in our class: sequence title (header) and
sequence. That&#8217;s it. Speeding up, our class will look like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Fasta</span><span class="p">:</span>
</span><span class='line'>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s
all we need for now. Of course we could include sequence length,
sequence type, or any other relevant information as a class attribute.
But we will stick with those for the time being. NOw that we have the
class declared, we need to read the file and create instances of the
<code>Fasta</code> class to store all our sequences. We already know how to open
and read a file and create functions. We put everything together in a
simple script:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
<span class='line-number'>47</span>
<span class='line-number'>48</span>
<span class='line-number'>49</span>
<span class='line-number'>50</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="c">#class declaration with both attributes we need </span>
</span><span class='line'><span class="k">class</span> <span class="nc">Fasta</span><span class="p">:</span>
</span><span class='line'>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>        <span class="c">#this will store the sequence name</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
</span><span class='line'>        <span class="c">#this will store the sequence itself </span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="c">#this function will receive the list with the file contents,</span>
</span><span class='line'><span class="c">#create instances of the Fasta class as it scans the list, putting the</span>
</span><span class='line'><span class="c">#sequence name on the first attribute and the sequence itself on the</span>
</span><span class='line'><span class="c">#second attribute </span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>    <span class="c">#we declare an empty list that will store all Fasta class instances generated </span>
</span><span class='line'>    <span class="n">items</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>        <span class="c">#we check to see if the line starts with a &gt; sign</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;&gt;&quot;</span><span class="p">):</span>
</span><span class='line'>            <span class="c">#if so and our counter is large than 1 we add the created class instance to our list </span>
</span><span class='line'>            <span class="c">#a counter larger than 1 means we are reading from sequences 2 and above </span>
</span><span class='line'>            <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>                 <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aninstance</span><span class="p">)</span>
</span><span class='line'>                 <span class="n">index</span><span class="o">+=</span><span class="mi">1</span>
</span><span class='line'>            <span class="c">#we add the line contents to a string</span>
</span><span class='line'>            <span class="n">name</span> <span class="o">=</span> <span class="n">line</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>            <span class="c">#and initialize the string to store the sequence </span>
</span><span class='line'>            <span class="n">seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>            <span class="c">#this creates a class instance and we add the attributes </span>
</span><span class='line'>            <span class="c">#which are the strings name and seq </span>
</span><span class='line'>            <span class="n">aninstance</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
</span><span class='line'>       <span class="k">else</span><span class="p">:</span>
</span><span class='line'>            <span class="c">#the line does not start with &gt; so it has to be #a sequence line, so we</span>
</span><span class='line'>            <span class="c">#increment the string and \#add it to the created instance </span>
</span><span class='line'>            <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>            <span class="n">aninstance</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
</span><span class='line'>            <span class="c">#the loop reads everything but the penultimate</span>
</span><span class='line'>            <span class="c">#sequence is added at the end, so we need to add it after the loop ends </span>
</span><span class='line'>       <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aninstance</span><span class="p">)</span>
</span><span class='line'>       <span class="c">#a list with all read sequences is returned </span>
</span><span class='line'>       <span class="k">return</span> <span class="n">items</span>
</span><span class='line'>
</span><span class='line'><span class="n">fastafile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">mysequences</span> <span class="o">=</span> <span class="n">read_fasta</span><span class="p">(</span><span class="n">fastafile</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">mysequences</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mysequences</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">i</span><span class="o">.</span><span class="n">name</span>
</span></code></pre></td></tr></table></div></figure>


<p>At
first, it looks scary. But is not. There are many ways to create this
loop and to read the sequences, and many ways to make this loop shorter.
We will get there, eventually, but for starters this is OK. Basically,
in the above script we have a <code>Fasta</code> class, a <code>read_fasta</code> function and
a couple of lines to read a print the results. The <code>read_fasta</code> function
basically checks all items of a list and see which is the first
character of each item: a <code>&gt;</code> sign indicates that the item should be
parsed to the temporary name string, another character redirects the
item to the sequence string. Instances are created on the fly and the
attributes assigned as what was file content is being scanned. In the
end, just to be sure of what we accomplished with the function we print
the list and loop through it and print the name of each sequence read.
Notice that the loop counter/index is the one that receives the class
attribute. The biggest advantage of this way of reading FASTA files is
that the class and the function are reusable. Basically we can create a
file (for instance called readfasta.py) and import into another script
and we have a FASTA reader ready to rock. We would need to tweak a
little bit in order to catch exceptions, but with consistent FASTA files
this would work fine.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/07/03/reading-fasta-files/">Reading FASTA files: introduction</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-07-03T16:39:37-04:00" pubdate data-updated="true">Jul 3<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Again after a long period we are back. We already have most of the
knowledge to create very useful scripts and small programs in Python.
And in this post we will create a routine that will make the biologist&#8217;s
life even easier (when programming Python). A great part of
bioinformatics is to store data, how to store it and which format to
use. Anyone working in a wetlab or on a pure bioinformatics lab had had
problems with file formats one day or another. In fact, you don&#8217;t have
to be in a bioinformatics environment to have such problems, but in
biology people tend to create a file format for everyt program they
write. We can call it a lack of standards, sometimes. But one very well
established format is the FASTA (pronounced FAST-Aye, according to the
EMBl-EBI page for the software with identical name. This format is
really simple and easy to manipulate with most computer languages and
being text-based adds an extra advantage of portability for the files.
Usually a FASTA file has a structure like this</p>

<pre><code>&gt;title/name/extra information about the sequence  
sequence in one or many lines
</code></pre>

<p>There is
no limit on the number of sequences that can be stored in a file,
neither on the size of each sequence. Usually sequences that are larger
than 70-80 nucleotides/amino acids are displayed in multiple lines.</p>

<p><code>&gt;Q15465|SHH_HUMAN Sonic hedgehog protein - Homo sapiens (Human).
MLLLARCLLLVLVSSLLVCSGLACGPGRGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASG
RYEGKISRNSERFKELTPNYNPDIIFKDEENTGADRLMTQRCKDKLNALAISVMNQWPGV
KLRVTEGWDEDGHHSEESLHYEGRAVDITTSDRDRSKYGMLARLAVEAGFDWVYYESKAH
IHCSVKAENSVAAKSGGCFPGSATVHLEQGGTKLVKDLSPGDRVLAADDQGRLLYSDFLT
FLDRDDGAKKVFYVIETREPRERLLLTAAHLLFVAPHNDSATGEPEASSGSGPPSGGALG
PRALFASRVRPGQRVYVVAERDGDRRLLPAAVHSVTLSEEAAGAYAPLTAQGTILINRVL
ASCYAVIEEHSWAHRAFAPFRLAHALLAALAPARTDRGGDSGGGDRGGGGGRVALTAPGA
ADAPGAGATAGIHWYSQLLYQIGTWLLDSEALHPLGMAVKSS</code></p>

<p>So, let&#8217;s write a function
and a small script to read a typical FASTA file and display the output.
Later we will see how to manipulate the file and in the end we will have
created a list of scripts that will be useful for the everyday
laboratory life. But before we get to our final goal, we need to learn
some new features of Python. In our case <strong>Classes</strong>. Classes in Python
are very similar to classes in C++, and they are the building blocks of
Object Oriented Programming (OOP). This guide will only scratch the
surface of OOP, as it is a very complex subject, requiring a guide of
its own. There are plenty of good introductory material online and any
Google search will return dozens of links. Let&#8217;s focus on some basic
concepts, what will be exactly what we will need here. Classes are
basically objects with associated properties (attributes) and methods. A
traditional introductory example of classes is the employee list. In a
company all employees are registered and they basic information are
stored in the human resources file system. Let&#8217;s call the main class
<code>Employee</code> with three attributes: <code>name</code>, <code>room_number</code> and
<code>favourite_colour</code>. This defines a class. Let&#8217;s see how to do that in
Python:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Employee</span><span class="p">:</span>
</span><span class='line'>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">room</span><span class="p">,</span> <span class="n">colour</span><span class="p">):</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">room</span> <span class="o">=</span> <span class="n">room</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">colour</span> <span class="o">=</span> <span class="n">colour</span>
</span></code></pre></td></tr></table></div></figure>


<p>Let&#8217;s dissect this piece of code.
First we declare a class and give a name to it <code>Employee</code>. The next
thing we need to do is to define the initiation (constructor) method of
the class and give it the attributes we need, that&#8217;s the <code>__init__</code>
method. Every time we create a copy of the main class object the copy
will be initialized with the values we are passing to the method.
According to Dive into Python &#8220;The first argument of every class method,
including <code>__init__</code>, is always a reference to the current instance of
the class&#8221;. That&#8217;s the <code>self</code> on the method definition, which is
followed by what we want to store in the object: name, room and
favourite colour. The lines in the method are the ones assigning the
received values to the different class&#8217; attributes.</p>

<p>Again from Dive into
Python we have &#8220;To reference this attribute from code outside the class,
you qualify it with the instance name, instance.data, in the same way
that you qualify a function with its module name. To reference a data
attribute from within the class, you use self as the qualifier&#8221;. In
other words the attributes of the class are seen internally by it by the
use of <code>self</code> while from the outside (another part of the script) it is
seen by the name given to the instance of the class. It is a little bit
confusing, so let&#8217;s see an example below.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Employee</span><span class="p">:</span>
</span><span class='line'>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">room</span><span class="p">,</span> <span class="n">colour</span><span class="p">):</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">room</span> <span class="o">=</span> <span class="n">room</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">colour</span> <span class="o">=</span> <span class="n">colour</span>
</span><span class='line'>
</span><span class='line'>
</span><span class='line'><span class="n">employeename</span> <span class="o">=</span> <span class="s">&quot;Paulo&quot;</span>
</span><span class='line'><span class="n">roomnumber</span> <span class="o">=</span> <span class="s">&quot;21&quot;</span>
</span><span class='line'><span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;blue&quot;</span>
</span><span class='line'><span class="n">newemployee</span> <span class="o">=</span> <span class="n">Employee</span><span class="p">(</span><span class="n">employeename</span><span class="p">,</span> <span class="n">roomnumber</span><span class="p">,</span> <span class="n">colour</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">newemployee</span>
</span></code></pre></td></tr></table></div></figure>


<p>If we run this simple code we will get
something like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="o">&lt;</span><span class="n">__main__</span><span class="o">.</span><span class="n">Employee</span> <span class="n">instance</span> <span class="n">at</span> <span class="mh">0x7379cc</span><span class="o">&gt;</span>
</span></code></pre></td></tr></table></div></figure>


<p>And that&#8217;s exactly what we need here: that&#8217;s a class instance that was created by</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">newemployee</span> <span class="o">=</span> <span class="n">Employee</span><span class="p">(</span><span class="n">employeename</span><span class="p">,</span> <span class="n">roomnumber</span><span class="p">,</span> <span class="n">colour</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Printing this information is not very useful. We
don&#8217;t want to know the memory address of the object, we want to get the
attributes of it, so we need to change <code>print newemployee</code> by these
lines:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">newemployee</span><span class="o">.</span><span class="n">name</span>
</span><span class='line'><span class="k">print</span> <span class="n">newemployee</span><span class="o">.</span><span class="n">room</span>
</span><span class='line'><span class="k">print</span> <span class="n">newemployee</span><span class="o">.</span><span class="n">colour</span>
</span></code></pre></td></tr></table></div></figure>


<p>There they are,
run it and see what is being printed to the screen. Of course, as is
this script does not accomplish anything useful, but imagine that you
need to read a file with 1000 new employees everyt month. That&#8217;s what we
need to do with FASTA files everyday and we will see how to do that on
the next post.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/06/25/sciview-part-3-interview-with-jeremy-squire/">SciView part 3: Interview with Jeremy Squire</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-06-25T10:21:14-04:00" pubdate data-updated="true">Jun 25<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The SciView project continues. This time the interviewed was <a href="http://www.utoronto.ca/cancyto">Dr Jeremy Squire</a>, from the Ontario Cancer Institute, which by some freaky coincidence is where I work. The two previous interviews were with scientists, if not completely, somewhat related to bioinformatics. Not this time.</p>

<p>With this interview the first cycle, which I might call inaugural cycle, is complete. I would like to thank everyone for their support and if you have any comments or ideas just drop me a note. Probably in the coming months, SciView will have a home for itself and added features.</p>

<hr />

<p><em>You have started your scientific career with conventional cytogenetics and doing analysis with optical microscopes. What have changed since you started, with the advances in technology. new methods,new equipments? Do you still believe that basic cytogenetics analysis is still strong?</em></p>

<p>JS: The big advance was application of CCD cameras and the computer interface to the scopes. The basic analytical platform is still 70% determined by the human brain as computers  are unable to interpret the biological and technical variables of banded metaphase preparations correctly.</p>

<p>The other big advance was the development of FISH methods and the parallel availability of the toolbox of probes and techniques created by the human genome project. This has enhanced rather than replaced my area so that basic cytogenetics has wider applications now, than in 1973 when I first started looking at human chromosomes.</p>

<p><em>How important is bioinformatics for human genetics and oncology? With the ever increasing presence of computers in any research, how do you see the introduction of bioinformatics in your field?</em></p>

<p>JS: Bioinformatics in its various forms has played a vital role in bringing all the &#8220;omics areas&#8221; back to the bench and it is an essential ingredient in most new hypothesis driven lab projects in genetics and oncology. Like most scientists brought up on wet bench research I have a high level of skepticism of in silico projects that do not touch down to test simulations and speculations by experimental methods. This has been a weakness on some grant proposal that I have reviewed in recent years. I my field I have covered past and present issues above.  In the future I see bioinformatics helping understand the 3 D structure of genomes (in cancer and normal nuclei) being essential.</p>

<p><em>Another question that is already favourite of this interview series is about computer graphic interfaces. Clearly for your type of research the graphic interface is a must have, especially in cytogenetics and microarray applications. Based on your experience would you say that the adoption of an application in your lab could be determined by the quality of its graphic interface? What do you and your group look for in
a program and its graphic interface?</em></p>

<p>JS: Absolutely agree. The graphic interface has to mirror the scientific thought processes that we use when we conceptualize the data. Our experience with the Agilent array platform has been great and it has helped up move faster in integrating different genomic findings. Apart from the overall output being elegant and easy to understand it should be transparent and adaptable to our own needs. Also in more practical terms - working on Apple and PC would help a lot. Plus being able to take on your own laptop to play with whilst waiting at airports (I have done that a lot this month!)</p>

<p><em>What is your take on the main differences of oncology research between US and Canada? It seems that Canada even with a smaller number of research institutes have the lead in different fields of cancer research.</em></p>

<p>JS: In human cancer Canada has a stronger programs in signaling cancer research and cancer stem cell approaches. This relates to successes in the 1960s (stem cells) and then in the 1980s (signaling). Cancer genetics and genomics is stronger in the US, as is translational cancer research. Our funding system favours individual PI driven projects that tend to be more basic. Our genome project was poorly supported in the late 1990s.  We need to have NIH SPORE grants for clinical cancer research projects as these encourage team approaches of scientists and clinicians to work together. Canada is excellent in model organism (mouse, flies, yeast) approaches to cancer research. The weakness in applied cancer research in Canada is that the bench to bedside work is used to raise money from donors but there are few Canadian success stories to talk about.</p>

<p><em>You are still in Brazil doing a sabbatical. From what you have seen of Brazilian research, how do you compare the science in a country under development, with a young science sometimes underfunded, and the science in major centres like US, Canada and Europe? What is the main advice for a young scientist who wants to come to a major centre?</em></p>

<p>JS: This is a big topic and I would prefer to talk to you about it over a beer!  I think any serious Brazilian scientist has to do a postdoc in NAmerica or EU. If they love Brazil more than they love research they should develop some serious political connections with FAPESP before they leave to do their postdoc.  This will help them when they come back. However it will always be frustrating to do good research in Brazil. Too many PIs here are not serious about research.</p>

<p><em>I asked this question to Joe Felsenstein and Brian Golding: How do you see the &#8220;publish or perish&#8221; in science, with the increasing number of retracted papers, fake results, rushed publications and sometimes publication of extremely similar results in different papers and journals? What are your views, especially in a very competitive field as oncology?</em></p>

<p>JS: As long as grant panels judge researchers by productivity this situation will not change. To address this area I support the idea of declaring the identity of the reviewer of your grants or your submitted paper. I think more thought would go into the review process and to the quality of submitted papers if the process was more transparent. Some journal are doing this now.  In the oncology field I would like some of the journal to have part of the abstract devoted to the translational component of the study.</p>

<p><em>Do you have an advice for the young researcher in human genetics? Should s/he focus on a determined trend or be a generalist in the field?</em></p>

<p>I think this entirely depends on how the young researchers own neurons are wired. Some people function better as generalists and others need focus. There are of course intermediate versions. I would advise someone starting out to be opportunistic and to look very closely at emerging trends for the choice of a postdoc for example.</p>

<p><em>What in your opinion is your most important publication?</em></p>

<p>AS STUDENT:
1.    Godbout R, Dryja TP, Squire JA, Gallie BL, and Phillips RA. Somatic inactivation of genes on chromosome 13 is a common event in retinoblastoma.  Nature. 304: 451-453, 1983.
AS PI:
2.    Kolomietz E, Al-Maghrabi J, Brennan S, Karaskova J, Minkin S, Lipton J, and Squire JA.  Primary chromosomal rearrangements of leukemia are frequently accompanied by extensive submicroscopic</p>

<h2>deletions and may lead to altered prognosis.  Blood. 97(11): 3581-3588, 2001. </h2>

<p>I would like to thank Joe Felsenstein, Brian Golding and Jeremy Squire for taking their time to answer my questions. SciView will be back (hopefully) soon.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/05/29/sciview-scientific-interviews-part-2/">SciView: scientific interviews, part 2</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-05-29T09:01:58-04:00" pubdate data-updated="true">May 29<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>After the first interview with Joe Felsenstein, I am continuing with this small project, SciView, with an interview with <a href="http://helix.biology.mcmaster.ca/">Brian Golding</a>, from the McMaster University in Hamilton, Canada. For those who don&#8217;t know, Brian is the person behind <a href="http://evol.mcmaster.ca/brian/evoldir.html">EvolDir</a>, where many researchers find new jobs, solve problems and check for next meeting.
I would like to thank him for taking the time to answer my (sometimes) dull questions.</p>

<hr />

<p>_In your opinion is there any difference about the Bioinformatics research done in Canada and in the US? The number of companies, universities and institutes in Canada is evidently smaller than in the US, would that be an obstacle for the development of Canadian Bioinformatics? _</p>

<p>BG: Canada has often had a &#8220;branch-plant&#8221; economy for many companies. So little primary research is done for the major pharmaceutical companies here. Because genomics and pharmaceutical research is expensive, the research here is probably less per capita and of a different nature (for example more testing).  I don&#8217;t think that this is an obstacle, it simply means that the research for bioinformatics is mainly centered in universities and mainly in a few concentrated areas.</p>

<p><em>I asked this question to Joe Felsenstein: How do you see the &#8220;publish or perish&#8221; in science, with the increasing number of retracted papers, fake results, rushed publications and sometimes publication of extremely similar results in different papers and journals? What are your views?</em></p>

<p>BG:Yes, the paradigm to publish is getting worse than when I was younger. Indeed, I would not have had enough papers to get a job.  It has also lead to many more authors per paper than in the remote past when I was younger. The increased pressure has an obvious consequence that you note above. However, this is not limited to research. The need for &#8220;accountability&#8221; is a buzz word in politics today and any government venture has to be &#8220;made accountable&#8221; (since for some reason it is perceived not to have been in the past).  Research with government dollars is painted with the same brush. Unfortunately, for much of science, increased accountability in an already very open system, simply leads to more paperwork and less productivity (less accountability).</p>

<p><em>You have been an advocate of free/open source software. In your opinion what is the importance of this type of software for the Bioinformatics research? Do you agree of academic researcher to pursue copyright/patents of applications that are developed using public money?</em></p>

<p>BG: I do not agree with this though I find my opinion to be in the minority. I have always found open source software to be excellent work (people base their reputations rather than their cash on the software). You do require some willingness to accept strange bugs but people that pay for software seem to be willing to accept these same bugs that they are intolerant of in open source. I think that many are living by the adage that &#8220;you get what you pay for&#8221;. I think that there are many exceptions to this and a community based software project is one of these exceptions.</p>

<p><em>Another question that will become a favourite of this interview series is about computer graphic interfaces. There are many scientific software that does not have interfaces. Based on that, do you think that the lack of GUI would be an obstacle for the adoption of a program in research?</em></p>

<p>BG: Having a command line option for a program is a requirement for me. I find the command line to be the natural place to &#8220;get things done&#8221; in LINUX. It provides a way to use the power of the shell, the operating system and other programs to easily interface with whatever other program I may be running. Without it, you are limited to the interface that was created by the programmers.</p>

<p><em>How do you see that some bioinformatics companies create proprietary packages that are mere wrappers of open source academic software? Geneious, CLC Workbench, etc are excellent GUIs but the main engine is/was developed using academic research.</em></p>

<p>BG: This is common practice. They are selling value added software. So are the LINUX distribution packagers. They sell LINUX and a package of software that goes with it, each itself usually free but checked to make sure they all play nice together. The community is one where they usually also make available a lesser package that contians the essentials for free.  If it is important to people to have a package that they need not worry about piecing together then they can pay for it. I am however, surprised to find  any people that have paid big bucks for packages that I piece together for free often with a more powerful interface. Difference is, I am willing to spend some time to try to learn.</p>

<p><em>EvolDir is a great service for the scientific community (I found my first post-doc through it). I know that it is run using your own software, based on scripts and homebrew code. Why did you decide to rely on your own scripts instead of using Mailman or Majordomo? How EvolDir was started?</em></p>

<p>BG: Mailman and Majordomo did not exist when I started the EvolDir. The EvolDir was started simply as a file that contained a list of emails that I knew (this was before the days of web browsers or even gopher browsers and the only way to get an email was to ask someone for it (by voice only). A friend saw it and asked if he could get a copy. The list changed shortly after so I thought that I should send him another, updated copy.  Another person asked and so on; strictly via word-of-mouth (literally in those days). This is why it is called a Dir (directory; a list of names and emails). Notices of general interest to people came later.</p>

<p><em>What would be your advice for someone that starting in the field? You came from the wetlab and probably have used punch cards in the early days of informatics. Is the life of the young bioinformatician easier than in the old days?</em></p>

<p>BG: A bioinformatician today can do things that I could not imagine when I started as an undergraduate.  This morning I BLAST&#8217;ed over 5000 proteins against more than 500 complete genomes (in about 20 minutes). When one thinks of all that this entails it is staggering. The computers are so fast today that the trick to getting things done is no longer how quickly you can contrive to get back into the queue for those few precious cpu seconds. Now most people, except for some specific programs or gaming, cannot keep a computer busy &#8230; hence the creation of the grid to give this idleness to others that can.  The world is a much different place, biology is much different and what it even more exciting, I would surely anticipate (given the advances that one sees today) that in another ten/fifteen years, the world and biology will be different anew.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/36/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/34/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
