
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="We use our last code as a starting point in order to generate some real
information from our simulated sequence sets. Our approach here will be
the &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/35">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/10/still-simulation/">Still simulation</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-10T17:23:57-04:00" pubdate data-updated="true">Apr 10<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We use our last code as a starting point in order to generate some real
information from our simulated sequence sets. Our approach here will be
the same: functions to do all the work for us and a very simple main
code. We also reuse some code with applied before to count the
nucleotides. Let&#8217;s see the code, discussion just after it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
<span class='line-number'>47</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python</span>
</span><span class='line'>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">nucleotide_percentage</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; As &#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Cs &#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Gs &#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Ts&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">sequence_identity</span><span class="p">(</span><span class="nb">set</span><span class="p">):</span>
</span><span class='line'>    <span class="n">iden</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">count</span> <span class="o">=</span> <span class="mf">0.0</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
</span><span class='line'>        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">])):</span>
</span><span class='line'>            <span class="k">if</span> <span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">n</span><span class="p">]</span> <span class="o">==</span> <span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="n">n</span><span class="p">]:</span>
</span><span class='line'>                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>        <span class="n">iden</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">]))</span>
</span><span class='line'>        <span class="n">count</span> <span class="o">=</span> <span class="mf">0.0</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">iden</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>    <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>    <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>
</span><span class='line'><span class="n">identity</span> <span class="o">=</span> <span class="n">sequence_identity</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)):</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequenceset</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>
</span><span class='line'>        <span class="k">print</span> <span class="s">&#39;sequence identity to next sequence : &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">identity</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span class='line'>    <span class="n">nucleotide_percentage</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span class='line'>    <span class="k">print</span>
</span></code></pre></td></tr></table></div></figure>


<p>Well, not many
new things here. The main one might the be the assignment of the
variable <code>count</code>, which receives initially a value of <code>0.0</code>, which in
this case is a float. This variable is used the calculate the sequence
identity which is a percentage, or a relative value. If we had
initialized <code>count</code> with zero, our division <code>count/len(set(x)</code> would
always be zero, due to the fact that count would have been an integer.
The ideal input for this script would have equal minimum and maximum
sequence lengths, as it the simulated set would look more like an
alignment with no indels. Something like:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'><span class="nv">$&gt;</span>python simulation.py 10 30 30
</span></code></pre></td></tr></table></div></figure>


<p>As is
pointed out in BPB, this example is more an indication that we are able
to use our Python skills to actually make some real code, with some real
output. One issue with this example is the fact that we only calculate
sequence identity of two sequences at a time. It would be ideal to have
sequence identity between all simulated sequences. This would require
much more code, of course a good educational step, but it is something
that can be easily obtained with classes and we will see this later on.
With this entry, we finished our Section 4 and we will start Section 5
with Python&#8217;s dictionaries, moving to FASTA files and classes next.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/04/the-current-state-of-bioinformatics-software/">The current state of bioinformatics software</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-04T18:19:33-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have been planning to write about the current state of bioinformatics software development for quite sometime. Not something scientific, but something that I see along the days and weeks in the lab.</p>

<p>This post on <a href="http://www.ghastlyfop.com/blog/2007/03/software-availabilty-quick-survey-using.html">Flags and Lollipops</a> showed the current state of availability of bioinformatics software published in the past years. The post itself says it is not scientific, but give (along with the comments) an excellent base to what I see daily.</p>

<p>In no particular order, in my opinion there are three classes of bioinformatics software developed in academia:</p>

<ul>
<li><p><strong>for-publication</strong> programs - these are the ones developed with the sole purpose of publication. They are poorly and not actively maintained, sometimes (a small percentage) lack good documentation and usually perform one type of analysis that is available in many different packages</p></li>
<li><p><strong>closed source</strong> programs - these are the relatively well maintained packages, with good or sufficient documentation. The point here is the fact that they have source not available, and in some cases they require some extra installation in order to work, not mentioning the fact that you sometimes have to fax the license agreement in order to obtain it. If you work in a close system (my case) with no admin access, it might take a week or two to have such package installed, counting the time to send the fax, to receive the application and to ask the sys admin to install it and its requirements</p></li>
<li><p>and finally <strong>open source</strong> packages - these fall in a 50/50 tie regarding maintenance: they can bell well maintained or basically abandonware. The well maintained cluster is usually well documented, have a very active group of developers and users receive good support. The abandonware group in some cases are <strong>for-publication</strong> and in other cases they are abandoned because the grad student/post-doc left the lab/job and now has to pay the bills with a day job that does not give him/her time to code.</p></li>
</ul>


<p>I have developed and released software in two of the three categories above. Myself, I don&#8217;t like to release closed source programs, but sometimes the institution where you are working at make you do that. All my extra-lab software is open source, or at least you can ask me to get the source.</p>

<p>Regarding the for-publication software, this is the most pernicious category. Sometimes they are a by-product of some larger research, sometimes they are just bad software that replicate another widely used and available package, sometimes a tool that has not scientific merit. I respect large and small projects that release a very well programmed application, with a good documentation and thrives to give support to the users. Projects that clearly show that they had project, development and testing phases along the way, and actually enhances our scientific knowledge or make scientific analysis more straight forward. But lately, I have seen greasemonkey scripts being published, what can be compared to the MS Excel macro published some years ago in Bioinformatics.</p>

<p>As always, just my 2 cents.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/04/a-script-to-simulate-dna-sequence-sets/">A script to simulate DNA sequence sets</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-04T12:17:22-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In Beginning Perl for Bioinformatics the chapter that covers simulated
mutations on a DNA sequence is quite verbose and the code examples
employ some subroutines to do what we have done on the last post.
Basically the code example that generates a random DNA sequence is the
last one on the chapter, but it was the first one we covered. Also this
code example has a twist that our code from the last post does not have,
which is it allows you to generate a set of sequences with different
length instead of one sequence with fixed length that our script does.
So, the script that generates a user defined simulated DNA sequence set
is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>
</span><span class='line'><span class="c">#!/usr/bin/env python</span>
</span><span class='line'>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>    <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>    <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple and efficient. First we
define a function that generates a simulated DNA sequence from the four
nucleotides, again using the <code>random.choice</code>. The sequence length is
based on the parameter received by the function. The sequence length is
another random number defined in the main body of the script. This
random number is generated by <code>random.randint</code> with a range based on the
arguments given by the user when running the script. And finally, the
number of sequences to be simulated is define by the first parameter.
Seventeen lines. Next we will see how to draw some scientific
information about the sequences, such as sequence identity and
nucleotide frequency.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/02/random-sequences-in-python/">Random sequences in Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-02T12:57:24-04:00" pubdate data-updated="true">Apr 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>There is a reason to say that Python has batteries included. On the last
post we have seen a small example of randomization in Python, generating
random integer values for a (extremely) simple dice game. We move to
another example, still simple which will allow us to generate random DNA
sequences. The script is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">dnaseq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s">&#39;ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">result</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dnaseq</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">result</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not fancy at all, just plain simple (yet again). We
import a couple of modules, <code>sys</code> and <code>random</code>, and ask for the sequence
length as a script parameter. <code>dnaseq</code> is a list containing a tandem
repeat of ACGT, and from there we will extract our random nucleotides.
In fact <code>dnaseq</code> could have been &#8216;ACGT&#8217; only. The fact that we create a
string and convert it to a list, is just for convenience of writing
<code>'ACGT...'</code> easier than <code>['A', 'C' ...]</code>. We then declare an empty
string that will be used to store the random sequence. And inside the
loop, the command that does all the magic: <code>random.choice()</code>. This
command will return a random element from the list passed as subject.
Using it inside a loop we will get a random nucleotide on each iteration
and add it to our string. This is a very simple command, but at the same
time extremely powerful and easy to implement. A good exercise from this
would be to modify the <code>dnaseq</code> string and see if there is any change in
the final random sequence.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/30/ramdomization/">Ramdomization</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-30T15:19:30-04:00" pubdate data-updated="true">Mar 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We start section 4 with a very short introduction to randomization.
Chapter 7 of BPB discuss the use of randomization to obtain mutations in
DNA and protein sequences. The example given in the book is at the same
time simple and interesting, as it creates a paragraph from random
selections of nouns, adjectives, verbs and other grammar elements. Here
we are going to to create a very (stress on very) simple dice game,
where each time you run the script it will throw two dices for you and
two dices for the computer. It then prints the sum of the dices and
tells who won the match. Randomization is an important feature of
computer languages. There are different researchers involved in the
creation of the best approaches to generate random number in computers.
Random number are important in the simulation of different natural
processes, such as genetic mutation, gene drift, epidemiology, weather
forecast, etc. The better the generator, the better the simulation.
Python has a random module included, with a myriad of functions that can
perform different randomizations. In this post we will see the <code>integer</code>
randomization, and in later entries we will see some other powerful
functions. Our script is quite simple, and the only new aspect for us
here is the random module and the <code>randint</code> function.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'>
</span><span class='line'><span class="n">dice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">dice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">mine</span> <span class="o">=</span> <span class="n">dice1</span> <span class="o">+</span> <span class="n">dice2</span>
</span><span class='line'>
</span><span class='line'><span class="n">his</span> <span class="o">=</span> <span class="n">computerdice1</span> <span class="o">+</span> <span class="n">computerdice2</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&#39;mine = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mine</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; vs. computer = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">his</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">if</span> <span class="n">mine</span> <span class="o">&gt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;I won&quot;</span>
</span><span class='line'><span class="k">elif</span> <span class="n">mine</span> <span class="o">&lt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Computer won&quot;</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Tie. Try again&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>random.randint</code> is a function that generates an integer random number
between a range specified by the number between parentheses. In the
above case, we are using a dice of 6 values. So, for every call of
<code>random.randint(1,6)</code> a random number between 1 and 6 will be generated.
We store this number in a variable, sum the user&#8217;s dices and the
computer&#8217;s and check with a <code>if</code> clause to see the winner. A good
exercise would be to make this script interactive, allowing multiple
matches. Next we will see another random module function and then we
will generate mutations on DNA sequences.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/26/command-line-arguments-and-a-second-take-on-functions/">Command line arguments and a second take on functions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-26T13:18:48-04:00" pubdate data-updated="true">Mar 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen, briefly, how to define and use a function in Python. Now
we are going to jump forward a bit and create a new function and at the
same time take a look on command line parameters that can be passed to
the script. If you have used command line applications before, you might
have encountered programs that asks for a file name, a calculation
parameter, etc to be passed in the command line. Python scripts are no
different, they accept such parameters. For this we have the <code>sys</code>
module that has system specific parameters and functions. We have used
before the <code>sys.exit</code>, imported as an extra module function.</p>

<p>Every operating system (even Windows) has arguments in its command line, and
programming languages usually call such arguments <code>argv</code> (in the C/C++
you have argv in the parameters of the main function). Lists in Python
start at 0 (zero), and for the argument list the first item is the
script/program name. Basically if we have this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'><span class="nv">$&gt;</span> python myscript.py DNA.txt
</span></code></pre></td></tr></table></div></figure>


<p><code>myscript.py</code> is the argument 0 in the
list and <code>DNA.txt</code> is the argument 1. So whenever we create a script that
receives arguments in the command line, we have to start (in most cases,
be aware) from 1.</p>

<p>In Python using system arguments in the CLI will look
like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">filename</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">valueone</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>We will a variation of our
previous script that counts the bases, now with command line arguments
and a function (with no &#8220;error&#8221; checking at first)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python import sys </span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">count_nucleotide_types</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">totalA</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalC</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalG</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalT</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequencefile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequencefile</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">values</span> <span class="o">=</span> <span class="n">count_nucleotide_types</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;As&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Cs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Gs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Ts&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>Few new things here. We created a function <code>count_nucleotide_types</code> that should receive a string
containing the sequence. The &#8220;real&#8221; first line of the program flow is
the one that gets the name of the file from the command line argument,
open and read it. We then convert the list to a string, modify it a but
and throw it to the function. Get the result back, and done. With
functions we actually don&#8217;t save coding time/length (at least here), we
make out code more organized, easier to read and somewhat easier to
someone else read and understand it. It is not a good coding practice to
have long programs/scripts with no functions, no subdivision, no
structure. Functions are sometimes good program nuggets that can be
reused in the same application or even ported/copied to other
applications and reused indefinitely. Soon we will see a function and
class that reads a FASTA file in Python that can be used anywhere in any
program that needs such feature. Try the code and come back later for more.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/22/python-functions-simple-example/">Python functions: simple example</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-22T09:04:10-04:00" pubdate data-updated="true">Mar 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Python subroutines do not exist. <a href="http://www.diveintopython.org/getting_to_know_python/declaring_functions.html">Everything is a function, all
functions return a value (even if it&#8217;s None), and all functions start
with <code>def</code></a>.
This statement is from Dive into Python, a book on Python programming
available for free. As mentioned Python functions start with the word
<code>def</code>, which is followed by the function name that is followed by the
arguments the function receives in between parentheses. Something like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span></code></pre></td></tr></table></div></figure>


<p>Usually Python coders
(sometime called Pythonistas, among others), following the Python coding
style (that states: <em>Function names should be lowercase, with words
separated by underscores as necessary to improve readability</em>) name
their functions with words separated by underscores. And we are going to
use this style here, whenever a function becomes handy. The parameters
passed to the function (above <code>somevalue</code>) do not have a datatype,
Python should handle it whatever is being passed. It is also attribute
of your code to handle the parameter/value passed inside the function
and avoid errors. Functions also follow the same indentation of normal
programming and the line after the declaration should be indented with
four spaces</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span><span class='line'>    <span class="n">do_something</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, let&#8217;s
warmup with functions. The following script is just the start: it adds a
poly-T tail to a DNA sequence. We are going to use our old friend
AY162388.seq. I will be back after the script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">add_tail</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="n">seq</span> <span class="o">+</span> <span class="s">&#39;TTTTTTTTTTTTTTTTTTTTT&#39;</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&#39;AY162388.seq&#39;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">sequence</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">add_tail</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not very useful, at first sight, but gives us an
impression of what a function looks like. Basically we define a function
<code>add_tail</code> that receives <code>seq</code> as a parameter. Don&#8217;t worry about
variable scope now, we will see it later. The rest of the script is just
like things we saw before, except for the line
<code>sequence = add_tail(sequence)</code>. Here we are saving memory (yep, not
that much and not even impressive) by assigning the return value of the
function to the same string where we have the sequence stored. Run the
scritp and get ready for the command line arguments.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/16/reccommended-read/">Recommended read</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-16T13:20:13-04:00" pubdate data-updated="true">Mar 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the latest post I forgot to mention a book that is wonderful for those with no programming habits and for those who have (in this case more like vices).</p>

<p><a href="http://www.amazon.com/gp/product/1593271190?ie=UTF8&amp;tag=genedrift-20&amp;linkCode=as2&amp;camp=1789&amp;creative=9325&amp;creativeASIN=1593271190">Code Craft: The Practice of Writing Excellent Code</a><img src="http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=as2&amp;o=1&amp;a=1593271190" alt="" /></p>

<p>It is a very good read, with several insights on how to write programs.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/16/on-notes-to-a-young-computational-biologist/">On notes to a young computational biologist</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-16T11:14:15-04:00" pubdate data-updated="true">Mar 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Yes, I never blogged, never liked to. I always been a tad recluse when posting opinions, comments and other things on websites. But recently I have found many nice blogs and personal pages that have insightful posts, interesting information and tips-and-tricks for your daily routine. Pages that are really a good read. I have the feed to <a href="http://www.nodalpoint.org">nodalpoint</a> which is more a community than a personal page, and the latest feed brought me to this <a href="http://boscoh.com/protein/notes-to-a-young-computational-biologist">post</a> written by Bosco in his blog. The post gives a list of advices for young computational biologists.</p>

<p>I cannot agree more with his post, and I wish that I had the same information in 1995 when I started my internship in the lab where I would come to do my <a href="http://www.genedrift.org/winpop.php">masters</a>. Back then we did not have the internet the way it is today, with the amount of information, the large number of open-source software available, no easy Linux installation, among other things that make anyone&#8217;s life easier (sometimes) today. As for any kind of list, one cannot be satisfied until s/he gives her/his own opinion about it. All his eleven entries are great, and here I take the liberty to expand each one of them giving my humble opinion (as pointed <a href="http://www.craigslist.org/about/best/sfo/279126743.html">elsewhere</a>, 97% of all advice is worthless, use whatever is good for you and throw away the rest):</p>

<p>1- On scripting language: definitely a must, either one of them. But I would stress some knowledge of shell script as a key point too. You don&#8217;t have to be an expert on bash scripting, for instance, but some basic stuff might come handy when nothing else is available (such as a Solaris machine with no Python, Perl or Ruby). Use this <a href="http://tldp.org/LDP/abs/html/">link</a> as a reference guide (or have any published book at arm&#8217;s distance).</p>

<p>2- On programmable statistics package: just one addition R, free, open source and a great user base.</p>

<p>5 - On delete data that you don&#8217;t need: I would do that, but be careful. Backup, compress, move to obscure partitions, directories, whatever is easier before deleting data. You might think that the data is useless today, until tomorrow proves you wrong. The Gb price is going down every day, we are not in the age of <a href="http://www.findarticles.com/p/articles/mi_m0SMG/is_n9_v13/ai_13926185">doubledisk</a> anymore. But annotation of everything you have is crucial.</p>

<p>8- On command line plotting program: very nice advice, just pick one that fits you better. This is also a must if you don&#8217;t want to publish lame graphs in your papers.</p>

<p>9- On programming editor: I personally use <a href="http://kate-editor.org/">Kate</a> , but any other option is great. Be sure to find something you are comfortable with.</p>

<p>I cannot find further advices on the other points, but I would again take the liberty to add some points of my own (always follow the 97% rule):</p>

<p>1- As one of the comments in his post pointed out: subversion. Maybe not the greatest invention after sliced bread, but it really gives you a great deal of help when working in different places/machines. Say you have a great idea at home and want to update your scripts but forgot to copy everything to your usb disk. Yep, subversion is there.</p>

<p>2- Be a generalist: learn as many languages as you can. You don&#8217;t have to be an expert in any of them [after all you are a biologist, right?(??)], but sometimes one feature that will save your day is available in a language that you never noticed.</p>

<p>3- System administration skills: yep, you will need then. Not the crazy things as knowing where every log file is in your system, but at least some handy commands help you check your system or maybe give you more ability to use such system. One extra advice here: don&#8217;t buy any books on Windows system administration.</p>

<p>4- Three words: Linux, Linux and Mac OSX. That&#8217;s what you really need. Yep, you need to know your way on Windows, but when you master the power of the command line, new worlds will open for your. Always remember: you can type faster than you click.</p>

<p>Thanks for Bosco to post his insightful advices. I just added my two cents of worth(less) advice.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/15/interfaces/">Bioinformatics and interfaces</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-15T14:39:15-04:00" pubdate data-updated="true">Mar 15<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the reply to my second Geneious review, an interesting point was raised: building interfaces, &#8220;which is one of the biggest workloads in writing software&#8221;.</p>

<p>My first programming language was the old <a href="http://www.webdevelopersjournal.com/archive/vbasic.html">Visual Basic 3.0</a> which has a visual driven user interface and programming approach. To create windows and dialogs in VB is an extremely easy process, dragging and dropping objects in a window, easily setting properties, menus and features. The interface design is straightforward, and the burden of a user-friendly program lies on the programmer and his/her ability to set the objects on the screen, manipulate their function and add programming elements to operate them.</p>

<p>But Visual Basic has it drawbacks, first by not being a powerful language and restrictions of only running resulting programs on Windows machines, etc. In Bioinformatics, Visual Basic is great for simple applications where the processing speed, heavy number crunching, portability, intense I/O operations and other high-performance issues are not a requirement. Applications like <a href="http://www.genedrift.org/winpop.php">WinPop</a> (I will focus on my personal experience), that had version 1.0 released late 1998 (&#8220;compiled&#8221; with VB3.0) and version 2.5 which was released in 2004 and was &#8220;compiled&#8221; with the more advanced VB 6.0. WinPop is the perfect example of a good project for VB: simulates and represents simple Population Genetics phenomena with limited input and outputs nice looking graphs to help the user understands what is going on in each one of the models. WinPop is also geared to undergrad and grad students that need to learn Population Genetics, and can be used in any Windows machine (Windows 95 and up) and does not require a state-of-the-art box to run nor it will allow you to discover a major breakthrough in science.</p>

<p>VB programming is focused on the visual part. Your whole approach to develop the software has to consider the objects you see (and sometimes not) on the screen, how to link them if needed, which event will trigger what and so on. It is different than writing your script in Python, or your number crunching application in C++ or Java. In VB, there is the possibility of generating command line applications, but this is more difficult than creating one with a window. In Java you can opt to make a command line application or to create a GUI for it with Swing. In C++, everything is command line, unless you use one the GUI frameworks available, like <a href="http://www.trolltech.com">Qt</a> or <a href="http://www.wxwidgets.org">wxWidgets</a>, which in some case is like learning a new language by itself. Some of these frameworks were ported to other languages too, like wxPython, wxPerl, PyQt.</p>

<p>Until recently, advanced Bioinformatics applications were command-line only. Examples can be found everywhere: alignment packages (Clustal has), Blast (web-interface, not very powerful), TFBS determination, HMM, phylogenetics (PAUP for Mac and Mega are most recognizable exceptions), etc, etc. Nowadays it is becoming easier to find good programs with nice interfaces, like <a href="http://www.genedrift.org/sugg.php">Geneious</a>, which are portable and powerful. But still, in my opinion researchers don&#8217;t give much attention to the interface development.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/36/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/34/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
