
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="If you don&#8217;t like Snap preview tooltips (oh, they are so useful) you came to the right place. No, in fact it wasn&#8217;t me that invented it, &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/27">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/15/for-every-problem-there-is-a-solution-thankfully/">For every &#8220;problem&#8221; there is a &#8220;solution&#8221;. Thankfully.</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-15T14:07:48-05:00" pubdate data-updated="true">Feb 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>If you don&#8217;t like Snap preview tooltips (oh, they are <em>so useful</em>) you came to the right place. No, in fact it wasn&#8217;t me that invented it, the company that created Snap did it. You don&#8217;t need to thank me.</p>

<blockquote><p>How can I turn Snap Shots on or off?</p></blockquote>

<p>Click here (on their <a href="http://www.snap.com/snapshots_faq.php#2">website</a> FAQ) to disable Snap Shots. Please note you must reload the page or clear your cache for the change to take effect. Deactivation uses cookies, so if you delete your cookies, Snap Shots will be re-activated for all sites.</p>

<p>The JavaScript will save a cookie on your browser and you are free to surf without any further annoyance. Thankfully.</p>

<p><em>from <a href="http://www.ofzenandcomputing.com/zanswers/304">1</a> (Of Zen and Computing) and <a href="http://www.lifehacker.com/software/annoyances/disable-snap-previews-on-web-sites-232846.php">2</a> (Lifehacker)</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/15/parsing-a-genbank-file-another-approach/">Parsing a GenBank file, another approach</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-15T11:35:13-05:00" pubdate data-updated="true">Feb 15<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of entries we learnt a little bit about <code>sets</code> and that
newer Python versions make its use a lost faster than <code>dictionaries</code> to
uniquify <code>lists</code>. And soon we will see good examples of <code>sets</code> usage.
And now for some completely different &#8230; we go back to parsing GenBank
files. We saw sometime ago how to extract some simple information
contained in this type of files. I am working on a project that requires
a lot of work in GenBank files, and instead of learning how to use an
available parser I decided to write a simple one on my own. The
requirements were simple: extract all protein sequences from a whole
genome GenBank file and also the DNA from the sequence available in it
(this time we will see how to get the proteins and next time the
latter). We all know that GenBank files have an annotation section, with
IDs, references, organism and some other information at the top. The DNA
sequence at the bottom, after ORIGIN, and the protein sequences are in
annotated regions of CDSs. An example is below:</p>

<p><code>     gene            8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /db_xref="GeneID:855380"      CDS             8330..9475                      /gene="COS1"                      /locus_tag="YNL336W"                      /experiment="experimental evidence, no additional details                      recorded"                      /note="Protein of unknown function, member of the DUP380                      subfamily of conserved, often subtelomerically-encoded                      proteins"                      /codon_start=1                      /product="Cos1p"                      /protein_id="NP_014063.1"                      /db_xref="GI:6323993"                      /db_xref="SGD:S000005280"                      /db_xref="GeneID:855380"                      /translation="MKENELKNEKSVDVLSFKQLESQKIVLPQDLFRSSFTWFCYEIY                      KSLAFRIWMLLWLPLSVWWKLSNNCIYPLIVSLLVLFLGPIFVLVICGLSRKRSLSKQ                      LIQFCKEVTENTPSSDPHDWEVVAANLNSYLYENKAWNTKYFFFNAMVCQEAFRTTLL                      EPFSLKKDEAAKVKSFKDSVPYIEEALGVYFREVEKQWKLFNSEKSWSPVGLEDAKLP                      KEAYRFKLTWFLKRISNIFMLIPFLNFLCCIYVSRGMCLLLRTFYLGWILFMLVQGFQ                      NMRMIVLSVKMEHKMQFLSTIINEQESGANGWDEIAKKMNRYLFEKKVWKNEEFFFDG
IDCEWFFSHFFYRVLSAKKSMRALSLNVELWPYIKEAQLSCSEESLA"</code></p>

<p>In a GenBank file each coding gene corresponds to at least one protein.
The first line, gene, contains the nucleotide region, as the CDS line.
The other lines, starting with a slash provide information about the
gene and the last piece of information is the DNA translation with the
whole protein sequence. In my case, the first objective was to obtain
the protein sequence, get the protein ID and the gi ID from each entry
in a genome file. An initial assessment, tell us that creating a class
for each protein in the file is a good start.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Protein</span><span class="p">:</span> <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gi</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">gi</span> <span class="o">=</span> <span class="n">gi</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="nb">id</span>
</span><span class='line'>  <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>So we declare the class and have elements for the
IDs and the amino acid sequence. The next problem is how do get each
gene/CDS information, extract what we need and store in a list of class
instances? This is much like we do with our FASTA entries, reading line
by line and looking for FASTA sequence names, which start with <code>&gt;</code>. For
GenBank files we will look for lines containing the word gene. Of course
this would mean that every gene that has this word in its name will be a
problem, so we match against <code>'  gene  '</code> (the word gene flanked by two
spaces).</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39; gene &#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">if</span> <span class="n">index</span> \<span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>      <span class="c">#parses the CDS and appends to a list</span>
</span><span class='line'>      <span class="n">proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_entry</span><span class="p">(</span><span class="n">entry</span><span class="p">))</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>      <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="c">#found the DNA sequence, we can stop now </span>
</span><span class='line'>      <span class="k">break</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="n">entry</span> <span class="o">+=</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p> In the above excerpt we read
the file line by line, having an index just to keep the code aware of
which entry we are at. Every time we find a line with <code>'  gene  '</code>, we
check for the index and if it is larger than one we parse the entry that
has been compiled already and start a new one. If we find &#8216;ORIGIN&#8217; we
bail out, breaking the loop, and in every other case we add a line to
our current entry. When this is sent to the parsing function we clear
the string and start compiling the information again. Using a string for
an entry might be a second choice, as lists are easier to manipulate
after (and eventually we will transform our string in a list by
splitting it). Notice that we have a call to a function <code>parse_entry</code>,
let&#8217;s see how this function is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">parse_entry</span><span class="p">(</span><span class="n">gene_data</span><span class="p">):</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="n">gene_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">gi_id</span> <span class="o">=</span> <span class="s">&#39;gi&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;GI:&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>      <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>          <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">return</span> <span class="n">Protein</span><span class="p">(</span><span class="n">gi_id</span><span class="p">,</span> <span class="n">prot_id</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p> As mentioned
above, we split the string in a list (we will check if there is any
improvement in speed if we use a list from the beginning) and then
analyse each one of the lines. In this case, we only need the protein,
the gene name and, evidently, the protein sequence. Dissecting the
excerpt above we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/product&#39;</span><span class="p">)</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;protein_id&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">prot_id</span> <span class="o">+=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>This will get us the protein ID from
<code>/protein_id="NP_014063.1"</code> and also the gene name/info from
<code>/product="Cos1p"</code>, which them can be stored in the <code>prot_id</code> string
that will be passed to the class instance. The gi ID can be extracted in
similar fashion. In order to get the sequence we have to do a small
trick, that can be seen on the excerpt below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/translation&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">:]</span>
</span><span class='line'>  <span class="n">temp</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">temp</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">)):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">gene_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p> We find the beginning of the
translation information by checking the line that contains
<code>/translation</code> and we extract the initial part of the sequence my
finding the equal sign and getting everything after it. Then the trick.
We know that the end of the translation information will be the last
line of the whole gene/CDS entry. So we check for the index of the first
sequence line and store it in a temp variable. Using this variable as
the first value of a range we are able to get every other line of the
sequence. That simple. To do the output, we can just iterate the list of
class objects and print all the information in FASTA format or any other
that we wish. Put everything together and we have a simple but effective
GenBank parser. Please notice that this script wouldn&#8217;t work in every
possible scenario, but with minor tweaks it should be applicable in most
cases. Next time, we will see how we can get the DNA sequence in an
effective way. <em>An example of GenBank file can be found
<a href="http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?db=nuccore&amp;id=117937805">here</a></em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/14/3-masterpieces/">3 masterpieces</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-14T15:04:10-05:00" pubdate data-updated="true">Feb 14<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Leaving the Science behind, and thanks to the &#8220;beauty&#8221; of open office spaces fro bioinformatics development, I am able to enjoy three great music masterpieces (AKA the records I would take to an isolated island)</p>

<p>1- <a href="http://www.amazon.com/Black-Love-Afghan-Whigs/dp/B000002HLG/ref=pd_bbs_1?ie=UTF8&amp;s=music&amp;qid=1203019110&amp;sr=8-1">Black Love - Afghan Whighs</a></p>

<p>2- <a href="http://www.amazon.com/Heaven-Up-Here-Echo-Bunnymen/dp/B0000E2Y8Z/ref=sr_1_2?ie=UTF8&amp;s=music&amp;qid=1203019254&amp;sr=1-2">Heaven up Here - Echo and the Bunnymen</a></p>

<p>3- <a href="http://www.amazon.com/Starfish-Church/dp/B0009NR7VI/ref=sr_1_1?ie=UTF8&amp;s=music&amp;qid=1203019320&amp;sr=1-1">Starfish - The Church</a></p>

<p><em>Disclaimer: none of the links above are connected to the Amazon Associates program. I searched Amazon for the title and copied the url that appeared on my address bar.</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/07/we-are-doomed/">We are doomed!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-07T10:37:32-05:00" pubdate data-updated="true">Feb 7<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Tornadoes? No. Earthquakes? No. Zombies? No. <a href="http://news.yahoo.com/s/afp/20080207/ts_afp/fblengproverseas">They want to control our minds and lower the standards globally</a>. Enough of Beckhams, Rooneys and alikes. I want to see real football (soccer for the football-played-with-the-hand people).</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/03/tnt-review-of-the-graphical-interface/">TNT: review of the graphical interface</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-03T12:27:27-05:00" pubdate data-updated="true">Feb 3<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This time I am writing about the Windows version of TNT which is somewhat graphical. I am currently using Windows Vista Home Premium.</p>

<p><strong>Interface</strong></p>

<p>TNT&#8217;s interface is far from being stylish, and it is quite simple and intuitive. Most of the options you would need to calculate trees and modify your analysis are on a toolbar and on menus at the top.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2008/01/tnt1.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2008/01/tnt1.png" alt="Toolbar" /></a></p>

<p>The main area is occupied by a log/text area where the current actions/selections are displayed from time to time. The default font is clear and large and there are ways to format it as you wish. You also have buttons that allow you to increase or decrease the font size.</p>

<p><strong>Running TNT</strong></p>

<p>Running the program with the interface was clearly more pleasant than using the cryptic command line version. TNT starts with an blank area, toolbar and menu, as an usual Windows application. I decided to open one of the three example files that come with it, zilla.tnt.</p>

<p>First thing would be to obtain the tree, and intuitively I ended up pressing the button the toolbar with the man running. Bingo, that&#8217;s what I needed. TNT, then, showed me a options dialog.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2008/01/tnt2.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2008/01/tnt2.thumbnail.png" alt="tnt2.png" /></a></p>

<p>This makes life much easier. All options for tree calculation were right in front of me, and I even had access to specific settings for each method. Then I could see the full range of features that TNT offers. I set to run with default parameters adding the ability to use the ratchet. The run went smoothly, quite fast and an informative progrss dialog displayed the status. (One thing though: TNT developers have created some of the dialogs without a frame or border, what makes impossible to move then around. Sometimes they got in the way of information displayed on the log that might have been important to set the runs).</p>

<p>After the calculation ended, TNT told me that 4 trees were retained from my run. So I wanted to check the trees, not that I was worried if my analysis was right or wrong. One fault was evident: tool bar buttons don&#8217;t have a tooltip and this made a little bit difficult to understand some of the buttons&#8217; pictures. The one to create a consensus is pretty clear, as the one to trash all trees obtained. So I decided to check the menu.</p>

<p>As in any normal Windows program TNT&#8217;s menu is richer in options than the tool bar for obvious reasons. I went to the Trees menu and selected View. All four trees were then displayed on the screen with a scrollbar on the side, much better than the command line version. There is an option that allows the user to collapse the branches, but in my opinion it still a little bit awkward. You have first to unlock the tree and then the program allows you to &#8220;shrink&#8221; nodes. One click at a node adds a red spot to it, another puts a green spot, and a double-click &#8220;shrinks&#8221; the node.</p>

<p>There are some errors on some tree display options. For instance, when the consensus button is clicked it brings up a dialog to change the calculation and after you click OK, TNT generates the consensus and automatically displays a tree. But the window containing the tree is larger than my desktop, has no scrollbar and cannot be dragged. I don&#8217;t know if this an error, a bug or an incompatibility of TNT with Vista. The only way to exit is by using ESC. I will try on XP and update the review.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2008/02/tnt3.png">TNT displaying a consensus tree (full screen capture)</a></p>

<p><strong>Conclusions</strong></p>

<p>TNT seems to be a powerful software for phylogenetic analysis if you are looking mainly for parsimony. Also, its algorithms seems to have inherited the best of Hennig86. The graphic interface, even though not groundbreaking by any means, makes its use very easy. The program is fast, has many different options for parsimony tree calculation and it is an excellent option for determination of trees to be used as input for slower methods, as pointed out in one of the comments of the CLI review. There are still some errors and bugs and some features I would change in its GUI, but that&#8217;s true for most of the software available in any field.</p>

<p><em><strong>Update: another review is available <a href="http://blindscientist.genedrift.org/2008/01/30/tnt-a-comprehensive-review/">here</a>.</strong></em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/02/is-this-the-year-chelios-retires/">Is this the year Chelios retires?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-02T10:32:28-05:00" pubdate data-updated="true">Feb 2<span>nd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content">
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/01/powerpoint-tips/">Powerpoint tips</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-01T22:59:56-05:00" pubdate data-updated="true">Feb 1<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In the same BMJ issue that came out with a <em>ridiculous</em> article about testes, a scientist tries to give some Powerpoint tips. I couldn&#8217;t find the link so I am pasting it below:</p>

<blockquote><p>The main purpose of a PowerPoint presentation is entertainment. Intellectual content is an unwarranted distraction. In preparing a PowerPoint presentation, aesthetics should transcend substance. The background colour scheme and logo for your slides should be selected for maximum emetogenic potential. The first inverse ridicule rule of PowerPoint presentation states: “The more lines of writing that can be coerced onto a slide and the smaller the font, the lower the risk of anyone criticising any data which has accidentally been included.” The second rule states: “The number of slides you can show in your allotted time is inversely proportional to the number of awkward questions which can be asked at the end.” PowerPoint has superseded the carousel era, when presentations were severely limited by the number of slots in the slide carousel and the risk of dropping the lot seconds before your talk. Plagiarism laws do not apply to PowerPoint, so cartoons of marginal relevance but high entertainment value can be downloaded and shown at suitable intervals to maintain audience mirth while minimising critical capacity. Research has shown that the ideal cartoon: data ratio is 5:1.</p></blockquote>

<p>The seasoned PowerPoint artist or PowerPointilliste has refined the presentation into a son-et-lumiere extravaganza, in which scattered dots and luminescent clumps of meaningless datasets hurtle on to the screen from all points of the compass, to the strident strains of Handel’s Fireworks Music, building inexorably to a Fantasia-style Sorcerer’s Apprentice climax. This fulfils an important subsidiary purpose of the PowerPoint presentation—to act as a bioassay of the epileptic threshold of the audience, a form of PowerPoint EEG. PowerPoint has spawned a number of hitherto unrecognised diseases. These include PowerPoint phobia (PPP), PowerPoint stress disorder (PPSD), and a form of depression called PowerPointlessness. Yet another purpose of the PowerPoint presentation is to test the capacity of the regional electricity grid. In case this should be found wanting, and your presentation succeeds in fusing the power supply to the surrounding region, it is advisable to have a back-up presentation, a box of matches and a Chinese lantern.</p>

<p>David Isaacs (davidi@chw.edu.au) is a senior staff specialist in the department of immunology and infectious diseases and Dominic Fitzgerald is a senior staff specialist in the department of respiratory medicine, at the Children’s Hospital at Westmead, Sydney. Stephen Isaacs is a consultant at Waltham Forest Child and Family Consultation Service, London. at Children’s Hospital at Westmead.</p>

<p>I will try to follow that in the next ISMB.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/01/talking-about-chocolate/">Talking about chocolate &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-01T18:17:35-05:00" pubdate data-updated="true">Feb 1<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Don&#8217;t forget your hot cocoa when reading the last edition of the Bio::Blogs, at the <a href="http://www.bioinformaticszen.com/2008/02/february-2008-edition-of-bioblogs/">Bioinformatics Zen</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/01/thats-why-we-pay-to-access-scientific-journals/">That&#8217;s why we pay to access scientific journals?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-01T18:14:25-05:00" pubdate data-updated="true">Feb 1<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Alerted by a message sent to the EvolDir mailing list I went to check the <a href="http://scientific.thomson.com/products/scientificdirect/">Scientific Direct Customized Mailing Lists</a>. Very entertaining product.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/02/01/ummm-chocolate/">Ummm, chocolate</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-02-01T11:50:58-05:00" pubdate data-updated="true">Feb 1<span>st</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Apparently is not only in bioinformatics that you can publish Greasemonkey scripts, Excel macros and 40* errors on PubMed entries.</p>

<p>BMJ also has some <a href="http://www.bmj.com/cgi/content/full/335/7633/1287?fmr">very useful science</a> and with its impact factor higher than 9, I can see a lot of opportunity to publish there. I guess just this short article alone would be able to &#8220;provide a rare opportunity for the chocolate industry to become <strong>palpably</strong> involved in public health promotion&#8221;.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/28/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/26/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
