
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Apparently, upstate NY and southern Ontario did not get the memo or the report that the Global Warming is here.
">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/38">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/global-warming-no-memo-no-report/">Global warming: no memo, no report</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:44:52-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Apparently, upstate NY and southern Ontario did not get the memo or the report that the Global Warming is here.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/wikipedia/">Wikipedia</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:43:45-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Wikipedia is growing daily, and becoming a a major reference for many subjects, from arts and entertainment, sports to quantum physics and population genetics. Myself, I have been following its growth through websites like <a href="http://reddit.com">reddit.com</a> and by the increase of Google top hits directing to Wikipedia entries.</p>

<p>In December 2005, a small study compared the freely available encyclopedia to Britannica (summarized <a href="http://www.wired.com/news/culture/0,1284,69844,00.html">here</a>) and found errors on both, almost leveling them. I cannot say about the quality of the content of Wikipedia in English, I only use it to quick scientific searches on topics that I do not have great knowledge or to mundane topics, like writers&#8217; and bands&#8217; biographies.</p>

<p>Lately, I have focused on checking the content of Wikipedia in Portuguese, especially entries that relate to biology, bioinformatics and genetics, with an anonymous profile. I was appalled with the poor quality of the entries, and I decided to edit some articles, in some places correcting minor grammar mistakes and in other replacing entire sections that were completely off the mark. I can say that a gave a minor contribution, I could and can give more if I had more time to edit things and I also have some articles marked for revision.</p>

<p>But that is not my concern. Several entries in the Portuguese version were poor translated material from the English website. So poor that in places the content makes no sense in Portuguese. Checking the profile of &#8220;editors&#8221; that have translated such entries, you invariably see that they claim to have an advanced level (sometimes even native-like) of English, and sometimes advanced levels in other languages. Still you see very poor translations. This, coupled with the poor knowledge of the subject of the entry, is a recipe for disaster. I know by personal experience that is very difficult to master a language that is not your native one (I have good domain of English, but not advertised anywhere); even more difficult is to do a technical translation of a topic that you don&#8217;t master completely.</p>

<p>It has been discussed elsewhere (ie <a href="http://ascii.textfiles.com/archives/000276.html">here</a>) about the potential problems of Wikipedia and the prepotency of some editors (who are volunteers) in protecting the articles from the so-called vandalism. Maybe I should register myself and edit with an username, and be prepared to the criticism and sometimes the bullying in the lengthy discussions that can be found on Wikipedia. It will be a double work: edit the entry and then explain why is was edited.</p>

<p>I don&#8217;t think I have stomach and patience to do that again. I just fear for the high-school kid that will learn from there, thinking it is a great reference, and in the end will have to relearn it, due to poor content. Just my two cents.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/sad-news/">Sad news</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:43:29-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Sad news to any bioinformatician out there who believes in the support of companies to create nice applications. AbOrygen, creators of the fine software biOpen, very recently reviewed in this space had closed. A note left on the company&#8217;s website mentions the liquidation.</p>

<p>It is very sad to hear that and also a warning to the future to whoever is in the field.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/biopen-a-review/">biOpen: a review</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:40:28-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>After reviewing Geneious twice, I decided to search for other proprietary software that would be comparable in price, features and capabilities. I happened to find such software: <a href="http://www.aborygen.com/products/biOpen/">biOpen</a>, from a French company called AbOrygen.</p>

<p>biOpen motto is &#8220;Simply modular&#8221;, and in this case the modules come from where they should: the company that makes the software. If I am paying a considerable amount of money for a program that will help me do my research, I want any module/plugin to be reliable and take full advantage of the base program. AbOrygen&#8217;s website mentions that they have plans to release a SDK for biOpen, but no word on availability or features.</p>

<p>In a full comparison, biOpen lacks some of the eye catching details of Geneious, a little bit the ease of use and some functionality regarding PubMed searches. But biOpen does have some advantages if you are interested in doing actual science and do not care for the latest trend in design. Also, biOpen is a Mac only application, hence its interface using the Mac-favourite approach of a multi-window GUI (they claim a Windows and Linux version are coming). Basically the program has main project window, an analysis window and a small log, that can be scattered around on your desktop. The analysis is based on projects, not on folders as Geneious, and the data is organized as such.</p>

<p>On the scientific side, biOpen excels: contig assembly, vector cleaning, ORF search, PCR and other neat features for DNA/RNA/protein analysis. But the best is their pairwise/multiple alignment suite of programs that allows you to align your sequence with ClustalW (OK), <a href="http://www.drive5.com/muscle/">Muscle</a> (good) and <a href="http://probcons.stanford.edu/">ProbCons</a> (excellent). The latter is one of the best programs to align proteins available (see <a href="http://www.biomedcentral.com/1471-2105/7/471/">here</a>) and it is much better than the reliable old-timer ClustalW. On the other hand, on Geneious you are stuck with ClustalW and a proprietary alignment algorithm that no one tested, or maybe you will be waiting for some good heart develop a plugin that will allow the use of better alignment packages.</p>

<p>Also, by being &#8220;simply modular&#8221;, you can order only the modules you want, whenever you want. You can get biOpen for a spin before buying. If you are more interested on doing hard-core research, this is a very strong competition to Geneious. And if you are interested in having a GUI for two of the best programs for sequence alignment you definitely would give biOpen a try.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/global-warming-here-to-stay/">Global warming: here to stay</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:39:58-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>You know that global warming is here top stay when &#8230;</p>

<p>invariably everyday when you take the elevator, people only travel one floor. Call it waste of energy.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/geneious-a-reply-from-biomatters/">Geneious: a reply from Biomatters</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:27:54-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the Geneious discussion forum, Tobias, has replied to my review(s) with very interesting points that I think need some further argumentation. I start dissecting his post:</p>

<blockquote><p>I remember your old post from a year ago and just read your new review, and I thought I&#8217;d address a few of the points you raised.</p></blockquote>

<p>Although, I &#8220;raised more points&#8221; on my second review, which is based on a more mature version of the software, I had posted some interesting points on my initial assessment which haven&#8217;t been touched by Tobias on his reply. Maybe because my first review was less critical, less &#8220;negative&#8221; than the second one. Regardless, I still think (and I will have this opinion until I see some software better than) that Geneious is a great piece of software and deserve a very high praise.</p>

<blockquote><p>Of course I am 100% biased because I work at Biomatters, but I would like to give my personal view on some points that you raised. You may of course take my affiliation bias into account when reading those</p></blockquote>

<p>Bias acknowledged.</p>

<blockquote><ol>
<li>I agree that the free version gives you &#8220;limited resources&#8221; as you point out. But please do also consider that software is always limited, and in fact Geneious today gives you a lot more than the free version which you have positively reviewed a year ago. So the question here is whether you should compare today&#8217;s free version with Geneious Pro, or with last year&#8217;s free version and the other free software that is available. I lean towards the latter, and I think Geneious 2.5.3 scores very well in this comparison.</li>
</ol>
</blockquote>

<p>My first &#8220;positive&#8221; review praised the first version of Geneious. My second &#8220;negative&#8221; one &#8220;slammed&#8221; it, or maybe just gave a little bit of balance to the initial review. I agree that the number of free features in Geneious is good and software is always limiting. But with no bias, compare the pro and free versions.</p>

<blockquote><ol>
<li>The point of Geneious having &#8220;no groundbreaking or innovative feature that wasn&#8217;t seen in other molecular biology/phylogenetics package&#8221; is probably more true than others you raise, however it is also true that Geneious does offer some unique and innovative functionality that I don&#8217;t know from other bioinformatics software:</li>
<li>collaboration (obviously there is a chicken and egg problem,
and it will only become really useful once many people start
using it). The new version of collaboration for Geneious 3.0 has
much improved collaboration features such as one on one chat
and will hopefully greatly accelerate this development.</li>
<li>agents
Also, Geneious Pro offers many features otherwise found only in much more expensive software packages.</li>
</ol>
</blockquote>

<p>As I said, I could not test collaboration. I tried testing it with Daniel Batten, but we never found a right time to meet online. Regarding the interface, you can find similar or even better features in <a href="http://www.mbio.ncsu.edu/BioEdit/bioedit.html">BioEdit</a>, which is also free. But it has major disadvantages, as being Windows-only and being focused only on sequence alignment. Other software that can be comparable to Geneious is <a href="http://www.aborygen.com/downloads/biOpen/">bioOpen</a>, but falls in the same category as BioEdit: Mac-only and not even free. I don&#8217;t know the price of bioOpen but its ordering form gives you the option to select which functionalities you want.</p>

<blockquote><ol>
<li>You ask what incentive someone has to write a plugin for Geneious. If I was a bioinformatics student today and wrote software e.g. for my MSc project, I don&#8217;t see what reasons there are <em>not</em> to write it as a plugin for Geneious.</li>
</ol>
</blockquote>

<p>I was never a bioinformatics student, maybe because during my time there were not many bioinformatics courses available, even more where I did my graduate studies. So I cannot give you a personal view as a &#8220;bioinformatics student&#8221;. But first, we have to separate bioinformatics students in two classes (too few maybe): the computer scientist that understands biology and the biologist that programs computers. You can even include here a third class which is the computer-savvy biologist: he/she cannot program but have a high knowledge of computers.</p>

<p>With our categories defined I ask you again: what are the benefits of writing a plug-in for Geneious for each one of these classes of users?</p>

<blockquote><p>This is much easier than writing the functionality from scratch, because I don&#8217;t have to worry about the graphical user interface - which is one of the biggest workloads in writing software.</p></blockquote>

<p>I agree that writing plugins using a <strong>proprietary</strong> API is much easier than doing for scratch. I also agree that the interface is one of the biggest workloads in writing software. Being a interface programmer like myself, I understand how underrated the interface development is. But how many bioinformatics applications have interfaces? How many are cross-platform? And if I only need some functionality that Geneious does not give me but it would take a couple of hours to write in Perl/Python/Ruby (<a href="http://www.biomatters.com/userforum/comments.php?DiscussionID=87">like this</a>)? If I am a bench biologist with zero knowledge of programming do I have to wait for someone write me a plug-in or wait for Biomatters to include in Geneious?</p>

<blockquote><p>Also, it makes my program a lot more usable for a lot more people. Peer respect and publicity are among the strongest incentives - when people write free software, as many other people as possible using it is their <em>main</em> goal.</p></blockquote>

<p>Of course it makes a program more usable for a lot more people. People that are using Geneious, right? How do you measure peer respect? By doing a wonderful program from scratch or by adding to a <strong>proprietary</strong> software a plug-in? Or, wouldn&#8217;t be better to create a open source project (or join one) that works in bioinformatics development?</p>

<blockquote><p>I spent a considerable amount of time during my studies writing software, and I wish Geneious had been around back then - it would have made my software much easier to write and use, and it would have become known to a lot more people&lt;</p></blockquote>

<p>I did my Masters writing software and my PhD working on phylogenetics. Would Geneious help me in the first case? No (maybe with ideas). Would it help me in the latter? Definitely, yes. Would I spend time during my PhD to write plug-ins to Geneious? No. See the difference?</p>

<blockquote><ol>
<li>You suggest that we concentrate on writing plugins and sell licenses for using those. Really this isn&#8217;t very far from what we are already doing - the main difference between Geneious and Geneious Pro is that Geneious Pro has more features, and most of these are provided plugins. So really there are already <em>lots</em>of plugins - it&#8217;s just that most of them already come shipped with Geneious, instead of you having to download them in a separate file from our website. In fact, if other people were to write free plugins which provide all of the Geneious Pro functionality for free, the incentive for Geneious Pro would basically disappear. This is the only concern I personally have about our Public API - and it&#8217;s about the opposite from yours.</li>
</ol>
</blockquote>

<p>This is not the first time I suggested this. I had discussions with Daniel Batten about this subject, and even point him some licensing possibilities, such as <a href="http://www.mathworks.com/">MatLab</a>. I saw the list of add-ons when customizing Geneious. Regarding your concern, I don&#8217;t think it will be fulfilled.</p>

<blockquote><ol>
<li>The import feature hasn&#8217;t always worked correctly for you. We would like to keep Geneious as bug free as possible, so if you could please email any files upon which Geneious fails to e.g. tobias@biomatters.com, then that would be much appreciated and we will fix these problems as soon as possible.</li>
</ol>
</blockquote>

<p>I have sent samples, even add to other topics guidance to some biological file formats. I personally don&#8217;t have any file that I am working at the moment, but as everyone that releases a new bioinformatics application, releases a new filetype at the same time, the amount of filters and errors will be huge.</p>

<blockquote><p>We are typically quite responsive, even though we live in a different time zone to most other people.</p></blockquote>

<p>When I say that is a disadvantage to be in New Zealand, is that sometimes users (especially the ones that pay) tend to be quite demanding. As I am not a Geneious user <em>per se</em>, this does not affect me.</p>

<blockquote><ol>
<li>There are now 4 plugins available for download, and a fifth will possibly follow soon. Only three of the currently available ones work for Geneious 2.5.x, but this will probably be fixed at some point. As our API stabilises and more plugins become available, we hope to introduce a more comfortable way of installing them and keeping them up to date.</li>
</ol>
</blockquote>

<p>There were only two at the time of my review. It is good that it is increasing, 100% in a couple of days. And going back to the &#8220;peer respect&#8221;, I downloaded <a href="http://mrbayes.csit.fsu.edu/">MrBayes</a> plug-in and I couldn&#8217;t find the names of the developers. Also could not find any reference to MrBayes authors, papers, etc.</p>

<blockquote><ol>
<li>Scripting language support is already being developed. Update: It probably won&#8217;t make it into Geneious 3.0, but nevertheless Geneious will have scripting support at some point.</li>
</ol>
</blockquote>

<p>Great. That was another thing that Daniel and I discussed last year. I think this is nice for people trying to create molecular biology, phylogenetics, etc courses for a large audience or for some university program. I couldn&#8217;t find the Educational plug-in in Geneious.</p>

<p>Still, in my opinion Geneious is a great software.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/02/geneious-one-year-or-almost-later/">Geneious: one year (or almost) later</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-02T17:27:24-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I intend to cover bioinformatics things in this space. At least mainly. I will try to post here and there stuff about science in general, more specifically biology and genetics. On my first post I reused an old overview of Geneious, and now I will take a fresher look at the program.</p>

<p>Geneious, at the time of this post, is on its version 2.5.3. Since the first release the program has evolved a lot: new functions were added, a cleaner and leaner interface and major bug killing. My first review claimed that Geneious is the Holy Grail of bioinformatics application, and I still think it is. But &#8230;</p>

<p>Well, all stories have a &#8220;but&#8221;, and Geneious&#8217; could not be different. Indeed, its new interface is cleaner, buttons were redesigned, some menus received illustrative icons. But there is (or was) no groundbreaking  or innovative feature that wasn&#8217;t seen in other molecular biology/phylogenetics packages. The strength is still having all features and functions in one cross-platform application. Also the power of having database-like capabilities is another plus: all your sequences, protein structures, references, etc is there in one place, easily accessible and at hand. Sometimes the search is a little bit slow, sometimes the import feature does not work correctly and even sometimes the program indexes entries in a strange fashion.</p>

<p>About new functions, one that calls attention is the collaboration possibilities. Using the free <a href="http://www.jabber.org">Jabber</a> network it is quite easy to register a new user and connect in the wild. But the niceties stop there. In my lab I am the only using Geneious at the moment, so my options of collaboration are minimal. I have tried setting up a virtual meeting with the developers in order to try the feature, but the fact that Biomatters is on the other side of the world makes a little bit difficult to set up a decent time to connect. And this is another small disadvantage of Geneious: the soft-house is located in New Zealand so any hope of fast support for people in North America are at least thin.</p>

<p>The last releases were followed by an introduction of several small (but very useful) features. More types of files are now supported, small bells and whistles were also added. But (always the but) one of the most promising features seems to be underused. Last time I checked Geneious API does not have the same success as the main program. I have tried creating a plugin to share with its userbase, but not being a Java developer I found more obstacles than solutions and gave up on trying. Geneious website gives links to two (yes two) plugins, and this is the result of almost six months of the API in the wild (if you know any other Geneious plugin please let me know).</p>

<p>And here I have to draw a line. Geneious is &#8220;free&#8221;, but this version gives you limited resources. A pro version will make you at least $166 poorer and will give you a lot of power. So my question is: why would you spend time and effort to learn the API and create a plugin to a mostly paid software? I understand people developing Firefox extensions, Linux open-source packages, Python libraries, etc (they are free as in freedom), but I can&#8217;t see scientists creating add-ons to Geneious. I might be wrong and would like to be proven wrong, but at least until now I can say that the plug-in feature hasn&#8217;t taken off.</p>

<p>Maybe the whole plug-in thing wasn&#8217;t a good move for Geneious and Biomatters. Maybe they should invest more time making Geneious even better and dedicate its programming resources to create the plug-ins themselves and charge for them. Something that MatLab does very well, staying in the scientific software realm only. You might argue that MatLab is also a programming language, has a compiler and it light-years more powerful than Geneious. I agree. But MatLab tries to cover many different fields, while Geneious is &#8220;just&#8221; a molecular biology/phylogenetics package.  Most of bench biologists are not interested in understanding how computer programs work, they want a black-box-like application that uses an easy and user friendly input and spits out a pre-formatted result, simply and efficiently. Maybe a macro tool or similar script/automating programming in Geneious would be more than enough for most people.</p>

<p>Overall, I still think that Geneious is the Holy Grail of bioinformatics applications, as mentioned above. But like the quest for the Holy Grail, we still have to wait sometime to finally find it completely. Geneious is great and a little bit of fine tuning would certainly improve it. I think Biomatters should focus on some points:</p>

<ul>
<li>improving its coverage of biological formats, maybe even making the bold move of standardize the myriad of different file types out there</li>
<li>improving a little bit more the interface of Geneious, adding some minor features like right-click in some areas</li>
<li>add a simple and effective scripting/automating feature to the program, instead of relying on the API and the &#8220;heavy&#8221; Java to create plugins</li>
<li>add some database-like search and access to the data stored, something like a grep for finding motifs on sequences, or even sequences itself when the data stored is massive</li>
</ul>


<p>Although excellent, Geneious has room for improvement. We will keep searching for the Holy Grail.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/26/hello-world/">First post: a take on Geneious</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-26T21:34:28-05:00" pubdate data-updated="true">Feb 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.geneious.com">Geneious</a> is an excellent bioinformatics software created by Biomatters, a wonderful company from New Zealand. When they released the first version of the software I was delighted to write a review of it. So here is my first take on Geneious from 2006. I will be back with more comments on the program on the other side of this post.</p>

<p><a href="http://www.genedrift.org/geneious.php">Geneious review</a></p>

<p>After installing Geneious and running it from the command line (double clicking if you are a Windows user) you don&#8217;t see anything special about the software. Apart from the nice looking interface there are no fancy buttons, no pop-up messages, no hints (maybe just the tutorial tab on the right side) and no clutter.</p>

<p>After the first minute staring at its interface you wouldn&#8217;t say that this is one of the most innovative software to appear for molecular biology, phylogenetics, bioinformatics and other fields of science. You wouldn&#8217;t say, but after playing for an extra five minutes the only sound in your mind would be &#8220;Wow!!&#8221;. Geneious is wonderful work of a group of bioinformaticians that decided to make science easier to do, at least when doing computer work.</p>

<p>Geneious is really genial. Basically you don’t have to do anything, the program does for you. OK, maybe you have to do something but the program seems to know what you want to do. Import a sequence into it: Geneious knows if it is nucleotides or amino acids. Do you have an annotation for your sequence? Nice, Geneious will display it for you. Structure of your protein? There it is, just select the tab. You don’t have to save anything: it is already there automatically saved for you, just like your personal GenBank, PDB, Swissprot.</p>

<p>And the software can do more, much more. Have internet? (Who doesn’t?) There is a direct connection to NCBI and no worries about formatting searches, remembering parameters, Geneious does it all for you. PubMed? Check, same thing: filter your references, search by first author, date of publication, anything. If you are Linux geek, your BibTex entry is there, if you are Windows person just export it to EndNote.</p>

<p>Another nice addition to the program is the possibility to create agents. These agents are like your free personal search expert who is there 24/7/365. Set up an agent to a specific search pattern and it will get for you when it appears, keeping your records up-to-date even before you have breakfast.</p>

<p>Nothing bad can be said about Geneious, and with endless possibilities of improvement (with a looming API release soon) this is the Holy Grail of molecular biology, phylogenetics and bioinformatics software. Kudos to Geneious</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/22/counting-nucleotides-part-ii/">Counting nucleotides: part II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-22T14:51:58-05:00" pubdate data-updated="true">Feb 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Let&#8217;s check the &#8220;short&#8221; way, that is basically a method that avoid the
&#8220;explosion&#8221; of the string. Instead of transforming the sequence from the
file from a string to a list, we go and use the string directly,
applying one of the methods available to manipulate strings. This time
we read the file at once and convert the list to a string using <code>join</code>.
To count we simply use the method <code>count</code> on our string. Pretty nice.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="c">#still keep the file fixed </span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="c">#opening the file, reading the sequence and storing in a list</span>
</span><span class='line'><span class="n">seqlist</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="c">#let&#39;s join the the lines in a temporary string </span>
</span><span class='line'><span class="n">temp</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seqlist</span><span class="p">)</span>
</span><span class='line'><span class="c">#counting </span>
</span><span class='line'><span class="n">totalA</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalC</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalG</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">totalT</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'><span class="c">#printing results </span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; As found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Cs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Gs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Ts found&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>That&#8217;s the short way: using <code>count</code>. This is a method that when applied
on a string counts the number of times the substring appears in our
string. We can even set a start and ending point to count. Notice one
difference in this script to the previous examples: after we join the
items of the list into a string we do not remove the carriage returns.
Because in this case we don&#8217;t need it, as it will be an extra character
there that won&#8217;t make any harm.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/22/counting-nucleotides/">Counting nucleotides</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-22T14:48:35-05:00" pubdate data-updated="true">Feb 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are going to jump the regular expression explanation that you can
find in the book, and we will go directly to the section that introduces
string/list/array manipulation and adapt it to Python. We are going to
see two different methods: a &#8220;long&#8221; and a &#8220;short&#8221; one. In many places
and computer languages you will see that there are different ways of
doing the same thing, with advantages and disadvantages. It is up to you
to define which methods are better or worse, as this is a very personal
matter. Some people prefer the longer way because it might be clearer
and easier to understand, or it might be necessary to use it due to code
maintainability. Other people would rather use the shorter path because
they want to generate code faster. It does not matter the path you
select, as long as you get your task done. That&#8217;s the key: focus on the
end product not on how exactly got there. If you used 10 lines of code
more, or 10 less, that&#8217;s irrelevant as long as you did what you wanted.
Live and learn and someday you will use the even shorter way. So we
start with the long way. We want to count the individual number of
nucleotides in a sequence, determining they relative frequency. Remember
that we read the lines of a sequence file into a list, using
<code>readlines</code>. But on the long method we will read the file and store the
data into a string like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p>This way it will be easier to &#8220;explode&#8221; the sequence
in separated items. After the &#8220;explosion&#8221; we can check each item in the
list and get our result. The code is below, I will be back after it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python</span>
</span><span class='line'><span class="c">#let&#39;s keep the file fixed for now</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="c">#opening the file, reading the sequence and storing in a list </span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="c">#initialize a string to receive the data </span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>    <span class="c">#notice the strip, to remove \n</span>
</span><span class='line'>    <span class="c">#&quot;exploding&quot; the sequence in a list </span>
</span><span class='line'><span class="n">seqlist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'><span class="c">#initializing integers to store the counts </span>
</span><span class='line'><span class="n">totalA</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalC</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalG</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="n">totalT</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'><span class="c">#checking each item in the list and updating counts </span>
</span><span class='line'><span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seqlist</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;A&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalA</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;C&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalC</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;G&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalG</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s">&#39;T&#39;</span><span class="p">:</span>
</span><span class='line'>        <span class="n">totalT</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span class='line'><span class="c">#printing results </span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; As found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Cs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Gs found&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; Ts found&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>Most of the above code was already covered before.
Let&#8217;s look at the different stuff, like the &#8220;explosion line&#8221;</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">seqlist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Here we basically transform our string <code>sequence</code> into a list, by putting the
object type we want before the object we want converted, like we do here</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; As found&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>This construction <code>type_to_convert(to_be_converted)</code>
tells Python to get whatever is inside the parentheses and transform
into whatever is outside. Converting the string to a list will get</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">GTGACTTTGTTCAACGGCCGCGGTATCCTAACCGT</span>
</span><span class='line'><span class="n">GCGAAGGTAGCGTAATCACTTGTTCTTTAAATAAGGACTAGTATG</span>
</span><span class='line'><span class="n">AATGGCATCACGAGGGCTTTACTGTCTCCTTTTTCTAATCAGTGAA</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>and transform it into</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span>
</span><span class='line'><span class="s">&#39;T&#39;</span><span class="p">,</span><span class="o">...</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>After &#8220;exploding&#8221;, we use a <code>for</code> loop to iterate
over every item in the list and use conditional statements to do the
counts. Regarding the counts, we use this operator</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">totalT</span> <span class="o">+=</span> <span class="mi">1</span>
</span></code></pre></td></tr></table></div></figure>


<p>that, in C/C++, tells the
interpreter to get the value of <code>totalT</code> and add 1 to it.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/39/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/37/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
