
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="As mentioned we will see in this entry some other features of Python
lists. We will start with a similar example to the one in the book and
then use &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/38">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/12/manipulating-python-lists-ii/">&#8220;Manipulating&#8221; Python lists II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-12T14:18:03-05:00" pubdate data-updated="true">Feb 12<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As mentioned we will see in this entry some other features of Python
lists. We will start with a similar example to the one in the book and
then use our DNA file. So let&#8217;s assume we have this simple list</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>If we print it directly we would get something like
this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>which is fine for now, as we are not worried (yet) with the output (what
we will do further below). Let&#8217;s remove the last nucleotide. To
accomplish that, we use <code>pop</code> with no specific index</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>which gives us this when printed</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Remember that lists are mutable, so the removed item is lost. We can
also remove any other in the list, let&#8217;s say &#8216;C&#8217;. First, we reassign the
original list items and then remove the second item</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>The list when printed will return this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p> <code>pop</code> accepts any valid index of the list. Any index larger that the length of the
list will return an error. For future reference, remember that when any
item is removed (and inserted) the indexes change and the length also.
It may seems obvious but mistakes are common. Shifting from our
&#8216;destructive&#8217; mode, we cal also add elements to the list. Adding to the
end of the list is trivial, by using <code>append</code></p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>that returns</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Adding to any position is also very straightforward with <code>insert</code>, like
this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>where <code>insert</code> takes two
arguments: first is the index of the element before which to insert and
second the element to be inserted. So our line above will insert an &#8216;A&#8217;
just before the &#8216;A&#8217; at position zero. We can try this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;A1&#39;</span><span class="p">)</span> <span class="n">nucleotides</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s">&#39;C1&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">nucleotides</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="s">&#39;G1&#39;</span><span class="p">)</span> <span class="n">nucleotides</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="s">&#39;T1&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>that will result in</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">[</span><span class="s">&#39;A1&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C1&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T1&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G1&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that we add every new item at
an even position, due to the fact that for every insertion the list&#8217;s
length and indexes change. And for last, we will take care of the
output. Of course if are creating a script that requires a nicer output,
printing a list is not the best way. We could create a loop and merge
all entries in the list, but that would be a couple of lines and we
ought to have an easier way (otherwise we could be using C++ instead).
There is a way, by using the method <code>join</code>. This method will join all
the elements in a list into a single string, with a selected delimiter.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="o">.</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'><span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>will generate this output</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">ACGT</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>join</code> is a method that applies to
strings. The first &#8220;item&#8221; is a string, that could be anything (in our
case is an empty one). The code line tells Python to get the empty
string an join it to the list of strings that we call nucleotides. With
this we finish the first section of the site and we are moving to
chapter 5 in the book.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/09/manipulating-python-lists-i/">&#8220;Manipulating&#8221; Python lists I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-09T14:01:39-05:00" pubdate data-updated="true">Feb 9<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Now, I want to manipulate my DNA sequence, extract some nucleotides,
check lines, etc. Simple things. We start with the same basic code to
read the file:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Our nucleotides are stored in the variable
<code>file</code>. Remember that each line is one item of the list and the lines
still contain the carriage return present in the ASCII file. Let&#8217;s get
the first and the last lines of the sequence. The first line is easy to
get, as Python&#8217;s lists start at 0. To access one list item just add
square brackets with the index number of the item you want to get (this
is also known as slicing). Something like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>will return the item 0 from the
list, that in our case is the firs line of the sequence. If you add a
<code>print</code> command</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>you should expect <code>GTGACTTTGTTCAACGGCC....CGTAATCACTTGTTC</code> The last line
is a little bit trickier. Let&#8217;s assume that we don&#8217;t know the number of
lines in the list, and here we want to make our script as general as
possible, so it can handle some simple files later. It is also good code
practice to think ahead and plan what you want, first to have a detailed
project to follow, and second it allows you to be prepared to
errors/bugs that your code might have or situations not expected in your
original plan. In our file, we have <strong>eight</strong> lines of DNA, so it would
be just adding this <code>print file[7]</code> and we would output the last line.
But, the right way to do it is to check the length of the list and
output the item which has an index equal to the list length. In Python,
you can check the length of a list by adding the built-in function <code>len</code>
before the list name, like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">len</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>So who do we print the last
line of our sequence? Simple. <code>len(file)</code> should return an integer of
value 8, which is the actual number of elements in our list. We already
know that to access any item in a list we just add its index (that has
to be an integer) to the list name. One idea then would be to use
<code>len(file)</code> as the index, like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="nb">file</span><span class="p">)]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Why would that be wrong? Our list has eight
items, but the indexes are from 0 to <strong>7</strong>. So eight would be one index
over the list length, which is not accessible because it does not exist.
Solution? Let&#8217;s use the list length minus one:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>and there you
are, the last line of the sequence. But as we want the last line of the
file, which is a list there is an easier way to output just the last
line:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>which tells the interpreter to print the last item of a list. Putting
everything together, we have</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="k">print</span> <span class="s">&#39;I want the first line&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="k">print</span> <span class="s">&#39;now the last line&#39;</span>
</span><span class='line'><span class="k">print</span> <span class="nb">file</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Two points worth mentioning:
differently of strings, Python&#8217;s lists are mutable, items can be
removed, deleted, changed, and strings also can be sliced by using
indexes that access characters. Next we will see some more features of
lists and strings, and how to manipulate them. It will probably be the
last entry in the first section as we finish the Chapter 4 in the book.
As you may have noticed some items in the perl book will not be covered,
at least not immediately. We will jump back and forth sometimes.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/02/07/reading-files-in-python-using-lists/">Reading files in Python: using lists</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-02-07T13:56:59-05:00" pubdate data-updated="true">Feb 7<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Let&#8217;s improve our previous script and put the contents of the file in a
variable similar to an array. Python understands different formats of
compound data types, and <code>list</code> is the most versatile. A <code>list</code> in
Python can be assigned by a series of elements (or values) separated by
a comma and surrounded by square brackets</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">shoplist</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;milk&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;lettuce&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="s">&#39;coffee&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>Now, we are going to read the same file and store the
DNA sequence in a <code>list</code> and output this variable. The beginning of the
script is the same, where we basically tell Python that the file name is
AY162388.seq.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>We are going to change the way we
read the file. Instead of just opening and then reading line-by-line, we
are going to open it a read all the lines at once, by using this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice the part in bold? In the previous
script, we open and store the contents of the file in a <code>file object</code>.
Now , we are opening the file and just after it is opened, we are
reading all the lines of the file at once and storing them in
<code>file object</code>, but is a Python&#8217;s <code>list</code> of strings. Before, if we
wanted to manipulate our DNA sequence, we would had to read it, and then
in the loop store in a variable of our choice. In this script, we do
that all at once, and the result is a variable that we can change the
way we wanted. The code without the output part is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Try putting a <code>print</code>
statement after the last line to print the <code>file</code> list. You will get
something like this</p>

<p><code>['GTGACTT...TTGTTC\n', 'TTTAAATA....TAATC\n', 'AGTGA...CTATG\n', 'GAGCTCA....TATAGC\n', ...]</code></p>

<p>which is exactly the description of a Python&#8217;s list. You see all lines,
separated by comma and surrounded by square brackets. Notice that each
line has a carriage return (<code>\n</code>) symbol at the end. Let&#8217;s make
the output a little nicer including a loop. Remember when I introduced
loop I wrote that Python iterates over &#8220;items in a sequence of items&#8221;,
what is a good synonym for <code>list</code>. So the loop should be as
straightforward as</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p>Putting everything together gives us</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="k">print</span> <span class="nb">file</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/30/reading-files-in-python/">Reading files in Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-30T13:53:56-05:00" pubdate data-updated="true">Jan 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are currently following Chapter 4 of Beginning Perl for
Bioinformatics, which is the first chapter of the book that actually has
code snippets and real programming. The last exercises in this chapter
deal with the ability to read files and operate with information
extracted from these files, to create arrays and scalar list in perl. We
are going to check how to read files in python. The book tells you how
to read protein sequences. Here we are going to read DNA and protein
sequences from files and change them. Let say you have a file with a DNA
sequence in some directory in your hard disk. The file cannot be a FASTA
type (we will see later how to handle FASTA files), just pure sequence,
something like this:</p>

<p><code>GTGACTTTGTTCAACGGCCGCGGTATCCTAACCGTGCGAAGGTAGCGTAATCACTTGTTC TTTAAATAAGGACTAGTATGAATGGCATCACGAGGGCTTTACTGTCTCCTTTTTCTAATC AGTGAAACTAATCTCCCGTGAAGAAGCGGGAATTAACTTATAAGACGAGAAGACCCTATG GAGCTTTAAACCAAATAACATTTGCTATTTTACAACATTCAGATATCTAATCTTTATAGC ACTATGATTACAAGTTTTAGGTTGGGGTGACCGCGGAGTAAAAATTAACCTCCACATTGA AGGAATTTCTAAGCAAAAAGCTACAACTTTAAGCATCAACAAATTGACACTTATTGACCC AATATTTTGATCAACGAACCATTACCCTAGGGATAACAGCGCAATCCATTATGAGAGCTA TTATCGACAAGTGGGCTTACGACCTCGATGTTGGATCAGGG</code></p>

<p>You can download the file <a href="http://python.genedrift.org/codepy/AY162388.seq">here</a>. This is a
partial sequence of a mitochondrial gene from a South American frog
species called <a href="http://calphotos.berkeley.edu/imgs/128x192/0000_0000/1101/0201.jpeg"><em>Hylodes ornatus</em></a>.</p>

<p>For our purposes you can save the file in the same directory you are
going to run the script from, or if you are using the Python interpreter
start it in the directory that contains the file. The file name is not
important but we will use AY162388.seq from now on. The first thing we
have to do is to open the file for reading. We define a string variable
that will contain the file name.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="p">]</span><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>In order to open the file, we can use the
command <code>open</code>, that receives two strings: the first is the file name
(it can be the whole location too) to be opened and the <em>mode</em> to be
used, which is what you want to do with the file. The <em>mode</em> can be one
or more letters that tell the interpreter what to do. For now we are
going to use the <code>r</code> <em>mode</em> , which tells Python to read the file, and
only do that. So our code is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>file</code> is a file object that contains
the directives to read our DNA sequence file. Now, we have actually read
the contents of the file but they are stored in a file object and we did
not accessed it yet. We can achieve that by using a myriad of commands.
We will start with the commonest one: read the file line by line. <code>file</code>
is our file object. We have just opened it, but Python already knows
that any file contains lines (remember that this is a regular ASCII
file). We are going to use a loop to read each line of the file, one by
one</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>


<p>In Python, the <code>for</code> loop/statement iterates over
items in a sequence of items, usually a string or a list (we will see
Python&#8217;s <code>list</code> soon), instead of iterating over a progression of
numbers. Basically, our <code>for</code> above will iterate over each line in the
file until EOF (end-of-file) is reached. Our simple script to read a DNA
sequence from a file and output to the screen is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&quot;AY162388.seq&quot;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">line</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/23/transcribing-the-other-way/">Transcribing: the &#8220;other&#8221; way</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-23T13:52:30-05:00" pubdate data-updated="true">Jan 23<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen how to transcribe DNA using regular expression, even though
the regex we used cannot be considered a real one. Now we are going to
simplify our small script even more and take advantage of some string
capabilities of Python. Instead of using two lines, we are going to use
only one. And also we won&#8217;t need to import anything. Let&#8217;s start again
with the same DNA sequence</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTTGCAACGTTGCAACGTTGCA&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>This time we are going to use
<code>replace</code>. This is one of the Python&#8217;s methods to manipulate strings.
Basically, we are asking the interpreter to replace a certain string by
another. The method returns a new copy of your string:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myRNA</span> <span class="o">=</span> <span class="n">myDNA</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;U&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>This tells Python: <code>myRNA</code> will receive a copy of <code>myDNA</code> where all Ts were
changed by Us. the &#8220;dot&#8221; after <code>myDNA</code> means that the method <code>replace</code>
will get that variable as input on that variable. So our code from above
would like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTTGCAACGTTGCAACGTTGCA&#39;</span>
</span><span class='line'><span class="n">myRNA</span> <span class="o">=</span> <span class="n">myDNA</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;U&#39;</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">myRNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple and efficient. Next we will use the same approach on generating the reverse complement of a DNA sequence, with no
regex pattern.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/16/the-regular-expression/">The Regular Expression</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-16T13:49:07-05:00" pubdate data-updated="true">Jan 16<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As mentioned above, regex in Python are provided by the <code>re</code> module,
which provides an interface for the regular expression engine. First
thing we have to do is to tell the interpreter what to do and what
expression to use. Let&#8217;s start with a DNA sequence.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTTGCAACGTTGCAACGTTGCA&#39;</span>
</span></code></pre></td></tr></table></div></figure>


<p>How to
transcribe it to RNA? Transcription creates a single-strand RNA molecule
from the double-strand DNA; basically the final result is a similar
sequence, with all <code>T</code>&#8217;s changed to <code>U</code>&#8217;s. So our regular expression has
to find all <code>T</code> nucleotides in the above sequence and then replace them.
Regular expressions in Python need to be compiled into a RegexObject,
that contains all possible regular expression operations. In our case we
need to search and replace, what can be done by using the <code>sub()</code>
method. According to the Python&#8217;s <a href="http://www.amk.ca/python/howto/regex/regex.html">Regular Expression HOWTO</a> <code>sub()</code> returns
the string obtained by replacing the leftmost non-overlapping
occurrences of the RE in string by the replacement replacement. If the
pattern isn&#8217;t found, string is returned unchanged. Let&#8217;s put everything
above in real code. First we need to compile a new RegexObject that will
search for all thymines in our sequence. It can be achieved by using
this:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">regexp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple as that. This line of code tells the
Python interpreter that our &#8220;regular expression&#8221; is every T in our
string. Now, we have to make replace those Ts with Us. In order to do
that we just tell the interpreter:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myRNA</span> <span class="o">=</span> <span class="n">regexp</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&#39;U&#39;</span><span class="p">,</span> <span class="n">myDNA</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Let&#8217;s look at the last two lines of
code. On the first line we created a new RegexObject, <code>regexp</code> (that
could have any name, as any variable) and compiled it, making our
regular expression to be every T in our string. On the second line, we
assigned our soon to be created RNA sequence to a new string (remember
that strings in Python are immutable) and used the command <code>sub</code> to
replace in the Ts by Us present in our original DNA string. Putting all
together our transcription code will be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTTGCAACGTTGCAACGTTGCA&#39;</span>
</span><span class='line'><span class="n">regexp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">myRNA</span> <span class="o">=</span> <span class="n">regexp</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&#39;U&#39;</span><span class="p">,</span> <span class="n">myDNA</span><span class="p">)</span>
</span><span class='line'><span class="k">print</span> <span class="n">myRNA</span>
</span></code></pre></td></tr></table></div></figure>

</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/15/python-printing-statement/">Python printing statement</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-15T18:40:41-05:00" pubdate data-updated="true">Jan 15<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Just an apart from the bioinformatics aspect of programming: Python&#8217;s
<code>print</code> statement. As in most computer languages Python allows an easy
way to write to the standard output. Python&#8217;s <code>print</code> only accepts
output of strings, and if the variable sent to it is not a string it is
first converted and then output. The <code>print</code> always put a linebreak
(<code>'\n'</code> or <code>"\n"</code>) at the end of the expression to be output, except
when the <code>print</code> statement ends with a comma. For example:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="s">&quot;This is a&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;test&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>will print</p>

<p><strong>This is a</strong>
<strong>test</strong></p>

<p>while</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="s">&quot;This is a&quot;</span><span class="p">,</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;test&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>will print</p>

<p><strong>This is a test</strong></p>

<p>Of course Python&#8217;s <code>print</code> statement allows any programming escape character, such
as <code>'\n'</code> and <code>'\t'</code>.</p>

<p><strong>Concatenating strings on output</strong></p>

<p>To concatenate two strings on output there are two possible ways in Python. You can
either separate the strings with a comma, like we did here</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">myDNA</span><span class="p">,</span> <span class="n">myDNA2</span>
</span></code></pre></td></tr></table></div></figure>


<p> or you can use the &#8220;+&#8221; sign in roder to obtain almost the same result. This is similar to
what was used here</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA3</span> <span class="o">=</span> <span class="n">myDNA</span> <span class="o">+</span> <span class="n">myDNA2</span>
</span></code></pre></td></tr></table></div></figure>


<p>but instead we would use the <code>print</code> command as</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">myDNA3</span> <span class="o">+</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>In the latter case, both strings will not be separated by a space and will be
merged. To get the same result you would have to concatenate an extra
space between the strings like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">myDNA3</span> <span class="o">+</span> <span class="s">&quot; &quot;</span> <span class="o">+</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/15/importing-and-regular-expressions/">Importing and regular expressions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-15T13:47:54-05:00" pubdate data-updated="true">Jan 15<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Tisdall&#8217;s book on Perl introduces next the ability to transcripts DNA
sequences into RNA. In order to do that we need to check a different
aspect of programming: regular expressions (or regex). Regular
expressions is a pattern/string expression that works
matching/describing/filtering other strings. Let&#8217;s say you want to
examine or extract all vowels contained in one phrase, one page, one
word. Another example would be to remove all html tags from a downloaded
webpage. As HTML tags are encapsulated between <code>&lt;</code> and <code>&gt;</code> signs we can
create a regex that will search for any characters in between the signs
and remove (parse) them from our page. We will deal very briefly with
regex, and if you are interested in learning more about it you can
search for countless references on the internet (such as <a href="http://www.regular-expressions.info/">this one</a>).</p>

<p>In order to use regular
expression in Python we need to check another concept present in the
language: importing modules. Python core functionality provides most of
the usual operations and also comes with a built-in library of functions
that &#8220;access to operations that are not part of the core of the language
but are nevertheless built in&#8221;. One of this operations is the ability to
interpret regular expression that in Python is located in the <code>re</code>
module. Apart from the language core, built in modules, Python can be
further extended by using third-party modules imported into the
language. Anyone can create a module and distribute to every Python user
and programmer. So, in order to have regex capabilities we literally
have to <code>import</code> the regex module. We do that by entering the line:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span></code></pre></td></tr></table></div></figure>


<p>Python&#8217;s code style
guide suggests that <code>import</code> statements should be on separate lines</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">re</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>and
to be put on the top of the file (usually below the line that tells your
OS to use Python&#8217;s interpreter on the script). So the first two lines of
our new script would be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env</span>
</span><span class='line'><span class="n">python</span> <span class="kn">import</span> <span class="nn">re</span>
</span></code></pre></td></tr></table></div></figure>


<p>Now that we have the ability to use regex,we need to create one expression that will transcribe our DNA sequences
into RNA.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/14/sequences-and-strings-part-ii/">Sequences and Strings - part II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-14T15:30:32-05:00" pubdate data-updated="true">Dec 14<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Another important task for many biologists is to merge/concatenate
different strings of DNA in one unique sequence. We can modify the
previous script to concatenate two distinct DNA sequences in one. We
start using code_01 structure, adding some extra elements (line 3):</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTACGTACGTACGTACGTACGT&#39;</span>
</span><span class='line'><span class="n">myDNA2</span> <span class="o">=</span> <span class="s">&quot;TCGATCGATCGATCGATCGA&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA</span><span class="p">,</span> <span class="n">myDNA2</span>
</span></code></pre></td></tr></table></div></figure>


<p>So far, we added a new string containing an extra
DNA sequence and we <code>print</code> both sequences. In Python <code>print</code> statement
automatically adds a new line at the end of the string to be printed,
unless you add a comma (,) to the end. The comma is also needed if you
are going to print more than one string in order to separate them (<em>try
removing the comma from the code above</em>). Now, how do we merge myDNA and
myDNA2? Easy in Python: just <em>sum</em> them with a plus signal:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA3</span> <span class="o">=</span> <span class="n">myDNA</span> <span class="o">+</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA3</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that in Python strings are immutable, meaning they cannot be
changed. This immutability confers some advantages to the code where
strings (<em>in Python strings are not <strong>variables</strong></em>) cannot be modified
anywhere in the program and also allowing some performance gain in the
interpreter. So, in order to have our sequences merged we create a third
sequence that receives both strings. Finally our code will be (some
captions were added):</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env</span>
</span><span class='line'><span class="n">python</span> <span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span><span class='line'><span class="n">myDNA2</span> <span class="o">=</span> <span class="s">&quot;TCGATCGATCGATCGATCGA&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;First and Second sequences&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA</span><span class="p">,</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="n">myDNA3</span> <span class="o">=</span> <span class="n">myDNA</span> <span class="o">+</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Concatenated sequence&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA3</span>
</span></code></pre></td></tr></table></div></figure>


<p>Easy, eh? Of course these two simple scripts do no
scratch the surface of Python programming, but they are a start.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/13/hands-on-code-sequences-and-strings-part-i/">Hands on code: Sequences and strings - part I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-13T22:27:48-05:00" pubdate data-updated="true">Dec 13<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As pointed in <strong>Beginning Perl for Bioinformatics</strong>, a large percentage
of bioinformatics procedures deals with strings, especially DNA and
amino acids sequence data. As is largely known DNA is composed of four
different nucleotides: A, C, T and G and proteins can contain up to 20
amino acids. Each one of these elements have one letter of the alphabet
assigned to them. In the DNA case some letters represent one or more
nucleotides that can be present at some sequence position (click
<a href="http://www.cns.fr/externe/English/Projets/Resultats/iupaciub.html">here</a>
for more ).</p>

<p>So, as the amino acid is the basic building block of
proteins (AKA polypeptides), strings containing sequence is our most
basic block, from where all the bioinformatics magic will work on.
Usually in Perl a string is represented by the dollar sign in front of
the variable name, like this <code>$sequence</code>. Python is dynamically typed,
meaning variable types are assigned/discovered by the interpreter at run
time. This means that the value after the equal sign will tell the
interpreter what variable type you are declaring. So in Python if you
want to store a DNA sequence you can just enter</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">mydna</span><span class="o">=</span><span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p><em>a quick
note: Python can be used with the interpreter command line or by
previously saved scripts. I will try to use the latter in the code
examples.</em></p>

<p>OK, we are ready to create our first Bioinformatics Python
Hello World script. Let&#8217;s get the sequence above and print it on the
screen. The first line will tell the operating system what to use and
where to find the Python interpreter</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span></code></pre></td></tr></table></div></figure>


<p>Next we will create the variable myDNA
and assign the corresponding sequence</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>And finally, we will print the contents of the variable to the screen:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>As mentioned above, Python mandates that you have your code indented, but in our
final script this is not needed:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env </span>
</span><span class='line'><span class="n">python</span> <span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span> <span class="k">print</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>The first line tells your operating system that this
is a Python script and to use the interpreter located in that directory;
line two declares a variable called <code>myDNA</code> and assigns the sequence
string to it and the last line simply output this variable to the
screen. That simple! To run this (extremely simple) script you can copy
and paste the code above to your favourite text editor save the file
with a <code>.py</code> extension (recommended but not necessary). To run the
script, as long as you have Python installed, just open a shell and type
on the command line:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'>&gt; python code_01.py
</span></code></pre></td></tr></table></div></figure>

</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/39/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/37/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
