
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Changing gears now, leaving behind Pfam alignments. I decided to start a
new series of posts based on the conversion of some small C++ programs I &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/22">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/05/09/obtaining-overrepresented-motifs-in-dna-sequences-part-i/">Obtaining overrepresented motifs in DNA sequences, part 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-05-09T21:46:50-04:00" pubdate data-updated="true">May 9<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Changing gears now, leaving behind Pfam alignments. I decided to start a
new series of posts based on the conversion of some small C++ programs I
developed in the past. These small programs (I call them modules because
they were part of a larger application) were used to count motifs, short
nucleotide <em>words</em> up to 10-12 base pairs, and then calculate
statistical overrepresentation of these words by comparing a foreground
set of DNA sequences against a background set.</p>

<p>We will start comparing
the different approaches of the C++ and the
<a href="http://python.org/" title="Python (programming language)">Python</a> codes and
point out advantages and disadvantages of doing it in one language or
the other.</p>

<p>First thing we need to do is to count the motifs in all
sequences from our foreground and background sets. For the project I was
working on, the ideal word length was 10 nucleotides. Basically our C++
approach to increase speed was to transform the character DNA sequences
into numbers and then, while sliding a window with the desired word
length,
<a href="http://en.wikipedia.org/wiki/Hash_function" title="Hash function">hash</a> the
base-four numbers into base 10 and increment a vector position,
previously initialized with 0. For four nucleotides and a word size of
10 there are 1,048,576 permutations possible, from <code>AAAAAAAAAA</code> to
<code>TTTTTTTTTT</code>. Initially the C++ program would do</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
</pre></td><td class='code'><pre><code class='c'><span class='line'><span class="k">for</span><span class="p">(</span><span class="n">j</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">seqsize</span><span class="p">;</span> <span class="n">j</span><span class="o">++</span><span class="p">)</span>
</span><span class='line'><span class="p">{</span>
</span><span class='line'>  <span class="n">seqfile</span><span class="p">.</span><span class="n">get</span><span class="p">(</span><span class="n">base</span><span class="p">);</span>
</span><span class='line'>  <span class="k">if</span><span class="p">(</span><span class="n">base</span> <span class="o">!=</span> <span class="sc">&#39;\n&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="p">{</span>
</span><span class='line'>      <span class="k">switch</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
</span><span class='line'>      <span class="p">{</span>
</span><span class='line'>      <span class="k">case</span> <span class="sc">&#39;A&#39;</span>:
</span><span class='line'>          <span class="n">bseq</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="mi">0</span><span class="p">);</span>
</span><span class='line'>          <span class="k">break</span><span class="p">;</span>
</span><span class='line'>      <span class="k">case</span> <span class="sc">&#39;C&#39;</span>:
</span><span class='line'>          <span class="n">bseq</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="mi">1</span><span class="p">);</span>
</span><span class='line'>          <span class="k">break</span><span class="p">;</span>
</span><span class='line'>      <span class="k">case</span> <span class="sc">&#39;G&#39;</span>:
</span><span class='line'>          <span class="n">bseq</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="mi">2</span><span class="p">);</span>
</span><span class='line'>          <span class="k">break</span><span class="p">;</span>
</span><span class='line'>      <span class="k">case</span> <span class="sc">&#39;T&#39;</span>:
</span><span class='line'>          <span class="n">bseq</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="mi">3</span><span class="p">);</span>
</span><span class='line'>          <span class="k">break</span><span class="p">;</span>
</span><span class='line'>      <span class="p">}</span>
</span><span class='line'>  <span class="p">}</span>
</span><span class='line'><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p> reading all
sequences and pushing an figure for each nucleotide in a vector, and
then sliding a window on this vector and hashing the base-four number</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='c'><span class='line'><span class="kt">int</span> <span class="nf">hashSeq</span><span class="p">(</span><span class="n">vector</span> <span class="n">subseq</span><span class="p">)</span> <span class="p">{</span> <span class="kt">int</span> <span class="n">w</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span>
</span><span class='line'><span class="n">hashvalue</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">power</span><span class="p">;</span> <span class="n">w</span> <span class="o">=</span> <span class="n">subseq</span><span class="p">.</span><span class="n">size</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">;</span> <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">i</span> <span class="o">&lt;</span>
</span><span class='line'><span class="n">subseq</span><span class="p">.</span><span class="n">size</span><span class="p">();</span> <span class="n">i</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span> <span class="n">power</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">hashvalue</span> <span class="o">+=</span> <span class="n">subseq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="err">\</span><span class="o">*</span>
</span><span class='line'><span class="n">pow</span><span class="p">((</span><span class="kt">double</span><span class="p">)</span><span class="mi">4</span><span class="p">,(</span><span class="kt">double</span><span class="p">)</span><span class="n">w</span><span class="p">);</span> <span class="n">w</span><span class="o">--</span><span class="p">;</span> <span class="p">}</span> <span class="k">return</span> <span class="n">hashvalue</span><span class="p">;</span> <span class="p">}</span> <span class="k">if</span><span class="p">(</span><span class="n">binseq</span><span class="p">[</span><span class="n">i</span><span class="p">].</span><span class="n">size</span><span class="p">()</span>
</span><span class='line'><span class="err">\</span><span class="o">&gt;</span> <span class="n">motifwidth</span><span class="p">)</span> <span class="p">{</span> <span class="k">for</span><span class="p">(</span><span class="n">j</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">binseq</span><span class="p">[</span><span class="n">i</span><span class="p">].</span><span class="n">size</span><span class="p">()</span><span class="o">-</span><span class="n">motifwidth</span><span class="o">+</span><span class="mi">1</span><span class="p">;</span> <span class="n">j</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span>
</span><span class='line'><span class="n">sub</span><span class="p">.</span><span class="n">assign</span><span class="p">(</span><span class="n">binseq</span><span class="p">[</span><span class="n">i</span><span class="p">].</span><span class="n">begin</span><span class="p">()</span> <span class="o">+</span> <span class="n">j</span><span class="p">,</span> <span class="n">binseq</span><span class="p">[</span><span class="n">i</span><span class="p">].</span><span class="n">begin</span><span class="p">()</span> <span class="o">+</span> <span class="n">j</span><span class="o">+</span><span class="n">motifwidth</span><span class="p">);</span>
</span><span class='line'><span class="n">hashed</span> <span class="o">=</span> <span class="n">hashSeq</span><span class="p">(</span><span class="n">sub</span><span class="p">);</span> <span class="n">nmercount</span><span class="p">[</span><span class="n">hashed</span><span class="p">]</span><span class="o">++</span><span class="p">;</span> <span class="n">sub</span><span class="p">.</span><span class="n">clear</span><span class="p">();</span> <span class="p">}</span>
</span><span class='line'><span class="p">}</span>
</span></code></pre></td></tr></table></div></figure>


<p>The whole C++ code has about 400 lines, including all the
possible output formatting and printing. Timing with <code>time</code> the C++
executable takes a little bit less than 2 seconds to read, count and
output different files. For Python, we will use a different approach and
gain a lot in code simplicity. As we want to count the number of times
size-10 words appear in all sequences, we first need to generate all
possible permutations (with replacement) of four nucleotides. This can
be easily accomplished by using generator functions. Regular functions
run until completion and the return a value. For instance, a function
that calculates the factorial of 10 will return the last value only,
after multiplying 10.9.8.7.6.5.4.3.2.</p>

<p>A generator function runs until a
value is available to return, <code>yielding</code> it and then suspending its
operation until called again. The yielding part was emphasized because
<code>yield</code> is the command used by Python to return the value and suspend
the function until further notice. In the <a href="http://en.wikipedia.org/wiki/Factorial" title="Factorial">factorial function</a>, a
generator would return the intermediary factorial values up to 10. To
generate all 1 million plus permutations of 4 nucleotides we need a
function similar to the one below (modified from
<a href="http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/190465">here</a>)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">permutations</span><span class="p">(</span><span class="n">items</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">yield</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">items</span><span class="p">)):</span>
</span><span class='line'>          <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">permutations</span><span class="p">(</span><span class="n">items</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
</span><span class='line'>              <span class="k">yield</span> <span class="nb">str</span><span class="p">(</span><span class="n">items</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>Basically, what this generator function does is to combine all four
nucleotides in words of size 10. This is a <a href="http://en.wikipedia.org/wiki/Recursion_(computer_science">recursive
function</a> &#8220;Recursion (computer science)&#8221;),
where the result of the function is dependent on the n-1 value
calculated by the function until n equals 0. The first <code>for</code> loop over
the items that we want to permute (the nucleotides) and the second <code>for</code>
recursively calls <code>permutations starting with the initial n</code> passed (10)
until we reach 0. Debugging this function we will see that <code>i</code> is
constant for each iteration of the second loop and only <code>n</code> changes from
10 to 0, while one by one nucleotides are joined to form a motif. It
starts with <code>AAAAAAAAAA</code>, then <code>AAAAAAAAAC</code>, then <code>AAAAAAAAAG</code>, until it gets
to a poly-T. Our final code would look like the one below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="k">def</span> <span class="nf">permutations</span><span class="p">(</span><span class="n">items</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="k">yield</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">else</span><span class="p">:</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">items</span><span class="p">)):</span>
</span><span class='line'>          <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">permutations</span><span class="p">(</span><span class="n">items</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
</span><span class='line'>              <span class="k">yield</span> <span class="nb">str</span><span class="p">(</span><span class="n">items</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="n">seqs</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">get_seqs</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="n">length</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
</span><span class='line'><span class="n">nucleotides</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>
</span><span class='line'><span class="n">merged_seqs</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">:</span>
</span><span class='line'>  <span class="n">merged_seqs</span> <span class="o">+=</span> <span class="n">i</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">permutations</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">length</span><span class="p">)):</span>
</span><span class='line'>      <span class="k">print</span> <span class="n">i</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">merged_seqs</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>where
we read the input sequence(s), merge them in one long string and as we
generate all possible combinations we count the number of times they
appear. This code running on the same input file used on the C++
executable is 60 times slower, taking in average one full minute to
count motifs in 8 500 bp DNA sequences. The slowest section is the
count, as the generation of all possible combinations is
straightforward. We lose some speed, but gain a lot on code simplicity
and clarity. Next we will modify this code to output different counts
needed for the statistical analysis.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/05/05/still-on-merging-pfam-alignments/">Still on merging Pfam alignments &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-05-05T15:54:10-04:00" pubdate data-updated="true">May 5<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>One of the things I like about Python and the Python community is the
search for the making code simple and clear. Tal left a comment in the
last post about merging Pfam alignment sequences suggesting another
approach to our problem. The code is below</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">merge_seqs</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>  <span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">chain</span><span class="p">,</span> <span class="n">groupby</span>
</span><span class='line'>  
</span><span class='line'>  <span class="n">format</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s">-</span><span class="si">%s</span><span class="s">-\&gt;</span><span class="si">%d</span><span class="se">\\</span><span class="s">n</span><span class="si">%s%s</span><span class="s">&quot;</span>
</span><span class='line'>  <span class="n">flist</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="n">keyfunc</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">it</span><span class="p">:</span> <span class="n">it</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">it</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">it</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">it</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">groupby</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">chain</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">keyfunc</span><span class="p">),</span> <span class="n">keyfunc</span><span class="p">):</span>
</span><span class='line'>      <span class="n">values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
</span><span class='line'>      <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
</span><span class='line'>          <span class="n">jname</span><span class="p">,</span> <span class="n">jseq</span> <span class="o">=</span> <span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
</span><span class='line'>          <span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>          <span class="n">kname</span><span class="p">,</span> <span class="n">kseq</span> <span class="o">=</span> <span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>          <span class="n">flist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">format</span> <span class="o">%</span> <span class="p">(</span><span class="n">jname</span><span class="p">,</span> <span class="n">kname</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">jseq</span><span class="p">),</span> <span class="n">jseq</span><span class="p">,</span> <span class="n">kseq</span><span class="p">))</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">return</span> <span class="n">flist</span>
</span></code></pre></td></tr></table></div></figure>


<p>The code also uses the <code>itertools</code> module, importing
<code>chains</code> and <code>groupby</code>. We already saw chains in the previous post, but
<code>groupby</code> is new to us here. <code>groupby</code> was introduced in the 2.4 version
of Python and is a method returns keys and groups from an <code>iterable</code>. An
Python <code>iterable</code> is any object that can return its elements at given
time, for instance in a <code>for</code> loop, while the index of this loop is the
<code>iterator</code>. So, in our case <code>groupby</code> will return the sequence names
based on the <code>lambda</code> function defined before the <code>groupby</code> and the
<code>chain</code> method. Usually <code>groupby</code> has this syntax</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">groupby</span><span class="p">(</span><span class="n">iterable</span><span class="p">[,</span> <span class="n">key</span><span class="p">])</span>
</span></code></pre></td></tr></table></div></figure>


<p>The key is
optional, and in our case it is the lambda function. Another method new
to use that uses the same lambda function is <code>sorted</code>. As its name
hints, <code>sorted</code> returns a sorted list of iterables. The key in this case
is the sorting algorithm, that actually creates the comparison between
items. Basically in the code above, a lambda function extracts the
desired regions from the sequence names, which are them iterated in a
<code>groupby</code> method that returns they key values, one value when the sequence
is unique, two values when there are two sequences, of a sorted iterable
generated by a chain that read both input lists in one pass. After this
we just need to check the number of returned values and we have our list
of matching sequences.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/05/03/revisting-pfam-alignments-again/">Revisting Pfam alignments, again &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-05-03T08:39:31-04:00" pubdate data-updated="true">May 3<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Continuing on the <a href="http://en.wikipedia.org/wiki/Pfam" title="Pfam">Pfam</a>
alignment sequence merge, Luke provided two solutions, one for case
where we have duplicates in the file (that happens) and another one
where duplicates are not tested. First for the case with no duplicates:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">merge_seqs</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>  <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">()</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data1</span><span class="p">:</span>
</span><span class='line'>      <span class="n">first</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]]</span> <span class="o">=</span> <span class="n">i</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data2</span><span class="p">:</span>
</span><span class='line'>      <span class="n">second</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]]</span> <span class="o">=</span> <span class="n">i</span>
</span><span class='line'>      
</span><span class='line'>  <span class="n">shared_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">first</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">second</span><span class="p">))</span>
</span><span class='line'>  
</span><span class='line'>  <span class="n">flist</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">shared_ids</span><span class="p">:</span>
</span><span class='line'>      <span class="n">j</span> <span class="o">=</span> <span class="n">first</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span><span class='line'>      <span class="n">k</span> <span class="o">=</span> <span class="n">second</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span><span class='line'>      <span class="n">tempname</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-&#39;</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-\&gt;&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">j</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
</span><span class='line'>      <span class="n">tempseq</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>      <span class="n">flist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tempname</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span> <span class="o">+</span> <span class="n">tempseq</span><span class="p">)</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">return</span> <span class="n">flist</span>
</span></code></pre></td></tr></table></div></figure>


<p>Basically, his approach is to create two dictionaries, <code>first</code> and
<code>second</code> to store the data from our lists of
<a href="http://fasta.bioch.virginia.edu/" title="FASTA">FASTA</a> objects. The
dictionaries keys are the sections for each sequence name, and the values
are the FASTA object created when we read the file. After that, a <code>set</code>
of both dictionaries is created, and this represents everything we need
in order to find shared sequences between both files. The output is
trivial, basically iterating the set and getting the values from both
dictionaries. But, in some cases there are some duplicates in each Pfam
alignment, and this would make the above approach to miss some
sequences. Luke also provided a solution for this case</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
</span><span class='line'><span class="k">def</span> <span class="nf">merge_seqs</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>  <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">),</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data1</span><span class="p">:</span>
</span><span class='line'>      <span class="n">first</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data2</span><span class="p">:</span>
</span><span class='line'>      <span class="n">second</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'>  <span class="n">shared_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">first</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">second</span><span class="p">))</span>
</span><span class='line'>  <span class="n">flist</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">shared_ids</span><span class="p">:</span>
</span><span class='line'>      <span class="n">cross</span> <span class="o">=</span> <span class="p">[(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">first</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">second</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
</span><span class='line'>      <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">k</span> <span class="ow">in</span> <span class="n">cross</span><span class="p">:</span> <span class="n">tempname</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-&#39;</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s">&#39;-\&gt;&#39;</span><span class="err">’</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">j</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
</span><span class='line'>      <span class="n">tempseq</span> <span class="o">=</span> <span class="n">j</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="n">k</span><span class="o">.</span><span class="n">sequence</span>
</span><span class='line'>      <span class="n">flist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tempname</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span> <span class="o">+</span> <span class="n">tempseq</span><span class="p">)</span>
</span><span class='line'>      
</span><span class='line'><span class="k">return</span> <span class="n">flist</span>
</span></code></pre></td></tr></table></div></figure>


<p>which
is similar to what Mike suggested. In his approach <code>defaultdict</code> is used
instead of a regular dict, initialized as a list. Again we can take
advantage of the fact that we that the value of <code>defaultdict</code> has the
same features as the object that was used to initialized it. The same
<code>set</code> approach is used and we have a nice set of merged sequences from
two Pfam alignments. Remember that <code>defaultdict</code> were introduced in Python
2.5, it won&#8217;t work on older versions.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/05/02/revisiting-pfam-alignments/">Revisiting Pfam alignments: using defaultdicts, chains &#8230;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-05-02T10:11:47-04:00" pubdate data-updated="true">May 2<span>nd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I haven&#8217;t posted in a while, so let&#8217;s get back to the last topic covered
here, merging sequences from Pfam alignments. Two comments to my last
post suggested some changes to the original code, and both comments made
a considerable improvement to script. But following our line of thought
here, there were many things in both posts that we haven&#8217;t covered in
the series. In order to make it clear to anyone that is chronologically
following the entries, we will see what new things were suggested in the
comments. We will start with Mike&#8217;s comment. The code is below.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">merge_seqs</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>  <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
</span><span class='line'>  <span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">chain</span>
</span><span class='line'>  
</span><span class='line'>  <span class="n">data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">):</span>
</span><span class='line'>      <span class="n">ident</span> <span class="o">=</span> <span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)]</span>
</span><span class='line'>      <span class="n">data</span><span class="p">[</span><span class="n">ident</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
</span><span class='line'>      
</span><span class='line'>  <span class="n">format</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s">-</span><span class="si">%s</span><span class="s">-\&gt;</span><span class="si">%d</span><span class="se">\\</span><span class="s">n</span><span class="si">%s%s</span><span class="s">&quot;</span>
</span><span class='line'>  <span class="n">flist</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
</span><span class='line'>      <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
</span><span class='line'>          <span class="n">jname</span><span class="p">,</span> <span class="n">jseq</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'>          <span class="n">kname</span><span class="p">,</span> <span class="n">kseq</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>          <span class="n">flist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">format</span> <span class="o">%</span> <span class="p">(</span><span class="n">jname</span><span class="p">,</span> <span class="n">kname</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">jseq</span><span class="p">),</span> <span class="n">jseq</span><span class="p">,</span> <span class="n">kseq</span><span class="p">)</span> <span class="p">)</span>
</span><span class='line'>
</span><span class='line'>  <span class="k">return</span> <span class="n">flist</span>
</span></code></pre></td></tr></table></div></figure>


<p>Mike&#8217;s code would only run on Python 2.5, due to the
import of <code>collections</code>. He imports only <code>defaultdict</code> from collections,
which is similar to a regular dictionary except for the fact that is
takes an extra argument, a <code>factory function</code> (that we will examine
soon), and this <code>factory function</code> is called every time a <code>dict</code> key is
found for the the first time, initializing the object. The line</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>creates and intializes a <code>defaultdict</code> from an empty list. This
defaultdict is used to store data from a <code>chain</code> formed by both input
files being compared. A <code>chain</code> makes an iterator that &#8220;returns elements
from the first iterable until it is exhausted, the proceeds to the next
iterable&#8221; (from Python docs). Basically a <code>chain</code> in our case will make
the loop iterate first through <code>data1</code> until its last item then will
start the iteration through <code>data2</code>. In the loop it will use the region
in the FASTA sequence name as a defaultdict key, and store the fasta
name and sequence as the value. Notice that the defaultdict, as it has
been initialized as a list, allows appending more than one value to each
key, so every time that it finds a sequence with a name identical to
some key already present in the defaultdict, it will append the
information to the value of that key in a list.</p>

<p>So, all the matching is
done with this short loop, we only need then to iterate through the keys
and print the result. For the results, Mike set up a format string (we
will see in a future post) that receives the merged sequences and their
names in a formatted way. The iteration along the defaultdict checks for
both keys and values and every time it finds a value with a length of 2
(meaning two sequences shared identical names) it puts the items of the
defaultdict&#8217;s value list into two tuples. These tuples, containing
sequence and name, are then formatted for output and appended to a list
of merged sequences.</p>

<p>There are a lot of new concepts, objects and
methods but Mike&#8217;s suggestion shows how powerful Python is. For a set of
25 alignments, there is little performance improvement, but our coding
ability improves exponentially with clever suggestions. I will comment
and explain Luke&#8217;s comments later.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/29/habs/">Habs</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-29T15:29:14-04:00" pubdate data-updated="true">Apr 29<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://www.thehockeynews.com/articles/15664-New-poll-finds-Montreal-Canadiens-considered-Canadas-hockey-team.html">New poll finds Montreal Canadiens as Canada&#8217;s team</a>, more than Toronto Maple Leafs. Ok, next Saturday when the bandwagon suddenly stops, please try not to fall and hurt yourself.</p>

<p>Flyers in five.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/26/twitter-updates-for-2008-04-26/">Twitter Updates for 2008-04-26</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-26T23:59:59-04:00" pubdate data-updated="true">Apr 26<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><ul>
<li><p>@<a href="http://twitter.com/roguepuppet">roguepuppet</a> here in Toronto, there is a good place to find some different fruits. Supermarket bananas have a bland taste. <a href="http://twitter.com/nuin/statuses/797568309">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> I discuss anything, anything at all. I have an incommensurable amount of useless knowledge <a href="http://twitter.com/nuin/statuses/797567694">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> I discuss anything, anything at all. I have an incommensurable amount of useless knowledge <a href="http://twitter.com/nuin/statuses/797567695">#</a></p></li>
<li><p>going to eat some traditional japanese sushi and buy some gohan to stock <a href="http://twitter.com/nuin/statuses/797475856">#</a></p></li>
<li><p>@<a href="http://twitter.com/roguepuppet">roguepuppet</a> I am not fan of bananas sold in NA. Stay away from the supermarket bananas and get some apple-banana. Better taste <a href="http://twitter.com/nuin/statuses/797475446">#</a></p></li>
<li><p>@<a href="http://twitter.com/roguepuppet">roguepuppet</a> Magnolias are one of my favourite plants. <a href="http://twitter.com/nuin/statuses/797464033">#</a></p></li>
<li><p>@<a href="http://twitter.com/roguepuppet">roguepuppet</a> Magnolias are from a very old family of plants, Magnoliaceae, one the basal families of angiosperms. <a href="http://twitter.com/nuin/statuses/797450365">#</a></p></li>
<li><p>I&#8217;m never buying anything from Netgear again. <a href="http://twitter.com/nuin/statuses/797215074">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a>  @<a href="http://twitter.com/neilfwa">neilfwa</a> @<a href="http://twitter.com/rvidal">rvidal</a> I installed this one in a VMWare and wasn&#8217;t good at all. <a href="http://twitter.com/nuin/statuses/797191876">#</a></p></li>
</ul>


<p>Powered by <a href="http://alexking.org/projects/wordpress">Twitter Tools</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/25/twitter-updates-for-2008-04-25/">Twitter Updates for 2008-04-25</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-25T23:59:59-04:00" pubdate data-updated="true">Apr 25<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><ul>
<li><p>@<a href="http://twitter.com/ChristianBurns">ChristianBurns</a> what program are you using? <a href="http://twitter.com/nuin/statuses/797027006">#</a></p></li>
<li><p>why are Planet Nature and Postgenomic feeds &#8220;seasonal&#8221;? All of a sudden hundreds of entries appear &#8230; <a href="http://twitter.com/nuin/statuses/797007562">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> Nada como conhecer os patricios. <a href="http://twitter.com/nuin/statuses/796969162">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> indeed, and the anniversary of the Cravos. <a href="http://twitter.com/nuin/statuses/796941967">#</a></p></li>
<li><p>@<a href="http://twitter.com/michaelbarton">michaelbarton</a> OK, thanks a lot. <a href="http://twitter.com/nuin/statuses/796899983">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> thanks, will see what I can do &#8230; <a href="http://twitter.com/nuin/statuses/796899556">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> email me too, I&#8217;d take a look if you want &#8230; <a href="http://twitter.com/nuin/statuses/796816320">#</a></p></li>
<li><p>@<a href="http://twitter.com/michaelbarton">michaelbarton</a> sorry to ask, but what is ONS? <a href="http://twitter.com/nuin/statuses/796815735">#</a></p></li>
<li><p>@<a href="http://twitter.com/SciPhu">SciPhu</a> man, that&#8217;s sad &#8230; <a href="http://twitter.com/nuin/statuses/796789556">#</a></p></li>
<li><p>Sometimes I lose faith in the humanbeing: <a href="http://www.thestar.com/article/418415">http://www.thestar.com/article/418415</a> <a href="http://twitter.com/nuin/statuses/796785572">#</a></p></li>
<li><p>isn&#8217;t Neil awake today/tonight? <a href="http://twitter.com/nuin/statuses/796780529">#</a></p></li>
<li><p>@<a href="http://twitter.com/pierreyves">pierreyves</a> nice tune <a href="http://twitter.com/nuin/statuses/796772793">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a>  get the server already, more tools, more stuff to play around <a href="http://twitter.com/nuin/statuses/796753895">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> I think in Ubuntu adding is easier than removing. For something like Gentoo I&#8217;d say yes <a href="http://twitter.com/nuin/statuses/796737548">#</a></p></li>
<li><p>@<a href="http://twitter.com/hooande">hooande</a> the Raptors won? No wonder it is snowing in TO <a href="http://twitter.com/nuin/statuses/796428633">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> cheers mate, keep on going &#8230; <a href="http://twitter.com/nuin/statuses/796374052">#</a></p></li>
</ul>


<p>Powered by <a href="http://alexking.org/projects/wordpress">Twitter Tools</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/24/twitter-updates-for-2008-04-24/">Twitter Updates for 2008-04-24</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-24T23:59:59-04:00" pubdate data-updated="true">Apr 24<span>th</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><ul>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> Thanks a lot &#8230; <a href="http://twitter.com/nuin/statuses/796298220">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> can you add some tag to the email subject? Something like [BioCamp] or similar &#8230; <a href="http://twitter.com/nuin/statuses/796292302">#</a></p></li>
<li><p>gas will be 2.25 $CDN in 2012. Predicted by the same bank  that lost 5 blln with the house bubble. <a href="http://twitter.com/nuin/statuses/796259498">#</a></p></li>
<li><p>konqueror does not like FriendFeed &#8230; <a href="http://twitter.com/nuin/statuses/796221166">#</a></p></li>
<li><p>&#8220;molecular data from extinct organisms may have the potential for resolving relationships at critical areas of the vert. evolutionary tree&#8221; <a href="http://twitter.com/nuin/statuses/796137883">#</a></p></li>
<li><p>@<a href="http://twitter.com/pierreyves">pierreyves</a> thx, added to Google Reader &#8230; <a href="http://twitter.com/nuin/statuses/796106201">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> @<a href="http://twitter.com/freesci">freesci</a> I would say API too. <a href="http://twitter.com/nuin/statuses/796099470">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> 2:30 am madness, for sure <a href="http://twitter.com/nuin/statuses/796064280">#</a></p></li>
<li><p>@<a href="http://twitter.com/yokofakun">yokofakun</a> @<a href="http://twitter.com/covert">covert</a> basically you can merge all your &#8220;social web life&#8221; in one place. I find it useful to see what are being shared in GooRead <a href="http://twitter.com/nuin/statuses/796063240">#</a></p></li>
<li><p>@<a href="http://twitter.com/yokofakun">yokofakun</a> @<a href="http://twitter.com/covert">covert</a> like and see what everyone else is sharing &#8230; You can link a feed from your blog, twitter account, Flickr, etc <a href="http://twitter.com/nuin/statuses/796060765">#</a></p></li>
<li><p>@<a href="http://twitter.com/yokofakun">yokofakun</a> @<a href="http://twitter.com/covert">covert</a> It is a service that merges all the feeds that you generate  and allows interaction between friends, so you can comment <a href="http://twitter.com/nuin/statuses/796060188">#</a></p></li>
<li><p>@<a href="http://twitter.com/covert">covert</a> FriendFeed enables you to keep up-to-date on the web pages, photos, videos and music that your friends and family are sharing. <a href="http://twitter.com/nuin/statuses/796056053">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> twhirl, but not liking it. <a href="http://twitter.com/nuin/statuses/795976501">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> I unistalled on windows, my iMac is my social-hub. <a href="http://twitter.com/nuin/statuses/795970139">#</a></p></li>
<li><p>The Greenland ice cap&#8217;s apparently melting quickly that it&#8217;s triggering earthquakes, as pieces of ice several cubic kilometers break off. <a href="http://twitter.com/nuin/statuses/795952114">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> I had issues with Twhirl on Mac. It does not update sometimes, it hangs and sometimes does not display the entries <a href="http://twitter.com/nuin/statuses/795946181">#</a></p></li>
<li><p>@<a href="http://twitter.com/amberf">amberf</a> com esses links do demo &#8230; <a href="http://twitter.com/nuin/statuses/795945534">#</a></p></li>
<li><p>@<a href="http://twitter.com/amberf">amberf</a> pára com isso &#8230; <a href="http://twitter.com/nuin/statuses/795912461">#</a></p></li>
<li><p>Morpheus&#8217; calling <a href="http://twitter.com/nuin/statuses/795633525">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> I also have that feeling that every time I use R I have to relearn it. Maybe the syntax is not clear enough &#8230; <a href="http://twitter.com/nuin/statuses/795627772">#</a></p></li>
<li><p>shameless plug <a href="http://tinyurl.com/6557oy">http://tinyurl.com/6557oy</a> <a href="http://twitter.com/nuin/statuses/795557662">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> &#8220;Don&#8217;t worry about the world coming to an end today. It&#8217;s already
tomorrow in Australia.&#8221; (Charles Schultz) <a href="http://twitter.com/nuin/statuses/795550172">#</a></p></li>
</ul>


<p>Powered by <a href="http://alexking.org/projects/wordpress">Twitter Tools</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/23/twitter-updates-for-2008-04-23/">Twitter Updates for 2008-04-23</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-23T23:59:59-04:00" pubdate data-updated="true">Apr 23<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><ul>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> Thnx, because she refused to add my blog sometime ago. <a href="http://twitter.com/nuin/statuses/795496596">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> what is the requisite to have a blog added to the network? <a href="http://twitter.com/nuin/statuses/795479048">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> question: you are one of the organizers of the DNA network, right? <a href="http://twitter.com/nuin/statuses/795456865">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> I am fully engaged on the BioBarCamp. The only problem is getting there &#8230; <a href="http://twitter.com/nuin/statuses/795456346">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> I don&#8217;t even have an AppEngine account &#8230; <a href="http://twitter.com/nuin/statuses/795451794">#</a></p></li>
<li><p>@<a href="http://twitter.com/mza">mza</a> Damn, it beats my cassette player &#8230; <a href="http://twitter.com/nuin/statuses/795448825">#</a></p></li>
<li><p>early leave &#8230; <a href="http://twitter.com/nuin/statuses/795408410">#</a></p></li>
<li><p>@<a href="http://twitter.com/mza">mza</a> I think the repository is a major issue. The release policy is also non-existent, and regarding the busy I will blog about it &#8230; <a href="http://twitter.com/nuin/statuses/795404493">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> Yes, I should have put inside quotes. My bad. <a href="http://twitter.com/nuin/statuses/795395026">#</a></p></li>
<li><p>@<a href="http://twitter.com/mza">mza</a> @<a href="http://twitter.com/mndoci">mndoci</a> I agree with the time (lazyness) constraints, and timing. @<a href="http://twitter.com/yokofakun">yokofakun</a> I was thinking more of the finished product &#8230; <a href="http://twitter.com/nuin/statuses/795378116">#</a></p></li>
<li><p>why published (in sci journals) applications are not properly checked by the reviewers? Opinions &#8230; <a href="http://twitter.com/nuin/statuses/795364138">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> indeed, I hate it <a href="http://twitter.com/nuin/statuses/795187828">#</a></p></li>
<li><p>job description = bioinformatician <a href="http://twitter.com/nuin/statuses/795154295">#</a></p></li>
<li><p>@<a href="http://twitter.com/roguepuppet">roguepuppet</a> better than mine, I guess <a href="http://twitter.com/nuin/statuses/795150524">#</a></p></li>
<li><p>time to meet Morpheus <a href="http://twitter.com/nuin/statuses/794816738">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> did you read all six? <a href="http://twitter.com/nuin/statuses/794808549">#</a></p></li>
<li><p>@<a href="http://twitter.com/Scobleizer">Scobleizer</a> to celebrate she will nuke Iran. <a href="http://twitter.com/nuin/statuses/794807411">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> I asked two: one does not like to talk about himself, the other said he would do it but only after July. <a href="http://twitter.com/nuin/statuses/794803989">#</a></p></li>
<li><p>@<a href="http://twitter.com/rvidal">rvidal</a> good call too, I will add to my list (very short at the moment) <a href="http://twitter.com/nuin/statuses/794797404">#</a></p></li>
<li><p>@<a href="http://twitter.com/neilfws">neilfws</a> @<a href="http://twitter.com/mndoci">mndoci</a> I had thought of her a long time ago, but decided to interview Rosie Redfield first due to the open science thingy &#8230; <a href="http://twitter.com/nuin/statuses/794793964">#</a></p></li>
<li><p>how about Janet Thornton interviewed in SciView? What do you guys think? <a href="http://twitter.com/nuin/statuses/794791750">#</a></p></li>
<li><p>@<a href="http://twitter.com/mndoci">mndoci</a> that&#8217;s why I don&#8217;t have a calendar. <a href="http://twitter.com/nuin/statuses/794790816">#</a></p></li>
<li><p>Thanks Flyers, I will watch more than 7 meaningful playoff hockey games this year. <a href="http://twitter.com/nuin/statuses/794785793">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> a more psychological, catastrophic, biological sci-fi <a href="http://twitter.com/nuin/statuses/794785462">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> I think we are living the sci-fi of gadgets, and I think it is time to the sci-fi of Dick, Ballard, etc to revamp &#8230; <a href="http://twitter.com/nuin/statuses/794785155">#</a></p></li>
<li><p>looking for suggestions for a new SciView interview, last couple were declined &#8230; <a href="http://blindscientist.genedrift.org/category/sciview/">http://blindscientist.genedrift.org/category/sciview/</a> <a href="http://twitter.com/nuin/statuses/794738655">#</a></p></li>
<li><p>@<a href="http://twitter.com/attilacsordas">attilacsordas</a> Sunshine is a nice movie. <a href="http://twitter.com/nuin/statuses/794737794">#</a></p></li>
<li><p>2 x 2, c&#8217;mon Flyers <a href="http://twitter.com/nuin/statuses/794723793">#</a></p></li>
</ul>


<p>Powered by <a href="http://alexking.org/projects/wordpress">Twitter Tools</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2008/04/23/how-to-improve-scientific-software/">How to improve scientific software?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2008-04-23T19:45:15-04:00" pubdate data-updated="true">Apr 23<span>rd</span>, 2008</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Do you know the answer to the above question? No? Me neither, but I can offer some suggestions. On a daily basis, a bioinformatician is exposed to hundreds of applications, computer languages, websites, you name it. Some of them are commercial, some of them free and open source. Some of the academia-developed software are open-source, some of them are not.</p>

<p>A good portion of the academia-developed software are published in scientific journals, as an 2-page application note in <a href="http://bioinformatics.oxfordjournals.org/">Bioinformatics </a>, or on a longer paper on <a href="http://www.biomedcentral.com/bmcbioinformatics/">BMC Bioinformatics</a>, just to name two of the journals of the field.</p>

<p>I cannot complain of non-published applications. Usually they are free, open, and were developed during someone&#8217;s spare time. I have the option of not using them, or modify them or helping the developer to improve it. I cannot complain of lack of documentation, bugs, minor errors or even the lack of an interface.</p>

<p>On the other hand, I can complain about published applications. Usually they are also free, but not always open and they were developed with a publication in mind, or at least as a mean towards a publication. It should have proper documentation, be slightly portable (yes, that&#8217;s important, so if you are developing your next groundbreaking phylogenetic tool in <a href="http://caml.inria.fr/">OCaml</a>, distribute the executable, don&#8217;t ask me to &#8220;compile&#8221; or install OCaml) and be easy or moderately easy to use. I give a break for the lack of interface.</p>

<p>And, why should I expect such things from a published application? Because, apart from the developers, at least one editor and two reviewers have (supposedly) tested (or glanced over) the application. Usually, scientific journals ask the authors to provide a copy of the package that is being submitted to publication. Basically one should include everything that is going to be installed, compiled, etc. So, a simple manual or readme, installation instructions, source code, executable(s), you name it, should be in the package.</p>

<p>I can say, from personal experience, that the majority of the published applications will have most of the items required for a user-friendly experience. But many fail in at least one of these aspects:</p>

<ul>
<li><p>poor (or nonexistent) documentation</p></li>
<li><p>errors and bugs, too evident to be missed by the reviewers/editor</p></li>
<li><p>far from being user-friendly, even the command-line ones</p></li>
<li><p>non-portable (even between <a href="http://en.wikipedia.org/wiki/Linux">Linux</a> distros)</p></li>
</ul>


<p>This makes me ask myself: Are the editors/reviewers testing, checking, using these applications at all, before accepting the manuscript? I guess not, or it seems that they are not. Some errors you see in some applications are easy fix, and wouldn&#8217;t harm the program&#8217;s merit, but in the end would greatly improve user satisfaction and require less email exchanges with developers/scientists regarding bugs and errors. After all, you already published it, nobody told you that you have to support it, right?</p>

<p>This brings us to the original question: how to improve scientific software? Simple &#8230; no, not really.</p>

<p>Far from having the ideal solution, I will add my two cents:</p>

<p>1- Journals can to create a more rigorous publication process for applications, what would include more testing from the editor/reviewers. That would make the review process slower, will make editors and reviewers to spend time, that they already have in short supply, on a job that they are not  compensated for. Everybody will be unhappy and the process fails. So, if journals are willing to publish application only manuscripts, why not have an in house testing facility, at least to check for basic things that the authors claim their software should do? Too expensive for the journals? Maybe, maybe not.</p>

<p>2- Publication of applications <em>per se</em> should be abolished. If you want to release an application and publish it, be sure you prove the merits of it with a publication that includes the software, its application and results that are scientifically relevant. This way we also abolish the publication-of-the-main-project, along with the paper on the application-that-was-used-to-generate-the-results and the paper about the program-that-was-used-to-display-the-results-generated-by-the-other-application. This way we would also foster collaboration between pure bioinformatics/application development groups with wet-lab groups.</p>

<p>3- Create a centralized scientific software repository, something like Sourceforge, and let people contribute, collaborate, code and develop and when the application is mature enough, publish it and give credit to everyone that had some input. Depending on the quality and amount of time dedicated to a project, the person would be a co-author or cited in the acknowledgments. This would also increase collaboration, people interested can learn and teach, exchange ideas and develop good application development standards, different computer languages, etc. Maybe you might not even need to publish the software in a journal, the user input and evaluations would be enough to solidify a good developer/scientist CV.</p>

<p>All options above are not mutually exclusive, they can be implemented at the same time, or they can&#8217;t be implemented at all. Would Open Science benefit from this? I bet so.</p>

<p><a href="http://www.zemanta.com/"><img src="http://img.zemanta.com/pixie.png?x-id=c24ea5f5-9528-4569-ac1c-ce2cea8e07c7" alt="" /></a></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/23/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/21/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
