
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="&#8220;Mendeley seems to be choking while importing my 1039 pdfs from ~/Documents/Papers/. Beachballing, can&#8217;t tell what it is doing.&#8221; &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/13">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/25/its-not-just-me-victor/">It&#8217;s not just me, Victor.</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-25T09:27:13-04:00" pubdate data-updated="true">Apr 25<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>&#8220;<a href="http://en.wikipedia.org/wiki/Mendeley">Mendeley</a> seems to be choking while importing my 1039 pdfs from ~/Documents/Papers/. Beachballing, can&#8217;t tell what it is doing.&#8221; <a href="http://twitter.com/mja/status/1612595721">mja on Twitter</a>.</p>

<p>And in reply:</p>

<p>&#8220;@mja Mendeley did exactly the same to me. Hangs on every start up now.&#8221; <a href="http://twitter.com/sjcockell/status/1612725193">sjcockell on Twitter</a>.</p>

<p>So, I guess it&#8217;s not something personal to the nice chaps that work at Mendeley. The software is pretty bad, and that&#8217;s the only reality here. And I&#8217;m not criticizing, I&#8217;m just stating.</p>

<h6>Related articles by Zemanta</h6>

<ul>
<li><a href="http://my.biotechlife.net/2009/03/04/community-liaison-mendeley/">Community Liaison @ Mendeley</a> (my.biotechlife.net)</li>
<li><a href="http://blindscientist.genedrift.org/2009/03/18/i-havent-joined-mendeley-as-community-liaison/">I haven&#8217;t joined Mendeley as community liaison</a> (blindscientist.genedrift.org)</li>
<li><a href="http://blindscientist.genedrift.org/2009/04/03/mendeley-an-extremely-short-review-sort-of/"> Mendeley: an extremely short review (sort of) </a> (blindscientist.genedrift.org)</li>
</ul>


<p><a href="http://reblog.zemanta.com/zemified/7352cb9c-7ac1-4a61-800e-55eec6897904/"><img src="http://img.zemanta.com/reblog_a.png?x-id=7352cb9c-7ac1-4a61-800e-55eec6897904" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/22/managing-a-simple-database-with-python-sqlite-and-wxpython-8/">Managing a simple database with Python, SQLite and wxPython, 8</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-22T10:04:17-04:00" pubdate data-updated="true">Apr 22<span>nd</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Gene.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/0/07/Gene.png/200px-Gene.png" title="Diagram of the location of introns and exons w..." alt="Diagram of the location of introns and exons
w..." /></a>
  ~ Image via</p>

<pre><code>[Wikipedia](http://commons.wikipedia.org/wiki/Image:Gene.png)
</code></pre>

<p>Thanks to the comments and suggestions to the last post, it&#8217;s possible
to make now a more pythonic and clearly generic database update class.
Let&#8217;s check how the &#8220;generic&#8221; update/edit entry function is currently:
[sourcecode language=&#8217;python&#8217;]def update_data(self, values_list):
&#8221;&#8217;edits and updates fields&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor,
database) = link_db(self.db_path) else: (cursor, database) =
link_db() cursor.execute(&#8220;UPDATE bac SET projects = ?, comments = ?,
temperature = ?, cell = ?, box = ?, tubes = ?, chromosome = ?, sdate =
?, clone = ?, source = ?, location1 = ?, startpos = ?, endpos = ?, gene
= ?, genelink = ?, dnaex = ?, validation = ?, pcr = ?, refs = ?,
antibiotic = ? WHERE idbac = ?&#8221;, values_list[&#8216;projects&#8217;],
values_list[&#8216;comments&#8217;], values_list[&#8216;temperature&#8217;],
values_list[&#8216;cell&#8217;], values_list[&#8216;box&#8217;], values_list[&#8216;tubes&#8217;],
values_list[&#8216;chromo&#8217;], values_list[&#8216;date&#8217;], values_list[&#8216;clone&#8217;],
values_list[&#8216;source&#8217;], values_list[&#8216;location&#8217;], values_list[&#8216;start&#8217;]
values_list[&#8216;end&#8217;], values_list[&#8216;gene&#8217;], values_list[&#8216;genelink&#8217;],
values_list[&#8216;dna&#8217;], values_list[&#8216;validation&#8217;], values_list[&#8216;pcr&#8217;],
values_list[&#8216;refs&#8217;], values_list[&#8216;antibiotic&#8217;],
values_list[&#8216;idbac&#8217;])) database.commit() database.close()[/sourcecode]
which is really ugly and, although it works, is not really useful
outside this small project. Based on the comments the best option was to
use placeholders and a dictionary, similar to the approach used on the
insert data function. Pre-formatting a string to have both the field
name to be updated and a placeholder (for instance <code>:idbac</code>) that will
receive the values [sourcecode language=&#8217;python&#8217;]update =
&#8217;,&#8217;.join([&#8216;%s=:%s&#8217; % (y, y) for y in values_list])[/sourcecode] where
update is the string we want and values_list is the dictionary with all
the key-value pairs. I tried this approach, using this structure in the
generic function, but then I decided that the best alternative was to
put this <code>join</code> in the derived class function and pre-populate the
string with the values and then send this string directly to the update
function. In the end I opted to use this [sourcecode
language=&#8217;python&#8217;]update = &#8216;,&#8217;.join([&#8216;%s=&#34;%s&#34;&#8217; % (y,
values_list[y]) for y in values_list])[/sourcecode] The latter is
slightly different to what was suggested. The original one would create
a tuple with the keys from the dictionary, making for instance
<code>sdate:sdate</code>. With all these place holders just pass the dictionary and
you have all the values inserted. This would be handy if the insert
string was being created on the &#8220;generic&#8221; function. If we move this to
the derived class, we can use the the alternative, keeping in mind that
the values parsed should be surrounded by quotes, otherwise the SQL
UPDATE statement will have problems with spaces and other foreign
characters that should not be there. So instead of placeholders we will
have <code>gene:"PTEN"</code> and we can attache this joined string to the actual
commands. We then can move all the machinery from the &#8220;generic&#8221; function
that can be written as [sourcecode language=&#8217;python&#8217;]def
update_data(self, update_string): &#8221;&#8217;edits and updates fields&#8221;&#8217; if
sys.platform == &#8216;darwin&#8217;: (cursor, database) = link_db(self.db_path)
else: (cursor, database) = link_db() cursor.execute(update_string)
database.commit() database.close()[/sourcecode] That&#8217;s it, very elegant
(we will see the derived class in the next post). And finishing our
generic class, we would need a delete function, so the user can
eliminate entries that he/she doesn&#8217;t want anymore. It&#8217;s also a very
simple function [sourcecode language=&#8217;python&#8217;]def delete_data(self,
delete_string): &#8221;&#8217;deletes one field&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;:
(cursor, database) = link_db(self.db_path) else: (cursor, database) =
link_db() cursor.execute(delete_string) database.commit()
database.close()[/sourcecode] We will check the delete string next time.
Again, I would like to thank for all the comments, it has been really
helpful for me. Previously in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://python.genedrift.org/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Part
3</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-4/">Part
4</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-5/">Part
5</a>
<a href="http://python.genedrift.org/2009/03/31/managing-a-simple-database-with-python-sqlite-and-wxpython-6/">Part
6</a>
<a href="http://python.genedrift.org/2009/04/20/managing-a-simple-database-with-python-sqlite-and-wxpython-7-includes-a-question/">Part
7</a></p>

<p><a href="http://reblog.zemanta.com/zemified/e8dc77f5-e3de-4d4f-8ec1-8c0006225743/" title="Reblog this post [with Zemanta]"><img src="http://img.zemanta.com/reblog_a.png?x-id=e8dc77f5-e3de-4d4f-8ec1-8c0006225743" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/20/managing-a-simple-database-with-python-sqlite-and-wxpython-7-includes-a-question/">Managing a simple database with Python, SQLite and wxPython, 7 (includes a question)</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-20T12:21:59-04:00" pubdate data-updated="true">Apr 20<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>And we&#8217;re back. After a couple of weeks of inactivity we will get back
to our small soap-opera pf Python, wxPython and SQLite. Continuing in
our database management code let&#8217;s check two other functions that
changed since our first inception of the code. The first one is the
<code>insert_data</code> function that looks like this now [sourcecode
language=&#8217;python&#8217;]def insert_data(self, values_list, insert_string):
&#8221;&#8217;inserts data in the database&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor,
database) = link_db(self.db_path) else: (cursor, database) =
link_db() cursor.execute(insert_string % self.table_name,
values_list) database.commit() database.close()[/sourcecode] Basically
no changes, apart from the obvious check for the current running
operating system, which was explained in the last post. The other
function to check is the <code>update_data</code>. This function is new and it
wasn&#8217;t in the first version, but as it can be seen it has a problem
being a &#8220;generic&#8221; function, because it contains information pertained to
the table and database being used in the interface. This function
basically received information that needs to be updated in the table&#8217;s
fields and by using the SQL <code>UPDATE ... SET</code> edits and updates data in
the changed fields. I have tried several different syntaxes to make the
execute generic, mainly trying to pre-format the string without success.
IF anyone reading this can help, I&#8217;d appreciate. [sourcecode
language=&#8217;python&#8217;]def update_data(self, values_list): &#8221;&#8217;edits and
updates fields&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor, database) =
link_db(self.db_path) else: (cursor, database) = link_db()
cursor.execute(&#8220;UPDATE bac SET projects = ?, comments = ?, temperature =
?, cell = ?, box = ?, tubes = ?, chromosome = ?, sdate = ?, clone = ?,
source = ?, location1 = ?, startpos = ?, endpos = ?, gene = ?, genelink
= ?, dnaex = ?, validation = ?, pcr = ?, refs = ?, antibiotic = ? WHERE
idbac = ?&#8221;, values_list[&#8216;projects&#8217;], values_list[&#8216;comments&#8217;],
values_list[&#8216;temperature&#8217;], values_list[&#8216;cell&#8217;], values_list[&#8216;box&#8217;],
values_list[&#8216;tubes&#8217;], values_list[&#8216;chromo&#8217;], values_list[&#8216;date&#8217;],
values_list[&#8216;clone&#8217;], values_list[&#8216;source&#8217;], values_list[&#8216;location&#8217;],
values_list[&#8216;start&#8217;], values_list[&#8216;end&#8217;], values_list[&#8216;gene&#8217;],
values_list[&#8216;genelink&#8217;], values_list[&#8216;dna&#8217;],
values_list[&#8216;validation&#8217;], values_list[&#8216;pcr&#8217;], values_list[&#8216;refs&#8217;],
values_list[&#8216;antibiotic&#8217;], values_list[&#8216;idbac&#8217;])) database.commit()
database.close()[/sourcecode] Anyway, I will explain the logic of the
command (OK for a stop gap, but not as a definite solution).
<code>values_list</code> is a dictionary that is passed to the function and
contains the field names as keys and the new/changed information as
values. The execute method simply parses the values from each key in the
update string which is then sent to the database and table to be
changed. Everything is committed and the database is closed. As this is
a &#8220;generic&#8221; function from a &#8220;generic&#8221; class the ideal scenario would be
to the function to receive a pre-formatted string with all the
information, as in the insert data function, and update the information
in the database. I would like to thank in advance anyone that can
comment on this. Next time we will continue checking the generic class
and finalize this part in order to start with the interface build
process.\
\
 Previously in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://python.genedrift.org/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Part
3</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-4/">Part
4</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-5/">Part
5</a>
<a href="http://python.genedrift.org/2009/03/31/managing-a-simple-database-with-python-sqlite-and-wxpython-6/">Part
6</a>
<a href="http://reblog.zemanta.com/zemified/d0bb5d11-6f9d-8521-9a2f-6cd30868e375/" title="Reblog this post [with Zemanta]"><img src="http://img.zemanta.com/reblog_a.png?x-id=d0bb5d11-6f9d-8521-9a2f-6cd30868e375" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/19/thanks-jim/">Thanks, Jim!</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-19T19:35:46-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:EmpireOfTheSun%281stEd%29.jpg"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/0/0e/EmpireOfTheSun%281stEd%29.jpg/202px-EmpireOfTheSun%281stEd%29.jpg" alt="Empire of the Sun" /></a></p>

<pre><code>Image via [Wikipedia](http://en.wikipedia.org/wiki/Image:EmpireOfTheSun%281stEd%29.jpg)
</code></pre>

<p>It was a great ride. It all started (for me) with a Crash and it will never end. Thanks for all the metaphors and for the books, they will be sorely missed.</p>

<p><a href="http://reblog.zemanta.com/zemified/c9c943e8-3f4f-442c-8339-2a5a580bd314/"><img src="http://img.zemanta.com/reblog_a.png?x-id=c9c943e8-3f4f-442c-8339-2a5a580bd314" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/17/this-should-not-be-here-or-notes-to-a-bioinformatician-two-years-later/">This should not be here &#8230; or Notes to a bioinformatician - two years later</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-17T11:07:25-04:00" pubdate data-updated="true">Apr 17<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content">
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/04/03/mendeley-an-extremely-short-review-sort-of/">Mendeley: an extremely short review (sort of)</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-04-03T14:21:27-04:00" pubdate data-updated="true">Apr 3<span>rd</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Fivestagespipeline.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/2/21/Fivestagespipeline.png/202px-Fivestagespipeline.png" alt="Basic five-stage pipeline.  In the best case s..." /></a></p>

<pre><code>Image via [Wikipedia](http://commons.wikipedia.org/wiki/Image:Fivestagespipeline.png)
</code></pre>

<p>I decided to test <a href="http://en.wikipedia.org/wiki/Mendeley">Mendeley</a> again (Yes, I have some kind of a masochist behaviour for software), and as the <a href="http://blindscientist.genedrift.org/2008/08/25/mendeley-a-very-short-review/">last time I tried</a>, I can say I&#8217;m not impressed.</p>

<p>I tested on a directory containing mixed bag of PDFs, books, book chapters, articles and misc stuff. Most of the time the data was extracted but I ended with a wrong reference or garbage information about the file. Not mentoning that it took almost one full hour to analyse a little bit more than 500 files. I think I need to buy an eight-core machine now. The CPU peaked at 25-30% (one full core for it) on Vista, and the memory got up to 300 Mbs.</p>

<p>Not impressed, maybe one day.</p>

<p><a href="http://reblog.zemanta.com/zemified/f02ae34d-704f-429c-be7e-a17b33688633/"><img src="http://img.zemanta.com/reblog_a.png?x-id=f02ae34d-704f-429c-be7e-a17b33688633" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/31/phasing-out-scientific-software/">Phasing out scientific software</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-31T13:51:49-04:00" pubdate data-updated="true">Mar 31<span>st</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Hemagglutinin-alignments.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/b/b4/Hemagglutinin-alignments.png/202px-Hemagglutinin-alignments.png" alt="Sequence alignment of 27 H9N2 avian influenza ..." /></a></p>

<pre><code>Image via [Wikipedia](http://commons.wikipedia.org/wiki/Image:Hemagglutinin-alignments.png)
</code></pre>

<p>There&#8217;s always the right time for everything. And for some scientific software it&#8217;s time for retirement. I will give here a one examples, but I bet if you&#8217;re a bioinformatician you will have others, and if you&#8217;re something else you will scratch your head and think &#8220;This guy is right!&#8221;. So let&#8217;s see the case of <strong><a href="http://www.clustal.org">ClustalW</a></strong>.</p>

<p>ClustalW was the first massively used alignment program and it&#8217;s a great program. But in my opinion it was a great program and has to be remembered as such, a great tool that introduced the mysteries of sequence alignment to so many people. It had everything, an UI, a command line, and more important it converted and many file formats.</p>

<p>A quick search on Google Scholar shows about 45000 results for ClustalW, but I&#8217;m sorry, it has to go. You have to stop using it, please. If you&#8217;re a bioinformatician, stop supporting it, don&#8217;t create a ClustalW(X) wrapper on any tool, forget about it. Every time someone mentions the word Clustal in front of you, pretend that he/she is a creationist. If you are not a bioinformatician, ask your local bioinformatician what is the best tool to use (Muscle, <a href="http://align.bmr.kyushu-u.ac.jp/mafft/software/">MAFFT</a> and <a href="http://probcons.stanford.edu/">ProbCons</a> are my favourite), delete the ClustalW on your HD (save one for posterity) and start living the high-alignment-life.</p>

<p><a href="http://reblog.zemanta.com/zemified/825ffe3c-3961-4bb4-9dfe-a5332cf658cb/"><img src="http://img.zemanta.com/reblog_a.png?x-id=825ffe3c-3961-4bb4-9dfe-a5332cf658cb" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/31/managing-a-simple-database-with-python-sqlite-and-wxpython-6/">Managing a simple database with Python, SQLite and wxPython, 6</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-31T12:06:08-04:00" pubdate data-updated="true">Mar 31<span>st</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:SQLite_Logo_4.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/1/19/SQLite_Logo_4.png/202px-SQLite_Logo_4.png" title="The :en:SQLite logo as of 2007-12-15" alt="The :en:SQLite logo as of
2007-12-15" /></a>
  ~ Image via</p>

<pre><code>[Wikipedia](http://commons.wikipedia.org/wiki/Image:SQLite_Logo_4.png)
</code></pre>

<p>Let&#8217;s get back to our SQLite and wxPython project. We haven&#8217;t seen
anything on wxPython yet, and we will check the interface only on the
next post. For now, let&#8217;s see some extra code added to the SQLite access
class. Remember that we have a generic class and one class derived from
it that would work on accessing specific tables in our database file.
When we last covered the db access routines, there was no search for an
entry (the function returned everything in the table no matter what),
there was no update function in case someone would want to modify an
entry and there was no delete method if you wanted to delete something.
In the meantime, I added all of this functionality (and some other) to
the generic class and extended it to the class derived from it. Let&#8217;s
check how the generic class is now (you will notice that there is an
issue in one of the methods, if someone can help me I&#8217;d appreciate. More
details later.) [sourcecode language=&#8217;python&#8217;]class DB_Generic():
&#8221;&#8217;generic class to add DB functionality&#8221;&#8217; def __init__(self,
table_name, db_path = &#8221;): #par= name of the table to be used
self.table_name = table_name if len(db_path) > 0: self.db_path =
db_path print db_path def get_data_generic(self, range = 1,
bac_to_get = 0): &#8221;&#8217;gets the data from the database&#8221;&#8217; if sys.platform
== &#8216;darwin&#8217;: (cursor, database) = link_db(self.db_path) else: (cursor,
database) = link_db() if range == 1: cursor.execute(&#8220;&#8221;&#8220;SELECT * FROM
%s&#8221;&#8220;&#8221; % self.table_name) elif range == 2: cursor.execute(&#8220;&#8221;&#8220;SELECT *
FROM %s where idbac = %d&#8221;&#8220;&#8221; % (self.table_name, bac_to_get))
table_data = cursor.fetchall() raw_data = [] for i in table_data:
raw_data.append(list(i)) self.table_data = raw_data database.close()
def insert_data(self, values_list, insert_string): &#8221;&#8217;inserts data in
the database&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor, database) =
link_db(self.db_path) else: (cursor, database) = link_db()
cursor.execute(insert_string % self.table_name, values_list)
database.commit() database.close() def update_data(self, values_list):
&#8221;&#8217;edits and updates fields&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor,
database) = link_db(self.db_path) else: (cursor, database) =
link_db() #change this to generic!!!!!!!!!!!! cursor.execute(&#8220;UPDATE
bac SET projects = ?, comments = ?, temperature = ?, cell = ?, box = ?,
tubes = ?, chromosome = ?, sdate = ?, clone = ?, source = ?, location1 =
?, startpos = ?, endpos = ?, gene = ?, genelink = ?, dnaex = ?,
validation = ?, pcr = ?, refs = ?, antibiotic = ? WHERE idbac = ?&#8221;,
(values_list[&#8216;projects&#8217;], values_list[&#8216;comments&#8217;],
values_list[&#8216;temperature&#8217;], values_list[&#8216;cell&#8217;], values_list[&#8216;box&#8217;],
values_list[&#8216;tubes&#8217;], values_list[&#8216;chromo&#8217;], values_list[&#8216;date&#8217;],
values_list[&#8216;clone&#8217;], values_list[&#8216;source&#8217;], values_list[&#8216;location&#8217;],
values_list[&#8216;start&#8217;], values_list[&#8216;end&#8217;], values_list[&#8216;gene&#8217;],
values_list[&#8216;genelink&#8217;], values_list[&#8216;dna&#8217;],
values_list[&#8216;validation&#8217;], values_list[&#8216;pcr&#8217;], values_list[&#8216;refs&#8217;],
values_list[&#8216;antibiotic&#8217;], values_list[&#8216;idbac&#8217;])) database.commit()
database.close() def delete_data(self, delete_string): &#8221;&#8217;deletes one
field&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;: (cursor, database) =
link_db(self.db_path) else: (cursor, database) = link_db()
cursor.execute(delete_string) database.commit()
database.close()[/sourcecode] In the next couple of posts we&#8217;ll dissect
each function and see what&#8217;s going on. The class definition wasn&#8217;t
changed, so we start with <code>get_data_generic</code> [sourcecode
language=&#8217;python&#8217;]def get_data_generic(self, range = 1, bac_to_get =
0): &#8221;&#8217;gets the data from the database&#8221;&#8217; if sys.platform == &#8216;darwin&#8217;:
(cursor, database) = link_db(self.db_path) else: (cursor, database) =
link_db() if range == 1: cursor.execute(&#8220;&#8221;&#8220;SELECT * FROM %s&#8221;&#8220;&#8221; %
self.table_name) elif range == 2: cursor.execute(&#8220;&#8221;&#8220;SELECT * FROM %s
where idbac = %d&#8221;&#8220;&#8221; % (self.table_name, bac_to_get)) table_data =
cursor.fetchall() raw_data = [] for i in table_data:
raw_data.append(list(i)) self.table_data = raw_data
database.close()[/sourcecode] The first difference we notice here is the
<code>sys.platform</code> usage. This is required if we intend to package our
application as an OS X app, using py2app. When a Python/wxPython
application is packaged in OS X, the actual application executable is
inside the a directory named after the application (or whatever you set
up). In our case here we don&#8217;t provide a way for the Python script to
receive the path and name for the database on a command line, as we
expect it to be in the executable&#8217;s current directory. Because of that
we need to provide a &#8220;config&#8221; file (in our case here a one-line text
file with the database path) inside the application wrapper, something
we will see in the end of the series. Another modification here is the
<code>range</code> parameter and the addition of the <code>bac_to_get</code> parameter. Notice
that both parameters have a value assigned to it. This means that they
are optional, the function&#8217;s call can pass them or not. If it doesn&#8217;t
pass, their value will be the one assigned on the function declaration.
So, here if we are interested in getting all bacs, <code>range</code> will have the
value of 1 and we don&#8217;t need to worry about it. If we want an specific
bac we will pass <code>range</code> as 2 and then pass the <code>bac_to_get</code> ID to be
returned. A final change/addition is that we added a new select
statement for the cases when <code>range</code> equals 2. This time we are adding
the bac ID to be returned. Previously in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://python.genedrift.org/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Part
3</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-4/">Part
4</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-5/">Part
5</a></p>

<p><a href="http://reblog.zemanta.com/zemified/ea53b728-33c6-47db-aabf-0c695dcfabd8/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_a.png?x-id=ea53b728-33c6-47db-aabf-0c695dcfabd8" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/18/i-havent-joined-mendeley-as-community-liaison/">I haven&#8217;t joined Mendeley as community liaison</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-18T09:31:24-04:00" pubdate data-updated="true">Mar 18<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Despite rumors floating in the upper echelons on the interwebs, I must admit that I haven&#8217;t been approached, contacted, persuaded or forced to join <a href="http://en.wikipedia.org/wiki/Mendeley">Mendeley</a> as a Community Liaison. I understand the rumours were founded on the statement from one of the company&#8217;s staff member that my role in the software development is &#8220;crucial&#8221; (not his exact words, but in the ballpark).</p>

<p>I wish <a href="http://synthesis.williamgunn.org/2009/03/18/ive-joined-mendeley-as-community-liaison/">Gunn</a> and <a href="http://my.biotechlife.net/2009/03/04/community-liaison-mendeley/">Vidal</a> all the luck.</p>

<h6>Related articles by Zemanta</h6>

<ul>
<li><a href="http://uk.techcrunch.com/2009/03/12/plugg-wraps-with-two-very-capable-winnners/">European Startups Put To The Plugg Test</a> (uk.techcrunch.com)</li>
</ul>


<p><a href="http://reblog.zemanta.com/zemified/89e7ea92-5320-4f20-b528-257a4d183309/"><img src="http://img.zemanta.com/reblog_e.png?x-id=89e7ea92-5320-4f20-b528-257a4d183309" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/15/presenting-biovb6/">Presenting BioVB6</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-15T10:02:24-04:00" pubdate data-updated="true">Mar 15<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:VB_cover.png"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/9/92/VB_cover.png/202px-VB_cover.png" alt="The four colored boxes is the logo of VBA, and..." /></a></p>

<pre><code>Image via [Wikipedia](http://en.wikipedia.org/wiki/Image:VB_cover.png)
</code></pre>

<p>I&#8217;m proud to present this initiative to keep old-school programmers in the cusp of Bio* development. After long <a href="http://python.genedrift.org/2009/03/13/biopython-and-cvs/">discussions and controversy</a>, I decided that it would be extremely important to stop neglecting a <a href="http://www.itjobswatch.co.uk/jobs/uk/vb6%20developer.do">good portion of developers around the world</a> that are still trying to catch-up with new, but yet completely unreliable, software and technology.</p>

<p>So far, I haven&#8217;t contacted anyone regarding this project, but I guess people will be thrilled to learn about it and I expect massive signup to become a developer (first come, first served basis). The main goals and &#8220;project culture&#8221; (similar to a company culture, but in project, got it?) are:</p>

<ul>
<li>develop a number of <a href="http://en.wikipedia.org/wiki/Object_Linking_and_Embedding">OCX</a> and <a href="http://www.microsoft.com/windows/directx/">DirectX</a> and VB classes and VB modules that can be used by anyone in the bio community. Extreme focus on being backward compatible with VB 3 and 4 (both 8-bit and 16-bit versions)</li>
<li>We will not care about VB .NET</li>
<li>Goto is considered harmful elsewhere not here</li>
<li>improve ADO and <a href="http://en.wikipedia.org/wiki/Open_Database_Connectivity">ODBC</a> connection between VB6 and Access and <a href="http://en.wikipedia.org/wiki/FoxPro_2">FoxPro</a> in order to mirror biodatabases such as NCBI and UniProt</li>
<li>Sourcesafe is our King (or Queen? I always get confused with the lack of gender in English words)</li>
<li>Development will be focused solely on technology available to VB6 by MSFT, we don&#8217;t care about third-party modules</li>
</ul>


<p>These are the initial goals. As I mentioned before, with massive synergy between the expected hundreds of co-developers, I guess release will be in the second quarter of 2009. Early second quarter.</p>

<p>I keep you posted.</p>

<p><a href="http://reblog.zemanta.com/zemified/cb7502c2-35fc-47b9-b8e6-195c2b69c630/"><img src="http://img.zemanta.com/reblog_a.png?x-id=cb7502c2-35fc-47b9-b8e6-195c2b69c630" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/14/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/12/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
