
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Image via [Wikipedia](http://en.wikipedia.org/wiki/Image:VB_cover.png) I&#8217;m proud to present this initiative to keep old-school programmers in &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/14">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/15/presenting-biovb6/">Presenting BioVB6</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-15T10:02:24-04:00" pubdate data-updated="true">Mar 15<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:VB_cover.png"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/9/92/VB_cover.png/202px-VB_cover.png" alt="The four colored boxes is the logo of VBA, and..." /></a></p>

<pre><code>Image via [Wikipedia](http://en.wikipedia.org/wiki/Image:VB_cover.png)
</code></pre>

<p>I&#8217;m proud to present this initiative to keep old-school programmers in the cusp of Bio* development. After long <a href="http://python.genedrift.org/2009/03/13/biopython-and-cvs/">discussions and controversy</a>, I decided that it would be extremely important to stop neglecting a <a href="http://www.itjobswatch.co.uk/jobs/uk/vb6%20developer.do">good portion of developers around the world</a> that are still trying to catch-up with new, but yet completely unreliable, software and technology.</p>

<p>So far, I haven&#8217;t contacted anyone regarding this project, but I guess people will be thrilled to learn about it and I expect massive signup to become a developer (first come, first served basis). The main goals and &#8220;project culture&#8221; (similar to a company culture, but in project, got it?) are:</p>

<ul>
<li>develop a number of <a href="http://en.wikipedia.org/wiki/Object_Linking_and_Embedding">OCX</a> and <a href="http://www.microsoft.com/windows/directx/">DirectX</a> and VB classes and VB modules that can be used by anyone in the bio community. Extreme focus on being backward compatible with VB 3 and 4 (both 8-bit and 16-bit versions)</li>
<li>We will not care about VB .NET</li>
<li>Goto is considered harmful elsewhere not here</li>
<li>improve ADO and <a href="http://en.wikipedia.org/wiki/Open_Database_Connectivity">ODBC</a> connection between VB6 and Access and <a href="http://en.wikipedia.org/wiki/FoxPro_2">FoxPro</a> in order to mirror biodatabases such as NCBI and UniProt</li>
<li>Sourcesafe is our King (or Queen? I always get confused with the lack of gender in English words)</li>
<li>Development will be focused solely on technology available to VB6 by MSFT, we don&#8217;t care about third-party modules</li>
</ul>


<p>These are the initial goals. As I mentioned before, with massive synergy between the expected hundreds of co-developers, I guess release will be in the second quarter of 2009. Early second quarter.</p>

<p>I keep you posted.</p>

<p><a href="http://reblog.zemanta.com/zemified/cb7502c2-35fc-47b9-b8e6-195c2b69c630/"><img src="http://img.zemanta.com/reblog_a.png?x-id=cb7502c2-35fc-47b9-b8e6-195c2b69c630" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/13/biopython-and-cvs/">BioPython and CVS</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-13T14:25:38-04:00" pubdate data-updated="true">Mar 13<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:Bazaar_logo_(software_product"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/c/c1/Bazaar_logo_(software_product" alt="Bazaar" />.svg/125px-Bazaar_logo_(software_product).svg.png &#8220;Bazaar&#8221;)</a>.svg)
  ~ Image via</p>

<pre><code>[Wikipedia](http://en.wikipedia.org/wiki/Image:Bazaar_logo_(software_product).svg)
</code></pre>

<p>I start this post with an apology. I usually don&#8217;t rant or vent here,
which are feelings that I usually reserve to my personal blog. I don&#8217;t
use BioPython, never used it. I have it installed in my systems, but I
never wrote a piece of code importing BioPython routines. But I
subscribe to their mailing lists, both user and developer. I maybe have
written once to the list, and I just follow the discussions there. Since
last year one of the main topics has been the possibility of moving
BioPython from CVS to another version control system. Yes, you read it
right. It&#8217;s 2009 and BioPython uses CVS and their version control
system. Soon, CVS will be like typewriters and LPs to young developers.
Last stable release of CVS was sometime in 2005, what in interwebs time
is equivalent to something like 1972. Since 2005, Subversion has taken
the world of version control by storm, and Git is getting also very
strong, not to mention Bazaar, Darcs, Mercurial and some others that I
might not be aware of. This is a discussion that have been dragging for
sometime in the list. And it&#8217;s a shame, a clear lack of leadership from
whoever is (not) leading the project. BioRuby is Git, BioPerl SVN and
BioPython is CVS, because they &#8220;need to care for the legacy developers&#8221;.
It&#8217;s like MSFT keeping two copies of the Notepad executable because they
needed to cater to legacy applications, but with a different scale of
course. With the current Python steam in the non-bioinformatics and
bioinformatics community is very sad to see BioPython not evolving
(before you ask me, no, I&#8217;m not interested in helping, not the way
things are now). Perl which is language forever-in-waiting for its holy
grail (Perl 6) has a strong community behind it, and more important an
excellent leadership, that&#8217;s not scare of making decisions. So, if
you&#8217;re still using CVS, it&#8217;s 2009!</p>

<p><a href="http://reblog.zemanta.com/zemified/06117f0a-6d25-497d-b686-1a96e8a7d81c/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=06117f0a-6d25-497d-b686-1a96e8a7d81c" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-5/">Managing a simple database with Python, SQLite and wxPython, 5</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-02T19:23:42-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen how to connect, get and insert data (at least
theoretically) in the database. Now, a little not about the SQL engine
of choice here: SQLite. SQLite databases have the main characteristic
that they are self-contained files. Also it does not require an
installation, works without a server and works pretty well in most
operating systems. Basically for the type of application we&#8217;re
developing here, SQLite seems ideal. It eliminates a lot of
infrastructure that would be needed if we were working with MySQL or
postgresql. We don&#8217;t need a server or know how to configure users or
manage the databases and tables. All we need is contained in a single
file that can be transported from system to system and can be accesed
from the computers used in the lab, mainly XP and OS X. Also some web
frameworks (Rails and
<a href="http://www.djangoproject.com" title="Django (web framework)">Django</a>, for
instance) can use SQLite, so in the end we can have a desktop
application and a web application accessing the same file without extra
configuration. Now the database created for this application has 8
tables and almost no relationships among them. SQLite allows the
creation of relationships but in our case only a couple of cases were
required. For the table we are using at the moment (bac) there is no
need for relationships, although there are some fileds that can benefit
from a more relational structure. Also SQLite don&#8217;t have the same data
types that are found on the bigger SQL engines. All values can be stored
as text, integer, real (floating point numbers), null and blob (verbose
type, what you store is what you get). As actual types, you can set
columns as Boolean and Data for instance and SQLite will understand
them. If you have no experience in creating databases, let&#8217;s check again
the table we are using in this small project. First, I would recommend
the use of some SQLite database editor. You can find pretty good ones
for any computer system and there is even a Firefox extension that
allows you to edit some files. Editors make it easier to generate the
SQL table creation scripts and make easier to visualize what we are
doing. So, the table bac looks like [sourcecode language=&#8217;sql&#8217;]CREATE
TABLE bac (idbac INTEGER PRIMARY KEY, clone Text, sdate Date, source
Text, gene Text, chromosome Text, startpos Integer, endpos Integer,
antibiotic Text, location1 Text, temperature Integer, tubes Integer, box
Integer, cell Integer, dnaex Boolean, validation Boolean, pcr Boolean,
projects Text, comments Text, genelink Text, refs Text);[/sourcecode] If
you go back to our last post, you will see that in the insert statement
there is no mention of the <code>idbac</code> field. We don&#8217;t actually insert ay
value there, the values that populate this field are created
automatically. And <code>idbac</code> is our primary key, meaning it&#8217;s the unique
identifier of each bac we insert in this table. And in SQLite a integer
primary key is automatically incremented whenever values are inserted in
the table. So our first insertion will create <code>idbac</code> 1, the second will
create <code>idbac</code> 2 and so on. I&#8217;m not going to enter in details about
database development and administration, but it&#8217;s usual and safe to
create tables with an auto-incremental integer primary keys. These
fields, apart from make it easier t identify records, make access to
such records faster and are great when relationships among tables are
set. Let&#8217;s say that we had a column user in our bac table. And let&#8217;s say
we had an user table with two columns: user_id and name, user_id being
a auto-increment primary key. The user column in back could be linked
with the user_id column in the user table, in what we call a
one-to-many relationship (one user can insert as many bacs as he wants).
One day we want to know who is actually working in the lab and we want
to check how many bacs were catalogued by each user. We can easily
search the user table and extract information from bacs at the same time
thanks to the relationship between the tables. And the result should be
returned quite quickly, as we are only searching integers. All the other
fields/columns in our table are straightforward to understand. They are
basically related to the type of data they need to store. <code>validation</code>
is a boolean because the bac might have been validated or not, just as
<code>danex</code> (DNA extraction). At the same time, the number of tubes stored
in the freezer will always be an integer. So, why does temperature is an
integer? Because we can only store bacs in two type of freezers: -80
(ultra freezers) or -20 (regular freezer that we can have at home), and
we don&#8217;t need to worry about fractional numbers. Well, this is a very
short and limited explanation of tables and SQLite. The web is full of
resources about it, so next time we will get back to Python. Previously
in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://python.genedrift.org/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Part
3</a>
<a href="http://python.genedrift.org/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-4/">Part
4</a>
<a href="http://reblog.zemanta.com/zemified/4be1389f-5603-4b76-961b-b79d985066cc/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=4be1389f-5603-4b76-961b-b79d985066cc" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/03/02/managing-a-simple-database-with-python-sqlite-and-wxpython-4/">Managing a simple database with Python, SQLite and wxPython, 4</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-03-02T11:58:09-05:00" pubdate data-updated="true">Mar 2<span>nd</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:SQLite_Logo_4.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/1/19/SQLite_Logo_4.png/202px-SQLite_Logo_4.png" title="The :en:SQLite logo as of 2007-12-15" alt="The :en:SQLite logo as of
2007-12-15" /></a>
  ~ Image via</p>

<pre><code>[Wikipedia](http://commons.wikipedia.org/wiki/Image:SQLite_Logo_4.png)
</code></pre>

<p>Let&#8217;s continue building our small db app. As mentioned in the previous
post we need now to instantiate a specific class from our generic
<a href="http://sqlite.org/" title="SQLite">SQLite</a> access class. In order to do this
we just have to declare a new class and its type will be <code>DB_Generic</code>.
[sourcecode language=&#8217;python&#8217;]class Bac(DB_Generic)[/sourcecode] This
new class is called Bac because it&#8217;s linked to the bac table in our
database file. A side note, bacs are Bacterial Artificial Chromosomes
and are used in different molecular biology techniques. Mainly in our
case bacs have incorporated human DNA segments and are used as probes
for deletion, duplication, etc studies. Now, back to our Python code, as
soon as we instantiate our generic class, the object (class) we create
has access to all methods and functions from the parent class (by using
<code>self</code>), but we still need to create functionality and expose other
methods that can be accessed from a class object derived from <code>Bac</code>. Our
instantiated class will be [sourcecode language=&#8217;python&#8217;]class
Bac(DB_Generic): def __init_(self): self.bac_data = []
DB_Generic.__init__(self, &#8216;bac&#8217;) def get_data(self): return
self.get_data_generic() def load_data(self): pass def add_data(self,
values_list): insert_string = &#8220;&#8221;&#8220;INSERT INTO %s (projects, comments,
temperature, cell, box, tubes, chromosome, sdate, clone, source,
location1, startpos, endpos, gene, genelink, dnaex, validation, pcr,
refs, antibiotic) VALUES (:projects, :comments, :temperature, :cell,
:box, :tubes, :chromo, :date, :clone, :source, :location, :start, :end,
:gene, :genelink, :dna, :validation, :pcr, :refs, :antibiotic)&#8221;&#8220;&#8221;
self.insert_data(values_list, insert_string)[/sourcecode] Pretty
simple so far, as we don&#8217;t have a lot of declared methods. Let&#8217;s check
one by one [sourcecode language=&#8217;python&#8217;]def __init_(self):
DB_Generic.__init__(self, &#8216;bac&#8217;)[/sourcecode] The only line is the
initialization required by the parent class, and we&#8217;re passing the value
that is the table to be accessed. [sourcecode language=&#8217;python&#8217;]def
get_data(self): self.get_data_generic() return
self.table_data[/sourcecode] The <code>get_data</code> function returns the all
elements in our table (So far, we still don&#8217;t have an elegant range
option) and has one too many lines in it. We will get rid of some
useless code here in the future, but it&#8217;s OK the way it is. Basically
this code access the <code>get_data_generic</code> from the parent class and gets
all the values stored in the table. There is a function not yet complete
that will load data, and will be used in the future. And the last one is
the function that actually adds the data to the table with a SQL insert
statement [sourcecode language=&#8217;python&#8217;]def add_data(self,
values_list): insert_string = &#8220;&#8221;&#8220;INSERT INTO %s (projects, comments,
temperature, cell, box, tubes, chromosome, sdate, clone, source,
location1, startpos, endpos, gene, genelink, dnaex, validation, pcr,
refs, antibiotic) VALUES (:projects, :comments, :temperature, :cell,
:box, :tubes, :chromo, :date, :clone, :source, :location, :start, :end,
:gene, :genelink, :dna, :validation, :pcr, :refs, :antibiotic)&#8221;&#8220;&#8221;
self.insert_data(values_list, insert_string)[/sourcecode] In this
function, we have a large string with all the SQL insert options. A SQL
insert statement is divided into two parts, one where you point where to
insert the values and another where you input the values. Usually simple
insert statements will have this structure [sourcecode
language=&#8217;sql&#8217;]INSERT INTO my_table_name (table_column1,
table_column2) VALUES (value1, value2);[/sourcecode] So, we have the
table we want to insert values into, its columns and the values we set
for each column. After executed this will put value1 into table_column1
and value2 into table_column2. The actual syntax can vary a bit for
different SQL engines but the structure is identical in most cases.
Pretty simple. For our insert string above, there are some aspects to
call for attention. Again note the triple quote around the statement.
This make sure that it&#8217;s not changed and parsed correctly. We also have
a <code>%s</code> for the table name, which will be parsed by the parent class
function that insert values, then a list of all the tables in the
database and then a list of values to insert. And why the values to be
inserted have this <code>:value</code> syntax? Because we are previously storing
the values in a dictionary, and the &#8220;:&#8221; indicates that we need to get
the dictionary value for the correspondent key. The insert string, and
the list of values (actually a dictionary, not the best variable/object
name I must admit) is then sent to the parent class to be inserted.
Storing the values to be inserted in a dictionary is OK for a one time
insert case, where the values are obtained from a form. If you are
parsing a large CSV or TSV file, ideally it&#8217;s better to put it in a
list, and dump them at the same time. We&#8217;re progressing. Next we will
take a look on some simple SQL table structure and then move to create
the form to insert the values and check the table. Previously in the
series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://python.genedrift.org/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Part
3</a></p>

<p><a href="http://reblog.zemanta.com/zemified/4c2737a4-f1de-46bf-ae53-c5de040daf97/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=4c2737a4-f1de-46bf-ae53-c5de040daf97" alt="Reblog this post [with
Zemanta]" /></a></p>

<p>evi</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/21/why-do-i-blog-or-science-blogging-is-it-worth/">Why do I blog. Or: Science Blogging, is it worth?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-21T18:18:24-05:00" pubdate data-updated="true">Feb 21<span>st</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:Nshl.jpg"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/6/61/Nshl.jpg/202px-Nshl.jpg" title="High/Low album cover" alt="High/Low album
cover" /></a>
  ~ Image via <a href="http://en.wikipedia.org/wiki/Image:Nshl.jpg">Wikipedia</a></p>

<p><em>Mirroring the post that appeared on Blind.Scientist</em> Some time ago
there was a meme about science blogging and one of the questions were
&#8220;why do you blog&#8221;. Well, I do it because of the &#8221;<a href="http://www.nadasurf.com/" title="Nada Surf">Nada
Surf</a> effect&#8221;. You don&#8217;t know the
&#8220;Nada Surf effect&#8221;? Pity you weren&#8217;t in Washington, DC 2001. In March or
April of 2001, Nada Surf played a concert there. It was a small bar on
14th Street W, close to the more famous Black Cat. It was a spring
night, I was with a couple of Dutch friends that had told me about the
concert, if I&#8217;m not wrong, a couple of days before. It was also
mid-week, so you wouldn&#8217;t expect big crowds in most concerts. We left
ISH around 7 pm, with spare time for the 9 pm concert. We didn&#8217;t know
the venue, we got there and it was empty, just a couple of souls at the
bar. We sat and for about an hour we were pondering if we were in the
right place, until a guy came and asked if we were staying for the
concert. We said yes, paid th US$ 7.50 of the admittance and sipped our
beers waiting for the opening act. Soon after we paid, a van parked
outside and some guys started bringing music equipment inside. At that
time there must have been around 20 people there. The van guys set the
instruments, wasted 5 minutes soundchecking, and started. IT was Ashtray
Babyhead. They played for 40 minutes and as fast as they arrived they
left. Another van parked outside and this time Nada Surf members started
unloading and setting up the stage. Now roadies. OK, maybe one guy
helped, but I&#8217;m getting old and the memory sometimes falters. At that
point in time, almost 9 pm, the number of brave souls was at 50. They
played as they were playing for 50.000 people in Wembley. Nice set,
great songs, unforgettable night. After the show, they sold CDs at the
usual after-show gathering, we talked about New York, Brazil and
feijoada. And why do I call it the &#8220;Nada Surf effect&#8221;? A band that used
to play for thousands of people in festivals and stadiums, had a number
one video on MTV (Popular), played in a midweek night in a small bar in
Washington, DC as it was the band farewell. Every fan that night felt
that they were the most important one, maybe even the only one. And this
type of example is the one that brings me to write this, and to write
Beginning Python for Bioinformatics. Especially the latter (as I spend
too much time here, writing about non-important stuff). If I can make
one person have an idea, one person there to use Python, or least to
learn something extra in their lives, I&#8217;m happy. I don&#8217;t care if I have
a huge audience or if I&#8217;m famous. I care for the undergrad that is
starting today, the high school kid that is hacking at night, or the
Java coder that is looking for some better language (Ok, not really, but
I couldn&#8217;t resist). And this is one of the things that I learned in
Science, always give back and don&#8217;t expect anything in return. Just add
to the pile of knowledge. So, my advice for the three people that read
this site is: let Nada Surf in!</p>

<p><a href="http://reblog.zemanta.com/zemified/2db17c83-bb95-4371-911a-0b58b781409b/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=2db17c83-bb95-4371-911a-0b58b781409b" alt="Reblog this post [with
Zemanta]" /></a></p>

<p>\
\</p>

<p><img src="http://img.zemanta.com/pixy.gif?x-id=90407633-45c2-41dc-912f-4b3d57e7eb2b" alt="image" /></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/20/why-do-i-blog-or-science-blogging-is-it-worth/">Why do I blog. Or: Science Blogging, is it worth?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-20T14:42:59-05:00" pubdate data-updated="true">Feb 20<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://en.wikipedia.org/wiki/Image:Nshl.jpg"><img src="http://upload.wikimedia.org/wikipedia/en/thumb/6/61/Nshl.jpg/202px-Nshl.jpg" alt="High/Low album cover" /></a></p>

<pre><code>Image via [Wikipedia](http://en.wikipedia.org/wiki/Image:Nshl.jpg)
</code></pre>

<p>Some time ago there was a meme about science blogging and one of the questions were &#8220;why do you blog&#8221;. Well, I do it because of the &#8221;<a href="http://www.nadasurf.com/">Nada Surf</a> effect&#8221;. You don&#8217;t know the &#8220;Nada Surf effect&#8221;? Pity you weren&#8217;t in Washington, DC 2001.</p>

<p>In March or April of 2001, Nada Surf played a concert there. It was a small bar on 14th Street W, close to the more famous Black Cat. It was a spring night, I was with a couple of Dutch friends that had told me about the concert, if I&#8217;m not wrong, a couple of days before. It was also mid-week, so you wouldn&#8217;t expect big crowds in most concerts. We left ISH around 7 pm, with spare time for the 9 pm concert. We didn&#8217;t know the venue, we got there and it was empty, just a couple of souls at the bar. We sat and for about an hour we were pondering if we were in the right place, until a guy came and asked if we were staying for the concert. We said yes, paid th US$ 7.50 of the admittance and sipped our beers waiting for the opening act. Soon after we paid, a van parked outside and some guys started bringing music equipment inside. At that time there must have been around 20 people there. The van guys set the instruments, wasted 5 minutes soundchecking, and started. IT was Ashtray Babyhead.</p>

<p>They played for 40 minutes and as fast as they arrived they left. Another van parked outside and this time Nada Surf members started unloading and setting up the stage. Now roadies. OK, maybe one guy helped, but I&#8217;m getting old and the memory sometimes falters. At that point in time, almost 9 pm, the number of brave souls was at 50. They played as they were playing for 50.000 people in Wembley. Nice set, great songs, unforgettable night. After the show, they sold CDs at the usual after-show gathering, we talked about New York, Brazil and feijoada.</p>

<p>And why do I call it the &#8220;Nada Surf effect&#8221;? A band that used to play for thousands of people in festivals and stadiums, had a number one video on MTV (Popular), played in a midweek night in a small bar in Washington, DC as it was the band farewell. Every fan that night felt that they were the most important one, maybe even the only one.</p>

<p>And this type of example is the one that brings me to write this, and to write Beginning Python for Bioinformatics. Especially the latter (as I spend too much time here, writing about non-important stuff). If I can make one person have an idea, one person there to use Python, or least to learn something extra in their lives, I&#8217;m happy. I don&#8217;t care if I have a huge audience or if I&#8217;m famous. I care for the undergrad that is starting today, the high school kid that is hacking at night, or the Java coder that is looking for some better language (Ok, not really, but I couldn&#8217;t resist). And this is one of the things that I learned in Science, always give back and don&#8217;t expect anything in return. Just add to the pile of knowledge.</p>

<p>So, my advice for the three people that read this site is: let Nada Surf in!</p>

<p><a href="http://reblog.zemanta.com/zemified/2db17c83-bb95-4371-911a-0b58b781409b/"><img src="http://img.zemanta.com/reblog_e.png?x-id=2db17c83-bb95-4371-911a-0b58b781409b" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/18/managing-a-simple-database-with-python-sqlite-and-wxpython-3/">Managing a simple database with Python, SQLite and wxPython, 3</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-18T11:06:58-05:00" pubdate data-updated="true">Feb 18<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In the last post we saw how to connect to a
<a href="http://sqlite.org/" title="SQLite">SQLite</a> database file and generate a
cursor that would allow us to actually interact with such database. Now
we need some functionality that will interact with the data, add, read,
delete and search. As was mentioned before the idea is to have a generic
database interaction class and have unique instantiated class objects
for each database of the project. In the <code>db_obj.py</code> file we have an
initial structure set, so let&#8217;s check the <code>DB_Generic</code> class.
[sourcecode language=&#8217;python&#8217;]class DB_Generic(): &#8221;&#8217;generic class to
add DB functionality&#8221;&#8217; def __init__(self, table_name): #par= name
of the table to be used self.table_name = table_name def
delete_entry(self): pass def get_data_generic(self): &#8221;&#8217;gets the data
from the database generic so far, needs to be updated to include
range&#8221;&#8217; range = 1 (cursor, database) = link_db() if range == 1:
cursor.execute(&#8220;&#8221;&#8220;SELECT * from %s&#8221;&#8220;&#8221; % self.table_name) table_data =
cursor.fetchall() raw_data = [] for i in table_data:
raw_data.append(list(i)) self.table_data = raw_data database.close()
def insert_data(self, values_list, insert_string): &#8221;&#8217;inserts data in
the database&#8221;&#8217; (cursor, database) = link_db()
cursor.execute(insert_string % self.table_name, values_list)
database.commit() database.close()[/sourcecode] There are different
functions in this class, we will take a look at each one individually.
We can see that the class is far from being complete, something that
we&#8217;ll do in the next posts. We start with the class initialization:
[sourcecode language=&#8217;python&#8217;]def __init__(self, table_name):
#par= name of the table to be used self.table_name =
table_name[/sourcecode] Very simple and direct, it receives the table
name that is being used by the interface (in this case). The table name
is then stored in a object that can be accessed by other functions in
the class. The function to delete entries is not finished as we only
have a <code>pass</code> in it. We&#8217;ll will do it soon. Next we have a function that
gets the data from the table. [sourcecode language=&#8217;python&#8217;] def
get_data_generic(self): &#8221;&#8217;gets the data from the database generic so
far, needs to be updated to include range&#8221;&#8217; range = 1 (cursor,
database) = link_db() if range == 1: cursor.execute(&#8220;&#8221;&#8220;SELECT * from
%s&#8221;&#8220;&#8221; % self.table_name) table_data = cursor.fetchall() raw_data = []
for i in table_data: raw_data.append(list(i)) self.table_data =
raw_data database.close()[/sourcecode] So far it grabs everything,
there is no range selection based on any of the table fields, so it&#8217;s a
generic SQL <code>SELECT</code>. Let&#8217;s dissect it. The <code>range</code> object is a dummy
variable that at the moment is there only to remind us that we need to
include a range select. The next line is the most important in this
function: it will call the <code>link_db</code> function and start the connection.
Remember that <code>link_db</code> returns a tuple with the cursor and database
connection. Basically we will work with cursor methods to get the data,
and the first action is to execute a SQL <code>SELECT</code> stetement where we
select everything in the table [sourcecode
language=&#8217;python&#8217;]cursor.execute(&#8220;&#8221;&#8220;SELECT * from %s&#8221;&#8220;&#8221; %
self.table_name)[/sourcecode] Notice that the statement is a regular
string and we use string formating <code>%</code> in ordert o add the table that we
are searching, which was defined when we initialized the class object in
the first place. Also, notice the triple quotes around the select
statement: this will avoid any problems in parsing it when sending to
the database, making it a <a href="http://en.wikipedia.org/wiki/String_literal" title="String literal">string
literal</a>.
So this line executes the statement we pass to the method, but it does
not actually get the data <em>per se</em>. We need to use another method and
fetch everything returned by the select. This is done by [sourcecode
language=&#8217;python&#8217;]table_data = cursor.fetchall()[/sourcecode] A couple
of things here. The data fetched will be always (or in most cases) in
unicode, when it&#8217;s a string field. And the data returned will be in a
list of tuples, with the actual number of fields from the table. We know
that it&#8217;s easier to work with lists than tuples, so we code something to
convert types [sourcecode language=&#8217;python&#8217;]table_data =
cursor.fetchall() raw_data = [] for i in table_data:
raw_data.append(list(i)) self.table_data = raw_data[/sourcecode]
There are extra lines here that are not needed, and we will get rid of
them in a code refactoring soon. This short function is able to connect
to database, execute a SQL statement on a specified table and grab the
data selected, returning a list of lists with every field and value
available. We need to add a better selection statement later, and we
will do as soon as we have a good structure set. The last function in
this generic class is the one that inserts data into the table.
[sourcecode language=&#8217;python&#8217;]def insert_data(self, values_list,
insert_string): &#8221;&#8217;inserts data in the database&#8221;&#8217; cursor, database) =
link_db() cursor.execute(insert_string % self.table_name,
values_list) database.commit() database.close()[/sourcecode] Identical
procedure: connect, get a cursor, execute a statement. But in this case
the extra step is not to get the data, but to commit the data to the
table, which is done by the commit method. We will explain later how the
execute method works here and what are the <code>insert_string</code> and
<code>values_list</code>. Notice at the end that we need to close the connection to
the database, so we know that the data has been inserted properly. Next,
we will instantiate a class from this generic one and see how easy is to
manipulate the data. Stay tuned. Previously in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://python.genedrift.org/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Part
2</a>
<a href="http://reblog.zemanta.com/zemified/1663854e-5aba-4ff1-9a75-dffb8e6b7945/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=1663854e-5aba-4ff1-9a75-dffb8e6b7945" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/17/managing-a-simple-database-with-python-sqlite-and-wxpython-2/">Managing a simple database with Python, SQLite and wxPython, 2</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-17T11:04:18-05:00" pubdate data-updated="true">Feb 17<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Let&#8217;s continue coding our small Python + SQLite application. The initial
idea was to have a file for the interface and another file for the DB
access. We will start with the later. If you have access to the
repository you will see two Python files, <code>bac_form.py</code> and <code>db_obj.py</code>.
At the moment they are not well commented and have some junk lines at
the bottom, legacy from older versions. Take a look on <code>db_obj.py</code>. It
has two class declarations, one called <code>DB_Generic</code> and another one
called <code>Bac</code>. Remember in the last post where I mentioned that the idea
was to have different simple tables in the same SQLite database and each
table would have a simple input/output interface (If I didn&#8217;t mention
that, I just did!). So, we can create a generic DB access class and we
can subtype from it for every table that we will be using. In the
<code>db_obj.py</code> file we have at the moment the generic database management
class, a class derived from the generic to access the Bac database and
an initialization function, that opens the access to the SQLite file.
Let&#8217;s take a look at it: [sourcecode language=&#8217;python&#8217;]def link_db():
&#8221;&#8217;initializes the database file&#8221;&#8217; try: db =
sqlite3.connect(&#8220;samples.db&#8221;) except sqlite3.Error, errmsg: print &#8216;DB
not available &#8217; + str(errmsg) sys.exit() db_cursor = db.cursor() return
db_cursor, db[/sourcecode] In order to access a SQLite database file we
need initially the name of the file. After importing sqlite3 (we&#8217;re
using the latest version of SQLite here) Python has everything it needs
to access, change and manipulate data in a SQLite database. Just to be
sure the database file is there and we don&#8217;t get an error, we have the
initialization code inside an exception. We have seen exceptions before
and in this case we use it to be sure the database file has been
accessed with no problems. The exception structure looks like
[sourcecode language=&#8217;python&#8217;]try: #here we try to do something #the
code placed here would be executed #if no error reported it will go
until the end and exit #if not, some error (exception) raised except:
#the code under except will be executed[/sourcecode] So, the first step
is to connect to the database file [sourcecode language=&#8217;python&#8217;]db =
sqlite3.connect(&#8220;samples.db&#8221;)[/sourcecode] In our case it&#8217;s a fixed
file, but the connect method receives any kind of string. It could have
been some parameter obtained from the command line or a string from a
configuration file. If the connect is successful, no error will be
raised and we are safe to continue. We connected to database, now what?
We need a cursor, an object that will actually access the data and allow
us to execute <a href="http://en.wikipedia.org/wiki/SQL" title="SQL">SQL</a> commands on
it. [sourcecode language=&#8217;python&#8217;]db_cursor = db.cursor()[/sourcecode]
<code>cursor</code> method works on the database connection object that we created
previously. We&#8217;re set. This function returns the cursor and database
connection objects that we created, in a tuple and this function can be
called from the classes we are going to work. The classes will then have
connection to the database and a cursor that would manage, select,
delete and add data to it. Next time we&#8217;ll see how our generic table
class works. Previously in the series: <a href="http://python.genedrift.org/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Part
1</a>
<a href="http://reblog.zemanta.com/zemified/cbbab23e-338a-4756-8f4c-25cc0a5239dc/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=cbbab23e-338a-4756-8f4c-25cc0a5239dc" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/10/nostradamus/">Nostradamus?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-10T22:12:37-05:00" pubdate data-updated="true">Feb 10<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:Smartcard2.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/7/70/Smartcard2.png/202px-Smartcard2.png" alt="Basic creditcard / debitcard / smartcard graph..." /></a></p>

<pre><code>Image via [Wikipedia](http://commons.wikipedia.org/wiki/Image:Smartcard2.png)
</code></pre>

<p>&#8220;Owners of capital will stimulate working class to buy more and more of expensive goods, houses and technology, pushing them to take more and more expensive credits, until their debt becomes unbearable.
The unpaid debt will lead to bankruptcy of banks, which will have to be nationalized, and State will have to take the road which will eventually lead to communism.&#8221;</p>

<p>Nope. Karl. Marx,</p>

<p><a href="http://reblog.zemanta.com/zemified/0cca205d-2122-4b58-ab1c-8931386d8d2b/"><img src="http://img.zemanta.com/reblog_a.png?x-id=0cca205d-2122-4b58-ab1c-8931386d8d2b" alt="Reblog this post [with Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2009/02/09/managing-a-simple-database-with-python-sqlite-and-wxpython-1/">Managing a simple database with Python, SQLite and wxPython, 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2009-02-09T15:11:57-05:00" pubdate data-updated="true">Feb 9<span>th</span>, 2009</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p><a href="http://commons.wikipedia.org/wiki/Image:WxPython-logo.png"><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/c/c0/WxPython-logo.png/202px-WxPython-logo.png" title="The official wxPython logo" alt="The official wxPython
logo" /></a>
  ~ Image via</p>

<pre><code>[Wikipedia](http://commons.wikipedia.org/wiki/Image:WxPython-logo.png)
</code></pre>

<p>A little break from reviewing the book, let&#8217;s check some database topics
in Python. I was asked to create a simple database to organize wet-lab
stuff. No relationships needs, no relational tables required. Just a
simple table with determined columns, and a nice GUI to go with it so
people can edit, search and use. My first idea was to use SQLite
database, and I stuck with it. After the initial phase of &#8220;interviews&#8221;
to check database requirements, I ended up with a list of tables and
decided to start working on the table that organizes the BACs used in
the lab. BAC is a DNA vector into which large DNA fragments can be
inserted and cloned in a bacterial host, and are used mainly in
cytogenetics around here. In the end the table had this structure
[sourcecode language=&#8217;sql&#8217;]CREATE TABLE bac (idbac INTEGER PRIMARY KEY,
clone Text, sdate Date, source Text, gene TEXT, chromosome Text,
startpos Integer, endpos Integer, antibiotic Text, location1 Text,
temperature Integer, tubes Integer, box Integer, cell Integer, dnaex
Boolean, validation Boolean, pcr Boolean, projects Text, comments Text,
genelink Text, refs Text);[/sourcecode] I won&#8217;t explain in detail each
of the fields, but we can see that there is a mix of different types.
SQLite doesn&#8217;t allow many different field types, so we stick to the
basics. And why SQLite? The module to access it comes with Python 2.5,
the whole database is stored in one file that can be moved around and it
allows a full SQL query language, which is perfect for these simple
cases. So we will going to use Python, SQLite and wxPython to create a
simple application to manage our simple database. I already created a
GitHub
<a href="http://github.com/nuin/python_sqlite_wxpython/tree/master">repository</a>
for the upcoming code.</p>

<p><a href="http://reblog.zemanta.com/zemified/1a2359cd-d9c7-4cee-a26b-c41b281ec3d2/" title="Zemified by Zemanta"><img src="http://img.zemanta.com/reblog_e.png?x-id=1a2359cd-d9c7-4cee-a26b-c41b281ec3d2" alt="Reblog this post [with
Zemanta]" /></a></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/15/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/13/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
