
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="A commenter, Dilmurat, gave me an idea about the script
that generates random DNA sequence sets. Apparently it wasn&#8217;t clear that
the script was &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/31">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/26/generating-multiple-sets-of-random-dna-sequences-with-one-script-and-a-bash-one/">Generating multiple sets of random DNA sequences with one script (and a bash example) [updated]</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-26T12:15:47-04:00" pubdate data-updated="true">Oct 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A commenter, Dilmurat, gave me an idea about the <a href="http://python.genedrift.org/2007/04/04/a-script-to-simulate-dna-sequence-sets/">script</a>
that generates random DNA sequence sets. Apparently it wasn&#8217;t clear that
the script was intended to generate only one sequence set, and not
multiple sets. Dilmurat also offered his solution:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'><span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">rlength</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">))</span>
</span><span class='line'><span class="k">for</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">rlength</span><span class="p">:</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>
</span><span class='line'>      
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>I don&#8217;t know if the lack of
formatting in the comments might have generated a different result to
me, but let&#8217;s see how we can generate multiple sets of random DNA
sequence. Getting back to section 4, we can start with the original
script that creates only one set</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>  
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>There is a very easy way of
changing this code to make it output a definite number of sets. WE add a
new input parameter and also a loop to it, and that should take care of
everything.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>  
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">nsets</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nsets</span><span class="p">):</span>
</span><span class='line'>  <span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>      <span class="k">print</span> <span class="n">sequence</span> <span class="k">print</span>
</span></code></pre></td></tr></table></div></figure>


<p>A test with 10 as the
number of sequences, 20 as minimum and 30 as maximum length and 2 sets
output something like this</p>

<p><code>GAGGACGACGTGTCGGATTAGAT GCAGATAAACGGGTAAGATGA
ATTAGGAAGGGTCTAAGAGATCCCGGTAA ATCAGCCTCCGCACGCCTCGTGTA
AGTTTAGCTAGCCGTAGTACTTTC TTTGAAGCTCGGGGCTAAACTCCGACCAC
GACAGGGTAAGGTGACCTTCTCAG TGGTTGACTCTGTTCTCTGGT
CGCAATACGTCGTGGTGCCGCTCCACT TCCTGCAGGGAGGCAGCAGC
TTCGGAGGTATTGAGTACTGTTGCATCG CCCTTATAATACCTAGTGTAACATTC
TAGTCATGTCATATAACCTGTGTCG AGCGCGAACCCTCCAGTTTTTTT
AGGGCGTGTAAGCGACGTTGAACAAG CCATGCGTCTTTTTGGCGATCCGT
ACTCAGATCGGCTAGTATCCTTCGCACG CTATCGTCGTTCCCAGCAACGGCAGCAGA
CGTGCGATGGGGCGAGATTTTT ACGCTTTGGCTAGTGGGGAGTTTCGCCACT</code></p>

<p>what is basically
what we need. Quick &#8216;n&#8217; easy. Of course a <a href="eridanus.net/blog/">top notch bioinformatician</a> that uses Linux 100% of the time
would be able to create a simple bash script and use the initial script
to generate multiple random sets. bash is not a primary scope of the
blog, but let&#8217;s see how to do that. Saying the our initial script was
saved as <code>randomset.py</code> we would have a very short script in bash to
generate multiple sets. We want to generate 10 sets of sequences of
length between 20 and 30 nucleotides, and each set would need to have 10
different sequences (remember that our initial script only accepts 3
parameters).</p>

<p><strong>Off topic</strong></p>

<p><code>for ((i=0;i&lt;10;i++))
  do     
  min=20     
  max=30     
  nseqs=10
  python randomset.py $nseqs $min $max     
echo  done</code></p>

<p>That should do the trick. You can even type it from the command line.
The <code>for</code> syntax in bash is very similar to C/C++ but it does not accept
spaces, as all the variable assignments. When you assigning a variable
value in bash you omit the dollar sign and when you are reading the
variable value you access it with the dollar sign, like in the line that
runs the script. A <code>for</code> loop in bash starts with a <code>do</code> command and
ends in a <code>done</code> command, not brackets or parenthesis involved. The
<code>echo</code> at the end helps adding a separating blank line after each random
set. <strong>off topic</strong></p>

<p>Of course we could use &#8220;makenucseq from the EMBOSS
package&#8221;, we can even use <a href="http://http://scit.us/projects/dawg">Dawg</a>
from my friend Reed Cartwright, which I believe has a shorter learning
curve (and faster compilation) than EMBOSS, but that will be away from
the scope of this blog, which is to show how to use <strong>Python</strong> in
<strong>simple</strong> bioinformatics scripts. I allowed myself to post off topic
subjects (including this last paragraph) because a (very) small
percentage of the commenter(s) seems to misunderstand the real focus of
this tutorial. We don&#8217;t need fancy words or complex topics to show how
simple things can achieve good results in almost no time. My focus is
not give the reader the fish and chips ready to eat. My focus is to show
how to fish and peel, cut and fry the potato in order to cook the dish.
This way, next time he or she might be cooking better things to eat.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/26/nobel/">Nobel</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-26T08:48:51-04:00" pubdate data-updated="true">Oct 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I would kill for a Nobel peace prize.
<em>Steven Wright</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/25/so-he-is-gone/">So, he is gone</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-25T17:03:06-04:00" pubdate data-updated="true">Oct 25<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Yep, he is <a href="http://www.sciam.com/article.cfm?chanID=sa003&amp;articleID=D898BFCA-E7F2-99DF-39F1E89CA728AD43&amp;ref=rss">gone</a>. Until the day he comes back. And he will.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/24/using-subversion-and-svn-time-lapse-to-edit-a-manuscript-take-2-a-short-tutorial/">Using subversion and svn-time-lapse to edit a manuscript, take 2, a short tutorial - PART I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-24T22:25:26-04:00" pubdate data-updated="true">Oct 24<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last post I introduced svn-time-lapse, which is already in <a href="http://svn-time-lapse-view.googlecode.com/files/svn-time-lapse-view-1.3.zip">version 1.3</a>. Jon included the option to view only the changes what I think might be an extra advantage when you are editing a manuscript. I outline here a small tutorial on how to edit a manuscript with subversion and svn-time-lapse. This is based on a Windows system, but it can be used with small changes for Linux, which has the advantage of having (most of the time) subversion installed.</p>

<p>(Be aware that this won&#8217;t work for Word files, you need to be using a text based, ie Latex, document)</p>

<p>First you need to <a href="http://tortoisesvn.net/downloads">download </a>and install the TortoiseSVN interface for Windows. After installing it, the Windows Explorer right-click menu will have some extra options, as can be seen on the figure below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn1.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn1.png" alt="svn1.png" /></a></p>

<p>Checkout allows you to get the source from a repository, while TortoiseSVN gives a list of features to use with subversion. Now, we need to create a directory where the manuscript repository is going to reside. You can create the repository in any place, online, locally, etc. But if you don&#8217;t have an online place to create it, I suggest to create underneath a Documents type of location. On Windows you can create a new folder inside the (My) Documents directory. So, make the new folder <strong>manuscript</strong>. <a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn2.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn2.png" alt="svn2.png" /></a></p>

<p>and the right-click on it and navigate to <strong>TortoiseSVN>>Create repository here</strong>. You can select either format. Done, your repository is set.</p>

<p><strong>SVN</strong><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn3.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn3.png" alt="svn3.png" /></a></p>

<p>If you open the manuscript directory you will see different files, but of course it is empty of actual content. You need to add some content to it and the ideal way is to check out (create a copy for edition) the repository in another location, drop files into this copy and add them to the version control.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn4.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn4.png" alt="svn4.png" /></a></p>

<p>You created the repository in your &#8220;Documents&#8221; directory, so you can check out it to your Desktop (or to any location). This will be your &#8220;local&#8221; copy of the manuscript, the one that you will work on. Close all open windows and right click on your desktop and select <strong>SVN Checkout</strong> from the popup menu.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn5.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn5.png" alt="svn5.png" /></a></p>

<p>In the dialog that appear enter the location of the repository (in your &#8220;Documents&#8221; dir) and the location where you want to checkout to (your &#8220;Desktop&#8221; location), just like the figure below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn6.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn6.png" alt="svn6.png" /></a></p>

<p>Click OK and a &#8220;local&#8221; copy of the repository will be created.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn7.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn7.png" alt="svn7.png" /></a></p>

<p>To be continued.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/23/rowling-dumbledore-is-gay/">Rowling: &#8216;Dumbledore Is Gay&#8217;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-23T11:19:46-04:00" pubdate data-updated="true">Oct 23<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>At a public appearance at Carnegie Hall in New York, author J.K. Rowling announced that Dumbledore, the headmaster character in her popular Harry Potter series, is homosexual. What do you think?</p>

<p>Larry Hahn,
Water Delivery Man
&#8220;Wow. I haven&#8217;t heard that. I&#8217;ve been really busy lately not caring about the sexual preferences of fictional people.&#8221;</p>

<p>More <a href="http://www.theonion.com/content/node/68575">here</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/23/never-trust/">Never trust</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-23T10:10:40-04:00" pubdate data-updated="true">Oct 23<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Never trust the judgment  of a man (or a woman) driving with sunglasses on a rainy day.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/20/using-subversion-and-svn-time-lapse-to-edit-a-manuscript/">Using subversion and svn-time-lapse to edit a manuscript, take 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-20T00:41:26-04:00" pubdate data-updated="true">Oct 20<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Neil Saunders started the thread <a href="http://nsaunders.wordpress.com/2007/10/03/organise-your-bioinformatics-projects-using-subversion-and-trac-part-1/">here </a>and <a href="http://nsaunders.wordpress.com/2007/10/05/organise-your-bioinformatics-projects-using-subversion-and-trac-part-2/">here</a> of using subversion to maintain your files, projects and general stuff. Myself, I find trac a little bit clumsy to check differences between two versions. It is easy to see what has been changed but if the modifications are large it gets a little bit tedious to scroll down and search everything. An example is below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/trac_example.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/trac_example.thumbnail.png" alt="Example of file comparison in trac (2 versions, Python)" /></a></p>

<p>Enters <a href="http://jonaquino.blogspot.com/2007/10/svn-time-lapse-view.html">svn-time-lapse</a>, which seems to be a very interesting tool to compare two versions of a file. I decided to test it on a manuscript edition. I first created a a repository on my Desktop and added an old LaTex file. I replaced this file with a different version of the article (1 or 2 revisions ahead) and opened the repository with svn-time-lapse. The program instantly highlights the portions of the article that are different and display both side by side. One advantage is that svn-time-lapse already gets a certain number of revisions from the repository that can be compared to the current file. A scroll on the top gives an extra help moving from one revision to another.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn_time_lapse_example.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn_time_lapse_example.thumbnail.png" alt="svn-time-lapse example" /></a></p>

<p>The program is available from the developer&#8217;s website, Jon Aquino, and from Google Code. The release still needs some polishing but it is stable enough to be used extensively.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/18/watson-yet-again/">Watson, yet again</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-18T15:25:18-04:00" pubdate data-updated="true">Oct 18<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>So, Watson <a href="http://www.iht.com/articles/ap/2007/10/18/europe/EU-GEN-Britain-US-Scientist-Racism.php">said</a> what he said and then got some lectures <a href="http://www.cbc.ca/world/story/2007/10/18/science-watson-controversy.html">canceled</a>. Fine.</p>

<p>Now Cold Spring Harbor issued an <a href="http://www.cshl.edu/public/releases/07_statement.html">statement</a> basically distancing themselves from Watson and his comments. I fail to understand one thing: when he says good things, he is the Chancellor of CSHL and when they don&#8217;t like the comment, he is not &#8220;speaking on behalf of the institution&#8221;.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/18/off-topic/">Off topic</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-18T08:10:59-04:00" pubdate data-updated="true">Oct 18<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Mr Daniel Swan accused me of not publishing his
<a href="http://eridanus.net/blog/?p=233#respond">comment</a> or reply to his
comments. I have never seen his reply, either in the comments or as an
email to me, so to fulfill his desires it is published here.</p>

<blockquote><p>I admit that some of the things you have addressed in the past, whilst
‘reinventing’ the wheel, offer important insights into the general
problems of programming in general - this particular example was not
one of them. From following the course which I assume was outlined by
‘Beginning Perl for Bioinformatics’ (which I do not think is a good
reference for beginning Perl at all) a user should already be able to
perform the following operations - open a file, iterate over and parse
the contents, open a file for writing, and apply some rules for
formatting the output. Thus you have already given them the tools to
complete the operation should they wish to do so in a Python context.
The problem is that you are trying to educate potentially early
career/transitional bioinformaticians here - and showing them how to
complete tasks in X lines of Python when there is a single command
line application to do the trick I think is disingenuous. We wouldn’t
dream, for instance, of educating our MRes students in this way. I
also would take issue that diversity is a key. You can be as diverse
as you like but no-one is seriously going to suggest that you write
your next script to parse something out of a GenBank file in
assembler. One of the hallmarks of any computer professional in any
field is the ability to use the right tool for the job. You should be
technology agnostic and be prepared to string together applications or
scripts from anywhere to reach your goal - take a look at something
like Taverna for instance. I notice in the other post that you mention
there is a high proportion of Windows users to this site and therefore
command line *nix comments are of no interest. Although I know a
number of bioinformaticians who use Windows (generally to integrate
with an institutions existing Microsoft heavy infrastructure) they all
have logins to dozens of *nix machines, run *nix in a VM or at the
very least have a Cygwin installation with which to work with. Basing
the availability of technology on the OS of your page imprints is not
necessarily indicative. I hope you don’t mind me addressing these
points here rather in my blog, as my blog is personal (hence the lack
of bioinformatics content on there). I have actually been following
these tutorials for a while (as someone who came into Bioinformatics
some time ago - my first forays in programming were in Perl) it is
nice for me to see tasks that I find intuitive in one language written
in another. Consequently the fact that I felt need to comment, merely
suggests that I felt strongly on this issue, rather than having any
negative feelings towards what you are trying to achieve.</p></blockquote>

<p>I am sorry for the usual readers of this tutorial, for posting his
comments as an article here. We will be back to our uneducated, simple
and basic &#8220;tutorial for bioinformatics in X lines for the noobie in
Windows&#8221; soon. Thanks Daniel.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/17/splitting-a-fasta-file/">Splitting a FASTA file</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-17T20:08:17-04:00" pubdate data-updated="true">Oct 17<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have posted in my other <a href="http://python.genedrift.org">site</a> different ways to split a FASTA file. <a href="http://python.genedrift.org/2007/10/10/alternative-methods-to-split-a-fasta-file/">Take a look</a>. The Python method is <a href="http://python.genedrift.org/2007/10/10/splitting-multiple-fasta-file-or-why-shouldnt-we-reinvent-the-wheel/">here</a>.</p>

<p>I am planning to include as many languages as possible.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/32/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/30/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'nuin';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
