
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="On the last entry we saw how to split a multiple FASTA file with our
previous FASTA module. The previous script saved a sequence per file,
where the &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.heroku.com/blog/page/31">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-30723948-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
    <h2>The most influential bioinformatician in the world (TM)</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.heroku.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/09/splitting-a-multiple-sequence-fasta-file-making-it-better/">Splitting a multiple sequence FASTA file, making it better</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-09T16:13:30-04:00" pubdate data-updated="true">Oct 9<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>On the last entry we saw how to split a multiple FASTA file with our
previous FASTA module. The previous script saved a sequence per file,
where the filename was identical to the original file with the exception
of a number added to the end. Let&#8217;s say we have the file mysequences.fa,
our script would save mysequences.fa1, mysequences.fa2, &#8230;,
mysequences.fa<em>n</em>. This is fine, but we need to make it better. Ideally
we should ask the user for a prefix or a filename for the output. On the
other hand, it would be an improvement if we put the number before the
file tag. We will do that in a couple of lines. First we need to find
the tag and the actual filename. We could go through a longer route if
we used the <code>find</code> method to find the dot and then extract portions of
the string to the actual name and tag.</p>

<p>But we can use <code>split</code></p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">temp</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;.&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">filename</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="n">tag</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span></code></pre></td></tr></table></div></figure>


<p>We split the original
string at the dot and the first item or the returned list is the file
name and the second item is the tag. Now we just need to change a little
bit the code for the output and we are done. The final script is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#!/usr/bin/env python</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">temp</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;.&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">filename</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span class='line'><span class="n">tag</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="n">count</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
</span><span class='line'>  <span class="c">#this is the only different line </span>
</span><span class='line'>  <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="o">+</span><span class="s">&#39;_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">count</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="n">tag</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\\</span><span class="s">n&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>  <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/09/splitting-a-multiple-sequence-fasta-file/">Splitting a multiple sequence FASTA file</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-09T15:22:39-04:00" pubdate data-updated="true">Oct 9<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Starting the new phase of the website, we
will see the easiest way to split a FASTA file. There are many ways to
name each file after the split, and we will use a simple <em>original
filename</em> + number. We also going to use the FASTA class and functions
defined previously in <code>fasta.py</code>.</p>

<p>To refresh our memories here is the code</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">class</span> <span class="nc">Fasta</span><span class="p">:</span>
</span><span class='line'>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
</span><span class='line'>        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
</span><span class='line'>    <span class="n">items</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;&gt;&quot;</span><span class="p">):</span>
</span><span class='line'>           <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
</span><span class='line'>               <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aninstance</span><span class="p">)</span>
</span><span class='line'>           <span class="n">index</span><span class="o">+=</span><span class="mi">1</span>
</span><span class='line'>           <span class="n">name</span> <span class="o">=</span> <span class="n">line</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>           <span class="n">seq</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>           <span class="n">aninstance</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
</span><span class='line'>        <span class="k">else</span><span class="p">:</span>
</span><span class='line'>           <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'>           <span class="n">aninstance</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'>    <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aninstance</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">items</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, basically we need to create a script that reads
the file and then in a loop saves every sequence in a separate file. For
the sake of simplicity all saved files will be in the same directory. To
the script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'>
</span><span class='line'><span class="c">#define a variable with the input filename </span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="c">#open the file and read the sequences in one line </span>
</span><span class='line'><span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'><span class="c">#counter to keep the number the sequences and to #be written to the output files </span>
</span><span class='line'><span class="n">count</span> <span class="o">=</span> <span class="mi">1</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
</span><span class='line'>  <span class="c">#open a file to output output =</span>
</span><span class='line'>  <span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">count</span><span class="p">),</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="c">#write the name and sequence</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>  <span class="c">#increment the counter </span>
</span><span class='line'>  <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
</span></code></pre></td></tr></table></div></figure>


<p>Very simple, but nice script. In the end
we will have a list of files with only one sequence in it, with
identical name of the original file but with a number in the end. Next
we will see how to modify the filename and even extract some information
from the FASTA sequence to use as a filename.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/05/on-computer-viruses-and-viruses/">On computer viruses and viruses</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-05T13:27:02-04:00" pubdate data-updated="true">Oct 5<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Bruce Schneier, a computer security expert, <a href="http://www.schneier.com/blog/archives/2007/10/the_storm_worm.html">posted</a> in his blog new findings of a powerful worm that is silently spreading through the internet. Technically, a worm is a virus that crawls the internet infecting hosts and moving to others, just like last week&#8217;s flu virus. It is believed that the Storm Worm infected from 1 to 50 million computers and no remedy was found to &#8220;kill&#8221; it or stop its spread. Storm is 3-in-1 computer application: it is a worm, a trojan horse and a bot. Some of its characteristics are:</p>

<ul>
<li><p>it attacks and then shuts off for a certain amount of time. Just like herpes.</p></li>
<li><p>not every host spreads the worm, some hosts only receive orders</p></li>
<li><p>your computer is not damaged, no performance impact. Just like a cold virus.</p></li>
<li><p>it is like bittorrent or Kazaa, it is peer-to-peer. Just like any virus.</p></li>
<li><p>its code mutates every 30 minutes. Just like any virus might (maybe not every 30 min)</p></li>
<li><p>Storm can be spread by different forms. Have you heard of any virus with multiple infection strategies? Yes, I guess so.</p></li>
</ul>


<p>It is the old cliche of life imitating &#8230; No it is not. It is normal evolution. the first computer viruses (virii) simply destroy a machine, with nasty consequences making any recovery impossible and maybe not spreading further. It is just like a new virus. Think Ebola. When it was first released in the human population it killed in a matter of days and maybe when the host died, the virus died too. When was the last time you heard that someone was killed by a cold? How many people worldwide are infected by the cold virus annually?</p>

<p>At the same time that <a href="http://tiago.org/ps/">Tiago</a> pointed <a href="http://tiago.org/ps/2007/10/03/evolution-is-a-route-to-genetic-stability/">here</a> and <a href="http://tiago.org/ps/2007/10/04/cheap-philosophy/">here</a> about the genome stability, what about the code stability of this highly technological computer viruses? Would they achieve a point where every computer in the world, at least the ones connected to the internets, would be infected by a virus like this with no &#8220;symptom&#8221;? Would we have &#8220;good&#8221; computer viruses (virii) that would be like the bacteria in our gut?</p>

<p>Tiago was looking, if I haven&#8217;t misinterpreted him, for the perfect genome, maybe the minimal one, where all the needs of an individual to adapt to an ever changing environment are contained in one solution. Maybe a virus has the perfect genome, as everything superfluous is removed and if it &#8220;needs&#8221; to change, it can get some extra genes and DNA from the hosts.</p>

<p>Maybe the Stormworm is the perfect computer program, or it is just the beginning to solve NP-complete problems.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/03/genbank-parsing-some-features-and-a-change-for-the-future/">GenBank: parsing some features (and a change for the future)</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-03T16:25:16-04:00" pubdate data-updated="true">Oct 3<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This is the last entry based on the book. In my opinion, further topics
in the book are a little bit redundant and can be accomplished quite
easily if you have followed the tutorial here.</p>

<p>If a good number of
people have interest in checking the remainder of the book, just let me
know and I will get back and follow the book. At the same time I am
accepting suggestions on topics to be covered (send me an email or leave
a comment). I already have some in mind and I am preparing a couple for
the next phase of the website. So, here is the last entry. Last time we
saw how to extract the sequence from a GenBank file.</p>

<p>This time we are
going to parse some other information from these files. Basically we
will use the same idea of our last post to extract the Organism name,
the Locus and the Accession number of the item. From our last entry we
have to remember this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="n">issequence</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">issequence</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&#39;0123456789 &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">issequence</span> <span class="o">=</span> <span class="bp">True</span>
</span></code></pre></td></tr></table></div></figure>


<p>and modify it to our needs. Looks simple, and it is. Let&#8217;s see</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span> <span class="nn">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">locus</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="n">organism</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'> <span class="n">accession</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>  <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;LOCUS&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">locus</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ACCESSION&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">accession</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>  <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORGANISM&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>      <span class="n">organism</span> <span class="o">=</span> <span class="n">line</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">locus</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'><span class="k">print</span> <span class="n">organism</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'><span class="k">print</span> <span class="n">accession</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Just add a flag for each entry you want
to parse and that&#8217;s it. For longer entries, such as the sequence, we
have to use the same approach used before, with a boolean flag and
concatenating the lines until another flag is found.</p>

<hr />

<p>Well, that&#8217;s it. After 46 entries we start a new phase.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/02/running-vista/">Running Vista</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-02T13:20:39-04:00" pubdate data-updated="true">Oct 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Recently I purchased a new system. It is a &#8220;slimline&#8221; HP computer that comes pre-loaded with Windows Vista Home whatever. My first idea was to partition the drive and install XP and a Linux distribution. While I was organizing my files, transferring stuff from my old box to the new one I got used to Vista. It is not a flawless system, like any other, it has its share of problems but has some nice things too.</p>

<p>On the upside, it is nice to the eyes. It is a beautiful desktop system, with some eye candy displays and some nice features, that are an upgrade over XP. The interface is cleaner and the Start button- Programs button-bar is much more functional. Not to mention the search which is vastly improved over its predecessor.</p>

<p>On the downside, there still some obstacles to pass. This system comes equipped with WiFi Internet and I decided to use only that due to the amount of cables needed. Even though the router is just inches from the mini-tower I stayed with wireless. For a couple of days it was fine, then Vista lost it and cannot find the wireless card. Gone, kaput. It happened to me once with XP: the system refused to enable the on-board ethernet card, which was working perfectly under Linux.</p>

<p>Accomplishing simple things on Vista are very annoying. To run applications and setup packages you are greeted with dialogs asking if you really want to run the program or not. Even Microsoft applications are not spared of the pre-check. Sharing folder on a intranet is a breeze, but not if you have a Mac on the side. My version of Vista does not come with some Administrator tools, which are only available on the top version (extra 200 bucks) and some user security settings can only be set via registry. It took me more than minutes to find a solution online that allowed me to connect from Vista to my iMac.</p>

<p>I think I am keeping Vista for a while. At least it saves me some disk space, not having to install XP too. Let&#8217;s see how it goes.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/02/i-cringe/">I cringe</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-02T12:34:39-04:00" pubdate data-updated="true">Oct 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I cringe every time I hear a company PR (or similar product) or advertisement saying that X or Y is &#8220;in the company&#8217;s DNA&#8221;.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/28/on-punching-the-liver-quotes/">On &#8220;punching the liver&#8221; quotes</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-28T10:13:55-04:00" pubdate data-updated="true">Sep 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>When asked how he would fight the largest news broadcast channel in Brazil, the owner (and self acclaimed <em>bishop</em> of the Church of the Universal Kingdom of God and of a newly released new channel replied:</p>

<blockquote><p>We punch them in the liver until they fall</p></blockquote>

<p>So long for the religion that asks they followers to &#8220;pray for peace&#8221;.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/28/hockey/">Hockey</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-28T10:08:39-04:00" pubdate data-updated="true">Sep 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The most hardcore hockey team in Canada, instead of having blue vs white practices, had shirts vs skins.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/28/genbank-files-take-two/">GenBank files: take two</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-28T09:50:27-04:00" pubdate data-updated="true">Sep 28<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last time we saw how to extract the sequence from our GenBank file, but
the final result might not be as nice as we wanted it to be. So, we need
to make it prettier. Running last entry&#8217;s script we end up with this
sequence</p>

<blockquote><p>1 gtttggtcct aaccttgtaa tcaattttta cttaatatac acatgcaagt ctccgcaccc
61 ctgtgaaaac gcccttaaat cccccatggg ataaggagct ggtatcaggc acgaaaatct <br/>
121 gcccaaaaca cctagctatg ccacacccac aagggtactc agcagtgattgacattattt <br/>
181 ataagcgcca gcttgactca gttaaagtaa agagagccgg caaatctggt gccagccgcc <br/>
241 gcggttaccc cacgtggctc aaattgattt ctttcggcgt aaagcgtgat taaagtgccc <br/>
301 atcaacattg gagttaaact aaaattaagc tgtgacacgc ttattttaca gaaaagcaca <br/>
361 aacgaaagtt acttcaattt aaacaacttg aattcacgac agtcaggaca caaactggga <br/>
421 ttagataccc cactatgccc gaccgtaaac tttaatttac accatcaccg ccagagaact <br/>
481 acgagcaaag cttaaaactc aaaggacttg acggtccccc acatccccct agaggagcct <br/>
541 gtcctttaat cgataatccc cgcttaacct caccattctt agtctttcag cctgtatacc <br/>
601 tccgtcgtca gcttaccccg tgagcgaaaa ttagtgagct taatgtccac acgtctacac <br/>
661 gtcaggtcaa ggtgcagcaa atataatggg aagagatggg ctacactttc tagtctagaa <br/>
721 tatacgaaag accacctatg aaacctggtc agaaggcgga tttagaagta aaaggaaacc <br/>
781 agagcatccc ttttaatttg gcactggggc atgtacacac cgcccgtcac cctcttcaaa <br/>
841 gcctaatttt agtatctaac caactaacgc ctagtagaag aggcaagtcg taacatggta <br/>
901 agtataccgg aaggtgtgct tggaaacaaa atatagccta atcaaagcat ttcgcttaca <br/>
961 ccgaaaagtt atctgtgaaa ttcagattat tttgagctaa aaatctagcc ccactttatt <br/>
1021 ctataatccc ttatcactta aattcatgaa tcaaaacatt ttaataatca agtaaaggcg <br/>
1081 attgaaaaat taataggagc aatatatact gtaccgcaag ggaaagatga aatagaaatg <br/>
1141 aaataataat taaagcataa aaaagtaaag attaaatctt gtaccttttg catcatgatt <br/>
1201 taactagtct acccaggcaa aatgatttta agtctgacct cccgaaacta agtgagctac <br/>
1261 ttcaaggcag cttaatgagc aaatccgtct ctgtcgcaaa agagtggaga gaccttcaag <br/>
1321 tagaagtgat agacctaacg aacttagtaa tagctggtta ttcaagaaaa ggatctcagt <br/>
1381 ccaacctaaa gtcaaattaa tgtttaaaaa taaaaattct gaccttagag taattcaatt <br/>
1441 aaggtacagc ctacttgaaa caggatacaa ccttaactaa tgggtaactt accccttcat <br/>
1501 cttttaagtg ggcctaaaag cagccacctt taaaatagcg tcaaagctta gccgtcctat <br/>
1561 acatctaata ccaaaaacat ctatgaaccc tatactcata ttgaataatt ctatattatt <br/>
1621 atagagattt ttatgttaaa actagtaaca agaattaaat tttctctatt atgttcgtgt <br/>
1681 acatcagaaa ggataaacca ctgataattg acatgcatga gtaaaaagca gtaacttaac <br/>
1741 aagaaaaccc tcctaactct aatgttaacc taacacaagt acatctcaag aaagatttaa <br/>
1801 agaaaaagaa ggaactcggc aaacattaac ctcgcctgtt taccaaaaac atcgcctctt <br/>
1861 gtcaaaattt aagaggtcca gcctgcccag tgaccctgtt caacggccgc ggtatcctaa <br/>
1921 ccgtgcgaag gtagcgtaat cacttgttct ttaaataagg actagtatga atggcaccac <br/>
1981 gagggttata ctgtctcctt tttctaatca gtgaaactaa tcttcccgtg aagaagcggg <br/>
2041 aatttttata taagacgaga agaccctatg gagctttaga cgagtaacaa ctgctaattt <br/>
2101 tataatattt cagataatat ctctatccta gcattatgat tataagtctt tggttggggt <br/>
2161 gaccgcggag aaaaaaataa cctccacatt gaaagaatat tattctaagc aaaaagacac <br/>
2221 atctttaagc atcaacaaat tgacatctat tgacccaata ttttgatcaa cgaaccaagt <br/>
2281 taccctaggg ataacagcgc aatccacttc gagagctctt atcgacaagt gggcttacga <br/>
2341 cctcgatgtt ggatcagggt atcctagtgg tgtagccgct actaaaggtt cgtttgttca <br/>
2401 acgattaaaa ccct //</p></blockquote>

<p>Notice that there were no mention to
the <code>//</code> at the end of the file This double slash represents the EOF of
a GenBank file. Ideally we need to remove it from the final output, as
we need to remove the nucleotide number at the beginning of each line.
In order to do that we need to modify our last script. First we will add
a condition on the <code>for</code> loop that will take care of the double slash at
the end. Then we need to use a trick to remove the numbers from the
lines.</p>

<p>We should expect the numbers to be at the beginning of the line,
and never in another place. The long way would be to create a regular
expression and then replace every number occurrence with an empty
string. But Python comes with batteries included, so we will take the
short path. We use the <code>lstrip</code> string method that will remove
characters from the left part of the string, and set it to remove
numbers from 0 to 9 and the extra space before the nucleotides. And our
script will like this:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="n">issequence</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">issequence</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&#39;0123456789 &#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="n">issequence</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that the first <code>if</code>
condition was modified to include a condition that we do <strong>not</strong> find a
slash. And at the line where we concatenate the sequence we added a
<code>lstrip('0123456789 ')</code> to modify the string on the fly. The final
result looks much better now. But we still have a small &#8220;problem&#8221;: there
are spaces between groups of 10 nucleotides and we want to get rid of
them. We would be surprised if it was difficult, and as expected it is
not. We used <a href="http://python.genedrift.org/2007/09/07/restrinction-enzymes-second-take/">here</a>
the replace method and we can use it here too. We only need to modify
the line that concatenates the sequence, and our final script will be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="n">issequence</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">issequence</span> <span class="o">==</span> <span class="bp">True</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&#39;0123456789 &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="n">issequence</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that we add the <code>replace</code> method after the <code>lstrip</code>, but it can either way. The output
should be only the nucleotides, with no space or numbers.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/09/27/genbank-files/">GenBank files: take one</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-09-27T16:13:14-04:00" pubdate data-updated="true">Sep 27<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We are back, moving to a new chapter of the book and starting a new
section on the site. This chapter deals with the manipulation of
<a href="http://www.ncbi.nlm.nih.gov">GenBank</a> files. These files are used by
NCBI to store information about RNA, DNA and protein sequences. It is
usually composed of an annotation section, that gives information about
the sequence present in the particular file. I won&#8217;t spend much time
explaining the GenBank format, because it is not the goal of the site.
The perl book has some good explanation about it and you can also find
more information
<a href="http://www.umanitoba.ca/afs/plant_science/psgendb/doc/GenBank/gbrel.txt">here</a>.
Also, we are going to see here some of the characteristics of such
files. The GenBank file we are going to manipulate from now on is this
one</p>

<p>`LOCUS       DQ283072                2414 bp    DNA     linear   VRT 12-MAR-2010
DEFINITION  Megaelosia goeldii voucher Paulo Nuin 12S ribosomal RNA gene,</p>

<pre><code>        partial sequence; tRNA-Val gene, complete sequence; and 16S
        ribosomal RNA gene, partial sequence; mitochondrial.
</code></pre>

<p>ACCESSION   DQ283072
VERSION     DQ283072.1  GI:90296241
KEYWORDS    .
SOURCE      mitochondrion Megaelosia goeldii (Rio big-tooth frog)
  ORGANISM  Megaelosia goeldii</p>

<pre><code>        Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
        Amphibia; Batrachia; Anura; Neobatrachia; Hyloidea; Hylodidae;
        Megaelosia.
</code></pre>

<p>REFERENCE   1  (bases 1 to 2414)
  AUTHORS   Frost,D.R., Grant,T., Faivovich,J., Bain,R., Haas,A.,</p>

<pre><code>        Haddad,C.F.B., de Sa,R.O., Channing,A., Wilkinson,M.,
        Donnellan,S.C., Raxworthy,C., Campbell,J.A., Blotto,B.L., Moler,P.,
        Drewes,R.C., Nussbaum,R.A., Lynch,J.D., Green,D.M. and Wheeler,W.C.
</code></pre>

<p>  TITLE     The Amphibian Tree of Life
  JOURNAL   Bull. Am. Mus. Nat. Hist. 297, 1-291 (2006)
REFERENCE   2  (bases 1 to 2414)
  AUTHORS   Frost,D.R., Grant,T., Faivovich,J., Bain,R., Haas,A.,</p>

<pre><code>        Haddad,C.F.B., de Sa,R.O., Channing,A., Wilkinson,M.,
        Donnellan,S.C., Raxworthy,C., Campbell,J.A., Blotto,B.L., Moler,P.,
        Drewes,R.C., Nussbaum,R.A., Lynch,J.D., Green,D.M. and Wheeler,W.C.
</code></pre>

<p>  TITLE     Direct Submission
  JOURNAL   Submitted (26-OCT-2005) Herpetology, Division of Vertebrate</p>

<pre><code>        Zoology, American Museum of Natural History, Central Park West at
        79th Street, New York, NY 10024, USA
</code></pre>

<p>FEATURES             Location/Qualifiers</p>

<pre><code> source          1..2414
                 /organism="Megaelosia goeldii"
                 /organelle="mitochondrion"
                 /mol_type="genomic DNA"
                 /specimen_voucher="Paulo Nuin"
                 /db_xref="taxon:209670"
                 /country="Brazil: Rio de Janeiro, Teresopolis, Rio Beija
                 Flor, 910 m, 22'24'S, 42'69'W"
 misc_feature    &lt;1..&gt;2414
                 /note="contains 12S ribosomal RNA, tRNA-Val, and 16S
                 ribosomal RNA"
</code></pre>

<p>ORIGIN</p>

<pre><code>    1 gtttggtcct aaccttgtaa tcaattttta cttaatatac acatgcaagt ctccgcaccc
   61 ctgtgaaaac gcccttaaat cccccatggg ataaggagct ggtatcaggc acgaaaatct
  121 gcccaaaaca cctagctatg ccacacccac aagggtactc agcagtgatt gacattattt
  181 ataagcgcca gcttgactca gttaaagtaa agagagccgg caaatctggt gccagccgcc
  241 gcggttaccc cacgtggctc aaattgattt ctttcggcgt aaagcgtgat taaagtgccc
  301 atcaacattg gagttaaact aaaattaagc tgtgacacgc ttattttaca gaaaagcaca
  361 aacgaaagtt acttcaattt aaacaacttg aattcacgac agtcaggaca caaactggga
  421 ttagataccc cactatgccc gaccgtaaac tttaatttac accatcaccg ccagagaact
  481 acgagcaaag cttaaaactc aaaggacttg acggtccccc acatccccct agaggagcct
  541 gtcctttaat cgataatccc cgcttaacct caccattctt agtctttcag cctgtatacc
  601 tccgtcgtca gcttaccccg tgagcgaaaa ttagtgagct taatgtccac acgtctacac
  661 gtcaggtcaa ggtgcagcaa atataatggg aagagatggg ctacactttc tagtctagaa
  721 tatacgaaag accacctatg aaacctggtc agaaggcgga tttagaagta aaaggaaacc
  781 agagcatccc ttttaatttg gcactggggc atgtacacac cgcccgtcac cctcttcaaa
  841 gcctaatttt agtatctaac caactaacgc ctagtagaag aggcaagtcg taacatggta
  901 agtataccgg aaggtgtgct tggaaacaaa atatagccta atcaaagcat ttcgcttaca
  961 ccgaaaagtt atctgtgaaa ttcagattat tttgagctaa aaatctagcc ccactttatt
 1021 ctataatccc ttatcactta aattcatgaa tcaaaacatt ttaataatca agtaaaggcg
 1081 attgaaaaat taataggagc aatatatact gtaccgcaag ggaaagatga aatagaaatg
 1141 aaataataat taaagcataa aaaagtaaag attaaatctt gtaccttttg catcatgatt
 1201 taactagtct acccaggcaa aatgatttta agtctgacct cccgaaacta agtgagctac
 1261 ttcaaggcag cttaatgagc aaatccgtct ctgtcgcaaa agagtggaga gaccttcaag
 1321 tagaagtgat agacctaacg aacttagtaa tagctggtta ttcaagaaaa ggatctcagt
 1381 ccaacctaaa gtcaaattaa tgtttaaaaa taaaaattct gaccttagag taattcaatt
 1441 aaggtacagc ctacttgaaa caggatacaa ccttaactaa tgggtaactt accccttcat
 1501 cttttaagtg ggcctaaaag cagccacctt taaaatagcg tcaaagctta gccgtcctat
 1561 acatctaata ccaaaaacat ctatgaaccc tatactcata ttgaataatt ctatattatt
 1621 atagagattt ttatgttaaa actagtaaca agaattaaat tttctctatt atgttcgtgt
 1681 acatcagaaa ggataaacca ctgataattg acatgcatga gtaaaaagca gtaacttaac
 1741 aagaaaaccc tcctaactct aatgttaacc taacacaagt acatctcaag aaagatttaa
 1801 agaaaaagaa ggaactcggc aaacattaac ctcgcctgtt taccaaaaac atcgcctctt
 1861 gtcaaaattt aagaggtcca gcctgcccag tgaccctgtt caacggccgc ggtatcctaa
 1921 ccgtgcgaag gtagcgtaat cacttgttct ttaaataagg actagtatga atggcaccac
 1981 gagggttata ctgtctcctt tttctaatca gtgaaactaa tcttcccgtg aagaagcggg
 2041 aatttttata taagacgaga agaccctatg gagctttaga cgagtaacaa ctgctaattt
 2101 tataatattt cagataatat ctctatccta gcattatgat tataagtctt tggttggggt
 2161 gaccgcggag aaaaaaataa cctccacatt gaaagaatat tattctaagc aaaaagacac
 2221 atctttaagc atcaacaaat tgacatctat tgacccaata ttttgatcaa cgaaccaagt
 2281 taccctaggg ataacagcgc aatccacttc gagagctctt atcgacaagt gggcttacga
 2341 cctcgatgtt ggatcagggt atcctagtgg tgtagccgct actaaaggtt cgtttgttca
 2401 acgattaaaa ccct
</code></pre>

<p>//`</p>

<p>which stores a sequence of a mitochondrial gene of a stream frog from
South America, called <em>Megaelosia goeldii</em>, also known as Rio big-tooth
frog. (get a better formatted file <a href="http://python.genedrift.org/megaelosia.gbk">here</a>) Our fast task will
be to extract the DNA sequence from the file. This sounds easy, and not
surprisingly it is. If we take a closer look at the file we will see
that the sequence starts after the mark <strong>ORIGIN</strong>. From what we have
seen before we just need to read the file, the a boolean variable that
checks for <strong>ORIGIN</strong> and concatenate everything after that. Something
like this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="n">gbfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="n">issequence</span> <span class="o">=</span> <span class="bp">False</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gbfile</span><span class="p">:</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">issequence</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span>
</span><span class='line'>    <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&#39;ORIGIN&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
</span><span class='line'>        <span class="n">issequence</span> <span class="o">=</span> <span class="bp">True</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">sequence</span>
</span><span class='line'>      
</span></code></pre></td></tr></table></div></figure>


<p>Quick and easy. When we find <strong>ORIGIN</strong>, <code>issequence</code> has its state
changed to True and the lines below will concatenated into a string. We
print at the end. Next time we will do more fancy things.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/32/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/30/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/06/03/installing-circos-on-os-x/">Installing Circos on OS X</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/04/09/new-blog/">New blog</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>Github Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/nuin">@nuin</a> on Github
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'nuin',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("nuin", 4, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/nuin" class="twitter-follow-button" data-show-count="false">Follow @nuin</a>
  
</section>


<section>
  <h1>On Delicious</h1>
  <div id="delicious"></div>
  <script type="text/javascript" src="http://feeds.delicious.com/v2/json/nuin?count=3&amp;sort=date&amp;callback=renderDeliciousLinks"></script>
  <p><a href="http://delicious.com/nuin">My Delicious Bookmarks &raquo;</a></p>
</section>


  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'zientzilaria';
      
        
        var disqus_script = 'count.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
