
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Last couple of entries we saw how to split a multiple FASTA file. And
how can we achieve the opposite, merge single (or sometimes multiples)
sequence &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/31">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/29/merging-single-or-multiple-sequence-fasta-files/">Merging single (or multiple) sequence FASTA files</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-29T15:17:02-04:00" pubdate data-updated="true">Oct 29<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last couple of entries we saw how to split a multiple FASTA file. And
how can we achieve the opposite, merge single (or sometimes multiples)
sequence files in on larger multiple sequence file? In Python is very
simple, but we need to introduce a new module that would allow us to get
the desired result. In most scripts we have seen in previous posts, we
end up reading a single file from the input (in some cases two) and
that&#8217;s not enough in the case of merging multiple files.</p>

<p>Python has a
module called <code>glob</code> that is very powerful for reading contents of a
directory. We can use any type of pattern matching to find only the
files we want, wildcards allowed. Let&#8217;s test glob and see what we can
get. We start with a simple script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">glob</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&#39;\*.txt&#39;</span><span class="p">):</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">i</span>
</span></code></pre></td></tr></table></div></figure>


<p>If we run this
script in a directory (Windows or Linux or Mac) we should see a list of
the filenames that match the patter passed to glob.glob. In this case
every file with a txt extension. This should be enough for us, for now.
Basically we need to create a list of all FASTA files, that match a
certain name pattern, and then take care of the sequences using our
already known FASTA module. To make things more flexible, we should ask
in the command line for a pattern to match, and maybe a output filename.
Our script would look like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">glob</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">fasta</span>
</span><span class='line'><span class="n">filepattern</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="c">#initialize lists </span>
</span><span class='line'><span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="c">#glob.glob returns a list of files that match the pattern </span>
</span><span class='line'><span class="k">for</span> <span class="nb">file</span> <span class="ow">in</span>  <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">filepattern</span><span class="p">):</span>
</span><span class='line'>  <span class="c">#we read the contents and an instance of the class is returned </span>
</span><span class='line'>  <span class="n">contents</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
</span><span class='line'>  <span class="c">#a file can contain more than one sequence so we read them in a loop</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">contents</span><span class="p">:</span>
</span><span class='line'>      <span class="n">names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span class='line'>      <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="c">#we print the output </span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">names</span><span class="p">)):</span>
</span><span class='line'>  <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">+</span> <span class="n">seqs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</span></code></pre></td></tr></table></div></figure>


<p>Except for the
glob part, we&#8217;ve already seen the rest in previous entries. Here, the
input <code>filepattern</code> should be enclosed in single or double quotes in
order to be transmitted properly (for instance when you include
wildcards). In the first <code>for</code> loop we access the list of files created
by <code>glob.glob</code> and extract the sequence name(s) and sequence(s) from
each file and parse them to a list which will serve to print the final
output. An example command line would look like
<code>python merge.py '*.fas' merged.fa</code> Of course, there are several ways of
merging FASTA files, but not every time you have Linux&#8217;s <code>cat</code> available.
Next time we will (try) to have fun with Python functional programming.
Stay tuned.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/29/splitting-a-fasta-file-using-awk-no-sed-required-or-do-we-care-about-csplit/">Splitting a FASTA file using awk (no sed required), or do we care about csplit?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-29T11:37:07-04:00" pubdate data-updated="true">Oct 29<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We saw that &#8221;<a href="http://eridanus.net/">top notch bioinformaticians</a>&#8221; use
csplit to split FASTA files, so I decided to
<a href="http://python.genedrift.org/2007/10/10/alternative-methods-to-split-a-fasta-file/">post as many as possible alternatives</a>
to split these files. As csplit, awk is something found with more
frequency in Linux machines than Windows, but it can be installed on
Windows (even Vista) and it runs fine.</p>

<p>Awk is usually shown in parallel
with sed, as both utilities have the main objective of text processing.
I browsed the web searching for a sed/awk solution to split FASTA files
something that I thought could be done in one line. And I found
<a href="http://www.unix.com/shell-programming-scripting/14576-split-file-using-awk.html">it</a>.</p>

<p>With a little tweak I ended up with a one-liner using awk that split the
sequences in multiple files:</p>

<p><code>awk '/^&gt;/{f="fasta."++d} {print &gt; f}'</code></p>

<p>Awk breaks each line of its input into fields that can be checked,
printed or ignored. In the command line above. awk&#8217;s commands are in
between the single quotes. The first command is a regex that matches
initial <code>&gt;</code>. Inside the curly braces we assign a variable <code>f</code> to store
the output filename we want and also have an integer that is incremented
every time we pass through it. The second curly braces are for the
actual output, where we use awk&#8217;s <code>print</code> command and redirect output to
a file named after the variable we created in the first commands in
between curly braces. Finally, we assign an input file, outside the
region delimited by single quotes. And that&#8217;s it. As output we will have
a series of files called fasta each one containing one sequence.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/26/generating-multiple-sets-of-random-dna-sequences-with-one-script-and-a-bash-one/">Generating multiple sets of random DNA sequences with one script (and a bash example) [updated]</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-26T12:15:47-04:00" pubdate data-updated="true">Oct 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A commenter, Dilmurat, gave me an idea about the <a href="http://python.genedrift.org/2007/04/04/a-script-to-simulate-dna-sequence-sets/">script</a>
that generates random DNA sequence sets. Apparently it wasn&#8217;t clear that
the script was intended to generate only one sequence set, and not
multiple sets. Dilmurat also offered his solution:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'><span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">rlength</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">))</span>
</span><span class='line'><span class="k">for</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">rlength</span><span class="p">:</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>
</span><span class='line'>      
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>I don&#8217;t know if the lack of
formatting in the comments might have generated a different result to
me, but let&#8217;s see how we can generate multiple sets of random DNA
sequence. Getting back to section 4, we can start with the original
script that creates only one set</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>  
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>  <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>There is a very easy way of
changing this code to make it output a definite number of sets. WE add a
new input parameter and also a loop to it, and that should take care of
everything.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>  <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>  <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>      <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>  <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>  
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">nsets</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nsets</span><span class="p">):</span>
</span><span class='line'>  <span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>  <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>  <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>  
</span><span class='line'>  <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>      <span class="k">print</span> <span class="n">sequence</span> <span class="k">print</span>
</span></code></pre></td></tr></table></div></figure>


<p>A test with 10 as the
number of sequences, 20 as minimum and 30 as maximum length and 2 sets
output something like this</p>

<p><code>GAGGACGACGTGTCGGATTAGAT GCAGATAAACGGGTAAGATGA
ATTAGGAAGGGTCTAAGAGATCCCGGTAA ATCAGCCTCCGCACGCCTCGTGTA
AGTTTAGCTAGCCGTAGTACTTTC TTTGAAGCTCGGGGCTAAACTCCGACCAC
GACAGGGTAAGGTGACCTTCTCAG TGGTTGACTCTGTTCTCTGGT
CGCAATACGTCGTGGTGCCGCTCCACT TCCTGCAGGGAGGCAGCAGC
TTCGGAGGTATTGAGTACTGTTGCATCG CCCTTATAATACCTAGTGTAACATTC
TAGTCATGTCATATAACCTGTGTCG AGCGCGAACCCTCCAGTTTTTTT
AGGGCGTGTAAGCGACGTTGAACAAG CCATGCGTCTTTTTGGCGATCCGT
ACTCAGATCGGCTAGTATCCTTCGCACG CTATCGTCGTTCCCAGCAACGGCAGCAGA
CGTGCGATGGGGCGAGATTTTT ACGCTTTGGCTAGTGGGGAGTTTCGCCACT</code></p>

<p>what is basically
what we need. Quick &#8216;n&#8217; easy. Of course a <a href="eridanus.net/blog/">top notch bioinformatician</a> that uses Linux 100% of the time
would be able to create a simple bash script and use the initial script
to generate multiple random sets. bash is not a primary scope of the
blog, but let&#8217;s see how to do that. Saying the our initial script was
saved as <code>randomset.py</code> we would have a very short script in bash to
generate multiple sets. We want to generate 10 sets of sequences of
length between 20 and 30 nucleotides, and each set would need to have 10
different sequences (remember that our initial script only accepts 3
parameters).</p>

<p><strong>Off topic</strong></p>

<p><code>for ((i=0;i&lt;10;i++))
  do     
  min=20     
  max=30     
  nseqs=10
  python randomset.py $nseqs $min $max     
echo  done</code></p>

<p>That should do the trick. You can even type it from the command line.
The <code>for</code> syntax in bash is very similar to C/C++ but it does not accept
spaces, as all the variable assignments. When you assigning a variable
value in bash you omit the dollar sign and when you are reading the
variable value you access it with the dollar sign, like in the line that
runs the script. A <code>for</code> loop in bash starts with a <code>do</code> command and
ends in a <code>done</code> command, not brackets or parenthesis involved. The
<code>echo</code> at the end helps adding a separating blank line after each random
set. <strong>off topic</strong></p>

<p>Of course we could use &#8220;makenucseq from the EMBOSS
package&#8221;, we can even use <a href="http://http://scit.us/projects/dawg">Dawg</a>
from my friend Reed Cartwright, which I believe has a shorter learning
curve (and faster compilation) than EMBOSS, but that will be away from
the scope of this blog, which is to show how to use <strong>Python</strong> in
<strong>simple</strong> bioinformatics scripts. I allowed myself to post off topic
subjects (including this last paragraph) because a (very) small
percentage of the commenter(s) seems to misunderstand the real focus of
this tutorial. We don&#8217;t need fancy words or complex topics to show how
simple things can achieve good results in almost no time. My focus is
not give the reader the fish and chips ready to eat. My focus is to show
how to fish and peel, cut and fry the potato in order to cook the dish.
This way, next time he or she might be cooking better things to eat.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/26/nobel/">Nobel</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-26T08:48:51-04:00" pubdate data-updated="true">Oct 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I would kill for a Nobel peace prize.
<em>Steven Wright</em></p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/25/so-he-is-gone/">So, he is gone</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-25T17:03:06-04:00" pubdate data-updated="true">Oct 25<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Yep, he is <a href="http://www.sciam.com/article.cfm?chanID=sa003&amp;articleID=D898BFCA-E7F2-99DF-39F1E89CA728AD43&amp;ref=rss">gone</a>. Until the day he comes back. And he will.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/24/using-subversion-and-svn-time-lapse-to-edit-a-manuscript-take-2-a-short-tutorial/">Using subversion and svn-time-lapse to edit a manuscript, take 2, a short tutorial - PART I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-24T22:25:26-04:00" pubdate data-updated="true">Oct 24<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Last post I introduced svn-time-lapse, which is already in <a href="http://svn-time-lapse-view.googlecode.com/files/svn-time-lapse-view-1.3.zip">version 1.3</a>. Jon included the option to view only the changes what I think might be an extra advantage when you are editing a manuscript. I outline here a small tutorial on how to edit a manuscript with subversion and svn-time-lapse. This is based on a Windows system, but it can be used with small changes for Linux, which has the advantage of having (most of the time) subversion installed.</p>

<p>(Be aware that this won&#8217;t work for Word files, you need to be using a text based, ie Latex, document)</p>

<p>First you need to <a href="http://tortoisesvn.net/downloads">download </a>and install the TortoiseSVN interface for Windows. After installing it, the Windows Explorer right-click menu will have some extra options, as can be seen on the figure below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn1.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn1.png" alt="svn1.png" /></a></p>

<p>Checkout allows you to get the source from a repository, while TortoiseSVN gives a list of features to use with subversion. Now, we need to create a directory where the manuscript repository is going to reside. You can create the repository in any place, online, locally, etc. But if you don&#8217;t have an online place to create it, I suggest to create underneath a Documents type of location. On Windows you can create a new folder inside the (My) Documents directory. So, make the new folder <strong>manuscript</strong>. <a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn2.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn2.png" alt="svn2.png" /></a></p>

<p>and the right-click on it and navigate to <strong>TortoiseSVN>>Create repository here</strong>. You can select either format. Done, your repository is set.</p>

<p><strong>SVN</strong><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn3.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn3.png" alt="svn3.png" /></a></p>

<p>If you open the manuscript directory you will see different files, but of course it is empty of actual content. You need to add some content to it and the ideal way is to check out (create a copy for edition) the repository in another location, drop files into this copy and add them to the version control.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn4.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn4.png" alt="svn4.png" /></a></p>

<p>You created the repository in your &#8220;Documents&#8221; directory, so you can check out it to your Desktop (or to any location). This will be your &#8220;local&#8221; copy of the manuscript, the one that you will work on. Close all open windows and right click on your desktop and select <strong>SVN Checkout</strong> from the popup menu.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn5.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn5.png" alt="svn5.png" /></a></p>

<p>In the dialog that appear enter the location of the repository (in your &#8220;Documents&#8221; dir) and the location where you want to checkout to (your &#8220;Desktop&#8221; location), just like the figure below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn6.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn6.png" alt="svn6.png" /></a></p>

<p>Click OK and a &#8220;local&#8221; copy of the repository will be created.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn7.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn7.png" alt="svn7.png" /></a></p>

<p>To be continued.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/23/rowling-dumbledore-is-gay/">Rowling: &#8216;Dumbledore Is Gay&#8217;</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-23T11:19:46-04:00" pubdate data-updated="true">Oct 23<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>At a public appearance at Carnegie Hall in New York, author J.K. Rowling announced that Dumbledore, the headmaster character in her popular Harry Potter series, is homosexual. What do you think?</p>

<p>Larry Hahn,
Water Delivery Man
&#8220;Wow. I haven&#8217;t heard that. I&#8217;ve been really busy lately not caring about the sexual preferences of fictional people.&#8221;</p>

<p>More <a href="http://www.theonion.com/content/node/68575">here</a>.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/23/never-trust/">Never trust</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-23T10:10:40-04:00" pubdate data-updated="true">Oct 23<span>rd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Never trust the judgment  of a man (or a woman) driving with sunglasses on a rainy day.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/20/using-subversion-and-svn-time-lapse-to-edit-a-manuscript/">Using subversion and svn-time-lapse to edit a manuscript, take 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-20T00:41:26-04:00" pubdate data-updated="true">Oct 20<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Neil Saunders started the thread <a href="http://nsaunders.wordpress.com/2007/10/03/organise-your-bioinformatics-projects-using-subversion-and-trac-part-1/">here </a>and <a href="http://nsaunders.wordpress.com/2007/10/05/organise-your-bioinformatics-projects-using-subversion-and-trac-part-2/">here</a> of using subversion to maintain your files, projects and general stuff. Myself, I find trac a little bit clumsy to check differences between two versions. It is easy to see what has been changed but if the modifications are large it gets a little bit tedious to scroll down and search everything. An example is below.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/trac_example.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/trac_example.thumbnail.png" alt="Example of file comparison in trac (2 versions, Python)" /></a></p>

<p>Enters <a href="http://jonaquino.blogspot.com/2007/10/svn-time-lapse-view.html">svn-time-lapse</a>, which seems to be a very interesting tool to compare two versions of a file. I decided to test it on a manuscript edition. I first created a a repository on my Desktop and added an old LaTex file. I replaced this file with a different version of the article (1 or 2 revisions ahead) and opened the repository with svn-time-lapse. The program instantly highlights the portions of the article that are different and display both side by side. One advantage is that svn-time-lapse already gets a certain number of revisions from the repository that can be compared to the current file. A scroll on the top gives an extra help moving from one revision to another.</p>

<p><a href="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn_time_lapse_example.png"><img src="http://blindscientist.genedrift.org/wordpress/wp-content/uploads/2007/10/svn_time_lapse_example.thumbnail.png" alt="svn-time-lapse example" /></a></p>

<p>The program is available from the developer&#8217;s website, Jon Aquino, and from Google Code. The release still needs some polishing but it is stable enough to be used extensively.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/10/18/watson-yet-again/">Watson, yet again</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-10-18T15:25:18-04:00" pubdate data-updated="true">Oct 18<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>So, Watson <a href="http://www.iht.com/articles/ap/2007/10/18/europe/EU-GEN-Britain-US-Scientist-Racism.php">said</a> what he said and then got some lectures <a href="http://www.cbc.ca/world/story/2007/10/18/science-watson-controversy.html">canceled</a>. Fine.</p>

<p>Now Cold Spring Harbor issued an <a href="http://www.cshl.edu/public/releases/07_statement.html">statement</a> basically distancing themselves from Watson and his comments. I fail to understand one thing: when he says good things, he is the Chancellor of CSHL and when they don&#8217;t like the comment, he is not &#8220;speaking on behalf of the institution&#8221;.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/32/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/30/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/09/06/goodbye-and-so-long/">Goodbye and so long</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/03/06/testing-feed/">Testing feed</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
