
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Browsing the &#8220;internets&#8221;, I came upon new bioinformatics package called CLC Workbench. This is new at least for me: the site informs that &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/36">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/30/new-review-clc-workbench-free-edition-and-more/">New review: CLC Workbench, free edition, and more</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-30T16:42:55-04:00" pubdate data-updated="true">Apr 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Browsing the &#8220;internets&#8221;, I came upon new bioinformatics package called CLC Workbench. This is new at least for me: the site informs that more than 36000 users have downloaded the application.  The program is developed by the Danish company <a href="http://www.clcbio.com/">CLC bio</a> and they have some other nice products that I will mention in other posts.</p>

<p>I downloaded the free edition, the one that will give me everything for free and won&#8217;t put limits on this review. The order price for the full version is a hefty 4500 dollars, although you can rent the package monthly or purchase just a limited set of functionalities. The download went smoothly and installation was easy on my Fedora Core and XP boxes. As <a href="http://www.geneious.com">Geneious</a>, CLC workbench is a Java application, and this is evident during the installation and first run. But Geneious interface, even though resembles a bit the &#8220;old&#8221; Java interfaces, looks cleaner and better designed than CLC Workbench.</p>

<p>On the first run on Linux, the program asks me if I want to download a example data set. Sure. It starts downloading and fails with an error message. There is an advance tab on the message but it does not work, the program seems frozen with a &#8220;Downloading data set&#8221; notice. I killed the process and I wondered why they did not include the data set with the download. The package is 25 Mb in size, a couple more won&#8217;t make a difference in high bandwidth times. I tried running again and no automatic message greeted me, so I decided downloading the data set manually. It failed again. I tried setting up the proxy and then searching NCBI: nothing.</p>

<p>I then moved back to my XP box and run the program. The same message greeted me and the this the download went smoothly. The interface is very similar on Linux and Windows, and the program&#8217;s approach is very similar to Geneious, as you have the same &#8220;feel&#8221; when running the program. There is no innovative GUI or analytical features on CLC Workbench, it is basically the same thing you see in other free packages. Basically you can align sequences, create extremely simple trees, analyse your DNA and protein and so on.</p>

<p>Most of its features overlaps with Geneious or BioEdit, and the late biOpen. Some random impressions on CLC Workbench:</p>

<ul>
<li><p>alignment procedure extremely slow</p></li>
<li><p>interface is sometimes difficult to understand</p></li>
<li><p>it is hard to get used to double click on the left hand side toolbox</p></li>
<li><p>font size is large and clean</p></li>
<li><p>database search is fast and results are cleaner (visually) than Geneious</p></li>
</ul>


<p>Based on the list above I would say that Geneious is better to the zero dollars you have to pay. It is good to see competition to Geneious, which I believe is gaining more ground in the market, especially due to its nice price for the registered version. I will test the paid versions of the other CLC&#8217;s software packages and post some reviews here.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/19/genetic-code-part-i/">Genetic code: part I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-19T16:53:29-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>The Python dictionary data-type is like hash in Perl. It is a very
similar structure, where each element in the variable is composed of two
values, more specifically a key-value pair. This is the ideal data type
to store the genetic code. As you might know the genetic code governs
the translation of DNA into proteins, where each codon (3 bases or
nucleotides in the DNA sequence) correspond to an amino acid in the
protein sequence. So for every sequence of 3 nucleotides (key) will
represent an amino acid (value). Important things: dictionaries do not
accept duplicated key values, and every time a new value is assigned to
a key the old value is erased. To create a new dictionary use the curly
brackets</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">first_dictionary</span> <span class="o">=</span> <span class="p">{}</span>
</span></code></pre></td></tr></table></div></figure>


<p> inside the curly braces we first assign a key and
separated by a colon (:), while multiple pairs should be separated by
comma. Both key and value have to be between single or double quotes.
Let&#8217;s see how we will represent the genetic code in a Python dictionary,
assigning values to keys</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
<span class='line-number'>44</span>
<span class='line-number'>45</span>
<span class='line-number'>46</span>
<span class='line-number'>47</span>
<span class='line-number'>48</span>
<span class='line-number'>49</span>
<span class='line-number'>50</span>
<span class='line-number'>51</span>
<span class='line-number'>52</span>
<span class='line-number'>53</span>
<span class='line-number'>54</span>
<span class='line-number'>55</span>
<span class='line-number'>56</span>
<span class='line-number'>57</span>
<span class='line-number'>58</span>
<span class='line-number'>59</span>
<span class='line-number'>60</span>
<span class='line-number'>61</span>
<span class='line-number'>62</span>
<span class='line-number'>63</span>
<span class='line-number'>64</span>
<span class='line-number'>65</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">gencode</span> <span class="o">=</span> <span class="p">{</span>
</span><span class='line'><span class="s">&#39;ATA&#39;</span><span class="p">:</span><span class="s">&#39;I&#39;</span><span class="p">,</span> <span class="c">#Isoleucine </span>
</span><span class='line'><span class="s">&#39;ATC&#39;</span><span class="p">:</span><span class="s">&#39;I&#39;</span><span class="p">,</span> <span class="c">#Isoleucine </span>
</span><span class='line'><span class="s">&#39;ATT&#39;</span><span class="p">:</span><span class="s">&#39;I&#39;</span><span class="p">,</span> <span class="c"># Isoleucine</span>
</span><span class='line'><span class="s">&#39;ATG&#39;</span><span class="p">:</span><span class="s">&#39;M&#39;</span><span class="p">,</span> <span class="c"># Methionine </span>
</span><span class='line'><span class="s">&#39;ACA&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="c"># Threonine </span>
</span><span class='line'><span class="s">&#39;ACC&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="c"># Threonine</span>
</span><span class='line'><span class="s">&#39;ACG&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="c"># Threonine </span>
</span><span class='line'><span class="s">&#39;ACT&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="c"># Threonine </span>
</span><span class='line'><span class="s">&#39;AAC&#39;</span><span class="p">:</span><span class="s">&#39;N&#39;</span><span class="p">,</span> <span class="c"># Asparagine</span>
</span><span class='line'><span class="s">&#39;AAT&#39;</span><span class="p">:</span><span class="s">&#39;N&#39;</span><span class="p">,</span> <span class="c"># Asparagine </span>
</span><span class='line'><span class="s">&#39;AAA&#39;</span><span class="p">:</span><span class="s">&#39;K&#39;</span><span class="p">,</span> <span class="c"># Lysine </span>
</span><span class='line'><span class="s">&#39;AAG&#39;</span><span class="p">:</span><span class="s">&#39;K&#39;</span><span class="p">,</span> <span class="c"># Lysine</span>
</span><span class='line'><span class="s">&#39;AGC&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine </span>
</span><span class='line'><span class="s">&#39;AGT&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine </span>
</span><span class='line'><span class="s">&#39;AGA&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine</span>
</span><span class='line'><span class="s">&#39;AGG&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine </span>
</span><span class='line'><span class="s">&#39;CTA&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine </span>
</span><span class='line'><span class="s">&#39;CTC&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine</span>
</span><span class='line'><span class="s">&#39;CTG&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine </span>
</span><span class='line'><span class="s">&#39;CTT&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine </span>
</span><span class='line'><span class="s">&#39;CCA&#39;</span><span class="p">:</span><span class="s">&#39;P&#39;</span><span class="p">,</span> <span class="c"># Proline</span>
</span><span class='line'><span class="s">&#39;CCC&#39;</span><span class="p">:</span><span class="s">&#39;P&#39;</span><span class="p">,</span> <span class="c"># Proline </span>
</span><span class='line'><span class="s">&#39;CCG&#39;</span><span class="p">:</span><span class="s">&#39;P&#39;</span><span class="p">,</span> <span class="c"># Proline </span>
</span><span class='line'><span class="s">&#39;CCT&#39;</span><span class="p">:</span><span class="s">&#39;P&#39;</span><span class="p">,</span> <span class="c"># Proline</span>
</span><span class='line'><span class="s">&#39;CAC&#39;</span><span class="p">:</span><span class="s">&#39;H&#39;</span><span class="p">,</span> <span class="c"># Histidine </span>
</span><span class='line'><span class="s">&#39;CAT&#39;</span><span class="p">:</span><span class="s">&#39;H&#39;</span><span class="p">,</span> <span class="c"># Histidine </span>
</span><span class='line'><span class="s">&#39;CAA&#39;</span><span class="p">:</span><span class="s">&#39;Q&#39;</span><span class="p">,</span> <span class="c"># Glutamine</span>
</span><span class='line'><span class="s">&#39;CAG&#39;</span><span class="p">:</span><span class="s">&#39;Q&#39;</span><span class="p">,</span> <span class="c"># Glutamine </span>
</span><span class='line'><span class="s">&#39;CGA&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine </span>
</span><span class='line'><span class="s">&#39;CGC&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine</span>
</span><span class='line'><span class="s">&#39;CGG&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine </span>
</span><span class='line'><span class="s">&#39;CGT&#39;</span><span class="p">:</span><span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="c"># Arginine </span>
</span><span class='line'><span class="s">&#39;GTA&#39;</span><span class="p">:</span><span class="s">&#39;V&#39;</span><span class="p">,</span> <span class="c"># Valine</span>
</span><span class='line'><span class="s">&#39;GTC&#39;</span><span class="p">:</span><span class="s">&#39;V&#39;</span><span class="p">,</span> <span class="c"># Valine </span>
</span><span class='line'><span class="s">&#39;GTG&#39;</span><span class="p">:</span><span class="s">&#39;V&#39;</span><span class="p">,</span> <span class="c"># Valine </span>
</span><span class='line'><span class="s">&#39;GTT&#39;</span><span class="p">:</span><span class="s">&#39;V&#39;</span><span class="p">,</span> <span class="c"># Valine</span>
</span><span class='line'><span class="s">&#39;GCA&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="c"># Alanine </span>
</span><span class='line'><span class="s">&#39;GCC&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="c"># Alanine </span>
</span><span class='line'><span class="s">&#39;GCG&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="c"># Alanine</span>
</span><span class='line'><span class="s">&#39;GCT&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="c"># Alanine </span>
</span><span class='line'><span class="s">&#39;GAC&#39;</span><span class="p">:</span><span class="s">&#39;D&#39;</span><span class="p">,</span> <span class="c"># Aspartic Acid </span>
</span><span class='line'><span class="s">&#39;GAT&#39;</span><span class="p">:</span><span class="s">&#39;D&#39;</span><span class="p">,</span> <span class="c"># Aspartic Acid </span>
</span><span class='line'><span class="s">&#39;GAA&#39;</span><span class="p">:</span><span class="s">&#39;E&#39;</span><span class="p">,</span> <span class="c"># Glutamic Acid </span>
</span><span class='line'><span class="s">&#39;GAG&#39;</span><span class="p">:</span><span class="s">&#39;E&#39;</span><span class="p">,</span> <span class="c"># Glutamic Acid </span>
</span><span class='line'><span class="s">&#39;GGA&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="c"># Glycine </span>
</span><span class='line'><span class="s">&#39;GGC&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="c"># Glycine </span>
</span><span class='line'><span class="s">&#39;GGG&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="c"># Glycine </span>
</span><span class='line'><span class="s">&#39;GGT&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="c"># Glycine </span>
</span><span class='line'><span class="s">&#39;TCA&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine </span>
</span><span class='line'><span class="s">&#39;TCC&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine </span>
</span><span class='line'><span class="s">&#39;TCG&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine</span>
</span><span class='line'><span class="s">&#39;TCT&#39;</span><span class="p">:</span><span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="c"># Serine </span>
</span><span class='line'><span class="s">&#39;TTC&#39;</span><span class="p">:</span><span class="s">&#39;F&#39;</span><span class="p">,</span> <span class="c"># Phenylalanine </span>
</span><span class='line'><span class="s">&#39;TTT&#39;</span><span class="p">:</span><span class="s">&#39;F&#39;</span><span class="p">,</span> <span class="c"># Phenylalanine </span>
</span><span class='line'><span class="s">&#39;TTA&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine </span>
</span><span class='line'><span class="s">&#39;TTG&#39;</span><span class="p">:</span><span class="s">&#39;L&#39;</span><span class="p">,</span> <span class="c"># Leucine </span>
</span><span class='line'><span class="s">&#39;TAC&#39;</span><span class="p">:</span><span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="c"># Tyrosine </span>
</span><span class='line'><span class="s">&#39;TAT&#39;</span><span class="p">:</span><span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="c"># Tyrosine </span>
</span><span class='line'><span class="s">&#39;TAA&#39;</span><span class="p">:</span><span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="c"># Stop </span>
</span><span class='line'><span class="s">&#39;TAG&#39;</span><span class="p">:</span><span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="c"># Stop</span>
</span><span class='line'><span class="s">&#39;TGC&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="c"># Cysteine </span>
</span><span class='line'><span class="s">&#39;TGT&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="c"># Cysteine </span>
</span><span class='line'><span class="s">&#39;TGA&#39;</span><span class="p">:</span><span class="s">&#39;</span><span class="se">\&#39;</span><span class="s">, # Stop</span>
</span><span class='line'><span class="s">&#39;TGG&#39;</span><span class="p">:</span><span class="s">&#39;W&#39;</span><span class="p">,</span> <span class="c"># Tryptophan }</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple, yet efficient. But this
is the type of functionality that would be great to have at hand
every time you write a script to translate DNA into proteins. And it is
not something that you would like to type (or even copy-and-paste) all
the time. On the next post we will create the translation script and
will also create our first Python module.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/19/sciview-scientific-interviews-part-1/">SciView: scientific interviews, part 1</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-19T15:26:33-04:00" pubdate data-updated="true">Apr 19<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Trying to generate content that would appeal to a wider range of scientists, I decided to do some email interviews with prominent scientists, in order to get their opinions on different aspects of science and research. The first interview was done with <a href="http://evolution.genetics.washington.edu/phylip/felsenstein.html">Joe Felsenstein</a> , from the University of Washington. I would like to thank him for taking the time to answer my (sometimes) dull questions.</p>

<hr />

<p><em>How do you see the &#8220;publish or perish&#8221; in science, with the increasing number of retracted papers, fake results, rushed publications and sometimes publication of extremely similar results in different papers and journals?</em></p>

<p>JF: As you imply, &#8220;publish or perish&#8221; is a Bad Thing, which leads to all these negative results.  We should all be given a lifetime stipend with freely available grant funds, in hopes that before retirement we will produce at least one passable paper.    ;-)</p>

<p><em>Phylip is a great piece of software, in most cases better than PAUP and other similar applications. Can this be linked to the fact that it is open source and freely available? Have this helped to keep the software updated and portable to any system?</em></p>

<p>JF: You will find that opinions vary greatly as to which is better, PHYLIP or PAUP<em>.  More people find PAUP</em> better than find PHYLIP better.  So your premise may be wrong, which leaves us without a need to find the reason. The availability of the source code allows people to suggest changes that increase robustness and portability.  But another reason is simply that I tend to write (and urge my programmers to write) in a &#8220;paranoid&#8221; style of C which deliberately avoids local features of one operating system or features of C that are sometimes not implemented correctly.  One correction: PHYLIP is <strong><em>not</em></strong> Open Source.  It is copyrighted to the University of Washington, and resale of it is potentially subject to a license fee.</p>

<p>Actually PHYLIP is not the result of a large community of users.  Mostly it&#8217;s me and my student programmers, with some important pieces added by others. So it is not an example of the strength of the Open Source or GNU approach.</p>

<p><em>Recently, I have focused on improving my interface design skills, trying to make the bioinformatics software-user interaction more friendly. Do you think that due to the fact that Phylip does not have a graphical interface was an obstacle to make it more widespread?</em></p>

<p>JF: People often ask me why it doesn&#8217;t have a GUI.  They also frequently request the opposite: a command-line interface with all options invoked from the command line. The problem is not having some GUI, it is having a good, intuitive GUI. That is a matter of design (Intelligent Design, I guess).  I once assigned a very good programmer to make a demonstration GUI for one of my programs, mimicking the menu that we already had (and actually then doing nothing, just showing what it would look like).  The result was so totally boring that it alerted me that the issue was not a GUI but what kind of GUI.</p>

<p><em>How do you see that some bioinformatics companies create proprietary packages that are mere wrappers of open source academic software? And what about academic software that is sold? Most of them are funded by public agencies and tax money.</em></p>

<p>JF: In the USA in the 1970s the standard assumption was that publicly funded research should not be exploited commercially without a royalty coming back to the funding agency and the researcher.  Under the Reagan administration in the 1980s the view changed to being that the duty of the government was to subsidize the private sector by having their results used without compensation. The very same administrators who, in the 1970s, checked to see that you had not secretly privatized the products of research were now assigned to make sure that these results did get used privately (without compensation).  So it is hard to say what the legal status of research products such as software is. (There is an older tradition that books written by faculty members at universities can result in royalties to the author, without the university getting any payment).</p>

<p>One problem with Open Source (and with GNU licenses) is that they permit and encourage commercial exploitation without compensating anyone.  They love this but I would rather see some of the money made getting back to the original software authors.  That is one reason we aren&#8217;t licensed under Open Source or under any of the GNU licenses.  I also have questions about whether these licenses permit the original author to say yea or nay to a proposed added feature &#8211; I would prefer to maintain that degree of control.</p>

<p><em>Your seminal work in 1981, introduced the maximum likelihood method for the calculation of phylogenetic trees. I do believe that a scientist&#8217;s work has to be evaluated in its entirety, but that Journal of Molecular Evolution publication has influenced thousands of researchers for almost 26 years. Would you consider it your most important publication?</em></p>

<p>JF:  The maximum likelihood method for phylogenies was introduced by Edwards and Cavalli-Sforza in 1964, and for molecular sequences it was first done by the great statistician Jerzy Neyman in 1971.  My 1981 paper did cite Neyman (it didn&#8217;t credit Edwards and Cavalli-Sforza but at least I have cited their pioneering work on this many times before and since).  What I did that was new was to make it practical to do multi-species cases by introducing the pruning algorithm.  I also promoted and explained ML for years.  In the 1980s I gave many talks at universities.  I usually had to explain what likelihood was because the statistics courses that biologists took didn&#8217;t bother to teach them this.  I also wrote a number of review articles and made programs available to do ML. Of my papers, the 1981 ML paper is the third most cited so far (after the 1985 bootstrap paper, a paper in 1989 in Cladistics describing PHYLIP, and just ahead of my 1985 paper on phylogenetic comparative methods).  Basically they are all part of the process of promoting a statistical modelling approach to phylogenies, so it is hard to label one as more important.  Possibly my most important paper in the long run will be my 1978 paper on a model of macroevolution, which has had no citations in the past 22 years.</p>

<p><em>Sometime ago I attended a Hennig Society meeting, mainly because it was organized by the same institute where I was doing my PhD and because I was curious to meet the researchers behind some of the papers I was reading at the time. Later when I was doing part of my work at the National Museum of Natural History, I attended a lecture by James Farris and what happened in that presentation made me coin (for my own usage) the term &#8220;cult&#8221; for cladists and cladistics. What is your take on the usual &#8220;cult&#8221;-followers narrow-minded scientific views, where no method is better than them and sometimes even alignment software is demonized.</em></p>

<p>JF: Now you have some sense of what the atmosphere was like in morphological systematics in the early 1980s.  At that time most young systematists had joined the Hennig Society in a fit of enthusiasm for phylogenetic systematics, and work coming from any other intellectual source stood a good chance of being rejected out of hand as &#8220;not being science&#8221;. My own guess is that this level of excessive zeal was a reaction to the inertia in systematics, as all the main professorships and curatorships then were in the hands of followers of Mayr and Simpson, the &#8220;evolutionary systematists&#8221;, who would not make their methods explicit.  The way to intimidate them was to up the temperature of the debate. This attracted people with certain kinds of personalities.</p>

<p>You aren&#8217;t the first person to analogize the behavior of this group to an evangelical religion. The Hennig Society types have always wanted to be seen as having a key to understanding everything phylogenetic, a key that everyone else lacked. In the longer run, their behavior alienated their own base, and there were internal conflicts. Most young systematists deserted the Hennig Society and started coming to the Evolution Meetings, where the discussion is fairly normal in tone, personal attacks are discouraged, and there is a lot of good-natured discussion among people with different views.</p>

<p>(I don&#8217;t think the issue is whether or not one uses cladistic approaches to classification, as most of the reasonable young systematists at the Evolution meetings take that position).</p>

<p>What interests me is that the hard core of the Hennig Society has not mellowed or compromised. I was expecting that they would, but they have not only remained as ferocious, but continued to hold views very distinct from everyone else&#8217;s. They have circled the wagons and are still holding out, though they must be getting a bit lonely by now.</p>

<p><em>Lately two big projects related to phylogenetics are getting some steam: the Tree of Life and the DNA barcoding. I am more aware of the latter, due to the fact that I attended a presentation about it and through a friend that woks on the project. I think that the goals of the
DNA barcoding are great for the field, wet-lab and theoretical scientists. What is your opinion about this type of project that tries to address all subjects in a big umbrella? Wouldn&#8217;t the support of a larger number of small projects be better?</em></p>

<p>JF: I don&#8217;t really object to big projects as long as they don&#8217;t suck up the money for everything else.  I was on the Genbank Advisory Committee some years ago, and I think the DNA and protein databases and Pubmed are examples of important and necessary large-scale projects.  So although I worry about money for small projects being soaked up by the big ones, I don&#8217;t object to the latter in principle.  In Canada they used to have a system of giving out small amounts of money to everybody.  They are moving away from that now, but I don&#8217;t think it should disappear entirely.</p>

<p><em>You have in Phylip&#8217;s grant webpage a &#8220;no thanks&#8221; section listing everyone that refused supporting the program along the years. In my opinion this is a bold statement and not very common in the scientific environment (at least not online). What would be your advice for the young scientist that is searching for financial support for his/her research? Should s/he be vocal when a strong application is rejected?</em></p>

<p>JF: Generally, no.  I could do this because I felt confident enough in my reputation.  It was fun.  These granting agencies can say all sorts of fatuous things in their evaluations and they are never called on this. A web page is editor-free publication so I had fun with that. I had a hope that it might cause grant reviewers to think twice about getting themselves on that list by coming up with arbitrary and ill-thought-out objections. (Based on our lab&#8217;s experience since then, alas, this isn&#8217;t working). But for a vulnerable young researcher I would say no, don&#8217;t succumb to that temptation.  You may get a reputation as a sorehead.  I know a few young scientists who are quite combative about any negative evaluation, and they do get a reputation as someone you don&#8217;t want to deal with. However in these days of increasing difficulty in finding funds, it is tough for newcomers as the agencies are looking for any reason to say no.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/10/still-simulation/">Still simulation</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-10T17:23:57-04:00" pubdate data-updated="true">Apr 10<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We use our last code as a starting point in order to generate some real
information from our simulated sequence sets. Our approach here will be
the same: functions to do all the work for us and a very simple main
code. We also reuse some code with applied before to count the
nucleotides. Let&#8217;s see the code, discussion just after it.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
<span class='line-number'>26</span>
<span class='line-number'>27</span>
<span class='line-number'>28</span>
<span class='line-number'>29</span>
<span class='line-number'>30</span>
<span class='line-number'>31</span>
<span class='line-number'>32</span>
<span class='line-number'>33</span>
<span class='line-number'>34</span>
<span class='line-number'>35</span>
<span class='line-number'>36</span>
<span class='line-number'>37</span>
<span class='line-number'>38</span>
<span class='line-number'>39</span>
<span class='line-number'>40</span>
<span class='line-number'>41</span>
<span class='line-number'>42</span>
<span class='line-number'>43</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">nucleotide_percentage</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; As&#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Cs &#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Gs &#39;</span><span class="p">,</span>
</span><span class='line'>    <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="s">&#39; Ts&#39;</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">sequence_identity</span><span class="p">(</span><span class="nb">set</span><span class="p">):</span>
</span><span class='line'>    <span class="n">iden</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">count</span> <span class="o">=</span> <span class="mf">0.0</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
</span><span class='line'>        <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
</span><span class='line'>        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">])):</span>
</span><span class='line'>            <span class="k">if</span> <span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">n</span><span class="p">]</span> <span class="o">==</span> <span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="n">n</span><span class="p">]:</span>
</span><span class='line'>                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span> <span class="n">iden</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">[</span><span class="n">x</span><span class="p">]))</span>
</span><span class='line'>                <span class="n">count</span> <span class="o">=</span> <span class="mf">0.0</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">iden</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>    <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>    <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>    <span class="n">identity</span> <span class="o">=</span> <span class="n">sequence_identity</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)):</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequenceset</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</span><span class='line'>    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>
</span><span class='line'>        <span class="k">print</span> <span class="s">&#39;sequence identity to next sequence : &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">identity</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">nucleotide_percentage</span><span class="p">(</span><span class="n">sequenceset</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">print</span>
</span></code></pre></td></tr></table></div></figure>


<p>Well, not many
new things here. The main one might the be the assignment of the
variable <code>count</code>, which receives initially a value of <code>0.0</code>, which in
this case is a float. This variable is used the calculate the sequence
identity which is a percentage, or a relative value. If we had
initialized <code>count</code> with zero, our division <code>count/len(set(x)</code> would
always be zero, due to the fact that count would have been an integer.
The ideal input for this script would have equal minimum and maximum
sequence lengths, as it the simulated set would look more like an
alignment with no indels. Something like:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'><span class="nv">$&gt;</span>python simulation.py 10 30 30
</span></code></pre></td></tr></table></div></figure>


<p>As is
pointed out in BPB, this example is more an indication that we are able
to use our Python skills to actually make some real code, with some real
output. One issue with this example is the fact that we only calculate
sequence identity of two sequences at a time. It would be ideal to have
sequence identity between all simulated sequences. This would require
much more code, of course a good educational step, but it is something
that can be easily obtained with classes and we will see this later on.
With this entry, we finished our Section 4 and we will start Section 5
with Python&#8217;s dictionaries, moving to FASTA files and classes next.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/04/the-current-state-of-bioinformatics-software/">The current state of bioinformatics software</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-04T18:19:33-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>I have been planning to write about the current state of bioinformatics software development for quite sometime. Not something scientific, but something that I see along the days and weeks in the lab.</p>

<p>This post on <a href="http://www.ghastlyfop.com/blog/2007/03/software-availabilty-quick-survey-using.html">Flags and Lollipops</a> showed the current state of availability of bioinformatics software published in the past years. The post itself says it is not scientific, but give (along with the comments) an excellent base to what I see daily.</p>

<p>In no particular order, in my opinion there are three classes of bioinformatics software developed in academia:</p>

<ul>
<li><p><strong>for-publication</strong> programs - these are the ones developed with the sole purpose of publication. They are poorly and not actively maintained, sometimes (a small percentage) lack good documentation and usually perform one type of analysis that is available in many different packages</p></li>
<li><p><strong>closed source</strong> programs - these are the relatively well maintained packages, with good or sufficient documentation. The point here is the fact that they have source not available, and in some cases they require some extra installation in order to work, not mentioning the fact that you sometimes have to fax the license agreement in order to obtain it. If you work in a close system (my case) with no admin access, it might take a week or two to have such package installed, counting the time to send the fax, to receive the application and to ask the sys admin to install it and its requirements</p></li>
<li><p>and finally <strong>open source</strong> packages - these fall in a 50/50 tie regarding maintenance: they can bell well maintained or basically abandonware. The well maintained cluster is usually well documented, have a very active group of developers and users receive good support. The abandonware group in some cases are <strong>for-publication</strong> and in other cases they are abandoned because the grad student/post-doc left the lab/job and now has to pay the bills with a day job that does not give him/her time to code.</p></li>
</ul>


<p>I have developed and released software in two of the three categories above. Myself, I don&#8217;t like to release closed source programs, but sometimes the institution where you are working at make you do that. All my extra-lab software is open source, or at least you can ask me to get the source.</p>

<p>Regarding the for-publication software, this is the most pernicious category. Sometimes they are a by-product of some larger research, sometimes they are just bad software that replicate another widely used and available package, sometimes a tool that has not scientific merit. I respect large and small projects that release a very well programmed application, with a good documentation and thrives to give support to the users. Projects that clearly show that they had project, development and testing phases along the way, and actually enhances our scientific knowledge or make scientific analysis more straight forward. But lately, I have seen greasemonkey scripts being published, what can be compared to the MS Excel macro published some years ago in Bioinformatics.</p>

<p>As always, just my 2 cents.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/04/a-script-to-simulate-dna-sequence-sets/">A script to simulate DNA sequence sets</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-04T12:17:22-04:00" pubdate data-updated="true">Apr 4<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>In Beginning Perl for Bioinformatics the chapter that covers simulated
mutations on a DNA sequence is quite verbose and the code examples
employ some subroutines to do what we have done on the last post.
Basically the code example that generates a random DNA sequence is the
last one on the chapter, but it was the first one we covered. Also this
code example has a twist that our code from the last post does not have,
which is it allows you to generate a set of sequences with different
length instead of one sequence with fixed length that our script does.
So, the script that generates a user defined simulated DNA sequence set
is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">simulate_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>        <span class="n">sequence</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">sequence</span>
</span><span class='line'>
</span><span class='line'><span class="n">setsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">minlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span class='line'><span class="n">maxlength</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequenceset</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">setsize</span><span class="p">):</span>
</span><span class='line'>    <span class="n">rlength</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minlength</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)</span>
</span><span class='line'>    <span class="n">sequenceset</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simulate_sequence</span><span class="p">(</span><span class="n">rlength</span><span class="p">))</span>
</span><span class='line'>
</span><span class='line'><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequenceset</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Simple and efficient. First we
define a function that generates a simulated DNA sequence from the four
nucleotides, again using the <code>random.choice</code>. The sequence length is
based on the parameter received by the function. The sequence length is
another random number defined in the main body of the script. This
random number is generated by <code>random.randint</code> with a range based on the
arguments given by the user when running the script. And finally, the
number of sequences to be simulated is define by the first parameter.
Seventeen lines. Next we will see how to draw some scientific
information about the sequences, such as sequence identity and
nucleotide frequency.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/04/02/random-sequences-in-python/">Random sequences in Python</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-04-02T12:57:24-04:00" pubdate data-updated="true">Apr 2<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>There is a reason to say that Python has batteries included. On the last
post we have seen a small example of randomization in Python, generating
random integer values for a (extremely) simple dice game. We move to
another example, still simple which will allow us to generate random DNA
sequences. The script is</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span class='line'><span class="n">dnaseq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s">&#39;ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">result</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">+=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">dnaseq</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="n">result</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not fancy at all, just plain simple (yet again). We
import a couple of modules, <code>sys</code> and <code>random</code>, and ask for the sequence
length as a script parameter. <code>dnaseq</code> is a list containing a tandem
repeat of ACGT, and from there we will extract our random nucleotides.
In fact <code>dnaseq</code> could have been &#8216;ACGT&#8217; only. The fact that we create a
string and convert it to a list, is just for convenience of writing
<code>'ACGT...'</code> easier than <code>['A', 'C' ...]</code>. We then declare an empty
string that will be used to store the random sequence. And inside the
loop, the command that does all the magic: <code>random.choice()</code>. This
command will return a random element from the list passed as subject.
Using it inside a loop we will get a random nucleotide on each iteration
and add it to our string. This is a very simple command, but at the same
time extremely powerful and easy to implement. A good exercise from this
would be to modify the <code>dnaseq</code> string and see if there is any change in
the final random sequence.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/30/ramdomization/">Ramdomization</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-30T15:19:30-04:00" pubdate data-updated="true">Mar 30<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We start section 4 with a very short introduction to randomization.
Chapter 7 of BPB discuss the use of randomization to obtain mutations in
DNA and protein sequences. The example given in the book is at the same
time simple and interesting, as it creates a paragraph from random
selections of nouns, adjectives, verbs and other grammar elements. Here
we are going to to create a very (stress on very) simple dice game,
where each time you run the script it will throw two dices for you and
two dices for the computer. It then prints the sum of the dices and
tells who won the match. Randomization is an important feature of
computer languages. There are different researchers involved in the
creation of the best approaches to generate random number in computers.
Random number are important in the simulation of different natural
processes, such as genetic mutation, gene drift, epidemiology, weather
forecast, etc. The better the generator, the better the simulation.
Python has a random module included, with a myriad of functions that can
perform different randomizations. In this post we will see the <code>integer</code>
randomization, and in later entries we will see some other powerful
functions. Our script is quite simple, and the only new aspect for us
here is the random module and the <code>randint</code> function.</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python </span>
</span><span class='line'><span class="kn">import</span> <span class="nn">random</span>
</span><span class='line'>
</span><span class='line'><span class="n">dice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">dice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">computerdice2</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</span><span class='line'><span class="n">mine</span> <span class="o">=</span> <span class="n">dice1</span> <span class="o">+</span> <span class="n">dice2</span>
</span><span class='line'>
</span><span class='line'><span class="n">his</span> <span class="o">=</span> <span class="n">computerdice1</span> <span class="o">+</span> <span class="n">computerdice2</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&#39;mine = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mine</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39; vs. computer = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">his</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">if</span> <span class="n">mine</span> <span class="o">&gt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;I won&quot;</span>
</span><span class='line'><span class="k">elif</span> <span class="n">mine</span> <span class="o">&lt;</span> <span class="n">his</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Computer won&quot;</span>
</span><span class='line'><span class="k">else</span><span class="p">:</span>
</span><span class='line'>    <span class="k">print</span> <span class="s">&quot;Tie. Try again&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p><code>random.randint</code> is a function that generates an integer random number
between a range specified by the number between parentheses. In the
above case, we are using a dice of 6 values. So, for every call of
<code>random.randint(1,6)</code> a random number between 1 and 6 will be generated.
We store this number in a variable, sum the user&#8217;s dices and the
computer&#8217;s and check with a <code>if</code> clause to see the winner. A good
exercise would be to make this script interactive, allowing multiple
matches. Next we will see another random module function and then we
will generate mutations on DNA sequences.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/26/command-line-arguments-and-a-second-take-on-functions/">Command line arguments and a second take on functions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-26T13:18:48-04:00" pubdate data-updated="true">Mar 26<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>We have seen, briefly, how to define and use a function in Python. Now
we are going to jump forward a bit and create a new function and at the
same time take a look on command line parameters that can be passed to
the script. If you have used command line applications before, you might
have encountered programs that asks for a file name, a calculation
parameter, etc to be passed in the command line. Python scripts are no
different, they accept such parameters. For this we have the <code>sys</code>
module that has system specific parameters and functions. We have used
before the <code>sys.exit</code>, imported as an extra module function.</p>

<p>Every operating system (even Windows) has arguments in its command line, and
programming languages usually call such arguments <code>argv</code> (in the C/C++
you have argv in the parameters of the main function). Lists in Python
start at 0 (zero), and for the argument list the first item is the
script/program name. Basically if we have this</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'><span class="nv">$&gt;</span> python myscript.py DNA.txt
</span></code></pre></td></tr></table></div></figure>


<p><code>myscript.py</code> is the argument 0 in the
list and <code>DNA.txt</code> is the argument 1. So whenever we create a script that
receives arguments in the command line, we have to start (in most cases,
be aware) from 1.</p>

<p>In Python using system arguments in the CLI will look
like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span>
</span><span class='line'><span class="n">filename</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</span><span class='line'><span class="n">valueone</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>We will a variation of our
previous script that counts the bases, now with command line arguments
and a function (with no &#8220;error&#8221; checking at first)</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env python import sys </span>
</span><span class='line'>
</span><span class='line'><span class="k">def</span> <span class="nf">count_nucleotide_types</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
</span><span class='line'>    <span class="n">totalA</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalC</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;C&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalG</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;G&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">totalT</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalA</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalC</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalG</span><span class="p">)</span>
</span><span class='line'>    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">totalT</span><span class="p">)</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">sequencefile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequencefile</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">values</span> <span class="o">=</span> <span class="n">count_nucleotide_types</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;As&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Cs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Gs&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Found &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;Ts&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>Few new things here. We created a function <code>count_nucleotide_types</code> that should receive a string
containing the sequence. The &#8220;real&#8221; first line of the program flow is
the one that gets the name of the file from the command line argument,
open and read it. We then convert the list to a string, modify it a but
and throw it to the function. Get the result back, and done. With
functions we actually don&#8217;t save coding time/length (at least here), we
make out code more organized, easier to read and somewhat easier to
someone else read and understand it. It is not a good coding practice to
have long programs/scripts with no functions, no subdivision, no
structure. Functions are sometimes good program nuggets that can be
reused in the same application or even ported/copied to other
applications and reused indefinitely. Soon we will see a function and
class that reads a FASTA file in Python that can be used anywhere in any
program that needs such feature. Try the code and come back later for more.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/03/22/python-functions-simple-example/">Python functions: simple example</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-03-22T09:04:10-04:00" pubdate data-updated="true">Mar 22<span>nd</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Python subroutines do not exist. <a href="http://www.diveintopython.org/getting_to_know_python/declaring_functions.html">Everything is a function, all
functions return a value (even if it&#8217;s None), and all functions start
with <code>def</code></a>.
This statement is from Dive into Python, a book on Python programming
available for free. As mentioned Python functions start with the word
<code>def</code>, which is followed by the function name that is followed by the
arguments the function receives in between parentheses. Something like</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span></code></pre></td></tr></table></div></figure>


<p>Usually Python coders
(sometime called Pythonistas, among others), following the Python coding
style (that states: <em>Function names should be lowercase, with words
separated by underscores as necessary to improve readability</em>) name
their functions with words separated by underscores. And we are going to
use this style here, whenever a function becomes handy. The parameters
passed to the function (above <code>somevalue</code>) do not have a datatype,
Python should handle it whatever is being passed. It is also attribute
of your code to handle the parameter/value passed inside the function
and avoid errors. Functions also follow the same indentation of normal
programming and the line after the declaration should be indented with
four spaces</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">def</span> <span class="nf">my_first_function</span><span class="p">(</span><span class="n">somevalue</span><span class="p">):</span>
</span><span class='line'>    <span class="n">do_something</span>
</span></code></pre></td></tr></table></div></figure>


<p>So, let&#8217;s
warmup with functions. The following script is just the start: it adds a
poly-T tail to a DNA sequence. We are going to use our old friend
AY162388.seq. I will be back after the script</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span><span class='line'><span class="k">def</span> <span class="nf">add_tail</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
</span><span class='line'>    <span class="n">result</span> <span class="o">=</span> <span class="n">seq</span> <span class="o">+</span> <span class="s">&#39;TTTTTTTTTTTTTTTTTTTTT&#39;</span>
</span><span class='line'>    <span class="k">return</span> <span class="n">result</span>
</span><span class='line'>
</span><span class='line'><span class="n">dnafile</span> <span class="o">=</span> <span class="s">&#39;AY162388.seq&#39;</span>
</span><span class='line'><span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dnafile</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
</span><span class='line'><span class="n">sequence</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
</span><span class='line'><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">file</span><span class="p">:</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span><span class='line'>    <span class="n">sequence</span> <span class="o">=</span> <span class="n">add</span>\<span class="n">_tail</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span class='line'>    <span class="k">print</span> <span class="n">sequence</span>
</span></code></pre></td></tr></table></div></figure>


<p>Not very useful, at first sight, but gives us an
impression of what a function looks like. Basically we define a function
<code>add_tail</code> that receives <code>seq</code> as a parameter. Don&#8217;t worry about
variable scope now, we will see it later. The rest of the script is just
like things we saw before, except for the line
<code>sequence = add_tail(sequence)</code>. Here we are saving memory (yep, not
that much and not even impressive) by assigning the return value of the
function to the same string where we have the sequence stored. Run the
scritp and get ready for the command line arguments.</p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
      <a class="prev" href="/blog/page/37/">&larr; Older</a>
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/35/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
