
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Zientzilaria</title>
  <meta name="author" content="Paulo Nuin">

  
  <meta name="description" content="Tisdall&#8217;s book on Perl introduces next the ability to transcripts DNA
sequences into RNA. In order to do that we need to check a different &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://zientzilaria.genedrift.org/blog/page/40">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="Zientzilaria" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">Zientzilaria</a></h1>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:zientzilaria.genedrift.org" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2007/01/15/importing-and-regular-expressions/">Importing and regular expressions</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2007-01-15T13:47:54-05:00" pubdate data-updated="true">Jan 15<span>th</span>, 2007</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Tisdall&#8217;s book on Perl introduces next the ability to transcripts DNA
sequences into RNA. In order to do that we need to check a different
aspect of programming: regular expressions (or regex). Regular
expressions is a pattern/string expression that works
matching/describing/filtering other strings. Let&#8217;s say you want to
examine or extract all vowels contained in one phrase, one page, one
word. Another example would be to remove all html tags from a downloaded
webpage. As HTML tags are encapsulated between <code>&lt;</code> and <code>&gt;</code> signs we can
create a regex that will search for any characters in between the signs
and remove (parse) them from our page. We will deal very briefly with
regex, and if you are interested in learning more about it you can
search for countless references on the internet (such as <a href="http://www.regular-expressions.info/">this one</a>).</p>

<p>In order to use regular
expression in Python we need to check another concept present in the
language: importing modules. Python core functionality provides most of
the usual operations and also comes with a built-in library of functions
that &#8220;access to operations that are not part of the core of the language
but are nevertheless built in&#8221;. One of this operations is the ability to
interpret regular expression that in Python is located in the <code>re</code>
module. Apart from the language core, built in modules, Python can be
further extended by using third-party modules imported into the
language. Anyone can create a module and distribute to every Python user
and programmer. So, in order to have regex capabilities we literally
have to <code>import</code> the regex module. We do that by entering the line:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">re</span>
</span></code></pre></td></tr></table></div></figure>


<p>Python&#8217;s code style
guide suggests that <code>import</code> statements should be on separate lines</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="kn">import</span> <span class="nn">sys</span> <span class="nn">import</span> <span class="nn">re</span> <span class="o">...</span>
</span></code></pre></td></tr></table></div></figure>


<p>and
to be put on the top of the file (usually below the line that tells your
OS to use Python&#8217;s interpreter on the script). So the first two lines of
our new script would be</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'>\<span class="c">#! /usr/bin/env</span>
</span><span class='line'><span class="n">python</span> <span class="kn">import</span> <span class="nn">re</span>
</span></code></pre></td></tr></table></div></figure>


<p>Now that we have the ability to use regex,we need to create one expression that will transcribe our DNA sequences
into RNA.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/14/sequences-and-strings-part-ii/">Sequences and Strings - part II</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-14T15:30:32-05:00" pubdate data-updated="true">Dec 14<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Another important task for many biologists is to merge/concatenate
different strings of DNA in one unique sequence. We can modify the
previous script to concatenate two distinct DNA sequences in one. We
start using code_01 structure, adding some extra elements (line 3):</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python </span>
</span><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&#39;ACGTACGTACGTACGTACGTACGT&#39;</span>
</span><span class='line'><span class="n">myDNA2</span> <span class="o">=</span> <span class="s">&quot;TCGATCGATCGATCGATCGA&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA</span><span class="p">,</span> <span class="n">myDNA2</span>
</span></code></pre></td></tr></table></div></figure>


<p>So far, we added a new string containing an extra
DNA sequence and we <code>print</code> both sequences. In Python <code>print</code> statement
automatically adds a new line at the end of the string to be printed,
unless you add a comma (,) to the end. The comma is also needed if you
are going to print more than one string in order to separate them (<em>try
removing the comma from the code above</em>). Now, how do we merge myDNA and
myDNA2? Easy in Python: just <em>sum</em> them with a plus signal:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA3</span> <span class="o">=</span> <span class="n">myDNA</span> <span class="o">+</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA3</span>
</span></code></pre></td></tr></table></div></figure>


<p>Notice that in Python strings are immutable, meaning they cannot be
changed. This immutability confers some advantages to the code where
strings (<em>in Python strings are not <strong>variables</strong></em>) cannot be modified
anywhere in the program and also allowing some performance gain in the
interpreter. So, in order to have our sequences merged we create a third
sequence that receives both strings. Finally our code will be (some
captions were added):</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env</span>
</span><span class='line'><span class="n">python</span> <span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span><span class='line'><span class="n">myDNA2</span> <span class="o">=</span> <span class="s">&quot;TCGATCGATCGATCGATCGA&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;First and Second sequences&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA</span><span class="p">,</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="n">myDNA3</span> <span class="o">=</span> <span class="n">myDNA</span> <span class="o">+</span> <span class="n">myDNA2</span>
</span><span class='line'><span class="k">print</span> <span class="s">&quot;Concatenated sequence&quot;</span>
</span><span class='line'><span class="k">print</span> <span class="n">myDNA3</span>
</span></code></pre></td></tr></table></div></figure>


<p>Easy, eh? Of course these two simple scripts do no
scratch the surface of Python programming, but they are a start.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/13/hands-on-code-sequences-and-strings-part-i/">Hands on code: Sequences and strings - part I</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-13T22:27:48-05:00" pubdate data-updated="true">Dec 13<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>As pointed in <strong>Beginning Perl for Bioinformatics</strong>, a large percentage
of bioinformatics procedures deals with strings, especially DNA and
amino acids sequence data. As is largely known DNA is composed of four
different nucleotides: A, C, T and G and proteins can contain up to 20
amino acids. Each one of these elements have one letter of the alphabet
assigned to them. In the DNA case some letters represent one or more
nucleotides that can be present at some sequence position (click
<a href="http://www.cns.fr/externe/English/Projets/Resultats/iupaciub.html">here</a>
for more ).</p>

<p>So, as the amino acid is the basic building block of
proteins (AKA polypeptides), strings containing sequence is our most
basic block, from where all the bioinformatics magic will work on.
Usually in Perl a string is represented by the dollar sign in front of
the variable name, like this <code>$sequence</code>. Python is dynamically typed,
meaning variable types are assigned/discovered by the interpreter at run
time. This means that the value after the equal sign will tell the
interpreter what variable type you are declaring. So in Python if you
want to store a DNA sequence you can just enter</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">mydna</span><span class="o">=</span><span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p><em>a quick
note: Python can be used with the interpreter command line or by
previously saved scripts. I will try to use the latter in the code
examples.</em></p>

<p>OK, we are ready to create our first Bioinformatics Python
Hello World script. Let&#8217;s get the sequence above and print it on the
screen. The first line will tell the operating system what to use and
where to find the Python interpreter</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#! /usr/bin/env python</span>
</span></code></pre></td></tr></table></div></figure>


<p>Next we will create the variable myDNA
and assign the corresponding sequence</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span>
</span></code></pre></td></tr></table></div></figure>


<p>And finally, we will print the contents of the variable to the screen:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="k">print</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>As mentioned above, Python mandates that you have your code indented, but in our
final script this is not needed:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='python'><span class='line'><span class="c">#!/usr/bin/env </span>
</span><span class='line'><span class="n">python</span> <span class="n">myDNA</span> <span class="o">=</span> <span class="s">&quot;ACGTACGTACGTACGTACGTACGT&quot;</span> <span class="k">print</span> <span class="n">myDNA</span>
</span></code></pre></td></tr></table></div></figure>


<p>The first line tells your operating system that this
is a Python script and to use the interpreter located in that directory;
line two declares a variable called <code>myDNA</code> and assigns the sequence
string to it and the last line simply output this variable to the
screen. That simple! To run this (extremely simple) script you can copy
and paste the code above to your favourite text editor save the file
with a <code>.py</code> extension (recommended but not necessary). To run the
script, as long as you have Python installed, just open a shell and type
on the command line:</p>

<figure class='code'> <div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='bash'><span class='line'>&gt; python code_01.py
</span></code></pre></td></tr></table></div></figure>

</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/13/4/">Why Python and not Perl?</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-13T22:00:09-05:00" pubdate data-updated="true">Dec 13<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>According to the official Python
<a href="http://www.python.org/doc/essays/comparisons.html">website</a>:</p>

<blockquote><p>Python and Perl come from a similar background (Unix scripting, which
both have long outgrown) [to learn more about that check this
tutorial], and sport many similar features, but have a different
philosophy. Perl emphasizes support for common application-oriented
tasks, e.g. by having built-in regular expressions, file scanning and
report generating features. Python emphasizes support for common
programming methodologies such as data structure design and
object-oriented programming, and encourages programmers to write
readable (and thus maintainable) code by providing an elegant but not
overly cryptic notation. As a consequence, Python comes close to Perl
but rarely beats it in its original application domain; however Python
has an applicability well beyond Perl&#8217;s niche.</p></blockquote>

<p>I couldn&#8217;t explain better than that. But still I have to give my take on
why I prefer Python over Perl, and why I decided to use it in my
day-to-day programming. First I have to admit that I am lousy Perl
programmer (not even close to an apprentice monger) and I always get
confused by its syntax. Second I come from a Basic/Pascal/C++
background, all of them having slightly better syntaxes than Perl. Thus,
it was natural to get on the Python bandwagon, and as the paragraph
above states Python code is &#8220;extremely&#8221; readable (emphasis are mine); in
no-time you can grasp it completely. OK, I admit that it has at least
one odd feature : the &#8220;mandatory&#8221; indentation. In Python you have to
tabulate (using tabs or space &lt;- recommended) loops, if clauses,
functions, anything. Maybe this is the first and only hard step to get,
but after a couple of hours of coding you will be satisfied on how good
your code look.</p>
</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2006/12/13/3/">Beginning the begin</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2006-12-13T21:36:29-05:00" pubdate data-updated="true">Dec 13<span>th</span>, 2006</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>This website uses as a premise the book: Beginning Perl for
Bioinformatics by James Tisdal (click
<a href="" title="http://www.amazon.com/gp/product/0596000804/ref=as_li_tf_tl?ie=UTF8&amp;tag=genedrift-20&amp;linkCode=as2&amp;camp=1789&amp;creative=9325&amp;creativeASIN=0596000804&quot;&gt;here&lt;/a&gt;&lt;img src=&quot;http://www.assoc-amazon.com/e/ir?t=genedrift-20&amp;l=as2&amp;o=1&amp;a=0596000804&quot; width=&quot;1&quot; height=&quot;1&quot; border=&quot;0&quot; alt=&quot;&quot; style=&quot;border:none !important; margin:0px !important;">here</a>
to buy it). which was published in 2001. My idea here is to follow the
structure of the book, analyzing each chapter and converting the Perl
scripts into Python. The original book is very well written and an
excellent starting point for any aspiring bioinformatician, either if
you are a biologist that does not understand programming or a computer
scientist that does not know a lot of biology and maybe even Perl. In no
way this website/tutorial tries to plagiarize the book and I will try to
include a minimum amount Perl code, as the book is only used as an
starting point (a very good one indeed) to this journey into Python.
Here you will not find biological concept explanations and criticisms
towards Perl. Making this clear, I will start from the beginning. <strong>Why
Python (and not Perl)?</strong></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
    <a href="/blog/archives">Blog Archives</a>
    
    <a class="next" href="/blog/page/39/">Newer &rarr;</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2011/02/23/scientists-are-human-beings/">Scientists are human beings</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/16/a-plea-to-mankind-news-outlets-and-even-scientists/">A plea to mankind, news outlets and even scientists</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/15/there-are-idiots-everywhere/">There are idiots everywhere &#8230;</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/10/bioconductor/">Bioconductor</a>
      </li>
    
      <li class="post">
        <a href="/blog/2011/02/09/theres-dumb-and-theres/">There&#8217;s dumb, and there&#8217;s &#8230;</a>
      </li>
    
  </ul>
</section>




  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - Paulo Nuin -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  





  <script type="text/javascript">
    (function() {
      var script = document.createElement('script'); script.type = 'text/javascript'; script.async = true;
      script.src = 'https://apis.google.com/js/plusone.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(script, s);
    })();
  </script>



  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
